{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "20f17ea3",
   "metadata": {},
   "source": [
    "# Bio Bootcamp by justinbois"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6dbc0f3",
   "metadata": {},
   "source": [
    "## Lesson 0: Configuring your computer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd5048cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import bokeh.plotting\n",
    "import bokeh.io\n",
    "\n",
    "bokeh.io.output_notebook()\n",
    "\n",
    "# Generate plotting values\n",
    "t = np.linspace(0, 2 * np.pi, 200)\n",
    "x = 16 * np.sin(t) ** 3\n",
    "y = 13 * np.cos(t) - 5 * np.cos(2 * t) - 2 * np.cos(3 * t) - np.cos(4 * t)\n",
    "\n",
    "p = bokeh.plotting.figure(height=250, width=275)\n",
    "p.line(x, y, color=\"red\", line_width=3)\n",
    "text = bokeh.models.Label(x=0, y=0, text=\"bootcamp\", text_align=\"center\")\n",
    "p.add_layout(text)\n",
    "\n",
    "bokeh.io.show(p)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "bb44834b-944e-454d-97f4-636a2ffa846f",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        .bk-notebook-logo {\n",
       "            display: block;\n",
       "            width: 20px;\n",
       "            height: 20px;\n",
       "            background-image: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAABx0RVh0U29mdHdhcmUAQWRvYmUgRmlyZXdvcmtzIENTNui8sowAAAOkSURBVDiNjZRtaJVlGMd/1/08zzln5zjP1LWcU9N0NkN8m2CYjpgQYQXqSs0I84OLIC0hkEKoPtiH3gmKoiJDU7QpLgoLjLIQCpEsNJ1vqUOdO7ppbuec5+V+rj4ctwzd8IIbbi6u+8f1539dt3A78eXC7QizUF7gyV1fD1Yqg4JWz84yffhm0qkFqBogB9rM8tZdtwVsPUhWhGcFJngGeWrPzHm5oaMmkfEg1usvLFyc8jLRqDOMru7AyC8saQr7GG7f5fvDeH7Ej8CM66nIF+8yngt6HWaKh7k49Soy9nXurCi1o3qUbS3zWfrYeQDTB/Qj6kX6Ybhw4B+bOYoLKCC9H3Nu/leUTZ1JdRWkkn2ldcCamzrcf47KKXdAJllSlxAOkRgyHsGC/zRday5Qld9DyoM4/q/rUoy/CXh3jzOu3bHUVZeU+DEn8FInkPBFlu3+nW3Nw0mk6vCDiWg8CeJaxEwuHS3+z5RgY+YBR6V1Z1nxSOfoaPa4LASWxxdNp+VWTk7+4vzaou8v8PN+xo+KY2xsw6une2frhw05CTYOmQvsEhjhWjn0bmXPjpE1+kplmmkP3suftwTubK9Vq22qKmrBhpY4jvd5afdRA3wGjFAgcnTK2s4hY0/GPNIb0nErGMCRxWOOX64Z8RAC4oCXdklmEvcL8o0BfkNK4lUg9HTl+oPlQxdNo3Mg4Nv175e/1LDGzZen30MEjRUtmXSfiTVu1kK8W4txyV6BMKlbgk3lMwYCiusNy9fVfvvwMxv8Ynl6vxoByANLTWplvuj/nF9m2+PDtt1eiHPBr1oIfhCChQMBw6Aw0UulqTKZdfVvfG7VcfIqLG9bcldL/+pdWTLxLUy8Qq38heUIjh4XlzZxzQm19lLFlr8vdQ97rjZVOLf8nclzckbcD4wxXMidpX30sFd37Fv/GtwwhzhxGVAprjbg0gCAEeIgwCZyTV2Z1REEW8O4py0wsjeloKoMr6iCY6dP92H6Vw/oTyICIthibxjm/DfN9lVz8IqtqKYLUXfoKVMVQVVJOElGjrnnUt9T9wbgp8AyYKaGlqingHZU/uG2NTZSVqwHQTWkx9hxjkpWDaCg6Ckj5qebgBVbT3V3NNXMSiWSDdGV3hrtzla7J+duwPOToIg42ChPQOQjspnSlp1V+Gjdged7+8UN5CRAV7a5EdFNwCjEaBR27b3W890TE7g24NAP/mMDXRWrGoFPQI9ls/MWO2dWFAar/xcOIImbbpA3zgAAAABJRU5ErkJggg==);\n",
       "        }\n",
       "    </style>\n",
       "    <div>\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-notebook-logo\"></a>\n",
       "        <span id=\"dacc8c24-abd4-4bcc-8a1b-5b9a8e7e7343\">Loading BokehJS ...</span>\n",
       "    </div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\nconst JS_MIME_TYPE = 'application/javascript';\n  const HTML_MIME_TYPE = 'text/html';\n  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n  const CLASS_NAME = 'output_bokeh rendered_html';\n\n  /**\n   * Render data to the DOM node\n   */\n  function render(props, node) {\n    const script = document.createElement(\"script\");\n    node.appendChild(script);\n  }\n\n  /**\n   * Handle when an output is cleared or removed\n   */\n  function handleClearOutput(event, handle) {\n    const cell = handle.cell;\n\n    const id = cell.output_area._bokeh_element_id;\n    const server_id = cell.output_area._bokeh_server_id;\n    // Clean up Bokeh references\n    if (id != null && id in Bokeh.index) {\n      Bokeh.index[id].model.document.clear();\n      delete Bokeh.index[id];\n    }\n\n    if (server_id !== undefined) {\n      // Clean up Bokeh references\n      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n      cell.notebook.kernel.execute(cmd_clean, {\n        iopub: {\n          output: function(msg) {\n            const id = msg.content.text.trim();\n            if (id in Bokeh.index) {\n              Bokeh.index[id].model.document.clear();\n              delete Bokeh.index[id];\n            }\n          }\n        }\n      });\n      // Destroy server and session\n      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n      cell.notebook.kernel.execute(cmd_destroy);\n    }\n  }\n\n  /**\n   * Handle when a new output is added\n   */\n  function handleAddOutput(event, handle) {\n    const output_area = handle.output_area;\n    const output = handle.output;\n\n    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n      return\n    }\n\n    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n\n    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n      // store reference to embed id on output_area\n      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n    }\n    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n      const bk_div = document.createElement(\"div\");\n      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n      const script_attrs = bk_div.children[0].attributes;\n      for (let i = 0; i < script_attrs.length; i++) {\n        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n      }\n      // store reference to server id on output_area\n      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n    }\n  }\n\n  function register_renderer(events, OutputArea) {\n\n    function append_mime(data, metadata, element) {\n      // create a DOM node to render to\n      const toinsert = this.create_output_subarea(\n        metadata,\n        CLASS_NAME,\n        EXEC_MIME_TYPE\n      );\n      this.keyboard_manager.register_events(toinsert);\n      // Render to node\n      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n      render(props, toinsert[toinsert.length - 1]);\n      element.append(toinsert);\n      return toinsert\n    }\n\n    /* Handle when an output is cleared or removed */\n    events.on('clear_output.CodeCell', handleClearOutput);\n    events.on('delete.Cell', handleClearOutput);\n\n    /* Handle when a new output is added */\n    events.on('output_added.OutputArea', handleAddOutput);\n\n    /**\n     * Register the mime type and append_mime function with output_area\n     */\n    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n      /* Is output safe? */\n      safe: true,\n      /* Index of renderer in `output_area.display_order` */\n      index: 0\n    });\n  }\n\n  // register the mime type if in Jupyter Notebook environment and previously unregistered\n  if (root.Jupyter !== undefined) {\n    const events = require('base/js/events');\n    const OutputArea = require('notebook/js/outputarea').OutputArea;\n\n    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n      register_renderer(events, OutputArea);\n    }\n  }\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(\"dacc8c24-abd4-4bcc-8a1b-5b9a8e7e7343\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.1.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.1.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.1.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.1.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.1.1.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\nif (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(\"dacc8c24-abd4-4bcc-8a1b-5b9a8e7e7343\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));",
      "application/vnd.bokehjs_load.v0+json": ""
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "\n",
       "  <div id=\"ce526b1e-4c1c-4a07-8762-483b2b5f1524\" data-root-id=\"p1001\" style=\"display: contents;\"></div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "(function(root) {\n  function embed_document(root) {\n  const docs_json = {\"03e4cfae-0886-48a9-8384-f20df2d84df0\":{\"version\":\"3.1.1\",\"title\":\"Bokeh Application\",\"defs\":[],\"roots\":[{\"type\":\"object\",\"name\":\"Figure\",\"id\":\"p1001\",\"attributes\":{\"width\":275,\"height\":250,\"x_range\":{\"type\":\"object\",\"name\":\"DataRange1d\",\"id\":\"p1002\"},\"y_range\":{\"type\":\"object\",\"name\":\"DataRange1d\",\"id\":\"p1003\"},\"x_scale\":{\"type\":\"object\",\"name\":\"LinearScale\",\"id\":\"p1014\"},\"y_scale\":{\"type\":\"object\",\"name\":\"LinearScale\",\"id\":\"p1016\"},\"title\":{\"type\":\"object\",\"name\":\"Title\",\"id\":\"p1004\"},\"renderers\":[{\"type\":\"object\",\"name\":\"GlyphRenderer\",\"id\":\"p1053\",\"attributes\":{\"data_source\":{\"type\":\"object\",\"name\":\"ColumnDataSource\",\"id\":\"p1047\",\"attributes\":{\"selected\":{\"type\":\"object\",\"name\":\"Selection\",\"id\":\"p1049\",\"attributes\":{\"indices\":[],\"line_indices\":[]}},\"selection_policy\":{\"type\":\"object\",\"name\":\"UnionRenderers\",\"id\":\"p1048\"},\"data\":{\"type\":\"map\",\"entries\":[[\"x\",{\"type\":\"ndarray\",\"array\":{\"type\":\"bytes\",\"data\":\"AAAAAAAAAADRwxpAin5AP+wA/sA6eHA/2uS6aCu5iz+argasEV+gP9JCLBT+1K8/JWBMKCdauz+9vJK/kZPFP33GiWNy988/tTWu1oOQ1j+fUBeKv6jePwr6dtqMMOQ/miFcRgTp6T8RVdu/4UPwP/hpzcI2CvQ/CthawhlK+D9+o8GW4QT9PwkZT7lOHQFAoRGC+wj1A0ArzlGlXAgHQNpaPBpjVQpA4UvKGZXZDUBtE6yh58gQQIsM3gMsvRJApuW6fXPHFED4gHW8VuUWQPoslz8wFBlAzmN62yFRG0BXh1zCGpkdQHwmUAbe6B9AblyhwoQeIUBtB6qXDkkiQNOSdktBciNA58CsXEqYJED0lhH3U7klQF2Vi/WI0yZATqpT5RjlJ0C/9SsBPOwoQAIOhRs35ylAUOagbl/UKkDobf1JHrIrQNjun5T0fixAvmk8HH45LUDsx6ipdOAtQLqykdOyci5AWfr8iDbvLkDOYMpNI1UvQEfbFSTEoy9AVFQcHo3aL0DU+QuVHPkvQL9c+QE8/y9A60MJd+DsL0CFH7q3KsIvQD2TFfBmfy9ABZVwCwwlL0AxKD2surMuQJDOScc7LC5AF3Ge5X6PLUAX2vIQmN4sQIgPfm+9GixANiuWlERFK0CyFESMn18qQBowjKhZaylA00a7FxRqKEC/W4RLgl0nQBSkJzlmRyZAv4sze4wpJUDEsbVdyAUkQISL4t3v3SJAk2pWptezIUAnuhkRT4kgQCe5CHE4wB5A+4HyP/BzHED1yHf6FTEaQB/uq1XR+hdAD/dIlxrUFUBQkWROtL8TQJ+yBYwlwBFATgt0UWmvD0DUJAtGxRAMQDBKTkLOpwhAJRJoPlp3BUB8NQKioIECQJsvxvZskP8/LLxiZBiY+j8TLJ512Rr2P/IUk5f8F/I/jJCMTRIb7T82dCafkfDmP139kKSjp+E/NLidSlBq2j/PPTZQoxfTP8S9Resncco/GPUu0m1VwT8hVb+N9CC1P/YSfv63Qqc/O19QwuT5lT9lGj9+yxCAP6Ns3fgY0Vs/0/ypcx6AED+B/alzHoAQvyps3fgY0Vu/Oho/fssQgL8RX1DC5PmVvxETfv63Qqe/OFW/jfQgtb8I9S7SbVXBv6y9Resnccq/wj02UKMX078juJ1KUGrav2f9kKSjp+G/KnQmn5Hw5r96kIxNEhvtv+oUk5f8F/K/Ciyeddka9r80vGJkGJj6v6QvxvZskP+/dTUCoqCBAsAeEmg+WncFwCxKTkLOpwjAyyQLRsUQDMBTC3RRaa8PwJqyBYwlwBHATpFkTrS/E8AJ90iXGtQVwBzuq1XR+hfA+8h3+hUxGsD0gfI/8HMcwCC5CHE4wB7AKboZEU+JIMCRalam17MhwIeL4t3v3SLAwrG1XcgFJMC9izN7jCklwBekJzlmRybAvFuES4JdJ8DWRrsXFGoowBUwjKhZaynAshREjJ9fKsA5K5aUREUrwIUPfm+9GizAGtryEJjeLMAUcZ7lfo8twJDOScc7LC7AMSg9rLqzLsAFlXALDCUvwECTFfBmfy/AhR+6tyrCL8DrQwl34OwvwL9c+QE8/y/A1PkLlRz5L8BUVBwejdovwEfbFSTEoy/AzmDKTSNVL8BZ+vyINu8uwL2ykdOyci7A7MeoqXTgLcC+aTwcfjktwNjun5T0fizA7W39SR6yK8BQ5qBuX9QqwAIOhRs35ynAv/UrATzsKMBLqlPlGOUnwGKVi/WI0ybA9JYR91O5JcDnwKxcSpgkwNWSdktBciPAbQeqlw5JIsByXKHChB4hwHwmUAbe6B/AUIdcwhqZHcDRY3rbIVEbwPcslz8wFBnA/oB1vFblFsCm5bp9c8cUwJYM3gMsvRLAchOsoefIEMDYS8oZldkNwOZaPBpjVQrAK85RpVwIB8CvEYL7CPUDwAkZT7lOHQHAdaPBluEE/b8X2FrCGUr4v/ZpzcI2CvS/IVXbv+FD8L+kIVxGBOnpvwL6dtqMMOS/tlAXir+o3r+uNa7Wg5DWv63GiWNy98+/xrySv5GTxb8GYEwoJ1q7vwJDLBT+1K+/ka4GrBFfoL9I5bpoK7mLv/sA/sA6eHC/VMMaQIp+QL/Soz3ndHl1tg==\"},\"shape\":[200],\"dtype\":\"float64\",\"order\":\"little\"}],[\"y\",{\"type\":\"ndarray\",\"array\":{\"type\":\"bytes\",\"data\":\"AAAAAAAAFEBRqPwN6BQUQG0hFSthUxRA9Dbf8q66FECWFkztmUkVQHapa0xz/hVALlAOFxrXFkDAR1quAdEXQFMBv5c56RhAW/BZcHYcGkAA1u3qG2cbQIHsDbZHxRxAZVkeJd0yHkAM31V0kasfQDhXBT98lSBAXQYm20hWIUA0PpCg6hUiQDgQ23sg0iJA00u0mLKII0Al11iTeDckQEXc8WJf3CRA7tjy6W51JUBkWwQbzwAmQIA9tKLMfCZAucoHCN3nJkD57x44okAnQBIxUXPthSdARvCElMG2J0BKeO6tVNInQObJ3PgQ2CdA2Oe0GJXHJ0DwvK+zs6AnQKltUWZyYydAUerZGAgQJ0CmVxXA2qYmQHhi75R8KCZAeYX3z6iVJUDvMZf3P+8kQLAXENFDNiRAWd9rBdNrI0DX/lOMJJEiQD10Te+CpyFAhdYPeEewIECPSkK5qlkfQC0ZEvcpPR1APhWic94NG0B1ckwnks4YQJBJo97+gRZALnrEgcYqFEDZKasWbMsRQCp2RimbzA5ATiGDNj37CUC+1402yCYFQIzwVvfhUgBAVMI/A48F9z+iekK8KeXqP2gXly4Ejc8/QhFwQGLl1b8YP6xGYJbtvz5ExUaFAfi/g//an1eNAMAeG/S81goFwI4+nXcKeQnAippitf3XDcAGXaJF+BMRwMkUDTKmNBPARiA04UtOFcAwQyPtNGEXwPFJwB+vbRnA3mhkcQR0G8B/tzlLdXQdwEdq9Cczbx/Ai3q21S2yIMDGnZQj+qkhwJFsiT/zniLA1xg+b/6QI8Azth1V8H8kwKNWjCyMayXA4LwRbINTJsBmJWXOdTcnwOOOa8HxFijAaUBmOHXxKMAE8cvcbsYpwFh6m5c/lSrAZIZqazxdK8A9wguWsB0swFi0bO/f1SzAt4xAegmFLcDxbTcaaiouwFQv5GI/xS7A9eEPcspUL8BH5BPIUtgvwI65dIiUJzDAauh36FRcMMCUVHv0H4owwK1X73KzsDDAoIONiNbPMMAlNUOcWucwwJP2qBIc9zDAbm+l3QL/MMBub6XdAv8wwJP2qBIc9zDAJTVDnFrnMMCgg42I1s8wwKxX73KzsDDAlFR79B+KMMBr6HfoVFwwwI65dIiUJzDASOQTyFLYL8D34Q9yylQvwFIv5GI/xS7A8203GmoqLsC5jEB6CYUtwFm0bO/f1SzAPsILlrAdLMBjhmprPF0rwFd6m5c/lSrABfHL3G7GKcBrQGY4dfEowOaOa8HxFijAZyVlznU3J8DfvBFsg1MmwKVWjCyMayXAM7YdVfB/JMDZGD5v/pAjwJRsiT/zniLAwp2UI/qpIcCOerbVLbIgwE1q9Cczbx/AfLc5S3V0HcDmaGRxBHQbwO1JwB+vbRnAN0Mj7TRhF8BLIDThS04VwMUUDTKmNBPAC12iRfgTEcCDmmK1/dcNwJ8+nXcKeQnAIBv0vNYKBcB5/9qfV40AwFpExUaFAfi/8j6sRmCW7b/qEXBAYuXVvzwXly4Ejc8/43pCvCnl6j8+wj8DjwX3P43wVvfhUgBAudeNNsgmBUBMIYM2PfsJQBV2RimbzA5A1SmrFmzLEUAzesSBxioUQItJo97+gRZAdnJMJ5LOGEAyFaJz3g0bQCoZEvcpPR1AkkpCuapZH0CB1g94R7AgQD10Te+CpyFA0/5TjCSRIkBZ32sF02sjQLEXENFDNiRA7TGX9z/vJEB6hffPqJUlQHdi75R8KCZAplcVwNqmJkBT6tkYCBAnQKhtUWZyYydA8Lyvs7OgJ0DX57QYlccnQOXJ3PgQ2CdAS3jurVTSJ0BH8ISUwbYnQBIxUXPthSdA/O8eOKJAJ0C2ygcI3ecmQIE9tKLMfCZAZFsEG88AJkDv2PLpbnUlQEbc8WJf3CRAJtdYk3g3JEDWS7SYsogjQDgQ23sg0iJAND6QoOoVIkBeBibbSFYhQDhXBT98lSBADd9VdJGrH0BlWR4l3TIeQH7sDbZHxRxACNbt6htnG0BY8FlwdhwaQFkBv5c56RhAvkdargHRF0AqUA4XGtcWQHepa0xz/hVAlxZM7ZlJFUD0Nt/yrroUQGshFSthUxRAUaj8DegUFEAAAAAAAAAUQA==\"},\"shape\":[200],\"dtype\":\"float64\",\"order\":\"little\"}]]}}},\"view\":{\"type\":\"object\",\"name\":\"CDSView\",\"id\":\"p1054\",\"attributes\":{\"filter\":{\"type\":\"object\",\"name\":\"AllIndices\",\"id\":\"p1055\"}}},\"glyph\":{\"type\":\"object\",\"name\":\"Line\",\"id\":\"p1050\",\"attributes\":{\"x\":{\"type\":\"field\",\"field\":\"x\"},\"y\":{\"type\":\"field\",\"field\":\"y\"},\"line_color\":\"red\",\"line_width\":3}},\"nonselection_glyph\":{\"type\":\"object\",\"name\":\"Line\",\"id\":\"p1051\",\"attributes\":{\"x\":{\"type\":\"field\",\"field\":\"x\"},\"y\":{\"type\":\"field\",\"field\":\"y\"},\"line_color\":\"red\",\"line_alpha\":0.1,\"line_width\":3}},\"muted_glyph\":{\"type\":\"object\",\"name\":\"Line\",\"id\":\"p1052\",\"attributes\":{\"x\":{\"type\":\"field\",\"field\":\"x\"},\"y\":{\"type\":\"field\",\"field\":\"y\"},\"line_color\":\"red\",\"line_alpha\":0.2,\"line_width\":3}}}}],\"toolbar\":{\"type\":\"object\",\"name\":\"Toolbar\",\"id\":\"p1008\",\"attributes\":{\"tools\":[{\"type\":\"object\",\"name\":\"PanTool\",\"id\":\"p1032\"},{\"type\":\"object\",\"name\":\"WheelZoomTool\",\"id\":\"p1033\"},{\"type\":\"object\",\"name\":\"BoxZoomTool\",\"id\":\"p1034\",\"attributes\":{\"overlay\":{\"type\":\"object\",\"name\":\"BoxAnnotation\",\"id\":\"p1035\",\"attributes\":{\"syncable\":false,\"level\":\"overlay\",\"visible\":false,\"left_units\":\"canvas\",\"right_units\":\"canvas\",\"bottom_units\":\"canvas\",\"top_units\":\"canvas\",\"line_color\":\"black\",\"line_alpha\":1.0,\"line_width\":2,\"line_dash\":[4,4],\"fill_color\":\"lightgrey\",\"fill_alpha\":0.5}}}},{\"type\":\"object\",\"name\":\"SaveTool\",\"id\":\"p1036\"},{\"type\":\"object\",\"name\":\"ResetTool\",\"id\":\"p1037\"},{\"type\":\"object\",\"name\":\"HelpTool\",\"id\":\"p1038\"}]}},\"left\":[{\"type\":\"object\",\"name\":\"LinearAxis\",\"id\":\"p1025\",\"attributes\":{\"ticker\":{\"type\":\"object\",\"name\":\"BasicTicker\",\"id\":\"p1027\",\"attributes\":{\"mantissas\":[1,2,5]}},\"formatter\":{\"type\":\"object\",\"name\":\"BasicTickFormatter\",\"id\":\"p1028\"},\"major_label_policy\":{\"type\":\"object\",\"name\":\"AllLabels\",\"id\":\"p1026\"}}}],\"below\":[{\"type\":\"object\",\"name\":\"LinearAxis\",\"id\":\"p1018\",\"attributes\":{\"ticker\":{\"type\":\"object\",\"name\":\"BasicTicker\",\"id\":\"p1020\",\"attributes\":{\"mantissas\":[1,2,5]}},\"formatter\":{\"type\":\"object\",\"name\":\"BasicTickFormatter\",\"id\":\"p1021\"},\"major_label_policy\":{\"type\":\"object\",\"name\":\"AllLabels\",\"id\":\"p1019\"}}}],\"center\":[{\"type\":\"object\",\"name\":\"Grid\",\"id\":\"p1024\",\"attributes\":{\"axis\":{\"id\":\"p1018\"}}},{\"type\":\"object\",\"name\":\"Grid\",\"id\":\"p1031\",\"attributes\":{\"dimension\":1,\"axis\":{\"id\":\"p1025\"}}},{\"type\":\"object\",\"name\":\"Label\",\"id\":\"p1056\",\"attributes\":{\"text\":\"bootcamp\",\"text_align\":\"center\",\"x\":0,\"y\":0}}]}}],\"callbacks\":{\"type\":\"map\"}}};\n  const render_items = [{\"docid\":\"03e4cfae-0886-48a9-8384-f20df2d84df0\",\"roots\":{\"p1001\":\"ce526b1e-4c1c-4a07-8762-483b2b5f1524\"},\"root_ids\":[\"p1001\"]}];\n  root.Bokeh.embed.embed_items_notebook(docs_json, render_items);\n  }\n  if (root.Bokeh !== undefined) {\n    embed_document(root);\n  } else {\n    let attempts = 0;\n    const timer = setInterval(function(root) {\n      if (root.Bokeh !== undefined) {\n        clearInterval(timer);\n        embed_document(root);\n      } else {\n        attempts++;\n        if (attempts > 100) {\n          clearInterval(timer);\n          console.log(\"Bokeh: ERROR: Unable to run BokehJS code because BokehJS library is missing\");\n        }\n      }\n    }, 10, root)\n  }\n})(window);",
      "application/vnd.bokehjs_exec.v0+json": ""
     },
     "metadata": {
      "application/vnd.bokehjs_exec.v0+json": {
       "id": "p1001"
      }
     },
     "output_type": "display_data"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "11d7ec85-21ce-4377-95ec-f3debb210ca8",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'watermark'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[4], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m get_ipython()\u001b[38;5;241m.\u001b[39mrun_line_magic(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mload_ext\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwatermark\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[0;32m      2\u001b[0m get_ipython()\u001b[38;5;241m.\u001b[39mrun_line_magic(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwatermark\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m-v -p numpy,bokeh,jupyterlab\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "File \u001b[1;32m~\\anaconda3\\Lib\\site-packages\\IPython\\core\\interactiveshell.py:2414\u001b[0m, in \u001b[0;36mInteractiveShell.run_line_magic\u001b[1;34m(self, magic_name, line, _stack_depth)\u001b[0m\n\u001b[0;32m   2412\u001b[0m     kwargs[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mlocal_ns\u001b[39m\u001b[38;5;124m'\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mget_local_scope(stack_depth)\n\u001b[0;32m   2413\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mbuiltin_trap:\n\u001b[1;32m-> 2414\u001b[0m     result \u001b[38;5;241m=\u001b[39m fn(\u001b[38;5;241m*\u001b[39margs, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39mkwargs)\n\u001b[0;32m   2416\u001b[0m \u001b[38;5;66;03m# The code below prevents the output from being displayed\u001b[39;00m\n\u001b[0;32m   2417\u001b[0m \u001b[38;5;66;03m# when using magics with decodator @output_can_be_silenced\u001b[39;00m\n\u001b[0;32m   2418\u001b[0m \u001b[38;5;66;03m# when the last Python token in the expression is a ';'.\u001b[39;00m\n\u001b[0;32m   2419\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(fn, magic\u001b[38;5;241m.\u001b[39mMAGIC_OUTPUT_CAN_BE_SILENCED, \u001b[38;5;28;01mFalse\u001b[39;00m):\n",
      "File \u001b[1;32m~\\anaconda3\\Lib\\site-packages\\IPython\\core\\magics\\extension.py:33\u001b[0m, in \u001b[0;36mExtensionMagics.load_ext\u001b[1;34m(self, module_str)\u001b[0m\n\u001b[0;32m     31\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m module_str:\n\u001b[0;32m     32\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m UsageError(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mMissing module name.\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m---> 33\u001b[0m res \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mshell\u001b[38;5;241m.\u001b[39mextension_manager\u001b[38;5;241m.\u001b[39mload_extension(module_str)\n\u001b[0;32m     35\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m res \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124malready loaded\u001b[39m\u001b[38;5;124m'\u001b[39m:\n\u001b[0;32m     36\u001b[0m     \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mThe \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m extension is already loaded. To reload it, use:\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;241m%\u001b[39m module_str)\n",
      "File \u001b[1;32m~\\anaconda3\\Lib\\site-packages\\IPython\\core\\extensions.py:76\u001b[0m, in \u001b[0;36mExtensionManager.load_extension\u001b[1;34m(self, module_str)\u001b[0m\n\u001b[0;32m     69\u001b[0m \u001b[38;5;250m\u001b[39m\u001b[38;5;124;03m\"\"\"Load an IPython extension by its module name.\u001b[39;00m\n\u001b[0;32m     70\u001b[0m \n\u001b[0;32m     71\u001b[0m \u001b[38;5;124;03mReturns the string \"already loaded\" if the extension is already loaded,\u001b[39;00m\n\u001b[0;32m     72\u001b[0m \u001b[38;5;124;03m\"no load function\" if the module doesn't have a load_ipython_extension\u001b[39;00m\n\u001b[0;32m     73\u001b[0m \u001b[38;5;124;03mfunction, or None if it succeeded.\u001b[39;00m\n\u001b[0;32m     74\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[0;32m     75\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m---> 76\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_load_extension(module_str)\n\u001b[0;32m     77\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mModuleNotFoundError\u001b[39;00m:\n\u001b[0;32m     78\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m module_str \u001b[38;5;129;01min\u001b[39;00m BUILTINS_EXTS:\n",
      "File \u001b[1;32m~\\anaconda3\\Lib\\site-packages\\IPython\\core\\extensions.py:91\u001b[0m, in \u001b[0;36mExtensionManager._load_extension\u001b[1;34m(self, module_str)\u001b[0m\n\u001b[0;32m     89\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mshell\u001b[38;5;241m.\u001b[39mbuiltin_trap:\n\u001b[0;32m     90\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m module_str \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m sys\u001b[38;5;241m.\u001b[39mmodules:\n\u001b[1;32m---> 91\u001b[0m         mod \u001b[38;5;241m=\u001b[39m import_module(module_str)\n\u001b[0;32m     92\u001b[0m     mod \u001b[38;5;241m=\u001b[39m sys\u001b[38;5;241m.\u001b[39mmodules[module_str]\n\u001b[0;32m     93\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_call_load_ipython_extension(mod):\n",
      "File \u001b[1;32m~\\anaconda3\\Lib\\importlib\\__init__.py:126\u001b[0m, in \u001b[0;36mimport_module\u001b[1;34m(name, package)\u001b[0m\n\u001b[0;32m    124\u001b[0m             \u001b[38;5;28;01mbreak\u001b[39;00m\n\u001b[0;32m    125\u001b[0m         level \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[1;32m--> 126\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m _bootstrap\u001b[38;5;241m.\u001b[39m_gcd_import(name[level:], package, level)\n",
      "File \u001b[1;32m<frozen importlib._bootstrap>:1204\u001b[0m, in \u001b[0;36m_gcd_import\u001b[1;34m(name, package, level)\u001b[0m\n",
      "File \u001b[1;32m<frozen importlib._bootstrap>:1176\u001b[0m, in \u001b[0;36m_find_and_load\u001b[1;34m(name, import_)\u001b[0m\n",
      "File \u001b[1;32m<frozen importlib._bootstrap>:1140\u001b[0m, in \u001b[0;36m_find_and_load_unlocked\u001b[1;34m(name, import_)\u001b[0m\n",
      "\u001b[1;31mModuleNotFoundError\u001b[0m: No module named 'watermark'"
     ]
    }
   ],
   "source": [
    "%load_ext watermark\n",
    "%watermark -v -p numpy,bokeh,jupyterlab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b49ccb5c-f049-47c9-a4a4-0b9dd8141e9b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The Zen of Python, by Tim Peters\n",
      "\n",
      "Beautiful is better than ugly.\n",
      "Explicit is better than implicit.\n",
      "Simple is better than complex.\n",
      "Complex is better than complicated.\n",
      "Flat is better than nested.\n",
      "Sparse is better than dense.\n",
      "Readability counts.\n",
      "Special cases aren't special enough to break the rules.\n",
      "Although practicality beats purity.\n",
      "Errors should never pass silently.\n",
      "Unless explicitly silenced.\n",
      "In the face of ambiguity, refuse the temptation to guess.\n",
      "There should be one-- and preferably only one --obvious way to do it.\n",
      "Although that way may not be obvious at first unless you're Dutch.\n",
      "Now is better than never.\n",
      "Although never is often better than *right* now.\n",
      "If the implementation is hard to explain, it's a bad idea.\n",
      "If the implementation is easy to explain, it may be a good idea.\n",
      "Namespaces are one honking great idea -- let's do more of those!\n"
     ]
    }
   ],
   "source": [
    "import this"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a2eedba6",
   "metadata": {},
   "source": [
    "## Lesson 1: Welcome and Introduction to JupyterLab"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1f8d6b2a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Beautiful is better than ugly.\n",
      "Explicit is better than implicit.\n",
      "Simple is better than complex.\n",
      "Complex is better than complicated.\n"
     ]
    }
   ],
   "source": [
    "# The first few lines from The Zen of Python by Tim Peters\n",
    "print('Beautiful is better than ugly.')\n",
    "print('Explicit is better than implicit.')\n",
    "print('Simple is better than complex.')\n",
    "print('Complex is better than complicated.')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "607a469a",
   "metadata": {},
   "source": [
    "test\n",
    "\n",
    "<table>\n",
    "    <tr>\n",
    "        <td>Foo</td>\n",
    "    </tr>\n",
    "</table>\n",
    "\n",
    "markdown syntax: https://daringfireball.net/projects/markdown/syntax\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3dedb0a8",
   "metadata": {},
   "source": [
    "Notebook shortcuts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c366def",
   "metadata": {},
   "source": [
    "<table class=\"docutils align-default\">\n",
    "<thead>\n",
    "<tr class=\"row-odd\"><th class=\"head\"><p>Quick keys</p></th>\n",
    "<th class=\"head\"><p>mode</p></th>\n",
    "<th class=\"head\"><p>action</p></th>\n",
    "</tr>\n",
    "</thead>\n",
    "<tbody>\n",
    "<tr class=\"row-even\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">Esc</span> <span class=\"pre\">-</span> <span class=\"pre\">m</span></code></p></td>\n",
    "<td><p>command</p></td>\n",
    "<td><p>switch cell to Markdown cell</p></td>\n",
    "</tr>\n",
    "<tr class=\"row-odd\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">Esc</span> <span class=\"pre\">-</span> <span class=\"pre\">y</span></code></p></td>\n",
    "<td><p>command</p></td>\n",
    "<td><p>switch cell to code cell</p></td>\n",
    "</tr>\n",
    "<tr class=\"row-even\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">Esc</span> <span class=\"pre\">-</span> <span class=\"pre\">a</span></code></p></td>\n",
    "<td><p>command</p></td>\n",
    "<td><p>insert cell above</p></td>\n",
    "</tr>\n",
    "<tr class=\"row-odd\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">Esc</span> <span class=\"pre\">-</span> <span class=\"pre\">b</span></code></p></td>\n",
    "<td><p>command</p></td>\n",
    "<td><p>insert cell below</p></td>\n",
    "</tr>\n",
    "<tr class=\"row-even\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">Esc</span> <span class=\"pre\">-</span> <span class=\"pre\">d</span> <span class=\"pre\">-</span> <span class=\"pre\">d</span> || <span class=\"pre\">del</span></code></p></td>\n",
    "<td><p>command</p></td>\n",
    "<td><p>delete cell</p></td>\n",
    "</tr>\n",
    "<tr class=\"row-odd\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">Alt</span> <span class=\"pre\">+</span> <span class=\"pre\">Enter</span></code></p></td>\n",
    "<td><p>edit</p></td>\n",
    "<td><p>execute cell and insert a cell below</p></td>\n",
    "</tr>\n",
    "</tbody>\n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ab7cd7d6",
   "metadata": {},
   "source": [
    "## Lesson 2: Basic command line skills "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b52ed343",
   "metadata": {},
   "source": [
    ">command\n",
    ">>description\n",
    "\n",
    "\n",
    ">pwd\n",
    ">>tells the path of current directory\n",
    "\n",
    ">ls\n",
    ">>list all files and folders in the current directory  \n",
    "  \n",
    "  \n",
    ">cd\n",
    ">>change directory\n",
    "        \n",
    "     \n",
    ">mkdir\n",
    ">>make new directory\n",
    "        \n",
    "     \n",
    ">rmdir\n",
    ">>delete an empty directory\n",
    "        \n",
    "\n",
    ">cd ~/...\n",
    ">>venture into that directory.\n",
    "\n",
    "\n",
    ">git pull upstream main  \n",
    ">>update directory, pulls in all of the changes from the upstream repository (from the forked one)\n",
    "\n",
    "\n",
    ">mv x y\n",
    ">>short for “move\", rename files. X: old file name; Y: new file name\n",
    "         \n",
    "\n",
    ">git mv 'x' y\n",
    ">>Remove spaces in file names! filename with space need to be in '' -> 'x'\n",
    ">>\n",
    ">>\n",
    ">>Avoid spaces in directory and file names!!!\n",
    ">>\n",
    ">>-> need git: Git will keep track of the naming changes you made. (version control)\n",
    "         \n",
    ">git mv x y/\n",
    ">>move x in folder y (/ is optional)\n",
    "        \n",
    ">more filename.x\n",
    ">>shows file content. exit with q button (visualize fasta e.g.)\n",
    ">    \n",
    ">    !type\n",
    ">>prints entire file to standard output (terminal) -> no feedback but the command is recognized.. need to put th amount of rows that want to visualize additionally\n",
    "        \n",
    ">gc filename.x -head 5\n",
    ">\n",
    ">gc myfile.txt -tail 5\n",
    ">>shows only five first or last rows of the file content\n",
    "        \n",
    ">cp\n",
    ">>Copying files and directories\n",
    ">>\n",
    ">>applied to directories it need a 'recursive' flag: cp -r\n",
    ">\n",
    "> e.g. cp dengue.fasta copy_of_dengue.fasta\n",
    ">\n",
    "> e.g. cp -r sequences copy_of_sequences | for directory\n",
    "       \n",
    ">cd ../\n",
    ">>go up one directory\n",
    ">>\n",
    ">>“../../” is two directories up, “../../../” is three directories up, and so on\n",
    "\n",
    "> rm filename.x\n",
    ">>remove a file\n",
    ">>\n",
    ">>use \"rm -i\" flag to ask first whether the file should be deleted\n",
    ">>\n",
    ">>use \"rm -r\" flag to remove directories\n",
    "        \n",
    ">git commit -m \"filename.x\"\n",
    ">\n",
    ">>git push origin main\n",
    ">>\n",
    ">>commit your change and then push the change to your main branch, which is your fork."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "026f7e06",
   "metadata": {},
   "source": [
    "## Lesson 3: Variables, operators, and types"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e63dbad4",
   "metadata": {},
   "source": [
    "see how a decimal number is represented in binary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d81caa97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'11000010'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "'{0:b}'.format(194) #2^7 + 2^6 + 2^1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be3f4d7d",
   "metadata": {},
   "source": [
    "perform bitwise XOR"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c100a3b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "194"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "10^200 "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f2f6809",
   "metadata": {},
   "source": [
    "compares the bits and gives a 1 as output if the output matches with XOR:"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1e4d8e9",
   "metadata": {},
   "source": [
    "i.e. 00001010 and 11001000 are compared, i.e. 10 and 200 and compared (..0+8+4*0+2*1+0=10 and 128+64+0+0+8+0..=200) and 194 is returned, i.e. 11000010 = 128+64+2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8da41cf0",
   "metadata": {},
   "source": [
    "represent a decimal number in binary:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7c8b8113",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'11000010'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "'{0:b}'.format(194)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "761773f0",
   "metadata": {},
   "source": [
    "These two are the same: aas the first n cubed integeres are equal to the sum of the n squared"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f5ea3c1d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "36"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "1**3+2**3+3**3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "762c523f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "36"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "(1+2+3)**2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "141cac99",
   "metadata": {},
   "source": [
    "variables: have an assigned value and a type, the increment operator adds to the variable the consequent value"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5179fce4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "a=5\n",
    "a+=5 # equal to a = a + 5\n",
    "a"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b08f645a",
   "metadata": {},
   "source": [
    "convert a string containing numbers from the actual string to an integer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5dfa8f79",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42 <class 'int'>\n"
     ]
    }
   ],
   "source": [
    "my_str = '42'\n",
    "my_int = int(my_str)\n",
    "print(my_int, type(my_int))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "021f3ef8",
   "metadata": {},
   "source": [
    "or convert the string to an integer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff6e24bd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42 <class 'int'>\n"
     ]
    }
   ],
   "source": [
    "str(my_int)\n",
    "print(my_int, type(my_int))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2e0e4d07",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "int"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "type(my_int) ## this didnt worked"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0008c9d",
   "metadata": {},
   "source": [
    "converting a float to an integer just gives us the floor and doesnt rounds the result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6394c8ec",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "int(2.999)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "111ee5d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "42\n",
      "6\n"
     ]
    }
   ],
   "source": [
    "#watch out for this too\n",
    "print('4' + '2')\n",
    "print(int('4') + int('2'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "52688efe",
   "metadata": {},
   "source": [
    "<i>Never use the == operator with floats.</i>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5c6a7bdf",
   "metadata": {},
   "outputs": [],
   "source": [
    "!= #not equal"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5edcbc1c",
   "metadata": {},
   "source": [
    "Unicode based strings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1a74522",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "ord('a') #assess number associated with character\n",
    "ord('λ')\n",
    "\n",
    "# relational operators compare the values that the ord() function is returning\n",
    "\n",
    "'a'<'λ' #true"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5c7d5937",
   "metadata": {},
   "source": [
    "Python allow chaining of relational operators."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75b7c27e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "4 < 6 < 6.1 < 9.3\n",
    "\n",
    "#important not to do the following, even though it is legal\n",
    "4 < 6.1 > 5"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd123fff",
   "metadata": {},
   "source": [
    "<i>Do not mix the directions of chained relational operators.</i>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5db30700",
   "metadata": {},
   "source": [
    "<b>identity operators</b>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4cf27c01",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(False, True)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "a=3.0\n",
    "b=3.0\n",
    "\n",
    "a is b, a is not b"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e00d3c6b",
   "metadata": {},
   "source": [
    "happens to store the two values in two different places in memory, so if one wants to compare the floats...."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "105ea967",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(True, True)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "a = 5.6\n",
    "b = a\n",
    "\n",
    "a == b, a is b"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7bc89454",
   "metadata": {},
   "source": [
    "for integers, Python employs integer caching for all integers between -5 and 256... so they are stored in the same place"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68b2e7d2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(False, True)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "a = 257\n",
    "b = 257\n",
    "c = 256\n",
    "d = 256\n",
    "\n",
    "a is b, c is d"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0740953b",
   "metadata": {},
   "source": [
    "same story for strings, it depends on which string to be stored either in the same place or not.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb62964a",
   "metadata": {},
   "source": [
    "## Lesson 4: More operators and conditionals"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1825d38",
   "metadata": {},
   "source": [
    "Different operators and their precedency"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e46d14d",
   "metadata": {},
   "source": [
    "<table class=\"docutils align-default\">\n",
    "<thead>\n",
    "<tr class=\"row-odd\"><th class=\"head\"><p>precedence</p></th>\n",
    "<th class=\"head\"><p>operators</p></th>\n",
    "</tr>\n",
    "</thead>\n",
    "<tbody>\n",
    "<tr class=\"row-even\"><td><p>1</p></td>\n",
    "<td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">**</span></code></p></td>\n",
    "</tr>\n",
    "<tr class=\"row-odd\"><td><p>2</p></td>\n",
    "<td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">*</span></code>, <code class=\"docutils literal notranslate\"><span class=\"pre\">/</span></code>, <code class=\"docutils literal notranslate\"><span class=\"pre\">//</span></code>, <code class=\"docutils literal notranslate\"><span class=\"pre\">%</span></code></p></td>\n",
    "</tr>\n",
    "<tr class=\"row-even\"><td><p>3</p></td>\n",
    "<td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">+</span></code>, <code class=\"docutils literal notranslate\"><span class=\"pre\">-</span></code></p></td>\n",
    "</tr>\n",
    "<tr class=\"row-odd\"><td><p>4</p></td>\n",
    "<td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">&lt;</span></code>, <code class=\"docutils literal notranslate\"><span class=\"pre\">&gt;</span></code>, <code class=\"docutils literal notranslate\"><span class=\"pre\">&lt;=</span></code>, <code class=\"docutils literal notranslate\"><span class=\"pre\">&gt;=</span></code></p></td>\n",
    "</tr>\n",
    "<tr class=\"row-even\"><td><p>5</p></td>\n",
    "<td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">==</span></code>, <code class=\"docutils literal notranslate\"><span class=\"pre\">!=</span></code></p></td>\n",
    "</tr>\n",
    "<tr class=\"row-odd\"><td><p>6</p></td>\n",
    "<td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">=</span></code>, <code class=\"docutils literal notranslate\"><span class=\"pre\">+=</span></code>, <code class=\"docutils literal notranslate\"><span class=\"pre\">-=</span></code>, <code class=\"docutils literal notranslate\"><span class=\"pre\">*=</span></code>, <code class=\"docutils literal notranslate\"><span class=\"pre\">/=</span></code>, <code class=\"docutils literal notranslate\"><span class=\"pre\">**=</span></code>, <code class=\"docutils literal notranslate\"><span class=\"pre\">%=</span></code>, <code class=\"docutils literal notranslate\"><span class=\"pre\">//=</span></code></p></td>\n",
    "</tr>\n",
    "<tr class=\"row-even\"><td><p>7</p></td>\n",
    "<td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">is</span></code>, <code class=\"docutils literal notranslate\"><span class=\"pre\">is</span> <span class=\"pre\">not</span></code></p></td>\n",
    "</tr>\n",
    "<tr class=\"row-odd\"><td><p>8</p></td>\n",
    "<td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">and</span></code>, <code class=\"docutils literal notranslate\"><span class=\"pre\">or</span></code>, <code class=\"docutils literal notranslate\"><span class=\"pre\">not</span></code></p></td>\n",
    "</tr>\n",
    "</tbody>\n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96f71dca",
   "metadata": {},
   "source": [
    "is in and is not is also very useful for conditional  if statements, they work with the booleans True and False-"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "49c55b70",
   "metadata": {},
   "source": [
    "## Lesson 5: Lists and tuples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8fc14d3d",
   "metadata": {},
   "source": [
    "membership operators return True or False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "865bb100",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_list = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5f2db20b",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_list_2 = [1, 2.4, 'a string', ['a string in another list', 5]]\n",
    "1 in my_list_2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3384323f",
   "metadata": {},
   "source": [
    "List indexing with [i:j] is inclusive for i but exclusive for j (goes to j-1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f74e0fd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_list[0:-3] #negative indices work as well\n",
    "[0, 1, 2, 3, 4, 5, 6, 7]\n",
    "Again, note that we only went to index -4."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4065c9c9",
   "metadata": {},
   "source": [
    "The stride comes after a second colon. For example, if we only wanted the even numbers, we could do the following."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0ab7c6aa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[0, 2, 4, 6, 8, 10]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "my_list[0::2]\n",
    "[0, 2, 4, 6, 8, 10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "77e76e74",
   "metadata": {},
   "source": [
    "General indexing scheme:\n",
    "\n",
    "my_list[start:end:stride]\n",
    "\n",
    "If there is one colon, stride is assumed to be 1.\n",
    "\n",
    "If start is not specified, it is assumed to be zero.\n",
    "\n",
    "If end is not specified, the interpreted assumed you want the entire list.\n",
    "\n",
    "If stride is not specified, it is assumed to be 1."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "052dd482",
   "metadata": {},
   "source": [
    "negative stride, which results in reversing the list."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c48738f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "\n",
    "my_list[::-1]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ce0d090",
   "metadata": {},
   "source": [
    "understand these examples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "68de9b2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2, 4, 6, 8, 10]\n",
      "[2, 4, 6, 8]\n",
      "[9, 7, 5, 3, 1]\n",
      "[9, 7, 5, 3]\n",
      "[]\n"
     ]
    }
   ],
   "source": [
    "print(my_list[2::2])\n",
    "print(my_list[2:-1:2])\n",
    "print(my_list[-2::-2])\n",
    "print(my_list[-2:2:-2])\n",
    "print(my_list[2:2:-2])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b887057f",
   "metadata": {},
   "source": [
    "but take care to make it readable and understandable, here instead of my_list[2:-1:2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3c83119",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2, 4, 6, 8]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Extract evens\n",
    "evens = my_list[::2]\n",
    "\n",
    "# Cut off end values\n",
    "evens_without_end_values = evens[1:-1]\n",
    "\n",
    "evens_without_end_values"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "461e4a85",
   "metadata": {},
   "source": [
    "use the id() function, which tells us where in memory that the variable is stored"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ded5bd5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1911101983696\n",
      "1911101983312\n"
     ]
    }
   ],
   "source": [
    "a = 689\n",
    "print(id(a))\n",
    "\n",
    "a = 690\n",
    "print(id(a))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c955b64c",
   "metadata": {},
   "source": [
    "didn’t actually change its value; we made a new variable. With lists, though, this is not the case"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "21c0cd42",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1911103017856\n",
      "1911103017856\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['zero', 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "print(id(my_list))\n",
    "\n",
    "my_list[0] = 'zero'\n",
    "print(id(my_list))\n",
    "my_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1085965f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Should it be a list?\n",
    "stop_codons = ['UAA', 'UAG', 'UGA']\n",
    "\n",
    "# or a tuple?\n",
    "stop_codons = ('UAA', 'UAG', 'UGA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5c754fe",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'UAG'"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "stop_codons[1]\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02f2f423",
   "metadata": {},
   "source": [
    "## Lesson 6: Iteration"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "13a22e9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.5517241379310345"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# The sequence we want to analyze, counting the GC content\n",
    "seq = 'GACAGACUCCAUGCACGUGGGUAUCUGUC'\n",
    "\n",
    "# Initialize GC counter\n",
    "n_gc = 0\n",
    "\n",
    "# Initialize sequence length\n",
    "len_seq = 0\n",
    "\n",
    "# Loop through sequence and count G's and C's\n",
    "for base in seq:\n",
    "    len_seq += 1\n",
    "    if base in 'GCgc':\n",
    "        n_gc += 1\n",
    "\n",
    "# Divide to get GC content\n",
    "n_gc / len_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec68c238",
   "metadata": {},
   "source": [
    "Python: <i>for base in seq:</i> #sequences, and python script in <i>italic</i>\n",
    "\n",
    "English: for each character in the string whose variable name is seq (len_seq counts), do the following, with that character taking the name base"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "82fe5051",
   "metadata": {},
   "source": [
    "## Lesson 7: Introduction to functions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3a398e6",
   "metadata": {},
   "source": [
    "just try to replicate"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "36a71c7a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def think_too_much():\n",
    "    \"\"\"Express Caesar's skepticism about Cassius\"\"\"\n",
    "    print(\"\"\"Yond Cassius has a lean and hungry look,\n",
    "He thinks too much; such men are dangerous.\"\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "61c9a22f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Yond Cassius has a lean and hungry look,\n",
      "He thinks too much; such men are dangerous.\n"
     ]
    }
   ],
   "source": [
    "think_too_much()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d396d818",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Yond Cassius has a lean and hungry look,\n",
      "He thinks too much; such men are dangerous.\n",
      "None\n"
     ]
    }
   ],
   "source": [
    "print(think_too_much())\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b890013e",
   "metadata": {},
   "source": [
    "To write a function that complements and reverses the bases of a sequence "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7763c70d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def complement_base(base):\n",
    "    \"\"\"Returns the Watson-Crick complement of a base.\"\"\"\n",
    "    if base in 'Aa':\n",
    "        return 'T'\n",
    "    elif base in 'Tt':\n",
    "        return 'A'\n",
    "    elif base in 'Gg':\n",
    "        return 'C'\n",
    "    else:\n",
    "        return 'G'\n",
    "\n",
    "\n",
    "def reverse_complement(seq):\n",
    "    \"\"\"Compute reverse complement of a sequence.\"\"\"\n",
    "    # Initialize reverse complement\n",
    "    rev_seq = ''\n",
    "\n",
    "    # Loop through and populate list with reverse complement\n",
    "    for base in reversed(seq):\n",
    "        rev_seq += complement_base(base)\n",
    "\n",
    "    return rev_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5744f2b2",
   "metadata": {},
   "source": [
    "We add a named keyword argument, also known as a named kwarg. The syntax for a named kwarg is\n",
    "\n",
    "-> could call the function with another material (RNA). Conveniently, when you call the function and omit the kwargs, they take on the default value within the function. So, if we wanted to use the default material of DNA, we don’t have to do anything different in the function call."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d3497b61",
   "metadata": {},
   "outputs": [],
   "source": [
    "def complement_base(base, material='DNA'):\n",
    "    \"\"\"Returns the Watson-Crick complement of a base.\"\"\"\n",
    "    if base in 'Aa':\n",
    "        if material == 'DNA':\n",
    "            return 'T'\n",
    "        elif material == 'RNA':\n",
    "            return 'U'\n",
    "    elif base in 'TtUu':\n",
    "        return 'A'\n",
    "    elif base in 'Gg':\n",
    "        return 'C'\n",
    "    else:\n",
    "        return 'G'\n",
    "    \n",
    "def reverse_complement(seq, material='DNA'):\n",
    "    \"\"\"Compute reverse complement of a sequence.\"\"\"\n",
    "    # Initialize reverse complement\n",
    "    rev_seq = ''\n",
    "\n",
    "    # Loop through and populate list with reverse complement\n",
    "    for base in reversed(seq):\n",
    "        rev_seq += complement_base(base, material=material)\n",
    "\n",
    "    return rev_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e97f6353",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('TGCAACTGC', 'UGCAACUGC')"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# default kwarg is DNA, but can also call it for RNA\n",
    "reverse_complement('GCAGTTGCA'), reverse_complement('GCAGTTGCA', material='RNA')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dfa93504",
   "metadata": {},
   "source": [
    "##### Splat operator"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47786894",
   "metadata": {},
   "source": [
    "For instance, if you have a function that takes separate arguments (like is_almost_right(a, b, c) in your example) and you have these arguments stored in a tuple (side_lengths = (13, 5, 12)), you can use the splat operator to unpack the tuple and pass its elements as separate arguments to the function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bc9e14d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_almost_right(a, b, c):\n",
    "    \"\"\"\n",
    "    Checks to see if a triangle with side lengths\n",
    "    `a`, `b`, and `c` is right.\n",
    "    \"\"\"\n",
    "    # Use sorted(), which gives a sorted list\n",
    "    a, b, c = sorted([a, b, c])\n",
    "\n",
    "    # Check to see if it is almost a right triangle\n",
    "    if abs(a**2 + b**2 - c**2) < 1e-12:\n",
    "        return True\n",
    "    else:\n",
    "        return False"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "33d65dfc",
   "metadata": {},
   "source": [
    "check if the triangle is according to the Pythagorean theorem"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01c64213",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "side_lengths = (13, 5, 12)\n",
    "is_almost_right(*side_lengths) # need to unpack the tuple because else it would be a single argument"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d88a35a9",
   "metadata": {},
   "source": [
    "Cant do it without unpacking the variables first"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4b8b0a8",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'is_almost_right' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m\n",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)\n",
      "Cell \u001b[1;32mIn[2], line 1\u001b[0m\n",
      "\u001b[1;32m----> 1\u001b[0m \u001b[43mis_almost_right\u001b[49m(side_lengths)\n",
      "\n",
      "\u001b[1;31mNameError\u001b[0m: name 'is_almost_right' is not defined"
     ]
    }
   ],
   "source": [
    "is_almost_right(side_lengths)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "972103c4",
   "metadata": {},
   "source": [
    "The lambda functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3636120f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0.6"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "f = lambda x, y: x / y # ration function\n",
    "\n",
    "f(3, 5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d949998",
   "metadata": {},
   "source": [
    "x in the lambda function acts as a key on which the function that we define is applied to it:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6906dd4a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Gareth Bale', 'Jesús David Murillo', 'Kellyn Acosta']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# from sorting the list by the first string\n",
    "sorted(['Kellyn Acosta', 'Jesús David Murillo', 'Gareth Bale'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "667c8457",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Kellyn Acosta', 'Gareth Bale', 'Jesús David Murillo']"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# to sort the list by last name\n",
    "# i.e. it takes the string after the lasta space and sorts by that\n",
    "sorted(\n",
    "    [\"Kellyn Acosta\", \"Jesús David Murillo\", \"Gareth Bale\"],\n",
    "    key=lambda x: x[x.rfind(\" \") + 1 :],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7ccc6dd1",
   "metadata": {},
   "source": [
    "## Lesson 8: String methods"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "536bc253",
   "metadata": {},
   "source": [
    "The seq.count() method enabled us to count the number times G and C occurred in the string seq. This notation is new. We have a variable, followed by a dot (.), and then a function. These functions are called methods in the language of object-oriented programming (OOP). If you have a string my_str, and you want to execute one of Python’s built-in string methods on it, the syntax is\n",
    "\n",
    "`my_str.string_method_of_choice(*args)`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d8f2a6f1",
   "metadata": {},
   "source": [
    "e.g. compute GC content in a sequence"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "7414bc6c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "16"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Define sequence\n",
    "seq = 'GACAGACUCCAUGCACGUGGGUAUCAUGUC'\n",
    "\n",
    "# Count G's and C's\n",
    "seq.count('G') + seq.count('C')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "67fa863e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Substrings cannot overlap\n",
    "'AAAAAAA'.count('AA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "1656c28d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find the start codon in a sequence\n",
    "seq.find('AUG')\n",
    "# if it doesnt exist, it returns -1"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "780fa547",
   "metadata": {},
   "source": [
    "Use the `in` operator to see whether there is a start codon or not"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2106709f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'AUG' in seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "803d042c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "25"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# find the first start codon from the right\n",
    "seq.rfind('AUG')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4b34f584",
   "metadata": {},
   "source": [
    "**few string methods here** and **[many more string methods](https://docs.python.org/3/library/stdtypes.html#string-methods)**\n",
    " "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "87191afa",
   "metadata": {},
   "source": [
    "find complementary base"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "140569f4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def complement_base(base):\n",
    "    \"\"\"Returns the Watson-Crick complement of a base.\"\"\"\n",
    "    # Convert to lowercase\n",
    "    base = base.lower()\n",
    "    # Or to uppercase\n",
    "    #base = base.upper()\n",
    "\n",
    "    if base == 'a':\n",
    "        return 'T'\n",
    "    elif base == 't':\n",
    "        return 'A'\n",
    "    elif base == 'g':\n",
    "        return 'C'\n",
    "    else:\n",
    "        return 'G'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e8af5ab",
   "metadata": {},
   "source": [
    "replace strings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "87d5704b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "('GACAGACTCCATGCACGTGGGTATCATGTC', 'GACAGACUCCAUGCACGUGGGUAUCAUGUC')"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# convert RNA to DNA, replace strings\n",
    "seq.replace('U', 'T'), seq # seq is unchanged, since strings are immutable"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a4c410f",
   "metadata": {},
   "source": [
    "join strings"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "16b6eb77",
   "metadata": {},
   "outputs": [],
   "source": [
    "# concatenate a tuple of words\n",
    "words = ('I', 'love', 'Python')\n",
    "\n",
    "# space as character to separate the words\n",
    "' '.join(words)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "daee2ba3",
   "metadata": {},
   "source": [
    "format strings with keyword arguments **kwargs (individual variables)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "7ff1ed8d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Let's do a Mad Lib!\n",
      "During this bootcamp, I feel truculent.\n",
      "The instructors give us haircuts.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "my_str = \"\"\"\n",
    "Let's do a Mad Lib!\n",
    "During this bootcamp, I feel {adjective}.\n",
    "The instructors give us {plural_noun}.\n",
    "\"\"\".format(adjective='truculent', plural_noun='haircuts') # kwargs\n",
    "\n",
    "print(my_str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "cf9ba7b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# string conversions, where kwargs are specified in braces"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "644d22cf",
   "metadata": {},
   "source": [
    "<table class=\"docutils align-default\">\n",
    "<thead>\n",
    "<tr class=\"row-odd\"><th class=\"head\"><p>conversion</p></th>\n",
    "<th class=\"head\"><p>description</p></th>\n",
    "</tr>\n",
    "</thead>\n",
    "<tbody>\n",
    "<tr class=\"row-even\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">d</span></code></p></td>\n",
    "<td><p>integer</p></td>\n",
    "</tr>\n",
    "<tr class=\"row-odd\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">04d</span></code></p></td>\n",
    "<td><p>integer with four digits, possibly with leading zeros</p></td>\n",
    "</tr>\n",
    "<tr class=\"row-even\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">f</span></code></p></td>\n",
    "<td><p>float, default to six digits after decimal</p></td>\n",
    "</tr>\n",
    "<tr class=\"row-odd\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">.8f</span></code></p></td>\n",
    "<td><p>float with 8 digits after the decimal</p></td>\n",
    "</tr>\n",
    "<tr class=\"row-even\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">e</span></code></p></td>\n",
    "<td><p>scientific notation, default to six digits after decimal</p></td>\n",
    "</tr>\n",
    "<tr class=\"row-odd\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">.16e</span></code></p></td>\n",
    "<td><p>scientific notation with 16 digits after the decimal</p></td>\n",
    "</tr>\n",
    "<tr class=\"row-even\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">s</span></code></p></td>\n",
    "<td><p>display as a string</p></td>\n",
    "</tr>\n",
    "</tbody>\n",
    "</table>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "e38abc17",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "There are 50 states in the US.\n",
      "Your file number is 23.\n",
      "π is approximately 3.140000.\n",
      "e is approximately 2.71828183.\n",
      "Avogadro's number is approximately 6.022000e+23.\n",
      "ε₀ is approximately 8.8541878170000005e-12 F/m.\n",
      "That rug really tied the room together.\n"
     ]
    }
   ],
   "source": [
    "# examples\n",
    "print('There are {n:d} states in the US.'.format(n=50))\n",
    "print('Your file number is {n:d}.'.format(n=23))\n",
    "print('π is approximately {pi:f}.'.format(pi=3.14))\n",
    "print('e is approximately {e:.8f}.'.format(e=2.7182818284590451))\n",
    "print(\"Avogadro's number is approximately {N_A:e}.\".format(N_A=6.022e23)) # for an apostrophe, use double quotes\n",
    "print('ε₀ is approximately {eps_0:.16e} F/m.'.format(eps_0=8.854187817e-12))\n",
    "print('That {thing:s} really tied the room together.'.format(thing='rug'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13a43623",
   "metadata": {},
   "source": [
    "## Lesson 9: Dictionaries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "30db4559",
   "metadata": {},
   "source": [
    "**dictionary syntax**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "5a57c523",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'a': 6, 'b': 7, 'c': 27.6}\n",
      "{'a': 6, 'b': 7, 'c': 27.6}\n",
      "{'a': 'yes', 'b': 'no', 'c': 'maybe'}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "7"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_dict = {'a': 6, 'b': 7, 'c': 27.6} # with curly braces\n",
    "print(my_dict),\n",
    "\n",
    "print(dict((('a', 6), ('b', 7), ('c', 27.6)))), # directly from tuples\n",
    "\n",
    "print(dict(a='yes', b='no', c='maybe')), # directly from keyword arguments (kwargs)\n",
    "\n",
    "# Pull out an entry\n",
    "my_dict['b']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed8e3bf0",
   "metadata": {},
   "source": [
    "the keys can also be immutable"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "4ec6d316",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{0: 'zero',\n",
       " 1.7: [1, 2, 3],\n",
       " (5, 6, 'dummy string'): 3.14,\n",
       " 'strings are immutable': 42}"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_dict = {\n",
    "    0: 'zero',\n",
    "    1.7: [1, 2, 3],\n",
    "    (5, 6, 'dummy string'): 3.14,\n",
    "    'strings are immutable': 42\n",
    "}\n",
    "my_dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "d1035ba4",
   "metadata": {},
   "outputs": [
    {
     "ename": "TypeError",
     "evalue": "unhashable type: 'list'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[36], line 1\u001b[0m\n\u001b[1;32m----> 1\u001b[0m my_dict \u001b[38;5;241m=\u001b[39m {\n\u001b[0;32m      2\u001b[0m     \n\u001b[0;32m      3\u001b[0m     \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mimmutable is ok\u001b[39m\u001b[38;5;124m\"\u001b[39m: \u001b[38;5;241m1\u001b[39m,\n\u001b[0;32m      4\u001b[0m     [\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmutable\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mnot\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mok\u001b[39m\u001b[38;5;124m\"\u001b[39m]: \u001b[38;5;241m5\u001b[39m\n\u001b[0;32m      5\u001b[0m }\n",
      "\u001b[1;31mTypeError\u001b[0m: unhashable type: 'list'"
     ]
    }
   ],
   "source": [
    "my_dict = {\n",
    "    \n",
    "    \"immutable is ok\": 1,\n",
    "    [\"mutable\", \"not\", \"ok\"]: 5\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bb69522",
   "metadata": {},
   "source": [
    "dictionary for aa three to one letter code and one with codons for amino acids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "797356b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "aa_dict = {\n",
    "    \"A\": \"Ala\",\n",
    "    \"R\": \"Arg\",\n",
    "    \"N\": \"Asn\",\n",
    "    \"D\": \"Asp\",\n",
    "    \"C\": \"Cys\",\n",
    "    \"Q\": \"Gln\",\n",
    "    \"E\": \"Glu\",\n",
    "    \"G\": \"Gly\",\n",
    "    \"H\": \"His\",\n",
    "    \"I\": \"Ile\",\n",
    "    \"L\": \"Leu\",\n",
    "    \"K\": \"Lys\",\n",
    "    \"M\": \"Met\",\n",
    "    \"F\": \"Phe\",\n",
    "    \"P\": \"Pro\",\n",
    "    \"S\": \"Ser\",\n",
    "    \"T\": \"Thr\",\n",
    "    \"W\": \"Trp\",\n",
    "    \"Y\": \"Tyr\",\n",
    "    \"V\": \"Val\",\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "ccdaf051",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Phe'"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aa_dict['F']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18cf27bc",
   "metadata": {},
   "source": [
    "note the zip() inner function of the dict() function ###"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "5301e074",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'TTT': 'F', 'TTC': 'F', 'TTA': 'L', 'TTG': 'L', 'TCT': 'S', 'TCC': 'S', 'TCA': 'S', 'TCG': 'S', 'TAT': 'Y', 'TAC': 'Y', 'TAA': '*', 'TAG': '*', 'TGT': 'C', 'TGC': 'C', 'TGA': '*', 'TGG': 'W', 'CTT': 'L', 'CTC': 'L', 'CTA': 'L', 'CTG': 'L', 'CCT': 'P', 'CCC': 'P', 'CCA': 'P', 'CCG': 'P', 'CAT': 'H', 'CAC': 'H', 'CAA': 'Q', 'CAG': 'Q', 'CGT': 'R', 'CGC': 'R', 'CGA': 'R', 'CGG': 'R', 'ATT': 'I', 'ATC': 'I', 'ATA': 'I', 'ATG': 'M', 'ACT': 'T', 'ACC': 'T', 'ACA': 'T', 'ACG': 'T', 'AAT': 'N', 'AAC': 'N', 'AAA': 'K', 'AAG': 'K', 'AGT': 'S', 'AGC': 'S', 'AGA': 'R', 'AGG': 'R', 'GTT': 'V', 'GTC': 'V', 'GTA': 'V', 'GTG': 'V', 'GCT': 'A', 'GCC': 'A', 'GCA': 'A', 'GCG': 'A', 'GAT': 'D', 'GAC': 'D', 'GAA': 'E', 'GAG': 'E', 'GGT': 'G', 'GGC': 'G', 'GGA': 'G', 'GGG': 'G'}\n"
     ]
    }
   ],
   "source": [
    "# The set of DNA bases\n",
    "bases = ['T', 'C', 'A', 'G']\n",
    "\n",
    "# Build list of codons\n",
    "codon_list = []\n",
    "for first_base in bases:\n",
    "    for second_base in bases:\n",
    "        for third_base in bases:\n",
    "            codon_list += [first_base + second_base + third_base]\n",
    "\n",
    "# The amino acids that are coded for (* = STOP codon)\n",
    "amino_acids = 'FFLLSSSSYY**CC*WLLLLPPPPHHQQRRRRIIIMTTTTNNKKSSRRVVVVAAAADDEEGGGG'\n",
    "\n",
    "# Build dictionary from tuple of 2-tuples (technically an iterator, but it works)\n",
    "codons = dict(zip(codon_list, amino_acids)) ###########\n",
    "\n",
    "# Show that we did it\n",
    "print(codons)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "14c9c4ac",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'I'"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codons['ATT']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "428d24a9",
   "metadata": {},
   "source": [
    "Dictionary keys and value are stored in memory by using a hash function, i.e. the implementation of a hash table.\n",
    "A hash function converts its input to an integer and having a collection of elements that can be indexed this way is named a hash table.\n",
    "\n",
    "Dictionaries themselves are mutable"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffebd5da",
   "metadata": {},
   "source": [
    "**Membership operators** can be used only with the keys of the dictionaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "9ade6488",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(True, False, True)"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Make a fresh my_dict\n",
    "my_dict = {'a': 1, 'b': 2, 'c': 3}\n",
    "\n",
    "# in works with keys\n",
    "'b' in my_dict, 'd' in my_dict, 'e' not in my_dict # if 'key' is in dict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "f83d7d2d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "a : 1\n",
      "b : 2\n",
      "c : 3\n"
     ]
    }
   ],
   "source": [
    "for key in my_dict:\n",
    "    print(key, ':', my_dict[key])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "183a4875",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "a : 1\n",
      "a : this string will not be in dictionary.\n",
      "b : 2\n",
      "b : this string will not be in dictionary.\n",
      "c : 3\n",
      "c : this string will not be in dictionary.\n",
      "\n",
      "{'a': 1, 'b': 2, 'c': 3}\n"
     ]
    }
   ],
   "source": [
    "# iterate over key,value pairs in a dictionary using the items() method of a dictionary\n",
    "for key, value in my_dict.items(): # items returns a copy of my_dict\n",
    "    print(key, ':', value)\n",
    "    value = 'this string will not be in dictionary.'\n",
    "    print(key, ':', value)\n",
    "    #my_dict[key] = 'this will be in the dictionary.'\n",
    "print()\n",
    "print(my_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2745b8d2",
   "metadata": {},
   "source": [
    "<ul class=\"simple\">\n",
    "<li><p><code class=\"docutils literal notranslate\"><span class=\"pre\">len(d)</span></code> gives the number of entries in dictionary <code class=\"docutils literal notranslate\"><span class=\"pre\">d</span></code></p></li>\n",
    "<li><p><code class=\"docutils literal notranslate\"><span class=\"pre\">del</span> <span class=\"pre\">d[k]</span></code> deletes entry with key <code class=\"docutils literal notranslate\"><span class=\"pre\">k</span></code> from dictionary <code class=\"docutils literal notranslate\"><span class=\"pre\">d</span></code></p></li>\n",
    "</ul>\n",
    "\n",
    "works for dicts and lists, but that doesnt work for tuples!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "82eb402e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "my_dict: {'a': 1, 'b': 2, 'c': 3, 'd': 4}\n",
      "my_list: [1, 2, 3, 4]\n",
      "length of my_dict: 4\n",
      "length of my_list: 4\n",
      "post-deleted my_dict: {'a': 1, 'c': 3, 'd': 4}\n",
      "post-deleted my_list: [1, 3, 4]\n"
     ]
    }
   ],
   "source": [
    "# Create my_list and my_dict for reference\n",
    "my_dict = dict(a=1, b=2, c=3, d=4)\n",
    "my_list = [1, 2, 3, 4]\n",
    "\n",
    "# Print them\n",
    "print('my_dict:', my_dict)\n",
    "print('my_list:', my_list)\n",
    "\n",
    "# Get lengths\n",
    "print('length of my_dict:', len(my_dict))\n",
    "print('length of my_list:', len(my_list))\n",
    "\n",
    "# Delete a key from my_dict\n",
    "del my_dict['b']\n",
    "\n",
    "# Delete entry from my_list\n",
    "del my_list[1]\n",
    "\n",
    "# Show post-deleted objects\n",
    "print('post-deleted my_dict:', my_dict)\n",
    "print('post-deleted my_list:', my_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "356bad5c",
   "metadata": {},
   "source": [
    "**Dictionary methods** and examples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2d2e3253",
   "metadata": {},
   "source": [
    "<table class=\"docutils align-default\">\n",
    "<thead>\n",
    "<tr class=\"row-odd\"><th class=\"head\"><p>method</p></th>\n",
    "<th class=\"head\"><p>effect</p></th>\n",
    "</tr>\n",
    "</thead>\n",
    "<tbody>\n",
    "<tr class=\"row-even\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">d.keys()</span></code></p></td>\n",
    "<td><p>return keys</p></td>\n",
    "</tr>\n",
    "<tr class=\"row-odd\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">d.pop(key)</span></code></p></td>\n",
    "<td><p>return value associated with <code class=\"docutils literal notranslate\"><span class=\"pre\">key</span></code> and delete <code class=\"docutils literal notranslate\"><span class=\"pre\">key</span></code> from <code class=\"docutils literal notranslate\"><span class=\"pre\">d</span></code></p></td>\n",
    "</tr>\n",
    "<tr class=\"row-even\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">d.values()</span></code></p></td>\n",
    "<td><p>return the values in <code class=\"docutils literal notranslate\"><span class=\"pre\">d</span></code></p></td>\n",
    "</tr>\n",
    "<tr class=\"row-odd\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">d.get(key,</span> <span class=\"pre\">None)</span></code></p></td>\n",
    "<td><p>Fetch a value in <code class=\"docutils literal notranslate\"><span class=\"pre\">d</span></code> by key giving a default value (the second argument) if <code class=\"docutils literal notranslate\"><span class=\"pre\">key</span></code> is missing</p></td>\n",
    "</tr>\n",
    "</tbody>\n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b32f7a0",
   "metadata": {},
   "source": [
    "If, say, we wanted to sort the keys and have them index-able, we would have to *convert them to a **list**.*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "f5123013",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['a', 'b', 'c', 'd'])\n",
      "['a', 'b', 'c', 'd']\n"
     ]
    }
   ],
   "source": [
    "my_dict = dict(a=1, b=2, c=3, d=4)\n",
    "\n",
    "print(my_dict.keys())\n",
    "print(sorted(list(my_dict.keys())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "01bcae1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# remove an entry from a dictionary\n",
    "my_dict.pop('c')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "0092e9e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(None, 'default value')"
      ]
     },
     "execution_count": 76,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# fetch an entry, if empty we can return a default value\n",
    "my_dict.get('c'), my_dict.get('c', 'default value')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f59993a1",
   "metadata": {},
   "source": [
    "**List methods**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bd77cdab",
   "metadata": {},
   "source": [
    "<table class=\"docutils align-default\">\n",
    "<thead>\n",
    "<tr class=\"row-odd\"><th class=\"head\"><p>method</p></th>\n",
    "<th class=\"head\"><p>effect</p></th>\n",
    "</tr>\n",
    "</thead>\n",
    "<tbody>\n",
    "<tr class=\"row-even\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">s.pop(i)</span></code></p></td>\n",
    "<td><p>return value at index <code class=\"docutils literal notranslate\"><span class=\"pre\">i</span></code> and delete it from the list  <code class=\"docutils literal notranslate\"><span class=\"pre\">s</span></code> </p></td>\n",
    "</tr>\n",
    "<tr class=\"row-odd\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">s.append(x)</span></code></p></td>\n",
    "<td><p>Put <code class=\"docutils literal notranslate\"><span class=\"pre\">x</span></code> at the end of the list <code class=\"docutils literal notranslate\"><span class=\"pre\">s</span></code> </p></td>\n",
    "</tr>\n",
    "<tr class=\"row-even\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">s.insert(i,</span> <span class=\"pre\">x)</span></code></p></td>\n",
    "<td><p>Insert <code class=\"docutils literal notranslate\"><span class=\"pre\">x</span></code> at index <code class=\"docutils literal notranslate\"><span class=\"pre\">i</span></code> in the list <code class=\"docutils literal notranslate\"><span class=\"pre\">s</span></code> </p></td>\n",
    "</tr>\n",
    "<tr class=\"row-odd\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">s.remove(x)</span></code></p></td>\n",
    "<td><p>Remove the first occurrence of <code class=\"docutils literal notranslate\"><span class=\"pre\">x</span></code> from the list <code class=\"docutils literal notranslate\"><span class=\"pre\">s</span></code> </p></td>\n",
    "</tr>\n",
    "<tr class=\"row-even\"><td><p><code class=\"docutils literal notranslate\"><span class=\"pre\">s.reverse()</span></code></p></td>\n",
    "<td><p>Reverse the order of items in the list <code class=\"docutils literal notranslate\"><span class=\"pre\">s</span></code> </p></td>\n",
    "</tr>\n",
    "</tbody>\n",
    "</table>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31bbc1a3",
   "metadata": {},
   "source": [
    "kwargs are passed into dictionaries if they're added as `**kwargs` in a function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "f3c61d77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TGACACCAGGGAGGGGGGGGGAAAATTTTT'"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# can accomodate arbitrary keyword arguments\n",
    "def concatenate_sequences(a, b, **kwargs):\n",
    "    \"\"\"Concatenate (combine) 2 or more sequences.\"\"\"\n",
    "    seq = a + b\n",
    "\n",
    "    for key in kwargs:\n",
    "        seq += kwargs[key]\n",
    "        #print(kwargs[key])\n",
    "\n",
    "    return seq\n",
    "\n",
    "concatenate_sequences('TGACAC', 'CAGGGA', c='GGGGGGGGG', d='AAAATTTTT')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3d58a4ea",
   "metadata": {},
   "source": [
    "When providing an own dictionary, we need to provide the arguments individually `**`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "id": "bc7f0c57",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'TGACACCAGGGAGGGGGGGGGAAAATTTTT'"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "my_dict = {\"a\": \"TGACAC\", \"b\": \"CAGGGA\", \"c\": \"GGGGGGGGG\", \"d\": \"AAAATTTTT\"}\n",
    "concatenate_sequences(**my_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c8775344",
   "metadata": {},
   "source": [
    "The risk is that one needs to stick to the convention set in the funciton\n",
    "\n",
    ">`TypeError`: concatenate_sequences() missing 1 required positional argument: 'a'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9d1ab605",
   "metadata": {},
   "source": [
    "**Merge dictionaries**\n",
    "\n",
    "can call `dict()` with `**`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "d0cefa49",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'a': 'TGACAC',\n",
       " 'b': 'CAGGGA',\n",
       " 'c': 'GGGGGGGGG',\n",
       " 'd': 'AAAATTTTT',\n",
       " 'KpnI': 'GGTACC',\n",
       " 'HindII': 'AAGCTT',\n",
       " 'ecoRI': 'GAATTC',\n",
       " 'another_seq': 'AGTGTAGTG'}"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "restriction_dict = {\"KpnI\": \"GGTACC\", \"HindII\": \"AAGCTT\", \"ecoRI\": \"GAATTC\"}\n",
    "\n",
    "dict(**my_dict, **restriction_dict, another_seq=\"AGTGTAGTG\") # basically creating a new dictionary with the two dictionaries and a new entry"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e3232e0",
   "metadata": {},
   "source": [
    "## Lesson 10: Packages and modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "63fcb963",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ebffb755",
   "metadata": {},
   "source": [
    "np.mean(),  np.median(), ..\n",
    "\n",
    "having an even amount in a list leads to interpolation (average of the two middle numbers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "88d25eee",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.85\n"
     ]
    }
   ],
   "source": [
    "# important thing about the median\n",
    "print(np.median((4.5, 1.2, -1.6, 9.0)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8668a212",
   "metadata": {},
   "source": [
    "access doc strings of functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "fb330615",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[1;31mSignature:\u001b[0m \u001b[0mnp\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mmedian\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0ma\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mNone\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mout\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mNone\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0moverwrite_input\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mFalse\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mkeepdims\u001b[0m\u001b[1;33m=\u001b[0m\u001b[1;32mFalse\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mDocstring:\u001b[0m\n",
      "Compute the median along the specified axis.\n",
      "\n",
      "Returns the median of the array elements.\n",
      "\n",
      "Parameters\n",
      "----------\n",
      "a : array_like\n",
      "    Input array or object that can be converted to an array.\n",
      "axis : {int, sequence of int, None}, optional\n",
      "    Axis or axes along which the medians are computed. The default\n",
      "    is to compute the median along a flattened version of the array.\n",
      "    A sequence of axes is supported since version 1.9.0.\n",
      "out : ndarray, optional\n",
      "    Alternative output array in which to place the result. It must\n",
      "    have the same shape and buffer length as the expected output,\n",
      "    but the type (of the output) will be cast if necessary.\n",
      "overwrite_input : bool, optional\n",
      "   If True, then allow use of memory of input array `a` for\n",
      "   calculations. The input array will be modified by the call to\n",
      "   `median`. This will save memory when you do not need to preserve\n",
      "   the contents of the input array. Treat the input as undefined,\n",
      "   but it will probably be fully or partially sorted. Default is\n",
      "   False. If `overwrite_input` is ``True`` and `a` is not already an\n",
      "   `ndarray`, an error will be raised.\n",
      "keepdims : bool, optional\n",
      "    If this is set to True, the axes which are reduced are left\n",
      "    in the result as dimensions with size one. With this option,\n",
      "    the result will broadcast correctly against the original `arr`.\n",
      "\n",
      "    .. versionadded:: 1.9.0\n",
      "\n",
      "Returns\n",
      "-------\n",
      "median : ndarray\n",
      "    A new array holding the result. If the input contains integers\n",
      "    or floats smaller than ``float64``, then the output data-type is\n",
      "    ``np.float64``.  Otherwise, the data-type of the output is the\n",
      "    same as that of the input. If `out` is specified, that array is\n",
      "    returned instead.\n",
      "\n",
      "See Also\n",
      "--------\n",
      "mean, percentile\n",
      "\n",
      "Notes\n",
      "-----\n",
      "Given a vector ``V`` of length ``N``, the median of ``V`` is the\n",
      "middle value of a sorted copy of ``V``, ``V_sorted`` - i\n",
      "e., ``V_sorted[(N-1)/2]``, when ``N`` is odd, and the average of the\n",
      "two middle values of ``V_sorted`` when ``N`` is even.\n",
      "\n",
      "Examples\n",
      "--------\n",
      ">>> a = np.array([[10, 7, 4], [3, 2, 1]])\n",
      ">>> a\n",
      "array([[10,  7,  4],\n",
      "       [ 3,  2,  1]])\n",
      ">>> np.median(a)\n",
      "3.5\n",
      ">>> np.median(a, axis=0)\n",
      "array([6.5, 4.5, 2.5])\n",
      ">>> np.median(a, axis=1)\n",
      "array([7.,  2.])\n",
      ">>> m = np.median(a, axis=0)\n",
      ">>> out = np.zeros_like(m)\n",
      ">>> np.median(a, axis=0, out=m)\n",
      "array([6.5,  4.5,  2.5])\n",
      ">>> m\n",
      "array([6.5,  4.5,  2.5])\n",
      ">>> b = a.copy()\n",
      ">>> np.median(b, axis=1, overwrite_input=True)\n",
      "array([7.,  2.])\n",
      ">>> assert not np.all(a==b)\n",
      ">>> b = a.copy()\n",
      ">>> np.median(b, axis=None, overwrite_input=True)\n",
      "3.5\n",
      ">>> assert not np.all(a==b)\n",
      "\u001b[1;31mFile:\u001b[0m      c:\\users\\fulci\\anaconda3\\lib\\site-packages\\numpy\\lib\\function_base.py\n",
      "\u001b[1;31mType:\u001b[0m      function"
     ]
    }
   ],
   "source": [
    "np.median? # see notes for info above with avg of two middle nrs"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9dc96e7",
   "metadata": {},
   "source": [
    "**import own modules**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "8349de6b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GACGAUCUAGGCGACCGACUGGCAUCG\n",
      "\n",
      "CGAUGCCAGUCGGUCGCCUAGAUCGUC\n"
     ]
    }
   ],
   "source": [
    "import na_utils as na\n",
    "\n",
    "# Sequence\n",
    "seq = 'GACGATCTAGGCGACCGACTGGCATCG'\n",
    "\n",
    "# Convert to RNA\n",
    "print(na.rna(seq))\n",
    "\n",
    "print()\n",
    "\n",
    "# compute the reverse RNA complement.\n",
    "print(na.reverse_rna_complement(seq))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66b04786",
   "metadata": {},
   "source": [
    "*Imports should be grouped in the following order:*\n",
    "\n",
    "1. standard library imports\n",
    "\n",
    "2. related third party imports\n",
    "\n",
    "3. local application/library specific imports"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "087a0ae9",
   "metadata": {},
   "source": [
    "automatically relading the imported packages and modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "8613c747",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "66536212",
   "metadata": {},
   "source": [
    "can also create own packages, a good habit is to create environments for projects, such that a specific version e.g. of a package is given"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ad327790",
   "metadata": {},
   "source": [
    "package manager [conda](https://docs.conda.io/en/latest/)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f454a0fd",
   "metadata": {},
   "source": [
    "## Lesson 11: File I/O"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d0a14608",
   "metadata": {},
   "source": [
    "open a file with context management (the with keyword)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "id": "68c5e0eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class '_io.TextIOWrapper'>\n"
     ]
    }
   ],
   "source": [
    "# read pdb file\n",
    "with open('data/1OLG.pdb', 'r') as f:   # f has the meaning of an open file (instance of the _io.TextIOWrapper class)\n",
    "    print(type(f))\n",
    "\n",
    "# 'r': open a text file for reading\n",
    "# 'w': create and open a text file for writing\n",
    "# 'a': append an existing text file\n",
    "# 'r+': open a text file for reading and writing\n",
    "# append 'b' to any of the above: same as above, except for binary files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5d621b19",
   "metadata": {},
   "source": [
    "f.read() reads in the whole file as a string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "id": "eefc0fc2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'HEADER    ANTI-ONCOGENE                           13-JUN-94   1OLG              \\nTITLE     HIGH-RESOLUTION SOLUTION STRUCTURE OF THE OLIGOMERIZATION             \\nTITLE    2 DOMAIN OF P53 BY MULTI-DIMENSIONAL NMR                               \\nCOMPND    MOL_ID: 1;                                                            \\nCOMPND   2 MOLECULE: TUMOR SUPPRESSOR P53 (OLIGOMERIZATION DOMAIN);             \\nCOMPND   3 CHAIN: A, B, C, D;                                                   \\nCOMPND   4 ENGINEERED: YES                                                      \\nSOURCE    MOL_ID: 1;                                                            \\nSOURCE   2 ORGANISM_SCIENTIFIC: HOMO SAPIENS;                                   \\nSOURCE   3 ORGANISM_COMMON: HUMAN;                                              \\nSOURCE   4 ORGANISM_TAXID: 9606                                                 \\nKEYWDS    ANTI-ONCOGENE                                                         \\nEXPDTA    SOLUTION NMR      '"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read file into string\n",
    "with open('data/1OLG.pdb', 'r') as f:\n",
    "    f_str = f.read()\n",
    "\n",
    "# Let's look at the first 1000 characters\n",
    "f_str[:1000]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "63e65688",
   "metadata": {},
   "source": [
    "f.readlines() reads in the file line by line into a list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "id": "86a78bc2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['HEADER    ANTI-ONCOGENE                           13-JUN-94   1OLG              \\n',\n",
       " 'TITLE     HIGH-RESOLUTION SOLUTION STRUCTURE OF THE OLIGOMERIZATION             \\n',\n",
       " 'TITLE    2 DOMAIN OF P53 BY MULTI-DIMENSIONAL NMR                               \\n',\n",
       " 'COMPND    MOL_ID: 1;                                                            \\n',\n",
       " 'COMPND   2 MOLECULE: TUMOR SUPPRESSOR P53 (OLIGOMERIZATION DOMAIN);             \\n',\n",
       " 'COMPND   3 CHAIN: A, B, C, D;                                                   \\n',\n",
       " 'COMPND   4 ENGINEERED: YES                                                      \\n',\n",
       " 'SOURCE    MOL_ID: 1;                                                            \\n',\n",
       " 'SOURCE   2 ORGANISM_SCIENTIFIC: HOMO SAPIENS;                                   \\n',\n",
       " 'SOURCE   3 ORGANISM_COMMON: HUMAN;                                              \\n']"
      ]
     },
     "execution_count": 128,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Read contents of the file in as a list\n",
    "with open('data/1OLG.pdb', 'r') as f:\n",
    "    f_list = f.readlines()\n",
    "\n",
    "# Look at the list (first ten entries)\n",
    "f_list[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e1cce32",
   "metadata": {},
   "source": [
    "rstrip() method strips all whitespace, including newlines, from the end of a string"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "id": "bfd44a77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'HEADER    ANTI-ONCOGENE                           13-JUN-94   1OLG'"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f_list[0].rstrip()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed676610",
   "metadata": {},
   "source": [
    "**Reading line-by-line**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "10abd7e0",
   "metadata": {},
   "source": [
    "Following there is examples of previewing the text files and eventually saving them with just a subset of the data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "005ea907",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HEADER    ANTI-ONCOGENE                           13-JUN-94   1OLG\n",
      "TITLE     HIGH-RESOLUTION SOLUTION STRUCTURE OF THE OLIGOMERIZATION\n",
      "TITLE    2 DOMAIN OF P53 BY MULTI-DIMENSIONAL NMR\n",
      "COMPND    MOL_ID: 1;\n",
      "COMPND   2 MOLECULE: TUMOR SUPPRESSOR P53 (OLIGOMERIZATION DOMAIN);\n",
      "COMPND   3 CHAIN: A, B, C, D;\n",
      "COMPND   4 ENGINEERED: YES\n",
      "SOURCE    MOL_ID: 1;\n",
      "SOURCE   2 ORGANISM_SCIENTIFIC: HOMO SAPIENS;\n",
      "SOURCE   3 ORGANISM_COMMON: HUMAN;\n",
      "SOURCE   4 ORGANISM_TAXID: 9606\n"
     ]
    }
   ],
   "source": [
    "# either reading it with a for loop\n",
    "with open('data/1OLG.pdb', 'r') as f:\n",
    "    for i, line in enumerate(f):\n",
    "        print(line.rstrip())\n",
    "        if i >= 10:\n",
    "            break"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "d6cc225e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HEADER    ANTI-ONCOGENE                           13-JUN-94   1OLG\n",
      "TITLE     HIGH-RESOLUTION SOLUTION STRUCTURE OF THE OLIGOMERIZATION\n",
      "TITLE    2 DOMAIN OF P53 BY MULTI-DIMENSIONAL NMR\n",
      "COMPND    MOL_ID: 1;\n",
      "COMPND   2 MOLECULE: TUMOR SUPPRESSOR P53 (OLIGOMERIZATION DOMAIN);\n",
      "COMPND   3 CHAIN: A, B, C, D;\n",
      "COMPND   4 ENGINEERED: YES\n",
      "SOURCE    MOL_ID: 1;\n",
      "SOURCE   2 ORGANISM_SCIENTIFIC: HOMO SAPIENS;\n",
      "SOURCE   3 ORGANISM_COMMON: HUMAN;\n"
     ]
    }
   ],
   "source": [
    "# or with a while loop and .readlines()\n",
    "with open('data/1OLG.pdb', 'r') as f:\n",
    "    i = 0\n",
    "    while i < 10:\n",
    "        print(f.readline().rstrip())        # each subsequent call read sin the next line of the file\n",
    "        i += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "b40cbe9b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(['HEADER    ANTI-ONCOGENE                           13-JUN-94   1OLG',\n",
       "  'TITLE     HIGH-RESOLUTION SOLUTION STRUCTURE OF THE OLIGOMERIZATION',\n",
       "  'TITLE    2 DOMAIN OF P53 BY MULTI-DIMENSIONAL NMR',\n",
       "  'COMPND    MOL_ID: 1;',\n",
       "  'COMPND   2 MOLECULE: TUMOR SUPPRESSOR P53 (OLIGOMERIZATION DOMAIN);',\n",
       "  'COMPND   3 CHAIN: A, B, C, D;',\n",
       "  'COMPND   4 ENGINEERED: YES',\n",
       "  'SOURCE    MOL_ID: 1;',\n",
       "  'SOURCE   2 ORGANISM_SCIENTIFIC: HOMO SAPIENS;',\n",
       "  'SOURCE   3 ORGANISM_COMMON: HUMAN;'],\n",
       " None,\n",
       " 'TITLE    2 DOMAIN OF P53 BY MULTI-DIMENSIONAL NMR')"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# try to create the file\n",
    "a=[]\n",
    "with open('data/1OLG.pdb', 'r') as f:\n",
    "    i = 0\n",
    "    while i < 10:\n",
    "        a = a + [f.readline().rstrip()]     # each subsequent call read sin the next line of the file\n",
    "        i += 1\n",
    "a, print(), a[2]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02f79ab4",
   "metadata": {},
   "source": [
    "**Writing files**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "f2c2f31d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# check available file\n",
    "print(os.path.isfile('data/1OLG.pdb'))\n",
    "print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "44690909",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write file and check at the same time if a filename is already used\n",
    "fname = 'yogi.txt'\n",
    "\n",
    "if os.path.isfile(fname):\n",
    "    raise RuntimeError(f'File {fname} already exists.')\n",
    "\n",
    "with open(fname, 'w') as f:\n",
    "    f.write('When you come to a fork in the road, take it.')\n",
    "    f.write('You can observe a lot by just watching.')\n",
    "    f.write('I never said most of the things I said.')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "497d33b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# preview the file (opens in a text editor)\n",
    "!>cat yogi.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b71d7104",
   "metadata": {},
   "source": [
    "Saving the file with new lines needs specification"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "f09ecd41",
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(fname, 'w') as f:\n",
    "    f.write('When you come to a fork in the road, take it.\\n')\n",
    "    f.write('You can observe a lot by just watching.\\n')\n",
    "    f.write('I never said most of the things I said.\\n')\n",
    "\n",
    "!>cat yogi.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "3e352077",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "When you come to a fork in the road, take it.\n",
      "You can observe a lot by just watching.\n",
      "I never said most of the things I said.\n"
     ]
    }
   ],
   "source": [
    "# can instead visualize the file with !type\n",
    "!type yogi.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b3350fd6",
   "metadata": {},
   "source": [
    "keep in mind that we can only write strings!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "75f98df6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The golden ratio is φ = 1.61803399\n"
     ]
    }
   ],
   "source": [
    "with open('gimme_phi.txt', 'w') as f:\n",
    "    f.write('The golden ratio is φ = ')\n",
    "    f.write('{phi:.8f}'.format(phi=1.61803398875)) # f.write(1.61803398875) wouldnt work\n",
    "\n",
    "!type gimme_phi.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bd06bdb",
   "metadata": {},
   "source": [
    "**Extract atomic coordinates for first chain in tetramer**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "6445c00c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HEADER    ANTI-ONCOGENE                           13-JUN-94   1OLG              \n",
      "TITLE     HIGH-RESOLUTION SOLUTION STRUCTURE OF THE OLIGOMERIZATION             \n",
      "TITLE    2 DOMAIN OF P53 BY MULTI-DIMENSIONAL NMR                               \n",
      "COMPND    MOL_ID: 1;                                                            \n",
      "COMPND   2 MOLECULE: TUMOR SUPPRESSOR P53 (OLIGOMERIZATION DOMAIN);             \n",
      "COMPND   3 CHAIN: A, B, C, D;                                                   \n",
      "COMPND   4 ENGINEERED: YES                                                      \n",
      "SOURCE    MOL_ID: 1;                                                            \n",
      "SOURCE   2 ORGANISM_SCIENTIFIC: HOMO SAPIENS;                                   \n",
      "SOURCE   3 ORGANISM_COMMON: HUMAN;                                              \n",
      "SOURCE   4 ORGANISM_TAXID: 9606                                                 \n",
      "KEYWDS    ANTI-ONCOGENE                                                         \n",
      "EXPDTA    SOLUTION NMR                                                          \n",
      "AUTHOR    G.M.CLORE,J.G.OMICHINSKI,A.M.GRONENBORN                               \n",
      "REVDAT   4   24-FEB-09 1OLG    1       VERSN                                    \n",
      "REVDAT   3   01-APR-03 1OLG    1       JRNL                                     \n",
      "REVDAT   2   08-MAR-95 1OLG    1       REMARK                                   \n",
      "REVDAT   1   26-JAN-95 1OLG    0                                                \n",
      "JRNL        AUTH   G.M.CLORE,J.G.OMICHINSKI,K.SAKAGUCHI,N.ZAMBRANO,             \n",
      "JRNL        AUTH 2 H.SAKAMOTO,E.APPELLA,A.M.GRONENBORN                          \n",
      "JRNL        TITL   HIGH-RESOLUTION STRUCTURE OF THE OLIGOMERIZATION             \n",
      "JRNL        TITL 2 DOMAIN OF P53 BY MULTIDIMENSIONAL NMR.                       \n",
      "JRNL        REF    SCIENCE                       V. 265   386 1994              \n",
      "JRNL        REFN                   ISSN 0036-8075                               \n",
      "JRNL        PMID   8023159                                                      \n",
      "REMARK   1                                                                      \n",
      "REMARK   2                                                                      \n",
      "REMARK   2 RESOLUTION. NOT APPLICABLE.                                          \n",
      "REMARK   3                                                                      \n",
      "REMARK   3 REFINEMENT.                                                          \n",
      "REMARK   3   PROGRAM     : NULL                                                 \n",
      "REMARK   3   AUTHORS     : NULL                                                 \n",
      "REMARK   3                                                                      \n",
      "REMARK   3  OTHER REFINEMENT REMARKS:                                           \n",
      "REMARK   3  THE 3D STRUCTURE OF THE OLIGOMERIZATION DOMAIN (RESIDUES            \n",
      "REMARK   3  319 - 360) OF P53 BY MULTI-DIMENSIONAL HETERONUCLEAR-EDITED         \n",
      "REMARK   3  AND -FILTERED NMR IS BASED ON 3824 EXPERIMENTAL RESTRAINTS          \n",
      "REMARK   3  COMPRISING THE FOLLOWING INTRA- AND INTER-SUBUNIT                   \n",
      "REMARK   3  RESTRAINTS:  (A) INTRASUBUNIT:  840 SEQUENTIAL (|I-J|=1),           \n",
      "REMARK   3  744 SHORT RANGE (1 < |I-J| >=5) AND 72 LONG RANGE                   \n",
      "REMARK   3  (|I-J| >5) INTERRESIDUES AND INTRARESIDUE APPROXIMATE               \n",
      "REMARK   3  INTERPROTON DISTANCE RESTRAINTS, 136 DISTANCE RESTRAINTS            \n",
      "REMARK   3  FOR 68 HYDROGEN BONDS, 268 TORSION ANGLE (144 PHI, 104 CHI1         \n",
      "REMARK   3  AND 20 CHI2) RESTRAINTS, AND 144 THREE-BOND HN-HA COUPLING          \n",
      "REMARK   3  CONSTANT RESTRAINTS.  (B) INTERSUBUNIT:  72 A-B/C-D,                \n",
      "REMARK   3  758 A-C/B-D, 10 A-D/B-C APPROXIMATE INTERPROTON DISTANCE            \n",
      "REMARK   3  RESTRAINTS, AND 24 DISTANCE RESTRAINTS FOR 12 HYDROGEN              \n",
      "REMARK   3  BONDS INVOLVING THE A-C/B-D SUBUNITS.  IN ADDITION, THERE           \n",
      "REMARK   3  ARE A TOTAL OF 38 CALPHA AND 38 CB CHEMICAL SHIFT                   \n",
      "REMARK   3  RESTRAINTS PER SUBUNIT THAT HAVE BEEN INCORPORATED                  \n",
      "REMARK   3  INTO THE REFINEMENT [J. KUSZWESKI, J. QIN, A.M. GRONENBORN          \n",
      "REMARK   3  AND G.M. CLORE, J. MAGN RESON. SER IN PRESS (1994)]                 \n",
      "REMARK   3                                                                      \n",
      "REMARK   3  THE STRUCTURES ARE CALCULATED USING THE HYBRID METRIC               \n",
      "REMARK   3  MATRIX DISTANCE GEOMETRY-DYNAMICAL SIMULATED ANNEALING              \n",
      "REMARK   3  METHOD DESCRIBED BY:  NILGES, M., CLORE, G.M.  AND                  \n",
      "REMARK   3  GRONENBORN, A.M. (1988) FEBS LETT. 29, 317-324.  ALL                \n",
      "REMARK   3  STRUCTURAL STATISTICS ARE GIVEN IN THE JRNL REFERENCE.              \n",
      "REMARK   3                                                                      \n",
      "REMARK   3  THE STRUCTURE PRESENTED IN THIS ENTRY IS THE RESTRAINED             \n",
      "REMARK   3  MINIMIZED AVERAGE STRUCTURE (SA)R.  THIS IS OBTAINED BY             \n",
      "REMARK   3  FIRST AVERAGING THE COORDINATES OF THE INDIVIDUAL 35                \n",
      "REMARK   3  DYNAMICAL SIMULATED ANNEALING SA STRUCTURES BEST FITTED TO          \n",
      "REMARK   3  RESIDUES 324 - 356 OF ALL FOUR SUBUNITS AND SUBJECTING THE          \n",
      "REMARK   3  RESULTING COORDINATES TO RESTRAINED MINIMIZATION.  THE              \n",
      "REMARK   3  QUANTITY PRESENTED IN COLUMNS 61 - 66 IN THIS SET OF                \n",
      "REMARK   3  COORDINATES (THE B-FACTOR FIELD IN X-RAY STRUCTURES) GIVES          \n",
      "REMARK   3  THE AVERAGE RMS DIFFERENCE BETWEEN THE INDIVIDUAL SA                \n",
      "REMARK   3  STRUCTURES AND THE MEAN STRUCTURE.  THE NUMBERS IN COLUMNS          \n",
      "REMARK   3  61 - 66 OF THE INDIVIDUAL STRUCTURES HAVE NO MEANING.  NOTE         \n",
      "REMARK   3  THAT RESIDUES 319 - 323 AT THE N-TERMINUS AND RESIDUES 357          \n",
      "REMARK   3  - 360 AT THE C-TERMINUS ARE DISORDERED.  THE SET OF 35              \n",
      "REMARK   3  STRUCTURES IS PRESENTED IN PROTEIN DATA BANK ENTRY 1OLH.            \n",
      "REMARK   4                                                                      \n",
      "REMARK   4 1OLG COMPLIES WITH FORMAT V. 3.15, 01-DEC-08                         \n",
      "REMARK 100                                                                      \n",
      "REMARK 100 THIS ENTRY HAS BEEN PROCESSED BY BNL.                                \n",
      "REMARK 210                                                                      \n",
      "REMARK 210 EXPERIMENTAL DETAILS                                                 \n",
      "REMARK 210  EXPERIMENT TYPE                : NMR                                \n",
      "REMARK 210  TEMPERATURE           (KELVIN) : NULL                               \n",
      "REMARK 210  PH                             : NULL                               \n",
      "REMARK 210  IONIC STRENGTH                 : NULL                               \n",
      "REMARK 210  PRESSURE                       : NULL                               \n",
      "REMARK 210  SAMPLE CONTENTS                : NULL                               \n",
      "REMARK 210                                                                      \n",
      "REMARK 210  NMR EXPERIMENTS CONDUCTED      : NULL                               \n",
      "REMARK 210  SPECTROMETER FIELD STRENGTH    : NULL                               \n",
      "REMARK 210  SPECTROMETER MODEL             : NULL                               \n",
      "REMARK 210  SPECTROMETER MANUFACTURER      : NULL                               \n",
      "REMARK 210                                                                      \n",
      "REMARK 210  STRUCTURE DETERMINATION.                                            \n",
      "REMARK 210   SOFTWARE USED                 : NULL                               \n",
      "REMARK 210   METHOD USED                   : NULL                               \n",
      "REMARK 210                                                                      \n",
      "REMARK 210 CONFORMERS, NUMBER CALCULATED   : NULL                               \n",
      "REMARK 210 CONFORMERS, NUMBER SUBMITTED    : 1                                  \n",
      "REMARK 210 CONFORMERS, SELECTION CRITERIA  : NULL                               \n",
      "REMARK 210                                                                      \n",
      "REMARK 210 BEST REPRESENTATIVE CONFORMER IN THIS ENSEMBLE : NULL                \n",
      "REMARK 210                                                                      \n",
      "REMARK 210 REMARK: NULL                                                         \n",
      "REMARK 215                                                                      \n",
      "REMARK 215 NMR STUDY                                                            \n",
      "REMARK 215 THE COORDINATES IN THIS ENTRY WERE GENERATED FROM SOLUTION           \n",
      "REMARK 215 NMR DATA.  PROTEIN DATA BANK CONVENTIONS REQUIRE THAT                \n",
      "REMARK 215 CRYST1 AND SCALE RECORDS BE INCLUDED, BUT THE VALUES ON              \n",
      "REMARK 215 THESE RECORDS ARE MEANINGLESS.                                       \n",
      "REMARK 500                                                                      \n",
      "REMARK 500 GEOMETRY AND STEREOCHEMISTRY                                         \n",
      "REMARK 500 SUBTOPIC: TORSION ANGLES                                             \n",
      "REMARK 500                                                                      \n",
      "REMARK 500 TORSION ANGLES OUTSIDE THE EXPECTED RAMACHANDRAN REGIONS:            \n",
      "REMARK 500 (M=MODEL NUMBER; RES=RESIDUE NAME; C=CHAIN IDENTIFIER;               \n",
      "REMARK 500 SSEQ=SEQUENCE NUMBER; I=INSERTION CODE).                             \n",
      "REMARK 500                                                                      \n",
      "REMARK 500 STANDARD TABLE:                                                      \n",
      "REMARK 500 FORMAT:(10X,I3,1X,A3,1X,A1,I4,A1,4X,F7.2,3X,F7.2)                    \n",
      "REMARK 500                                                                      \n",
      "REMARK 500 EXPECTED VALUES: GJ KLEYWEGT AND TA JONES (1996). PHI/PSI-           \n",
      "REMARK 500 CHOLOGY: RAMACHANDRAN REVISITED. STRUCTURE 4, 1395 - 1400            \n",
      "REMARK 500                                                                      \n",
      "REMARK 500  M RES CSSEQI        PSI       PHI                                   \n",
      "REMARK 500    LYS A 320      -74.36     68.94                                   \n",
      "REMARK 500    LYS A 321      135.74     67.72                                   \n",
      "REMARK 500    GLU A 326     -174.99    -58.70                                   \n",
      "REMARK 500    PHE A 328     -165.33   -112.87                                   \n",
      "REMARK 500    LYS B 320      -74.65     69.83                                   \n",
      "REMARK 500    LYS B 321      135.74     67.89                                   \n",
      "REMARK 500    GLU B 326     -175.08    -59.01                                   \n",
      "REMARK 500    PHE B 328     -165.39   -113.04                                   \n",
      "REMARK 500    LYS C 320      116.35     57.32                                   \n",
      "REMARK 500    LYS C 321     -127.71   -161.07                                   \n",
      "REMARK 500    PRO C 322     -137.97    -40.58                                   \n",
      "REMARK 500    GLU C 326     -173.91    -57.97                                   \n",
      "REMARK 500    PHE C 328     -162.84   -113.14                                   \n",
      "REMARK 500    LYS D 320      -75.43     69.86                                   \n",
      "REMARK 500    LYS D 321      135.67     68.29                                   \n",
      "REMARK 500    GLU D 326     -174.99    -58.62                                   \n",
      "REMARK 500    PHE D 328     -165.46   -112.99                                   \n",
      "REMARK 500                                                                      \n",
      "REMARK 500 REMARK: NULL                                                         \n",
      "REMARK 900                                                                      \n",
      "REMARK 900 RELATED ENTRIES                                                      \n",
      "REMARK 900 RELATED ID: 1OLH   RELATED DB: PDB                                   \n",
      "DBREF  1OLG A  319   360  UNP    P04637   P53_HUMAN      319    360             \n",
      "DBREF  1OLG B  319   360  UNP    P04637   P53_HUMAN      319    360             \n",
      "DBREF  1OLG C  319   360  UNP    P04637   P53_HUMAN      319    360             \n",
      "DBREF  1OLG D  319   360  UNP    P04637   P53_HUMAN      319    360             \n",
      "SEQRES   1 A   42  LYS LYS LYS PRO LEU ASP GLY GLU TYR PHE THR LEU GLN          \n",
      "SEQRES   2 A   42  ILE ARG GLY ARG GLU ARG PHE GLU MET PHE ARG GLU LEU          \n",
      "SEQRES   3 A   42  ASN GLU ALA LEU GLU LEU LYS ASP ALA GLN ALA GLY LYS          \n",
      "SEQRES   4 A   42  GLU PRO GLY                                                  \n",
      "SEQRES   1 B   42  LYS LYS LYS PRO LEU ASP GLY GLU TYR PHE THR LEU GLN          \n",
      "SEQRES   2 B   42  ILE ARG GLY ARG GLU ARG PHE GLU MET PHE ARG GLU LEU          \n",
      "SEQRES   3 B   42  ASN GLU ALA LEU GLU LEU LYS ASP ALA GLN ALA GLY LYS          \n",
      "SEQRES   4 B   42  GLU PRO GLY                                                  \n",
      "SEQRES   1 C   42  LYS LYS LYS PRO LEU ASP GLY GLU TYR PHE THR LEU GLN          \n",
      "SEQRES   2 C   42  ILE ARG GLY ARG GLU ARG PHE GLU MET PHE ARG GLU LEU          \n",
      "SEQRES   3 C   42  ASN GLU ALA LEU GLU LEU LYS ASP ALA GLN ALA GLY LYS          \n",
      "SEQRES   4 C   42  GLU PRO GLY                                                  \n",
      "SEQRES   1 D   42  LYS LYS LYS PRO LEU ASP GLY GLU TYR PHE THR LEU GLN          \n",
      "SEQRES   2 D   42  ILE ARG GLY ARG GLU ARG PHE GLU MET PHE ARG GLU LEU          \n",
      "SEQRES   3 D   42  ASN GLU ALA LEU GLU LEU LYS ASP ALA GLN ALA GLY LYS          \n",
      "SEQRES   4 D   42  GLU PRO GLY                                                  \n",
      "HELIX    1   1 GLY A  334  GLY A  356  1                                  23    \n",
      "HELIX    2   2 GLY B  334  GLY B  356  1                                  23    \n",
      "HELIX    3   3 GLY C  334  GLY C  356  1                                  23    \n",
      "HELIX    4   4 GLY D  334  GLY D  356  1                                  23    \n",
      "SHEET    1   A 2 TYR A 327  ARG A 333  0                                        \n",
      "SHEET    2   A 2 TYR C 327  ARG C 333 -1  N  PHE C 328   O  ILE A 332           \n",
      "SHEET    1   B 2 TYR B 327  ARG B 333  0                                        \n",
      "SHEET    2   B 2 TYR D 327  ARG D 333 -1  N  PHE D 328   O  ILE B 332           \n",
      "CRYST1    1.000    1.000    1.000  90.00  90.00  90.00 P 1           1          \n",
      "ORIGX1      1.000000  0.000000  0.000000        0.00000                         \n",
      "ORIGX2      0.000000  1.000000  0.000000        0.00000                         \n",
      "ORIGX3      0.000000  0.000000  1.000000        0.00000                         \n",
      "SCALE1      1.000000  0.000000  0.000000        0.00000                         \n",
      "SCALE2      0.000000  1.000000  0.000000        0.00000                         \n",
      "SCALE3      0.000000  0.000000  1.000000        0.00000                         \n",
      "ATOM      1  N   LYS A 319      18.634  25.437  10.685  1.00  4.81           N  \n",
      "ATOM      2  CA  LYS A 319      17.984  25.295   9.354  1.00  4.32           C  \n",
      "ATOM      3  C   LYS A 319      18.160  23.876   8.818  1.00  3.74           C  \n",
      "ATOM      4  O   LYS A 319      19.259  23.441   8.537  1.00  3.67           O  \n",
      "ATOM      5  CB  LYS A 319      18.609  26.282   8.371  1.00  4.67           C  \n",
      "ATOM      6  CG  LYS A 319      18.003  26.056   6.986  1.00  5.15           C  \n",
      "ATOM      7  CD  LYS A 319      16.476  26.057   7.091  1.00  5.90           C  \n",
      "ATOM      8  CE  LYS A 319      16.014  27.341   7.784  1.00  6.51           C  \n",
      "ATOM      9  NZ  LYS A 319      16.388  28.518   6.952  1.00  7.33           N  \n",
      "ATOM     10  H1  LYS A 319      18.414  24.606  11.281  1.00  5.09           H  \n",
      "ATOM     11  H2  LYS A 319      19.658  25.516  10.555  1.00  5.18           H  \n",
      "ATOM     12  H3  LYS A 319      18.282  26.296  11.150  1.00  4.96           H  \n",
      "ATOM     13  HA  LYS A 319      16.934  25.509   9.452  1.00  4.66           H  \n",
      "ATOM     14  HB2 LYS A 319      18.399  27.288   8.697  1.00  4.96           H  \n",
      "ATOM     15  HB3 LYS A 319      19.679  26.132   8.326  1.00  4.81           H  \n",
      "ATOM     16  HG2 LYS A 319      18.320  26.846   6.324  1.00  5.22           H  \n",
      "ATOM     17  HG3 LYS A 319      18.334  25.104   6.598  1.00  5.32           H  \n",
      "ATOM     18  HD2 LYS A 319      16.046  26.005   6.102  1.00  6.19           H  \n",
      "ATOM     19  HD3 LYS A 319      16.154  25.204   7.670  1.00  6.08           H  \n",
      "ATOM     20  HE2 LYS A 319      14.941  27.316   7.911  1.00  6.70           H  \n",
      "ATOM     21  HE3 LYS A 319      16.488  27.419   8.752  1.00  6.49           H  \n",
      "ATOM     22  HZ1 LYS A 319      17.395  28.460   6.701  1.00  7.60           H  \n",
      "ATOM     23  HZ2 LYS A 319      15.816  28.526   6.085  1.00  7.57           H  \n",
      "ATOM     24  HZ3 LYS A 319      16.214  29.392   7.491  1.00  7.66           H  \n",
      "ATOM     25  N   LYS A 320      17.063  23.177   8.658  1.00  3.85           N  \n",
      "ATOM     26  CA  LYS A 320      17.085  21.778   8.123  1.00  3.82           C  \n",
      "ATOM     27  C   LYS A 320      17.727  20.814   9.132  1.00  3.38           C  \n",
      "ATOM     28  O   LYS A 320      17.048  20.031   9.762  1.00  3.69           O  \n",
      "ATOM     29  CB  LYS A 320      17.844  21.740   6.797  1.00  4.21           C  \n",
      "ATOM     30  CG  LYS A 320      16.917  22.223   5.676  1.00  5.00           C  \n",
      "ATOM     31  CD  LYS A 320      17.701  22.331   4.365  1.00  5.81           C  \n",
      "ATOM     32  CE  LYS A 320      18.474  23.652   4.334  1.00  6.54           C  \n",
      "ATOM     33  NZ  LYS A 320      17.555  24.758   3.943  1.00  7.21           N  \n",
      "ATOM     34  H   LYS A 320      16.203  23.590   8.879  1.00  4.30           H  \n",
      "ATOM     35  HA  LYS A 320      16.067  21.461   7.949  1.00  4.36           H  \n",
      "ATOM     36  HB2 LYS A 320      18.707  22.383   6.855  1.00  4.26           H  \n",
      "ATOM     37  HB3 LYS A 320      18.159  20.728   6.592  1.00  4.38           H  \n",
      "ATOM     38  HG2 LYS A 320      16.104  21.520   5.553  1.00  5.26           H  \n",
      "ATOM     39  HG3 LYS A 320      16.516  23.192   5.934  1.00  5.13           H  \n",
      "ATOM     40  HD2 LYS A 320      18.394  21.505   4.293  1.00  5.89           H  \n",
      "ATOM     41  HD3 LYS A 320      17.015  22.299   3.532  1.00  6.15           H  \n",
      "ATOM     42  HE2 LYS A 320      18.883  23.852   5.313  1.00  6.83           H  \n",
      "ATOM     43  HE3 LYS A 320      19.278  23.581   3.617  1.00  6.61           H  \n",
      "ATOM     44  HZ1 LYS A 320      16.599  24.558   4.300  1.00  7.43           H  \n",
      "ATOM     45  HZ2 LYS A 320      17.897  25.653   4.348  1.00  7.46           H  \n",
      "ATOM     46  HZ3 LYS A 320      17.529  24.838   2.906  1.00  7.51           H  \n",
      "ATOM     47  N   LYS A 321      19.020  20.847   9.285  1.00  3.05           N  \n",
      "ATOM     48  CA  LYS A 321      19.688  19.926  10.251  1.00  2.81           C  \n",
      "ATOM     49  C   LYS A 321      19.570  18.479   9.751  1.00  2.50           C  \n",
      "ATOM     50  O   LYS A 321      18.512  18.062   9.322  1.00  2.62           O  \n",
      "ATOM     51  CB  LYS A 321      19.029  20.015  11.637  1.00  3.34           C  \n",
      "ATOM     52  CG  LYS A 321      18.339  21.373  11.842  1.00  4.02           C  \n",
      "ATOM     53  CD  LYS A 321      16.841  21.152  12.059  1.00  4.77           C  \n",
      "ATOM     54  CE  LYS A 321      16.501  21.307  13.544  1.00  5.69           C  \n",
      "ATOM     55  NZ  LYS A 321      17.082  20.148  14.319  1.00  6.47           N  \n",
      "ATOM     56  H   LYS A 321      19.556  21.470   8.761  1.00  3.26           H  \n",
      "ATOM     57  HA  LYS A 321      20.727  20.200  10.333  1.00  3.16           H  \n",
      "ATOM     58  HB2 LYS A 321      18.300  19.225  11.731  1.00  3.51           H  \n",
      "ATOM     59  HB3 LYS A 321      19.789  19.886  12.396  1.00  3.66           H  \n",
      "ATOM     60  HG2 LYS A 321      18.756  21.853  12.711  1.00  4.36           H  \n",
      "ATOM     61  HG3 LYS A 321      18.488  21.998  10.978  1.00  4.11           H  \n",
      "ATOM     62  HD2 LYS A 321      16.288  21.876  11.488  1.00  4.81           H  \n",
      "ATOM     63  HD3 LYS A 321      16.572  20.152  11.738  1.00  5.02           H  \n",
      "ATOM     64  HE2 LYS A 321      16.904  22.244  13.913  1.00  5.93           H  \n",
      "ATOM     65  HE3 LYS A 321      15.432  21.313  13.656  1.00  5.90           H  \n",
      "ATOM     66  HZ1 LYS A 321      17.162  19.350  13.688  1.00  6.74           H  \n",
      "ATOM     67  HZ2 LYS A 321      18.033  20.394  14.685  1.00  6.71           H  \n",
      "ATOM     68  HZ3 LYS A 321      16.460  19.881  15.120  1.00  6.81           H  \n",
      "ATOM     69  N   PRO A 322      20.660  17.750   9.832  1.00  2.87           N  \n",
      "ATOM     70  CA  PRO A 322      20.701  16.337   9.404  1.00  3.30           C  \n",
      "ATOM     71  C   PRO A 322      20.002  15.426  10.428  1.00  2.78           C  \n",
      "ATOM     72  O   PRO A 322      20.165  14.222  10.401  1.00  2.72           O  \n",
      "ATOM     73  CB  PRO A 322      22.198  16.001   9.352  1.00  4.33           C  \n",
      "ATOM     74  CG  PRO A 322      22.963  17.169  10.023  1.00  4.53           C  \n",
      "ATOM     75  CD  PRO A 322      21.944  18.275  10.333  1.00  3.61           C  \n",
      "ATOM     76  HA  PRO A 322      20.264  16.220   8.427  1.00  3.71           H  \n",
      "ATOM     77  HB2 PRO A 322      22.386  15.079   9.886  1.00  4.51           H  \n",
      "ATOM     78  HB3 PRO A 322      22.516  15.903   8.326  1.00  5.02           H  \n",
      "ATOM     79  HG2 PRO A 322      23.425  16.826  10.939  1.00  4.97           H  \n",
      "ATOM     80  HG3 PRO A 322      23.718  17.549   9.351  1.00  5.15           H  \n",
      "ATOM     81  HD2 PRO A 322      21.893  18.452  11.398  1.00  3.76           H  \n",
      "ATOM     82  HD3 PRO A 322      22.204  19.182   9.810  1.00  3.74           H  \n",
      "ATOM     83  N   LEU A 323      19.244  15.979  11.341  1.00  2.70           N  \n",
      "ATOM     84  CA  LEU A 323      18.560  15.130  12.361  1.00  2.39           C  \n",
      "ATOM     85  C   LEU A 323      17.248  14.572  11.801  1.00  1.78           C  \n",
      "ATOM     86  O   LEU A 323      16.196  14.744  12.384  1.00  1.89           O  \n",
      "ATOM     87  CB  LEU A 323      18.262  15.973  13.605  1.00  2.93           C  \n",
      "ATOM     88  CG  LEU A 323      19.571  16.320  14.320  1.00  3.65           C  \n",
      "ATOM     89  CD1 LEU A 323      19.263  16.901  15.702  1.00  4.35           C  \n",
      "ATOM     90  CD2 LEU A 323      20.417  15.054  14.480  1.00  3.83           C  \n",
      "ATOM     91  H   LEU A 323      19.132  16.949  11.365  1.00  3.05           H  \n",
      "ATOM     92  HA  LEU A 323      19.207  14.310  12.632  1.00  2.54           H  \n",
      "ATOM     93  HB2 LEU A 323      17.761  16.884  13.310  1.00  3.10           H  \n",
      "ATOM     94  HB3 LEU A 323      17.626  15.414  14.275  1.00  2.89           H  \n",
      "ATOM     95  HG  LEU A 323      20.116  17.048  13.738  1.00  3.76           H  \n",
      "ATOM     96 HD11 LEU A 323      18.450  16.350  16.151  1.00  4.70           H  \n",
      "ATOM     97 HD12 LEU A 323      20.139  16.824  16.328  1.00  4.63           H  \n",
      "ATOM     98 HD13 LEU A 323      18.982  17.939  15.601  1.00  4.59           H  \n",
      "ATOM     99 HD21 LEU A 323      19.783  14.235  14.789  1.00  4.03           H  \n",
      "ATOM    100 HD22 LEU A 323      20.881  14.811  13.535  1.00  3.92           H  \n",
      "ATOM    101 HD23 LEU A 323      21.180  15.220  15.225  1.00  4.16           H  \n",
      "ATOM    102  N   ASP A 324      17.296  13.893  10.687  1.00  1.51           N  \n",
      "ATOM    103  CA  ASP A 324      16.045  13.320  10.116  1.00  1.33           C  \n",
      "ATOM    104  C   ASP A 324      15.714  12.012  10.841  1.00  1.07           C  \n",
      "ATOM    105  O   ASP A 324      16.418  11.596  11.741  1.00  1.04           O  \n",
      "ATOM    106  CB  ASP A 324      16.239  13.043   8.624  1.00  1.76           C  \n",
      "ATOM    107  CG  ASP A 324      16.759  14.305   7.931  1.00  2.08           C  \n",
      "ATOM    108  OD1 ASP A 324      16.250  15.373   8.229  1.00  2.42           O  \n",
      "ATOM    109  OD2 ASP A 324      17.656  14.180   7.114  1.00  2.48           O  \n",
      "ATOM    110  H   ASP A 324      18.151  13.753  10.233  1.00  1.76           H  \n",
      "ATOM    111  HA  ASP A 324      15.234  14.021  10.251  1.00  1.53           H  \n",
      "ATOM    112  HB2 ASP A 324      16.951  12.243   8.496  1.00  1.90           H  \n",
      "ATOM    113  HB3 ASP A 324      15.294  12.758   8.185  1.00  2.04           H  \n",
      "ATOM    114  N   GLY A 325      14.651  11.359  10.459  1.00  0.98           N  \n",
      "ATOM    115  CA  GLY A 325      14.281  10.078  11.129  1.00  0.84           C  \n",
      "ATOM    116  C   GLY A 325      15.307   8.999  10.773  1.00  0.70           C  \n",
      "ATOM    117  O   GLY A 325      16.298   9.264  10.122  1.00  0.68           O  \n",
      "ATOM    118  H   GLY A 325      14.095  11.709   9.731  1.00  1.08           H  \n",
      "ATOM    119  HA2 GLY A 325      14.268  10.224  12.200  1.00  0.89           H  \n",
      "ATOM    120  HA3 GLY A 325      13.304   9.765  10.796  1.00  0.91           H  \n",
      "ATOM    121  N   GLU A 326      15.076   7.783  11.190  1.00  0.65           N  \n",
      "ATOM    122  CA  GLU A 326      16.040   6.691  10.869  1.00  0.58           C  \n",
      "ATOM    123  C   GLU A 326      16.181   6.571   9.351  1.00  0.52           C  \n",
      "ATOM    124  O   GLU A 326      15.633   7.358   8.605  1.00  0.51           O  \n",
      "ATOM    125  CB  GLU A 326      15.526   5.368  11.443  1.00  0.64           C  \n",
      "ATOM    126  CG  GLU A 326      15.658   5.387  12.967  1.00  0.75           C  \n",
      "ATOM    127  CD  GLU A 326      16.782   4.441  13.396  1.00  1.06           C  \n",
      "ATOM    128  OE1 GLU A 326      17.932   4.836  13.295  1.00  1.70           O  \n",
      "ATOM    129  OE2 GLU A 326      16.473   3.339  13.817  1.00  1.64           O  \n",
      "ATOM    130  H   GLU A 326      14.270   7.588  11.712  1.00  0.71           H  \n",
      "ATOM    131  HA  GLU A 326      17.002   6.922  11.303  1.00  0.59           H  \n",
      "ATOM    132  HB2 GLU A 326      14.489   5.237  11.171  1.00  0.68           H  \n",
      "ATOM    133  HB3 GLU A 326      16.109   4.552  11.043  1.00  0.65           H  \n",
      "ATOM    134  HG2 GLU A 326      15.887   6.391  13.296  1.00  0.92           H  \n",
      "ATOM    135  HG3 GLU A 326      14.730   5.065  13.414  1.00  0.97           H  \n",
      "ATOM    136  N   TYR A 327      16.912   5.596   8.884  1.00  0.49           N  \n",
      "ATOM    137  CA  TYR A 327      17.083   5.439   7.413  1.00  0.45           C  \n",
      "ATOM    138  C   TYR A 327      17.296   3.964   7.067  1.00  0.43           C  \n",
      "ATOM    139  O   TYR A 327      18.352   3.408   7.296  1.00  0.48           O  \n",
      "ATOM    140  CB  TYR A 327      18.294   6.252   6.953  1.00  0.47           C  \n",
      "ATOM    141  CG  TYR A 327      17.900   7.702   6.817  1.00  0.49           C  \n",
      "ATOM    142  CD1 TYR A 327      17.353   8.169   5.612  1.00  0.49           C  \n",
      "ATOM    143  CD2 TYR A 327      18.079   8.582   7.894  1.00  0.54           C  \n",
      "ATOM    144  CE1 TYR A 327      16.986   9.516   5.485  1.00  0.52           C  \n",
      "ATOM    145  CE2 TYR A 327      17.711   9.930   7.767  1.00  0.58           C  \n",
      "ATOM    146  CZ  TYR A 327      17.164  10.397   6.562  1.00  0.56           C  \n",
      "ATOM    147  OH  TYR A 327      16.801  11.723   6.437  1.00  0.61           O  \n",
      "ATOM    148  H   TYR A 327      17.349   4.971   9.500  1.00  0.51           H  \n",
      "ATOM    149  HA  TYR A 327      16.198   5.800   6.909  1.00  0.44           H  \n",
      "ATOM    150  HB2 TYR A 327      19.087   6.160   7.681  1.00  0.51           H  \n",
      "ATOM    151  HB3 TYR A 327      18.636   5.881   5.998  1.00  0.46           H  \n",
      "ATOM    152  HD1 TYR A 327      17.215   7.490   4.783  1.00  0.47           H  \n",
      "ATOM    153  HD2 TYR A 327      18.499   8.223   8.822  1.00  0.57           H  \n",
      "ATOM    154  HE1 TYR A 327      16.563   9.875   4.558  1.00  0.53           H  \n",
      "ATOM    155  HE2 TYR A 327      17.849  10.607   8.596  1.00  0.62           H  \n",
      "ATOM    156  HH  TYR A 327      17.140  12.045   5.598  1.00  1.04           H  \n",
      "ATOM    157  N   PHE A 328      16.303   3.329   6.508  1.00  0.39           N  \n",
      "ATOM    158  CA  PHE A 328      16.446   1.893   6.139  1.00  0.39           C  \n",
      "ATOM    159  C   PHE A 328      16.419   1.760   4.614  1.00  0.36           C  \n",
      "ATOM    160  O   PHE A 328      16.563   2.730   3.896  1.00  0.37           O  \n",
      "ATOM    161  CB  PHE A 328      15.290   1.090   6.741  1.00  0.40           C  \n",
      "ATOM    162  CG  PHE A 328      15.266   1.286   8.238  1.00  0.44           C  \n",
      "ATOM    163  CD1 PHE A 328      14.528   2.342   8.795  1.00  0.47           C  \n",
      "ATOM    164  CD2 PHE A 328      15.979   0.413   9.072  1.00  0.51           C  \n",
      "ATOM    165  CE1 PHE A 328      14.504   2.524  10.186  1.00  0.54           C  \n",
      "ATOM    166  CE2 PHE A 328      15.956   0.594  10.463  1.00  0.57           C  \n",
      "ATOM    167  CZ  PHE A 328      15.218   1.650  11.020  1.00  0.58           C  \n",
      "ATOM    168  H   PHE A 328      15.462   3.799   6.328  1.00  0.39           H  \n",
      "ATOM    169  HA  PHE A 328      17.384   1.514   6.518  1.00  0.42           H  \n",
      "ATOM    170  HB2 PHE A 328      14.357   1.431   6.317  1.00  0.40           H  \n",
      "ATOM    171  HB3 PHE A 328      15.425   0.042   6.518  1.00  0.43           H  \n",
      "ATOM    172  HD1 PHE A 328      13.978   3.014   8.152  1.00  0.48           H  \n",
      "ATOM    173  HD2 PHE A 328      16.547  -0.400   8.643  1.00  0.54           H  \n",
      "ATOM    174  HE1 PHE A 328      13.936   3.336  10.614  1.00  0.59           H  \n",
      "ATOM    175  HE2 PHE A 328      16.505  -0.078  11.105  1.00  0.65           H  \n",
      "ATOM    176  HZ  PHE A 328      15.199   1.790  12.090  1.00  0.64           H  \n",
      "ATOM    177  N   THR A 329      16.238   0.570   4.113  1.00  0.35           N  \n",
      "ATOM    178  CA  THR A 329      16.204   0.382   2.635  1.00  0.34           C  \n",
      "ATOM    179  C   THR A 329      15.386  -0.866   2.299  1.00  0.33           C  \n",
      "ATOM    180  O   THR A 329      15.063  -1.657   3.162  1.00  0.37           O  \n",
      "ATOM    181  CB  THR A 329      17.631   0.217   2.108  1.00  0.37           C  \n",
      "ATOM    182  OG1 THR A 329      18.349  -0.666   2.958  1.00  0.39           O  \n",
      "ATOM    183  CG2 THR A 329      18.326   1.580   2.079  1.00  0.42           C  \n",
      "ATOM    184  H   THR A 329      16.125  -0.201   4.707  1.00  0.35           H  \n",
      "ATOM    185  HA  THR A 329      15.748   1.246   2.174  1.00  0.35           H  \n",
      "ATOM    186  HB  THR A 329      17.602  -0.188   1.108  1.00  0.39           H  \n",
      "ATOM    187  HG1 THR A 329      18.897  -1.229   2.406  1.00  0.97           H  \n",
      "ATOM    188 HG21 THR A 329      17.726   2.277   1.513  1.00  1.08           H  \n",
      "ATOM    189 HG22 THR A 329      18.445   1.944   3.088  1.00  1.14           H  \n",
      "ATOM    190 HG23 THR A 329      19.296   1.480   1.615  1.00  1.09           H  \n",
      "ATOM    191  N   LEU A 330      15.043  -1.048   1.053  1.00  0.29           N  \n",
      "ATOM    192  CA  LEU A 330      14.242  -2.245   0.673  1.00  0.29           C  \n",
      "ATOM    193  C   LEU A 330      14.555  -2.641  -0.772  1.00  0.27           C  \n",
      "ATOM    194  O   LEU A 330      14.531  -1.823  -1.670  1.00  0.25           O  \n",
      "ATOM    195  CB  LEU A 330      12.754  -1.915   0.800  1.00  0.29           C  \n",
      "ATOM    196  CG  LEU A 330      11.923  -3.128   0.385  1.00  0.30           C  \n",
      "ATOM    197  CD1 LEU A 330      12.020  -4.207   1.465  1.00  0.36           C  \n",
      "ATOM    198  CD2 LEU A 330      10.461  -2.707   0.214  1.00  0.32           C  \n",
      "ATOM    199  H   LEU A 330      15.310  -0.398   0.369  1.00  0.27           H  \n",
      "ATOM    200  HA  LEU A 330      14.486  -3.065   1.332  1.00  0.32           H  \n",
      "ATOM    201  HB2 LEU A 330      12.529  -1.657   1.825  1.00  0.35           H  \n",
      "ATOM    202  HB3 LEU A 330      12.515  -1.080   0.158  1.00  0.29           H  \n",
      "ATOM    203  HG  LEU A 330      12.298  -3.519  -0.550  1.00  0.31           H  \n",
      "ATOM    204 HD11 LEU A 330      11.794  -3.773   2.428  1.00  1.08           H  \n",
      "ATOM    205 HD12 LEU A 330      11.315  -4.996   1.251  1.00  1.03           H  \n",
      "ATOM    206 HD13 LEU A 330      13.021  -4.612   1.479  1.00  1.09           H  \n",
      "ATOM    207 HD21 LEU A 330      10.240  -1.893   0.889  1.00  1.08           H  \n",
      "ATOM    208 HD22 LEU A 330      10.295  -2.387  -0.804  1.00  1.05           H  \n",
      "ATOM    209 HD23 LEU A 330       9.817  -3.545   0.438  1.00  1.08           H  \n",
      "ATOM    210  N   GLN A 331      14.840  -3.894  -1.002  1.00  0.29           N  \n",
      "ATOM    211  CA  GLN A 331      15.146  -4.349  -2.388  1.00  0.29           C  \n",
      "ATOM    212  C   GLN A 331      13.844  -4.741  -3.090  1.00  0.28           C  \n",
      "ATOM    213  O   GLN A 331      13.198  -5.703  -2.726  1.00  0.31           O  \n",
      "ATOM    214  CB  GLN A 331      16.081  -5.559  -2.331  1.00  0.34           C  \n",
      "ATOM    215  CG  GLN A 331      16.527  -5.928  -3.747  1.00  0.40           C  \n",
      "ATOM    216  CD  GLN A 331      18.038  -5.723  -3.876  1.00  0.64           C  \n",
      "ATOM    217  OE1 GLN A 331      18.809  -6.347  -3.175  1.00  1.37           O  \n",
      "ATOM    218  NE2 GLN A 331      18.495  -4.868  -4.749  1.00  0.62           N  \n",
      "ATOM    219  H   GLN A 331      14.848  -4.538  -0.263  1.00  0.32           H  \n",
      "ATOM    220  HA  GLN A 331      15.624  -3.549  -2.933  1.00  0.28           H  \n",
      "ATOM    221  HB2 GLN A 331      16.947  -5.316  -1.732  1.00  0.41           H  \n",
      "ATOM    222  HB3 GLN A 331      15.561  -6.396  -1.890  1.00  0.37           H  \n",
      "ATOM    223  HG2 GLN A 331      16.285  -6.963  -3.942  1.00  0.48           H  \n",
      "ATOM    224  HG3 GLN A 331      16.018  -5.298  -4.461  1.00  0.61           H  \n",
      "ATOM    225 HE21 GLN A 331      17.873  -4.364  -5.315  1.00  1.01           H  \n",
      "ATOM    226 HE22 GLN A 331      19.461  -4.729  -4.840  1.00  0.76           H  \n",
      "ATOM    227  N   ILE A 332      13.449  -4.001  -4.090  1.00  0.29           N  \n",
      "ATOM    228  CA  ILE A 332      12.184  -4.335  -4.806  1.00  0.31           C  \n",
      "ATOM    229  C   ILE A 332      12.506  -5.000  -6.146  1.00  0.32           C  \n",
      "ATOM    230  O   ILE A 332      12.959  -4.362  -7.075  1.00  0.33           O  \n",
      "ATOM    231  CB  ILE A 332      11.387  -3.052  -5.053  1.00  0.35           C  \n",
      "ATOM    232  CG1 ILE A 332      11.193  -2.311  -3.728  1.00  0.34           C  \n",
      "ATOM    233  CG2 ILE A 332      10.021  -3.405  -5.643  1.00  0.47           C  \n",
      "ATOM    234  CD1 ILE A 332      10.723  -0.882  -4.007  1.00  0.39           C  \n",
      "ATOM    235  H   ILE A 332      13.981  -3.226  -4.368  1.00  0.30           H  \n",
      "ATOM    236  HA  ILE A 332      11.597  -5.010  -4.202  1.00  0.33           H  \n",
      "ATOM    237  HB  ILE A 332      11.926  -2.422  -5.745  1.00  0.38           H  \n",
      "ATOM    238 HG12 ILE A 332      10.452  -2.825  -3.133  1.00  0.41           H  \n",
      "ATOM    239 HG13 ILE A 332      12.129  -2.282  -3.191  1.00  0.35           H  \n",
      "ATOM    240 HG21 ILE A 332       9.532  -4.131  -5.011  1.00  1.15           H  \n",
      "ATOM    241 HG22 ILE A 332       9.414  -2.513  -5.703  1.00  1.15           H  \n",
      "ATOM    242 HG23 ILE A 332      10.152  -3.819  -6.632  1.00  1.04           H  \n",
      "ATOM    243 HD11 ILE A 332       9.789  -0.909  -4.548  1.00  1.09           H  \n",
      "ATOM    244 HD12 ILE A 332      10.582  -0.361  -3.072  1.00  1.12           H  \n",
      "ATOM    245 HD13 ILE A 332      11.467  -0.367  -4.597  1.00  1.05           H  \n",
      "ATOM    246  N   ARG A 333      12.269  -6.279  -6.256  1.00  0.33           N  \n",
      "ATOM    247  CA  ARG A 333      12.555  -6.984  -7.537  1.00  0.36           C  \n",
      "ATOM    248  C   ARG A 333      11.514  -6.586  -8.584  1.00  0.38           C  \n",
      "ATOM    249  O   ARG A 333      10.403  -6.215  -8.260  1.00  0.42           O  \n",
      "ATOM    250  CB  ARG A 333      12.494  -8.496  -7.312  1.00  0.40           C  \n",
      "ATOM    251  CG  ARG A 333      13.339  -8.867  -6.092  1.00  0.43           C  \n",
      "ATOM    252  CD  ARG A 333      13.079 -10.327  -5.720  1.00  0.91           C  \n",
      "ATOM    253  NE  ARG A 333      13.936 -10.706  -4.561  1.00  1.05           N  \n",
      "ATOM    254  CZ  ARG A 333      13.682 -11.796  -3.888  1.00  1.50           C  \n",
      "ATOM    255  NH1 ARG A 333      12.677 -12.560  -4.228  1.00  2.10           N  \n",
      "ATOM    256  NH2 ARG A 333      14.433 -12.124  -2.873  1.00  2.10           N  \n",
      "ATOM    257  H   ARG A 333      11.899  -6.776  -5.496  1.00  0.34           H  \n",
      "ATOM    258  HA  ARG A 333      13.541  -6.712  -7.886  1.00  0.37           H  \n",
      "ATOM    259  HB2 ARG A 333      11.468  -8.793  -7.144  1.00  0.45           H  \n",
      "ATOM    260  HB3 ARG A 333      12.878  -9.006  -8.182  1.00  0.48           H  \n",
      "ATOM    261  HG2 ARG A 333      14.385  -8.733  -6.325  1.00  0.78           H  \n",
      "ATOM    262  HG3 ARG A 333      13.070  -8.232  -5.261  1.00  0.88           H  \n",
      "ATOM    263  HD2 ARG A 333      12.040 -10.451  -5.454  1.00  1.66           H  \n",
      "ATOM    264  HD3 ARG A 333      13.313 -10.960  -6.563  1.00  1.56           H  \n",
      "ATOM    265  HE  ARG A 333      14.690 -10.136  -4.304  1.00  1.63           H  \n",
      "ATOM    266 HH11 ARG A 333      12.099 -12.313  -5.005  1.00  2.21           H  \n",
      "ATOM    267 HH12 ARG A 333      12.487 -13.393  -3.709  1.00  2.79           H  \n",
      "ATOM    268 HH21 ARG A 333      15.202 -11.541  -2.611  1.00  2.40           H  \n",
      "ATOM    269 HH22 ARG A 333      14.239 -12.959  -2.357  1.00  2.59           H  \n",
      "ATOM    270  N   GLY A 334      11.862  -6.664  -9.839  1.00  0.39           N  \n",
      "ATOM    271  CA  GLY A 334      10.891  -6.296 -10.908  1.00  0.42           C  \n",
      "ATOM    272  C   GLY A 334      10.776  -4.774 -11.002  1.00  0.40           C  \n",
      "ATOM    273  O   GLY A 334      10.508  -4.099 -10.027  1.00  0.39           O  \n",
      "ATOM    274  H   GLY A 334      12.762  -6.970 -10.079  1.00  0.40           H  \n",
      "ATOM    275  HA2 GLY A 334      11.232  -6.691 -11.854  1.00  0.45           H  \n",
      "ATOM    276  HA3 GLY A 334       9.923  -6.712 -10.673  1.00  0.44           H  \n",
      "ATOM    277  N   ARG A 335      10.972  -4.228 -12.171  1.00  0.43           N  \n",
      "ATOM    278  CA  ARG A 335      10.869  -2.751 -12.332  1.00  0.45           C  \n",
      "ATOM    279  C   ARG A 335       9.397  -2.337 -12.280  1.00  0.45           C  \n",
      "ATOM    280  O   ARG A 335       9.056  -1.281 -11.786  1.00  0.44           O  \n",
      "ATOM    281  CB  ARG A 335      11.468  -2.343 -13.680  1.00  0.50           C  \n",
      "ATOM    282  CG  ARG A 335      11.267  -0.841 -13.897  1.00  0.58           C  \n",
      "ATOM    283  CD  ARG A 335      11.199  -0.548 -15.396  1.00  0.91           C  \n",
      "ATOM    284  NE  ARG A 335      11.634   0.854 -15.650  1.00  1.36           N  \n",
      "ATOM    285  CZ  ARG A 335      11.944   1.235 -16.860  1.00  1.83           C  \n",
      "ATOM    286  NH1 ARG A 335      11.879   0.388 -17.853  1.00  2.18           N  \n",
      "ATOM    287  NH2 ARG A 335      12.322   2.464 -17.079  1.00  2.68           N  \n",
      "ATOM    288  H   ARG A 335      11.183  -4.791 -12.945  1.00  0.47           H  \n",
      "ATOM    289  HA  ARG A 335      11.410  -2.263 -11.534  1.00  0.44           H  \n",
      "ATOM    290  HB2 ARG A 335      12.524  -2.570 -13.688  1.00  0.51           H  \n",
      "ATOM    291  HB3 ARG A 335      10.976  -2.888 -14.472  1.00  0.56           H  \n",
      "ATOM    292  HG2 ARG A 335      10.346  -0.528 -13.427  1.00  0.81           H  \n",
      "ATOM    293  HG3 ARG A 335      12.095  -0.300 -13.464  1.00  0.83           H  \n",
      "ATOM    294  HD2 ARG A 335      11.851  -1.228 -15.925  1.00  1.56           H  \n",
      "ATOM    295  HD3 ARG A 335      10.185  -0.679 -15.743  1.00  1.51           H  \n",
      "ATOM    296  HE  ARG A 335      11.685   1.492 -14.908  1.00  2.00           H  \n",
      "ATOM    297 HH11 ARG A 335      11.591  -0.556 -17.690  1.00  2.17           H  \n",
      "ATOM    298 HH12 ARG A 335      12.118   0.683 -18.777  1.00  2.88           H  \n",
      "ATOM    299 HH21 ARG A 335      12.373   3.114 -16.321  1.00  3.10           H  \n",
      "ATOM    300 HH22 ARG A 335      12.560   2.757 -18.005  1.00  3.16           H  \n",
      "ATOM    301  N   GLU A 336       8.524  -3.160 -12.791  1.00  0.50           N  \n",
      "ATOM    302  CA  GLU A 336       7.074  -2.816 -12.776  1.00  0.54           C  \n",
      "ATOM    303  C   GLU A 336       6.613  -2.596 -11.333  1.00  0.48           C  \n",
      "ATOM    304  O   GLU A 336       6.146  -1.533 -10.975  1.00  0.44           O  \n",
      "ATOM    305  CB  GLU A 336       6.274  -3.964 -13.395  1.00  0.62           C  \n",
      "ATOM    306  CG  GLU A 336       5.611  -3.488 -14.689  1.00  1.19           C  \n",
      "ATOM    307  CD  GLU A 336       4.989  -4.684 -15.413  1.00  1.58           C  \n",
      "ATOM    308  OE1 GLU A 336       4.286  -5.443 -14.766  1.00  2.22           O  \n",
      "ATOM    309  OE2 GLU A 336       5.225  -4.820 -16.602  1.00  2.04           O  \n",
      "ATOM    310  H   GLU A 336       8.821  -4.006 -13.187  1.00  0.53           H  \n",
      "ATOM    311  HA  GLU A 336       6.911  -1.915 -13.348  1.00  0.57           H  \n",
      "ATOM    312  HB2 GLU A 336       6.938  -4.788 -13.612  1.00  1.01           H  \n",
      "ATOM    313  HB3 GLU A 336       5.513  -4.288 -12.701  1.00  1.01           H  \n",
      "ATOM    314  HG2 GLU A 336       4.841  -2.767 -14.455  1.00  1.72           H  \n",
      "ATOM    315  HG3 GLU A 336       6.352  -3.030 -15.326  1.00  1.72           H  \n",
      "ATOM    316  N   ARG A 337       6.738  -3.595 -10.504  1.00  0.48           N  \n",
      "ATOM    317  CA  ARG A 337       6.303  -3.446  -9.086  1.00  0.46           C  \n",
      "ATOM    318  C   ARG A 337       7.106  -2.330  -8.418  1.00  0.39           C  \n",
      "ATOM    319  O   ARG A 337       6.628  -1.658  -7.526  1.00  0.36           O  \n",
      "ATOM    320  CB  ARG A 337       6.537  -4.762  -8.342  1.00  0.51           C  \n",
      "ATOM    321  CG  ARG A 337       5.607  -5.833  -8.912  1.00  0.63           C  \n",
      "ATOM    322  CD  ARG A 337       4.238  -5.725  -8.239  1.00  1.16           C  \n",
      "ATOM    323  NE  ARG A 337       3.657  -7.092  -8.051  1.00  1.43           N  \n",
      "ATOM    324  CZ  ARG A 337       3.584  -7.941  -9.044  1.00  2.15           C  \n",
      "ATOM    325  NH1 ARG A 337       3.827  -7.556 -10.268  1.00  2.74           N  \n",
      "ATOM    326  NH2 ARG A 337       3.203  -9.168  -8.818  1.00  2.77           N  \n",
      "ATOM    327  H   ARG A 337       7.115  -4.444 -10.814  1.00  0.53           H  \n",
      "ATOM    328  HA  ARG A 337       5.253  -3.199  -9.057  1.00  0.46           H  \n",
      "ATOM    329  HB2 ARG A 337       7.565  -5.071  -8.469  1.00  0.55           H  \n",
      "ATOM    330  HB3 ARG A 337       6.327  -4.626  -7.292  1.00  0.53           H  \n",
      "ATOM    331  HG2 ARG A 337       5.504  -5.682  -9.976  1.00  1.23           H  \n",
      "ATOM    332  HG3 ARG A 337       6.024  -6.811  -8.725  1.00  1.10           H  \n",
      "ATOM    333  HD2 ARG A 337       4.353  -5.276  -7.269  1.00  1.72           H  \n",
      "ATOM    334  HD3 ARG A 337       3.587  -5.101  -8.845  1.00  1.86           H  \n",
      "ATOM    335  HE  ARG A 337       3.376  -7.375  -7.156  1.00  1.69           H  \n",
      "ATOM    336 HH11 ARG A 337       4.062  -6.607 -10.459  1.00  2.62           H  \n",
      "ATOM    337 HH12 ARG A 337       3.777  -8.217 -11.017  1.00  3.54           H  \n",
      "ATOM    338 HH21 ARG A 337       2.968  -9.456  -7.890  1.00  2.78           H  \n",
      "ATOM    339 HH22 ARG A 337       3.145  -9.821  -9.573  1.00  3.47           H  \n",
      "ATOM    340  N   PHE A 338       8.322  -2.126  -8.840  1.00  0.39           N  \n",
      "ATOM    341  CA  PHE A 338       9.150  -1.053  -8.226  1.00  0.35           C  \n",
      "ATOM    342  C   PHE A 338       8.411   0.283  -8.320  1.00  0.33           C  \n",
      "ATOM    343  O   PHE A 338       8.162   0.935  -7.328  1.00  0.29           O  \n",
      "ATOM    344  CB  PHE A 338      10.483  -0.949  -8.967  1.00  0.39           C  \n",
      "ATOM    345  CG  PHE A 338      11.242   0.256  -8.464  1.00  0.37           C  \n",
      "ATOM    346  CD1 PHE A 338      11.823   0.234  -7.188  1.00  0.38           C  \n",
      "ATOM    347  CD2 PHE A 338      11.363   1.396  -9.272  1.00  0.42           C  \n",
      "ATOM    348  CE1 PHE A 338      12.526   1.354  -6.718  1.00  0.40           C  \n",
      "ATOM    349  CE2 PHE A 338      12.067   2.516  -8.803  1.00  0.43           C  \n",
      "ATOM    350  CZ  PHE A 338      12.648   2.494  -7.526  1.00  0.41           C  \n",
      "ATOM    351  H   PHE A 338       8.690  -2.679  -9.561  1.00  0.42           H  \n",
      "ATOM    352  HA  PHE A 338       9.333  -1.290  -7.188  1.00  0.35           H  \n",
      "ATOM    353  HB2 PHE A 338      11.065  -1.842  -8.790  1.00  0.42           H  \n",
      "ATOM    354  HB3 PHE A 338      10.300  -0.842 -10.026  1.00  0.43           H  \n",
      "ATOM    355  HD1 PHE A 338      11.729  -0.644  -6.566  1.00  0.42           H  \n",
      "ATOM    356  HD2 PHE A 338      10.915   1.412 -10.254  1.00  0.48           H  \n",
      "ATOM    357  HE1 PHE A 338      12.973   1.337  -5.735  1.00  0.45           H  \n",
      "ATOM    358  HE2 PHE A 338      12.160   3.394  -9.424  1.00  0.50           H  \n",
      "ATOM    359  HZ  PHE A 338      13.189   3.356  -7.164  1.00  0.44           H  \n",
      "ATOM    360  N   GLU A 339       8.065   0.699  -9.507  1.00  0.37           N  \n",
      "ATOM    361  CA  GLU A 339       7.351   1.998  -9.667  1.00  0.38           C  \n",
      "ATOM    362  C   GLU A 339       6.124   2.044  -8.750  1.00  0.33           C  \n",
      "ATOM    363  O   GLU A 339       5.763   3.085  -8.236  1.00  0.30           O  \n",
      "ATOM    364  CB  GLU A 339       6.901   2.152 -11.122  1.00  0.45           C  \n",
      "ATOM    365  CG  GLU A 339       8.128   2.207 -12.035  1.00  0.58           C  \n",
      "ATOM    366  CD  GLU A 339       7.740   2.830 -13.377  1.00  0.98           C  \n",
      "ATOM    367  OE1 GLU A 339       7.542   4.034 -13.413  1.00  1.66           O  \n",
      "ATOM    368  OE2 GLU A 339       7.646   2.094 -14.345  1.00  1.53           O  \n",
      "ATOM    369  H   GLU A 339       8.282   0.159 -10.295  1.00  0.41           H  \n",
      "ATOM    370  HA  GLU A 339       8.018   2.807  -9.412  1.00  0.38           H  \n",
      "ATOM    371  HB2 GLU A 339       6.283   1.310 -11.399  1.00  0.46           H  \n",
      "ATOM    372  HB3 GLU A 339       6.335   3.065 -11.229  1.00  0.48           H  \n",
      "ATOM    373  HG2 GLU A 339       8.898   2.805 -11.568  1.00  0.74           H  \n",
      "ATOM    374  HG3 GLU A 339       8.500   1.206 -12.199  1.00  0.81           H  \n",
      "ATOM    375  N   MET A 340       5.471   0.932  -8.550  1.00  0.33           N  \n",
      "ATOM    376  CA  MET A 340       4.260   0.923  -7.679  1.00  0.30           C  \n",
      "ATOM    377  C   MET A 340       4.637   1.300  -6.243  1.00  0.26           C  \n",
      "ATOM    378  O   MET A 340       4.032   2.165  -5.641  1.00  0.25           O  \n",
      "ATOM    379  CB  MET A 340       3.635  -0.475  -7.690  1.00  0.34           C  \n",
      "ATOM    380  CG  MET A 340       2.254  -0.422  -7.033  1.00  0.34           C  \n",
      "ATOM    381  SD  MET A 340       1.587  -2.100  -6.895  1.00  0.43           S  \n",
      "ATOM    382  CE  MET A 340       2.169  -2.437  -5.215  1.00  0.44           C  \n",
      "ATOM    383  H   MET A 340       5.771   0.104  -8.982  1.00  0.36           H  \n",
      "ATOM    384  HA  MET A 340       3.543   1.636  -8.057  1.00  0.31           H  \n",
      "ATOM    385  HB2 MET A 340       3.536  -0.816  -8.711  1.00  0.41           H  \n",
      "ATOM    386  HB3 MET A 340       4.266  -1.158  -7.142  1.00  0.34           H  \n",
      "ATOM    387  HG2 MET A 340       2.341   0.013  -6.048  1.00  0.35           H  \n",
      "ATOM    388  HG3 MET A 340       1.591   0.180  -7.636  1.00  0.44           H  \n",
      "ATOM    389  HE1 MET A 340       3.233  -2.257  -5.160  1.00  1.12           H  \n",
      "ATOM    390  HE2 MET A 340       1.661  -1.789  -4.519  1.00  1.15           H  \n",
      "ATOM    391  HE3 MET A 340       1.959  -3.468  -4.963  1.00  1.08           H  \n",
      "ATOM    392  N   PHE A 341       5.623   0.655  -5.685  1.00  0.24           N  \n",
      "ATOM    393  CA  PHE A 341       6.026   0.973  -4.285  1.00  0.22           C  \n",
      "ATOM    394  C   PHE A 341       6.504   2.425  -4.197  1.00  0.22           C  \n",
      "ATOM    395  O   PHE A 341       6.117   3.162  -3.313  1.00  0.22           O  \n",
      "ATOM    396  CB  PHE A 341       7.158   0.037  -3.859  1.00  0.22           C  \n",
      "ATOM    397  CG  PHE A 341       6.583  -1.308  -3.482  1.00  0.23           C  \n",
      "ATOM    398  CD1 PHE A 341       5.838  -1.447  -2.302  1.00  0.26           C  \n",
      "ATOM    399  CD2 PHE A 341       6.794  -2.419  -4.313  1.00  0.26           C  \n",
      "ATOM    400  CE1 PHE A 341       5.304  -2.696  -1.952  1.00  0.29           C  \n",
      "ATOM    401  CE2 PHE A 341       6.260  -3.668  -3.963  1.00  0.29           C  \n",
      "ATOM    402  CZ  PHE A 341       5.515  -3.806  -2.782  1.00  0.29           C  \n",
      "ATOM    403  H   PHE A 341       6.095  -0.045  -6.184  1.00  0.26           H  \n",
      "ATOM    404  HA  PHE A 341       5.180   0.834  -3.628  1.00  0.22           H  \n",
      "ATOM    405  HB2 PHE A 341       7.852  -0.084  -4.678  1.00  0.24           H  \n",
      "ATOM    406  HB3 PHE A 341       7.674   0.458  -3.009  1.00  0.23           H  \n",
      "ATOM    407  HD1 PHE A 341       5.675  -0.591  -1.662  1.00  0.30           H  \n",
      "ATOM    408  HD2 PHE A 341       7.368  -2.312  -5.222  1.00  0.30           H  \n",
      "ATOM    409  HE1 PHE A 341       4.730  -2.802  -1.043  1.00  0.33           H  \n",
      "ATOM    410  HE2 PHE A 341       6.423  -4.523  -4.603  1.00  0.34           H  \n",
      "ATOM    411  HZ  PHE A 341       5.104  -4.768  -2.513  1.00  0.32           H  \n",
      "ATOM    412  N   ARG A 342       7.346   2.839  -5.102  1.00  0.25           N  \n",
      "ATOM    413  CA  ARG A 342       7.857   4.239  -5.071  1.00  0.27           C  \n",
      "ATOM    414  C   ARG A 342       6.689   5.219  -4.938  1.00  0.25           C  \n",
      "ATOM    415  O   ARG A 342       6.760   6.185  -4.205  1.00  0.26           O  \n",
      "ATOM    416  CB  ARG A 342       8.617   4.527  -6.367  1.00  0.33           C  \n",
      "ATOM    417  CG  ARG A 342       9.268   5.909  -6.283  1.00  0.43           C  \n",
      "ATOM    418  CD  ARG A 342       9.680   6.364  -7.684  1.00  0.88           C  \n",
      "ATOM    419  NE  ARG A 342      10.307   7.714  -7.603  1.00  1.25           N  \n",
      "ATOM    420  CZ  ARG A 342      10.999   8.177  -8.610  1.00  1.72           C  \n",
      "ATOM    421  NH1 ARG A 342      11.142   7.464  -9.694  1.00  2.19           N  \n",
      "ATOM    422  NH2 ARG A 342      11.549   9.359  -8.532  1.00  2.44           N  \n",
      "ATOM    423  H   ARG A 342       7.647   2.227  -5.802  1.00  0.28           H  \n",
      "ATOM    424  HA  ARG A 342       8.524   4.359  -4.231  1.00  0.29           H  \n",
      "ATOM    425  HB2 ARG A 342       9.382   3.776  -6.510  1.00  0.37           H  \n",
      "ATOM    426  HB3 ARG A 342       7.931   4.505  -7.200  1.00  0.37           H  \n",
      "ATOM    427  HG2 ARG A 342       8.563   6.613  -5.867  1.00  0.80           H  \n",
      "ATOM    428  HG3 ARG A 342      10.142   5.857  -5.652  1.00  0.77           H  \n",
      "ATOM    429  HD2 ARG A 342      10.389   5.662  -8.096  1.00  1.51           H  \n",
      "ATOM    430  HD3 ARG A 342       8.808   6.410  -8.319  1.00  1.42           H  \n",
      "ATOM    431  HE  ARG A 342      10.202   8.254  -6.792  1.00  1.84           H  \n",
      "ATOM    432 HH11 ARG A 342      10.723   6.559  -9.759  1.00  2.21           H  \n",
      "ATOM    433 HH12 ARG A 342      11.673   7.823 -10.462  1.00  2.91           H  \n",
      "ATOM    434 HH21 ARG A 342      11.440   9.907  -7.703  1.00  2.77           H  \n",
      "ATOM    435 HH22 ARG A 342      12.079   9.715  -9.301  1.00  2.94           H  \n",
      "ATOM    436  N   GLU A 343       5.617   4.985  -5.644  1.00  0.25           N  \n",
      "ATOM    437  CA  GLU A 343       4.453   5.912  -5.558  1.00  0.25           C  \n",
      "ATOM    438  C   GLU A 343       3.963   5.993  -4.111  1.00  0.21           C  \n",
      "ATOM    439  O   GLU A 343       3.722   7.063  -3.587  1.00  0.22           O  \n",
      "ATOM    440  CB  GLU A 343       3.323   5.397  -6.451  1.00  0.28           C  \n",
      "ATOM    441  CG  GLU A 343       2.102   6.307  -6.305  1.00  0.33           C  \n",
      "ATOM    442  CD  GLU A 343       1.514   6.601  -7.687  1.00  1.05           C  \n",
      "ATOM    443  OE1 GLU A 343       1.553   5.716  -8.526  1.00  1.74           O  \n",
      "ATOM    444  OE2 GLU A 343       1.034   7.705  -7.882  1.00  1.74           O  \n",
      "ATOM    445  H   GLU A 343       5.579   4.203  -6.234  1.00  0.27           H  \n",
      "ATOM    446  HA  GLU A 343       4.753   6.895  -5.890  1.00  0.28           H  \n",
      "ATOM    447  HB2 GLU A 343       3.650   5.394  -7.481  1.00  0.35           H  \n",
      "ATOM    448  HB3 GLU A 343       3.058   4.393  -6.154  1.00  0.32           H  \n",
      "ATOM    449  HG2 GLU A 343       1.359   5.817  -5.693  1.00  0.69           H  \n",
      "ATOM    450  HG3 GLU A 343       2.398   7.234  -5.838  1.00  0.64           H  \n",
      "ATOM    451  N   LEU A 344       3.812   4.873  -3.460  1.00  0.20           N  \n",
      "ATOM    452  CA  LEU A 344       3.335   4.890  -2.048  1.00  0.19           C  \n",
      "ATOM    453  C   LEU A 344       4.345   5.636  -1.175  1.00  0.20           C  \n",
      "ATOM    454  O   LEU A 344       3.984   6.333  -0.248  1.00  0.23           O  \n",
      "ATOM    455  CB  LEU A 344       3.191   3.454  -1.542  1.00  0.20           C  \n",
      "ATOM    456  CG  LEU A 344       2.070   2.753  -2.309  1.00  0.24           C  \n",
      "ATOM    457  CD1 LEU A 344       2.177   1.241  -2.105  1.00  0.29           C  \n",
      "ATOM    458  CD2 LEU A 344       0.716   3.242  -1.790  1.00  0.30           C  \n",
      "ATOM    459  H   LEU A 344       4.010   4.020  -3.900  1.00  0.21           H  \n",
      "ATOM    460  HA  LEU A 344       2.378   5.387  -1.997  1.00  0.20           H  \n",
      "ATOM    461  HB2 LEU A 344       4.120   2.923  -1.695  1.00  0.22           H  \n",
      "ATOM    462  HB3 LEU A 344       2.951   3.465  -0.489  1.00  0.23           H  \n",
      "ATOM    463  HG  LEU A 344       2.158   2.981  -3.362  1.00  0.27           H  \n",
      "ATOM    464 HD11 LEU A 344       2.657   1.039  -1.158  1.00  1.02           H  \n",
      "ATOM    465 HD12 LEU A 344       1.189   0.806  -2.107  1.00  1.03           H  \n",
      "ATOM    466 HD13 LEU A 344       2.762   0.810  -2.904  1.00  1.00           H  \n",
      "ATOM    467 HD21 LEU A 344       0.835   4.216  -1.338  1.00  1.07           H  \n",
      "ATOM    468 HD22 LEU A 344       0.018   3.310  -2.612  1.00  1.04           H  \n",
      "ATOM    469 HD23 LEU A 344       0.340   2.547  -1.054  1.00  1.04           H  \n",
      "ATOM    470  N   ASN A 345       5.609   5.493  -1.460  1.00  0.24           N  \n",
      "ATOM    471  CA  ASN A 345       6.640   6.190  -0.643  1.00  0.29           C  \n",
      "ATOM    472  C   ASN A 345       6.376   7.697  -0.657  1.00  0.25           C  \n",
      "ATOM    473  O   ASN A 345       6.272   8.327   0.376  1.00  0.25           O  \n",
      "ATOM    474  CB  ASN A 345       8.026   5.913  -1.227  1.00  0.38           C  \n",
      "ATOM    475  CG  ASN A 345       9.084   6.627  -0.385  1.00  0.49           C  \n",
      "ATOM    476  OD1 ASN A 345       8.821   7.662   0.195  1.00  1.22           O  \n",
      "ATOM    477  ND2 ASN A 345      10.280   6.116  -0.292  1.00  0.58           N  \n",
      "ATOM    478  H   ASN A 345       5.880   4.926  -2.210  1.00  0.28           H  \n",
      "ATOM    479  HA  ASN A 345       6.599   5.827   0.373  1.00  0.32           H  \n",
      "ATOM    480  HB2 ASN A 345       8.214   4.849  -1.217  1.00  0.41           H  \n",
      "ATOM    481  HB3 ASN A 345       8.069   6.277  -2.242  1.00  0.41           H  \n",
      "ATOM    482 HD21 ASN A 345      10.493   5.282  -0.759  1.00  1.14           H  \n",
      "ATOM    483 HD22 ASN A 345      10.965   6.566   0.246  1.00  0.58           H  \n",
      "ATOM    484  N   GLU A 346       6.274   8.280  -1.819  1.00  0.27           N  \n",
      "ATOM    485  CA  GLU A 346       6.025   9.747  -1.898  1.00  0.28           C  \n",
      "ATOM    486  C   GLU A 346       4.693  10.081  -1.223  1.00  0.23           C  \n",
      "ATOM    487  O   GLU A 346       4.513  11.155  -0.687  1.00  0.25           O  \n",
      "ATOM    488  CB  GLU A 346       5.972  10.176  -3.367  1.00  0.34           C  \n",
      "ATOM    489  CG  GLU A 346       5.777  11.691  -3.452  1.00  0.40           C  \n",
      "ATOM    490  CD  GLU A 346       6.454  12.223  -4.717  1.00  1.00           C  \n",
      "ATOM    491  OE1 GLU A 346       7.602  11.878  -4.941  1.00  1.70           O  \n",
      "ATOM    492  OE2 GLU A 346       5.812  12.968  -5.440  1.00  1.72           O  \n",
      "ATOM    493  H   GLU A 346       6.367   7.754  -2.641  1.00  0.30           H  \n",
      "ATOM    494  HA  GLU A 346       6.824  10.274  -1.399  1.00  0.31           H  \n",
      "ATOM    495  HB2 GLU A 346       6.897   9.903  -3.854  1.00  0.40           H  \n",
      "ATOM    496  HB3 GLU A 346       5.147   9.681  -3.856  1.00  0.37           H  \n",
      "ATOM    497  HG2 GLU A 346       4.721  11.916  -3.486  1.00  0.82           H  \n",
      "ATOM    498  HG3 GLU A 346       6.218  12.161  -2.586  1.00  0.80           H  \n",
      "ATOM    499  N   ALA A 347       3.757   9.173  -1.250  1.00  0.22           N  \n",
      "ATOM    500  CA  ALA A 347       2.435   9.441  -0.614  1.00  0.23           C  \n",
      "ATOM    501  C   ALA A 347       2.619   9.699   0.884  1.00  0.22           C  \n",
      "ATOM    502  O   ALA A 347       2.233  10.730   1.398  1.00  0.24           O  \n",
      "ATOM    503  CB  ALA A 347       1.520   8.231  -0.814  1.00  0.27           C  \n",
      "ATOM    504  H   ALA A 347       3.921   8.315  -1.693  1.00  0.23           H  \n",
      "ATOM    505  HA  ALA A 347       1.986  10.309  -1.073  1.00  0.26           H  \n",
      "ATOM    506  HB1 ALA A 347       2.097   7.402  -1.194  1.00  1.00           H  \n",
      "ATOM    507  HB2 ALA A 347       1.074   7.957   0.131  1.00  1.01           H  \n",
      "ATOM    508  HB3 ALA A 347       0.741   8.481  -1.520  1.00  1.07           H  \n",
      "ATOM    509  N   LEU A 348       3.199   8.767   1.589  1.00  0.22           N  \n",
      "ATOM    510  CA  LEU A 348       3.399   8.953   3.055  1.00  0.24           C  \n",
      "ATOM    511  C   LEU A 348       4.295  10.169   3.306  1.00  0.27           C  \n",
      "ATOM    512  O   LEU A 348       4.025  10.983   4.166  1.00  0.29           O  \n",
      "ATOM    513  CB  LEU A 348       4.067   7.705   3.639  1.00  0.25           C  \n",
      "ATOM    514  CG  LEU A 348       3.030   6.592   3.810  1.00  0.24           C  \n",
      "ATOM    515  CD1 LEU A 348       1.852   7.111   4.634  1.00  0.29           C  \n",
      "ATOM    516  CD2 LEU A 348       2.532   6.142   2.435  1.00  0.24           C  \n",
      "ATOM    517  H   LEU A 348       3.497   7.940   1.156  1.00  0.22           H  \n",
      "ATOM    518  HA  LEU A 348       2.442   9.106   3.532  1.00  0.25           H  \n",
      "ATOM    519  HB2 LEU A 348       4.847   7.369   2.971  1.00  0.25           H  \n",
      "ATOM    520  HB3 LEU A 348       4.496   7.944   4.601  1.00  0.28           H  \n",
      "ATOM    521  HG  LEU A 348       3.484   5.755   4.321  1.00  0.27           H  \n",
      "ATOM    522 HD11 LEU A 348       2.222   7.696   5.463  1.00  1.08           H  \n",
      "ATOM    523 HD12 LEU A 348       1.221   7.727   4.012  1.00  1.06           H  \n",
      "ATOM    524 HD13 LEU A 348       1.280   6.275   5.011  1.00  1.03           H  \n",
      "ATOM    525 HD21 LEU A 348       3.363   6.103   1.748  1.00  1.01           H  \n",
      "ATOM    526 HD22 LEU A 348       2.086   5.162   2.517  1.00  1.05           H  \n",
      "ATOM    527 HD23 LEU A 348       1.796   6.843   2.072  1.00  1.03           H  \n",
      "ATOM    528  N   GLU A 349       5.363  10.296   2.567  1.00  0.30           N  \n",
      "ATOM    529  CA  GLU A 349       6.280  11.455   2.771  1.00  0.35           C  \n",
      "ATOM    530  C   GLU A 349       5.509  12.766   2.597  1.00  0.32           C  \n",
      "ATOM    531  O   GLU A 349       5.793  13.752   3.248  1.00  0.34           O  \n",
      "ATOM    532  CB  GLU A 349       7.418  11.392   1.749  1.00  0.40           C  \n",
      "ATOM    533  CG  GLU A 349       8.235  10.116   1.971  1.00  0.48           C  \n",
      "ATOM    534  CD  GLU A 349       9.727  10.437   1.863  1.00  1.13           C  \n",
      "ATOM    535  OE1 GLU A 349      10.067  11.338   1.115  1.00  1.91           O  \n",
      "ATOM    536  OE2 GLU A 349      10.503   9.773   2.531  1.00  1.68           O  \n",
      "ATOM    537  H   GLU A 349       5.567   9.626   1.883  1.00  0.31           H  \n",
      "ATOM    538  HA  GLU A 349       6.693  11.412   3.768  1.00  0.39           H  \n",
      "ATOM    539  HB2 GLU A 349       7.005  11.387   0.750  1.00  0.38           H  \n",
      "ATOM    540  HB3 GLU A 349       8.059  12.252   1.870  1.00  0.46           H  \n",
      "ATOM    541  HG2 GLU A 349       8.021   9.719   2.952  1.00  0.90           H  \n",
      "ATOM    542  HG3 GLU A 349       7.971   9.385   1.221  1.00  0.78           H  \n",
      "ATOM    543  N   LEU A 350       4.534  12.789   1.730  1.00  0.29           N  \n",
      "ATOM    544  CA  LEU A 350       3.751  14.041   1.526  1.00  0.29           C  \n",
      "ATOM    545  C   LEU A 350       3.021  14.391   2.822  1.00  0.28           C  \n",
      "ATOM    546  O   LEU A 350       3.004  15.527   3.252  1.00  0.32           O  \n",
      "ATOM    547  CB  LEU A 350       2.730  13.833   0.405  1.00  0.29           C  \n",
      "ATOM    548  CG  LEU A 350       2.280  15.193  -0.135  1.00  0.34           C  \n",
      "ATOM    549  CD1 LEU A 350       1.720  15.022  -1.548  1.00  0.41           C  \n",
      "ATOM    550  CD2 LEU A 350       1.192  15.767   0.775  1.00  0.38           C  \n",
      "ATOM    551  H   LEU A 350       4.316  11.985   1.216  1.00  0.30           H  \n",
      "ATOM    552  HA  LEU A 350       4.421  14.846   1.261  1.00  0.32           H  \n",
      "ATOM    553  HB2 LEU A 350       3.182  13.261  -0.392  1.00  0.33           H  \n",
      "ATOM    554  HB3 LEU A 350       1.875  13.300   0.791  1.00  0.28           H  \n",
      "ATOM    555  HG  LEU A 350       3.123  15.867  -0.161  1.00  0.49           H  \n",
      "ATOM    556 HD11 LEU A 350       1.317  14.026  -1.658  1.00  1.08           H  \n",
      "ATOM    557 HD12 LEU A 350       0.937  15.748  -1.715  1.00  1.15           H  \n",
      "ATOM    558 HD13 LEU A 350       2.510  15.173  -2.269  1.00  1.06           H  \n",
      "ATOM    559 HD21 LEU A 350       0.387  15.054   0.865  1.00  1.15           H  \n",
      "ATOM    560 HD22 LEU A 350       1.606  15.968   1.750  1.00  1.06           H  \n",
      "ATOM    561 HD23 LEU A 350       0.815  16.685   0.349  1.00  1.07           H  \n",
      "ATOM    562  N   LYS A 351       2.423  13.419   3.451  1.00  0.27           N  \n",
      "ATOM    563  CA  LYS A 351       1.698  13.688   4.722  1.00  0.31           C  \n",
      "ATOM    564  C   LYS A 351       2.680  14.257   5.748  1.00  0.38           C  \n",
      "ATOM    565  O   LYS A 351       2.368  15.175   6.480  1.00  0.44           O  \n",
      "ATOM    566  CB  LYS A 351       1.104  12.381   5.251  1.00  0.33           C  \n",
      "ATOM    567  CG  LYS A 351       0.134  12.681   6.394  1.00  0.40           C  \n",
      "ATOM    568  CD  LYS A 351      -0.618  11.402   6.767  1.00  0.45           C  \n",
      "ATOM    569  CE  LYS A 351      -1.894  11.761   7.531  1.00  0.77           C  \n",
      "ATOM    570  NZ  LYS A 351      -2.257  10.639   8.442  1.00  1.56           N  \n",
      "ATOM    571  H   LYS A 351       2.455  12.509   3.089  1.00  0.26           H  \n",
      "ATOM    572  HA  LYS A 351       0.907  14.401   4.544  1.00  0.31           H  \n",
      "ATOM    573  HB2 LYS A 351       0.576  11.879   4.453  1.00  0.35           H  \n",
      "ATOM    574  HB3 LYS A 351       1.898  11.745   5.612  1.00  0.38           H  \n",
      "ATOM    575  HG2 LYS A 351       0.687  13.038   7.251  1.00  0.53           H  \n",
      "ATOM    576  HG3 LYS A 351      -0.573  13.434   6.080  1.00  0.52           H  \n",
      "ATOM    577  HD2 LYS A 351      -0.876  10.862   5.867  1.00  0.55           H  \n",
      "ATOM    578  HD3 LYS A 351       0.010  10.783   7.389  1.00  0.63           H  \n",
      "ATOM    579  HE2 LYS A 351      -1.727  12.656   8.111  1.00  1.30           H  \n",
      "ATOM    580  HE3 LYS A 351      -2.698  11.930   6.830  1.00  1.15           H  \n",
      "ATOM    581  HZ1 LYS A 351      -1.861   9.751   8.074  1.00  2.14           H  \n",
      "ATOM    582  HZ2 LYS A 351      -1.872  10.825   9.391  1.00  2.02           H  \n",
      "ATOM    583  HZ3 LYS A 351      -3.292  10.557   8.497  1.00  2.00           H  \n",
      "ATOM    584  N   ASP A 352       3.866  13.716   5.804  1.00  0.42           N  \n",
      "ATOM    585  CA  ASP A 352       4.874  14.221   6.778  1.00  0.51           C  \n",
      "ATOM    586  C   ASP A 352       5.093  15.720   6.560  1.00  0.51           C  \n",
      "ATOM    587  O   ASP A 352       5.209  16.483   7.498  1.00  0.59           O  \n",
      "ATOM    588  CB  ASP A 352       6.196  13.479   6.573  1.00  0.58           C  \n",
      "ATOM    589  CG  ASP A 352       6.074  12.056   7.119  1.00  0.67           C  \n",
      "ATOM    590  OD1 ASP A 352       5.104  11.396   6.785  1.00  1.14           O  \n",
      "ATOM    591  OD2 ASP A 352       6.952  11.650   7.862  1.00  1.43           O  \n",
      "ATOM    592  H   ASP A 352       4.095  12.977   5.202  1.00  0.42           H  \n",
      "ATOM    593  HA  ASP A 352       4.518  14.052   7.783  1.00  0.56           H  \n",
      "ATOM    594  HB2 ASP A 352       6.427  13.441   5.518  1.00  0.56           H  \n",
      "ATOM    595  HB3 ASP A 352       6.985  13.997   7.096  1.00  0.68           H  \n",
      "ATOM    596  N   ALA A 353       5.155  16.147   5.328  1.00  0.49           N  \n",
      "ATOM    597  CA  ALA A 353       5.372  17.596   5.053  1.00  0.56           C  \n",
      "ATOM    598  C   ALA A 353       4.154  18.396   5.520  1.00  0.56           C  \n",
      "ATOM    599  O   ALA A 353       4.260  19.554   5.872  1.00  0.66           O  \n",
      "ATOM    600  CB  ALA A 353       5.574  17.805   3.550  1.00  0.64           C  \n",
      "ATOM    601  H   ALA A 353       5.062  15.516   4.585  1.00  0.49           H  \n",
      "ATOM    602  HA  ALA A 353       6.250  17.935   5.583  1.00  0.63           H  \n",
      "ATOM    603  HB1 ALA A 353       6.214  17.027   3.161  1.00  1.30           H  \n",
      "ATOM    604  HB2 ALA A 353       4.617  17.767   3.050  1.00  1.11           H  \n",
      "ATOM    605  HB3 ALA A 353       6.032  18.767   3.379  1.00  1.23           H  \n",
      "ATOM    606  N   GLN A 354       2.999  17.789   5.528  1.00  0.53           N  \n",
      "ATOM    607  CA  GLN A 354       1.780  18.519   5.976  1.00  0.62           C  \n",
      "ATOM    608  C   GLN A 354       1.680  18.457   7.502  1.00  0.68           C  \n",
      "ATOM    609  O   GLN A 354       0.906  19.166   8.113  1.00  0.88           O  \n",
      "ATOM    610  CB  GLN A 354       0.541  17.869   5.358  1.00  0.64           C  \n",
      "ATOM    611  CG  GLN A 354      -0.144  18.864   4.419  1.00  0.97           C  \n",
      "ATOM    612  CD  GLN A 354      -1.506  18.311   3.996  1.00  0.84           C  \n",
      "ATOM    613  OE1 GLN A 354      -2.415  19.062   3.702  1.00  1.18           O  \n",
      "ATOM    614  NE2 GLN A 354      -1.687  17.019   3.953  1.00  0.68           N  \n",
      "ATOM    615  H   GLN A 354       2.934  16.854   5.242  1.00  0.51           H  \n",
      "ATOM    616  HA  GLN A 354       1.841  19.550   5.661  1.00  0.71           H  \n",
      "ATOM    617  HB2 GLN A 354       0.836  16.991   4.801  1.00  0.85           H  \n",
      "ATOM    618  HB3 GLN A 354      -0.145  17.585   6.141  1.00  0.89           H  \n",
      "ATOM    619  HG2 GLN A 354      -0.280  19.806   4.931  1.00  1.39           H  \n",
      "ATOM    620  HG3 GLN A 354       0.469  19.014   3.544  1.00  1.42           H  \n",
      "ATOM    621 HE21 GLN A 354      -0.954  16.414   4.190  1.00  0.73           H  \n",
      "ATOM    622 HE22 GLN A 354      -2.556  16.655   3.683  1.00  0.79           H  \n",
      "ATOM    623  N   ALA A 355       2.458  17.612   8.121  1.00  0.70           N  \n",
      "ATOM    624  CA  ALA A 355       2.410  17.502   9.606  1.00  0.83           C  \n",
      "ATOM    625  C   ALA A 355       3.074  18.731  10.233  1.00  0.91           C  \n",
      "ATOM    626  O   ALA A 355       2.564  19.316  11.167  1.00  1.23           O  \n",
      "ATOM    627  CB  ALA A 355       3.151  16.237  10.044  1.00  1.02           C  \n",
      "ATOM    628  H   ALA A 355       3.075  17.049   7.609  1.00  0.74           H  \n",
      "ATOM    629  HA  ALA A 355       1.381  17.446   9.929  1.00  0.95           H  \n",
      "ATOM    630  HB1 ALA A 355       3.572  15.749   9.177  1.00  1.52           H  \n",
      "ATOM    631  HB2 ALA A 355       3.943  16.501  10.728  1.00  1.58           H  \n",
      "ATOM    632  HB3 ALA A 355       2.459  15.567  10.533  1.00  1.31           H  \n",
      "ATOM    633  N   GLY A 356       4.212  19.124   9.726  1.00  1.03           N  \n",
      "ATOM    634  CA  GLY A 356       4.917  20.314  10.289  1.00  1.23           C  \n",
      "ATOM    635  C   GLY A 356       4.315  21.606   9.726  1.00  1.19           C  \n",
      "ATOM    636  O   GLY A 356       4.896  22.667   9.839  1.00  1.54           O  \n",
      "ATOM    637  H   GLY A 356       4.606  18.636   8.973  1.00  1.24           H  \n",
      "ATOM    638  HA2 GLY A 356       4.817  20.311  11.365  1.00  1.44           H  \n",
      "ATOM    639  HA3 GLY A 356       5.963  20.267  10.026  1.00  1.53           H  \n",
      "ATOM    640  N   LYS A 357       3.162  21.535   9.118  1.00  1.33           N  \n",
      "ATOM    641  CA  LYS A 357       2.546  22.768   8.552  1.00  1.54           C  \n",
      "ATOM    642  C   LYS A 357       1.765  23.498   9.646  1.00  1.98           C  \n",
      "ATOM    643  O   LYS A 357       0.766  23.013  10.140  1.00  2.51           O  \n",
      "ATOM    644  CB  LYS A 357       1.599  22.388   7.412  1.00  1.88           C  \n",
      "ATOM    645  CG  LYS A 357       1.623  23.481   6.342  1.00  2.25           C  \n",
      "ATOM    646  CD  LYS A 357       0.212  23.683   5.784  1.00  2.96           C  \n",
      "ATOM    647  CE  LYS A 357       0.244  24.754   4.692  1.00  3.50           C  \n",
      "ATOM    648  NZ  LYS A 357      -1.141  25.009   4.205  1.00  4.18           N  \n",
      "ATOM    649  H   LYS A 357       2.704  20.675   9.028  1.00  1.62           H  \n",
      "ATOM    650  HA  LYS A 357       3.323  23.416   8.173  1.00  1.72           H  \n",
      "ATOM    651  HB2 LYS A 357       1.918  21.451   6.978  1.00  2.23           H  \n",
      "ATOM    652  HB3 LYS A 357       0.596  22.284   7.796  1.00  2.14           H  \n",
      "ATOM    653  HG2 LYS A 357       1.973  24.405   6.780  1.00  2.39           H  \n",
      "ATOM    654  HG3 LYS A 357       2.285  23.188   5.542  1.00  2.54           H  \n",
      "ATOM    655  HD2 LYS A 357      -0.146  22.753   5.368  1.00  3.43           H  \n",
      "ATOM    656  HD3 LYS A 357      -0.447  24.001   6.578  1.00  3.17           H  \n",
      "ATOM    657  HE2 LYS A 357       0.658  25.666   5.094  1.00  3.67           H  \n",
      "ATOM    658  HE3 LYS A 357       0.857  24.412   3.871  1.00  3.76           H  \n",
      "ATOM    659  HZ1 LYS A 357      -1.824  24.577   4.859  1.00  4.46           H  \n",
      "ATOM    660  HZ2 LYS A 357      -1.310  26.034   4.159  1.00  4.55           H  \n",
      "ATOM    661  HZ3 LYS A 357      -1.257  24.594   3.258  1.00  4.44           H  \n",
      "ATOM    662  N   GLU A 358       2.216  24.661  10.029  1.00  2.43           N  \n",
      "ATOM    663  CA  GLU A 358       1.505  25.424  11.092  1.00  3.19           C  \n",
      "ATOM    664  C   GLU A 358       0.008  25.487  10.760  1.00  3.44           C  \n",
      "ATOM    665  O   GLU A 358      -0.359  25.786   9.641  1.00  3.47           O  \n",
      "ATOM    666  CB  GLU A 358       2.071  26.845  11.161  1.00  3.91           C  \n",
      "ATOM    667  CG  GLU A 358       2.237  27.262  12.624  1.00  4.53           C  \n",
      "ATOM    668  CD  GLU A 358       2.598  28.747  12.693  1.00  5.25           C  \n",
      "ATOM    669  OE1 GLU A 358       1.947  29.526  12.015  1.00  5.65           O  \n",
      "ATOM    670  OE2 GLU A 358       3.517  29.080  13.422  1.00  5.71           O  \n",
      "ATOM    671  H   GLU A 358       3.024  25.032   9.618  1.00  2.59           H  \n",
      "ATOM    672  HA  GLU A 358       1.651  24.933  12.041  1.00  3.49           H  \n",
      "ATOM    673  HB2 GLU A 358       3.031  26.873  10.667  1.00  4.22           H  \n",
      "ATOM    674  HB3 GLU A 358       1.393  27.527  10.670  1.00  4.17           H  \n",
      "ATOM    675  HG2 GLU A 358       1.312  27.092  13.154  1.00  4.65           H  \n",
      "ATOM    676  HG3 GLU A 358       3.026  26.681  13.077  1.00  4.78           H  \n",
      "ATOM    677  N   PRO A 359      -0.816  25.205  11.741  1.00  4.08           N  \n",
      "ATOM    678  CA  PRO A 359      -2.278  25.228  11.562  1.00  4.74           C  \n",
      "ATOM    679  C   PRO A 359      -2.736  26.637  11.176  1.00  4.92           C  \n",
      "ATOM    680  O   PRO A 359      -1.965  27.576  11.189  1.00  4.94           O  \n",
      "ATOM    681  CB  PRO A 359      -2.849  24.827  12.930  1.00  5.61           C  \n",
      "ATOM    682  CG  PRO A 359      -1.655  24.581  13.885  1.00  5.57           C  \n",
      "ATOM    683  CD  PRO A 359      -0.364  24.845  13.097  1.00  4.60           C  \n",
      "ATOM    684  HA  PRO A 359      -2.579  24.513  10.813  1.00  4.84           H  \n",
      "ATOM    685  HB2 PRO A 359      -3.472  25.623  13.314  1.00  6.02           H  \n",
      "ATOM    686  HB3 PRO A 359      -3.428  23.921  12.834  1.00  6.09           H  \n",
      "ATOM    687  HG2 PRO A 359      -1.716  25.254  14.729  1.00  6.03           H  \n",
      "ATOM    688  HG3 PRO A 359      -1.665  23.558  14.230  1.00  6.01           H  \n",
      "ATOM    689  HD2 PRO A 359       0.183  25.663  13.542  1.00  4.71           H  \n",
      "ATOM    690  HD3 PRO A 359       0.246  23.956  13.062  1.00  4.53           H  \n",
      "ATOM    691  N   GLY A 360      -3.984  26.791  10.832  1.00  5.45           N  \n",
      "ATOM    692  CA  GLY A 360      -4.489  28.138  10.445  1.00  5.95           C  \n",
      "ATOM    693  C   GLY A 360      -5.981  28.236  10.765  1.00  6.77           C  \n",
      "ATOM    694  O   GLY A 360      -6.401  27.621  11.732  1.00  7.24           O  \n",
      "ATOM    695  OXT GLY A 360      -6.679  28.924  10.039  1.00  7.15           O  \n",
      "ATOM    696  H   GLY A 360      -4.589  26.020  10.828  1.00  5.72           H  \n",
      "ATOM    697  HA2 GLY A 360      -3.950  28.896  10.995  1.00  5.99           H  \n",
      "ATOM    698  HA3 GLY A 360      -4.341  28.288   9.386  1.00  6.05           H  \n",
      "TER     699      GLY A 360                                                      \n",
      "ATOM    700  N   LYS B 319     -18.744  25.982 -10.009  1.00  4.80           N  \n",
      "ATOM    701  CA  LYS B 319     -18.097  25.837  -8.674  1.00  4.31           C  \n",
      "ATOM    702  C   LYS B 319     -18.281  24.408  -8.180  1.00  3.72           C  \n",
      "ATOM    703  O   LYS B 319     -19.378  23.976  -7.897  1.00  3.65           O  \n",
      "ATOM    704  CB  LYS B 319     -18.706  26.802  -7.649  1.00  4.65           C  \n",
      "ATOM    705  CG  LYS B 319     -18.095  26.504  -6.275  1.00  5.14           C  \n",
      "ATOM    706  CD  LYS B 319     -16.566  26.502  -6.387  1.00  5.90           C  \n",
      "ATOM    707  CE  LYS B 319     -16.094  27.811  -7.022  1.00  6.52           C  \n",
      "ATOM    708  NZ  LYS B 319     -16.456  28.951  -6.135  1.00  7.34           N  \n",
      "ATOM    709  H1  LYS B 319     -18.362  25.255 -10.645  1.00  5.07           H  \n",
      "ATOM    710  H2  LYS B 319     -19.771  25.857  -9.912  1.00  5.18           H  \n",
      "ATOM    711  H3  LYS B 319     -18.543  26.923 -10.402  1.00  4.95           H  \n",
      "ATOM    712  HA  LYS B 319     -17.040  26.038  -8.769  1.00  4.66           H  \n",
      "ATOM    713  HB2 LYS B 319     -18.482  27.820  -7.932  1.00  4.93           H  \n",
      "ATOM    714  HB3 LYS B 319     -19.775  26.664  -7.603  1.00  4.78           H  \n",
      "ATOM    715  HG2 LYS B 319     -18.408  27.255  -5.564  1.00  5.20           H  \n",
      "ATOM    716  HG3 LYS B 319     -18.427  25.532  -5.940  1.00  5.32           H  \n",
      "ATOM    717  HD2 LYS B 319     -16.132  26.401  -5.403  1.00  6.20           H  \n",
      "ATOM    718  HD3 LYS B 319     -16.253  25.673  -7.004  1.00  6.08           H  \n",
      "ATOM    719  HE2 LYS B 319     -15.022  27.783  -7.151  1.00  6.70           H  \n",
      "ATOM    720  HE3 LYS B 319     -16.569  27.938  -7.984  1.00  6.51           H  \n",
      "ATOM    721  HZ1 LYS B 319     -17.458  28.880  -5.870  1.00  7.61           H  \n",
      "ATOM    722  HZ2 LYS B 319     -15.870  28.924  -5.277  1.00  7.57           H  \n",
      "ATOM    723  HZ3 LYS B 319     -16.293  29.848  -6.639  1.00  7.67           H  \n",
      "ATOM    724  N   LYS B 320     -17.196  23.692  -8.072  1.00  3.81           N  \n",
      "ATOM    725  CA  LYS B 320     -17.231  22.277  -7.594  1.00  3.78           C  \n",
      "ATOM    726  C   LYS B 320     -17.889  21.364  -8.638  1.00  3.34           C  \n",
      "ATOM    727  O   LYS B 320     -17.220  20.607  -9.305  1.00  3.67           O  \n",
      "ATOM    728  CB  LYS B 320     -17.982  22.185  -6.265  1.00  4.17           C  \n",
      "ATOM    729  CG  LYS B 320     -17.039  22.603  -5.134  1.00  4.95           C  \n",
      "ATOM    730  CD  LYS B 320     -17.810  22.662  -3.814  1.00  5.76           C  \n",
      "ATOM    731  CE  LYS B 320     -18.567  23.988  -3.722  1.00  6.50           C  \n",
      "ATOM    732  NZ  LYS B 320     -17.633  25.067  -3.290  1.00  7.17           N  \n",
      "ATOM    733  H   LYS B 320     -16.334  24.097  -8.305  1.00  4.27           H  \n",
      "ATOM    734  HA  LYS B 320     -16.215  21.946  -7.441  1.00  4.32           H  \n",
      "ATOM    735  HB2 LYS B 320     -18.841  22.837  -6.282  1.00  4.21           H  \n",
      "ATOM    736  HB3 LYS B 320     -18.302  21.167  -6.106  1.00  4.35           H  \n",
      "ATOM    737  HG2 LYS B 320     -16.238  21.883  -5.052  1.00  5.22           H  \n",
      "ATOM    738  HG3 LYS B 320     -16.627  23.577  -5.352  1.00  5.08           H  \n",
      "ATOM    739  HD2 LYS B 320     -18.511  21.842  -3.771  1.00  5.85           H  \n",
      "ATOM    740  HD3 LYS B 320     -17.117  22.588  -2.990  1.00  6.10           H  \n",
      "ATOM    741  HE2 LYS B 320     -18.980  24.235  -4.689  1.00  6.79           H  \n",
      "ATOM    742  HE3 LYS B 320     -19.367  23.896  -3.002  1.00  6.59           H  \n",
      "ATOM    743  HZ1 LYS B 320     -16.683  24.874  -3.668  1.00  7.39           H  \n",
      "ATOM    744  HZ2 LYS B 320     -17.970  25.982  -3.649  1.00  7.42           H  \n",
      "ATOM    745  HZ3 LYS B 320     -17.595  25.098  -2.252  1.00  7.46           H  \n",
      "ATOM    746  N   LYS B 321     -19.187  21.414  -8.780  1.00  3.01           N  \n",
      "ATOM    747  CA  LYS B 321     -19.873  20.539  -9.781  1.00  2.79           C  \n",
      "ATOM    748  C   LYS B 321     -19.770  19.069  -9.345  1.00  2.47           C  \n",
      "ATOM    749  O   LYS B 321     -18.714  18.621  -8.943  1.00  2.58           O  \n",
      "ATOM    750  CB  LYS B 321     -19.227  20.678 -11.171  1.00  3.32           C  \n",
      "ATOM    751  CG  LYS B 321     -18.522  22.030 -11.326  1.00  3.99           C  \n",
      "ATOM    752  CD  LYS B 321     -17.026  21.803 -11.565  1.00  4.75           C  \n",
      "ATOM    753  CE  LYS B 321     -16.694  22.026 -13.049  1.00  5.66           C  \n",
      "ATOM    754  NZ  LYS B 321     -17.291  20.894 -13.854  1.00  6.44           N  \n",
      "ATOM    755  H   LYS B 321     -19.717  22.018  -8.225  1.00  3.21           H  \n",
      "ATOM    756  HA  LYS B 321     -20.911  20.826  -9.845  1.00  3.14           H  \n",
      "ATOM    757  HB2 LYS B 321     -18.506  19.885 -11.306  1.00  3.51           H  \n",
      "ATOM    758  HB3 LYS B 321     -19.993  20.593 -11.925  1.00  3.64           H  \n",
      "ATOM    759  HG2 LYS B 321     -18.941  22.559 -12.168  1.00  4.33           H  \n",
      "ATOM    760  HG3 LYS B 321     -18.657  22.626 -10.434  1.00  4.10           H  \n",
      "ATOM    761  HD2 LYS B 321     -16.458  22.494 -10.953  1.00  4.80           H  \n",
      "ATOM    762  HD3 LYS B 321     -16.769  20.792 -11.292  1.00  5.01           H  \n",
      "ATOM    763  HE2 LYS B 321     -17.105  22.977 -13.367  1.00  5.90           H  \n",
      "ATOM    764  HE3 LYS B 321     -15.619  22.042 -13.175  1.00  5.87           H  \n",
      "ATOM    765  HZ1 LYS B 321     -17.390  20.052 -13.242  1.00  6.70           H  \n",
      "ATOM    766  HZ2 LYS B 321     -18.235  21.167 -14.207  1.00  6.70           H  \n",
      "ATOM    767  HZ3 LYS B 321     -16.675  20.651 -14.660  1.00  6.79           H  \n",
      "ATOM    768  N   PRO B 322     -20.870  18.355  -9.450  1.00  2.86           N  \n",
      "ATOM    769  CA  PRO B 322     -20.926  16.925  -9.082  1.00  3.29           C  \n",
      "ATOM    770  C   PRO B 322     -20.244  16.049 -10.150  1.00  2.77           C  \n",
      "ATOM    771  O   PRO B 322     -20.422  14.848 -10.173  1.00  2.72           O  \n",
      "ATOM    772  CB  PRO B 322     -22.426  16.604  -9.034  1.00  4.33           C  \n",
      "ATOM    773  CG  PRO B 322     -23.182  17.809  -9.649  1.00  4.54           C  \n",
      "ATOM    774  CD  PRO B 322     -22.151  18.915  -9.919  1.00  3.60           C  \n",
      "ATOM    775  HA  PRO B 322     -20.483  16.761  -8.114  1.00  3.69           H  \n",
      "ATOM    776  HB2 PRO B 322     -22.629  15.708  -9.605  1.00  4.53           H  \n",
      "ATOM    777  HB3 PRO B 322     -22.739  16.467  -8.010  1.00  5.03           H  \n",
      "ATOM    778  HG2 PRO B 322     -23.654  17.511 -10.575  1.00  4.99           H  \n",
      "ATOM    779  HG3 PRO B 322     -23.927  18.168  -8.956  1.00  5.15           H  \n",
      "ATOM    780  HD2 PRO B 322     -22.105  19.136 -10.976  1.00  3.76           H  \n",
      "ATOM    781  HD3 PRO B 322     -22.397  19.802  -9.357  1.00  3.73           H  \n",
      "ATOM    782  N   LEU B 323     -19.485  16.633 -11.042  1.00  2.68           N  \n",
      "ATOM    783  CA  LEU B 323     -18.819  15.822 -12.104  1.00  2.36           C  \n",
      "ATOM    784  C   LEU B 323     -17.509  15.224 -11.578  1.00  1.75           C  \n",
      "ATOM    785  O   LEU B 323     -16.458  15.412 -12.158  1.00  1.86           O  \n",
      "ATOM    786  CB  LEU B 323     -18.519  16.715 -13.312  1.00  2.90           C  \n",
      "ATOM    787  CG  LEU B 323     -19.828  17.106 -14.002  1.00  3.63           C  \n",
      "ATOM    788  CD1 LEU B 323     -19.521  17.743 -15.360  1.00  4.32           C  \n",
      "ATOM    789  CD2 LEU B 323     -20.688  15.858 -14.211  1.00  3.80           C  \n",
      "ATOM    790  H   LEU B 323     -19.362  17.601 -11.025  1.00  3.03           H  \n",
      "ATOM    791  HA  LEU B 323     -19.476  15.021 -12.406  1.00  2.53           H  \n",
      "ATOM    792  HB2 LEU B 323     -18.006  17.607 -12.980  1.00  3.07           H  \n",
      "ATOM    793  HB3 LEU B 323     -17.892  16.180 -14.009  1.00  2.86           H  \n",
      "ATOM    794  HG  LEU B 323     -20.362  17.815 -13.386  1.00  3.74           H  \n",
      "ATOM    795 HD11 LEU B 323     -18.718  17.202 -15.838  1.00  4.66           H  \n",
      "ATOM    796 HD12 LEU B 323     -20.403  17.703 -15.983  1.00  4.60           H  \n",
      "ATOM    797 HD13 LEU B 323     -19.228  18.772 -15.217  1.00  4.56           H  \n",
      "ATOM    798 HD21 LEU B 323     -20.065  15.046 -14.556  1.00  4.01           H  \n",
      "ATOM    799 HD22 LEU B 323     -21.153  15.582 -13.275  1.00  3.90           H  \n",
      "ATOM    800 HD23 LEU B 323     -21.452  16.064 -14.945  1.00  4.12           H  \n",
      "ATOM    801  N   ASP B 324     -17.558  14.497 -10.496  1.00  1.48           N  \n",
      "ATOM    802  CA  ASP B 324     -16.310  13.885  -9.959  1.00  1.30           C  \n",
      "ATOM    803  C   ASP B 324     -16.000  12.604 -10.740  1.00  1.04           C  \n",
      "ATOM    804  O   ASP B 324     -16.718  12.232 -11.648  1.00  1.01           O  \n",
      "ATOM    805  CB  ASP B 324     -16.497  13.549  -8.477  1.00  1.74           C  \n",
      "ATOM    806  CG  ASP B 324     -16.992  14.788  -7.729  1.00  2.06           C  \n",
      "ATOM    807  OD1 ASP B 324     -16.470  15.860  -7.987  1.00  2.39           O  \n",
      "ATOM    808  OD2 ASP B 324     -17.883  14.644  -6.909  1.00  2.47           O  \n",
      "ATOM    809  H   ASP B 324     -18.412  14.344 -10.043  1.00  1.74           H  \n",
      "ATOM    810  HA  ASP B 324     -15.492  14.581 -10.070  1.00  1.50           H  \n",
      "ATOM    811  HB2 ASP B 324     -17.221  12.755  -8.377  1.00  1.89           H  \n",
      "ATOM    812  HB3 ASP B 324     -15.554  13.231  -8.059  1.00  2.02           H  \n",
      "ATOM    813  N   GLY B 325     -14.939  11.926 -10.398  1.00  0.95           N  \n",
      "ATOM    814  CA  GLY B 325     -14.589  10.671 -11.124  1.00  0.81           C  \n",
      "ATOM    815  C   GLY B 325     -15.626   9.590 -10.808  1.00  0.65           C  \n",
      "ATOM    816  O   GLY B 325     -16.613   9.840 -10.145  1.00  0.63           O  \n",
      "ATOM    817  H   GLY B 325     -14.372  12.241  -9.663  1.00  1.06           H  \n",
      "ATOM    818  HA2 GLY B 325     -14.579  10.863 -12.188  1.00  0.86           H  \n",
      "ATOM    819  HA3 GLY B 325     -13.614  10.333 -10.810  1.00  0.89           H  \n",
      "ATOM    820  N   GLU B 326     -15.411   8.389 -11.275  1.00  0.61           N  \n",
      "ATOM    821  CA  GLU B 326     -16.386   7.297 -10.995  1.00  0.53           C  \n",
      "ATOM    822  C   GLU B 326     -16.517   7.112  -9.482  1.00  0.47           C  \n",
      "ATOM    823  O   GLU B 326     -15.953   7.858  -8.707  1.00  0.46           O  \n",
      "ATOM    824  CB  GLU B 326     -15.894   5.994 -11.630  1.00  0.61           C  \n",
      "ATOM    825  CG  GLU B 326     -16.039   6.081 -13.151  1.00  0.71           C  \n",
      "ATOM    826  CD  GLU B 326     -17.176   5.167 -13.611  1.00  0.97           C  \n",
      "ATOM    827  OE1 GLU B 326     -18.321   5.568 -13.482  1.00  1.58           O  \n",
      "ATOM    828  OE2 GLU B 326     -16.883   4.081 -14.085  1.00  1.61           O  \n",
      "ATOM    829  H   GLU B 326     -14.608   8.206 -11.806  1.00  0.68           H  \n",
      "ATOM    830  HA  GLU B 326     -17.348   7.559 -11.410  1.00  0.54           H  \n",
      "ATOM    831  HB2 GLU B 326     -14.856   5.838 -11.374  1.00  0.65           H  \n",
      "ATOM    832  HB3 GLU B 326     -16.484   5.169 -11.262  1.00  0.61           H  \n",
      "ATOM    833  HG2 GLU B 326     -16.259   7.101 -13.434  1.00  0.92           H  \n",
      "ATOM    834  HG3 GLU B 326     -15.117   5.769 -13.619  1.00  0.95           H  \n",
      "ATOM    835  N   TYR B 327     -17.259   6.127  -9.053  1.00  0.43           N  \n",
      "ATOM    836  CA  TYR B 327     -17.420   5.908  -7.588  1.00  0.39           C  \n",
      "ATOM    837  C   TYR B 327     -17.648   4.422  -7.306  1.00  0.37           C  \n",
      "ATOM    838  O   TYR B 327     -18.711   3.887  -7.554  1.00  0.41           O  \n",
      "ATOM    839  CB  TYR B 327     -18.618   6.714  -7.083  1.00  0.41           C  \n",
      "ATOM    840  CG  TYR B 327     -18.205   8.152  -6.887  1.00  0.44           C  \n",
      "ATOM    841  CD1 TYR B 327     -17.646   8.561  -5.667  1.00  0.43           C  \n",
      "ATOM    842  CD2 TYR B 327     -18.379   9.080  -7.925  1.00  0.49           C  \n",
      "ATOM    843  CE1 TYR B 327     -17.262   9.897  -5.484  1.00  0.47           C  \n",
      "ATOM    844  CE2 TYR B 327     -17.995  10.416  -7.742  1.00  0.53           C  \n",
      "ATOM    845  CZ  TYR B 327     -17.436  10.825  -6.522  1.00  0.52           C  \n",
      "ATOM    846  OH  TYR B 327     -17.057  12.140  -6.342  1.00  0.57           O  \n",
      "ATOM    847  H   TYR B 327     -17.709   5.536  -9.691  1.00  0.45           H  \n",
      "ATOM    848  HA  TYR B 327     -16.527   6.236  -7.077  1.00  0.38           H  \n",
      "ATOM    849  HB2 TYR B 327     -19.417   6.663  -7.808  1.00  0.45           H  \n",
      "ATOM    850  HB3 TYR B 327     -18.956   6.306  -6.143  1.00  0.40           H  \n",
      "ATOM    851  HD1 TYR B 327     -17.512   7.846  -4.868  1.00  0.41           H  \n",
      "ATOM    852  HD2 TYR B 327     -18.809   8.765  -8.865  1.00  0.52           H  \n",
      "ATOM    853  HE1 TYR B 327     -16.831  10.211  -4.545  1.00  0.48           H  \n",
      "ATOM    854  HE2 TYR B 327     -18.129  11.130  -8.541  1.00  0.58           H  \n",
      "ATOM    855  HH  TYR B 327     -17.386  12.429  -5.488  1.00  0.97           H  \n",
      "ATOM    856  N   PHE B 328     -16.658   3.753  -6.781  1.00  0.34           N  \n",
      "ATOM    857  CA  PHE B 328     -16.815   2.304  -6.473  1.00  0.34           C  \n",
      "ATOM    858  C   PHE B 328     -16.776   2.106  -4.956  1.00  0.31           C  \n",
      "ATOM    859  O   PHE B 328     -16.900   3.045  -4.196  1.00  0.32           O  \n",
      "ATOM    860  CB  PHE B 328     -15.673   1.516  -7.119  1.00  0.35           C  \n",
      "ATOM    861  CG  PHE B 328     -15.657   1.775  -8.606  1.00  0.39           C  \n",
      "ATOM    862  CD1 PHE B 328     -14.911   2.845  -9.124  1.00  0.42           C  \n",
      "ATOM    863  CD2 PHE B 328     -16.388   0.946  -9.471  1.00  0.46           C  \n",
      "ATOM    864  CE1 PHE B 328     -14.895   3.085 -10.506  1.00  0.49           C  \n",
      "ATOM    865  CE2 PHE B 328     -16.372   1.187 -10.853  1.00  0.52           C  \n",
      "ATOM    866  CZ  PHE B 328     -15.626   2.256 -11.371  1.00  0.53           C  \n",
      "ATOM    867  H   PHE B 328     -15.812   4.206  -6.584  1.00  0.34           H  \n",
      "ATOM    868  HA  PHE B 328     -17.760   1.952  -6.859  1.00  0.37           H  \n",
      "ATOM    869  HB2 PHE B 328     -14.733   1.828  -6.689  1.00  0.36           H  \n",
      "ATOM    870  HB3 PHE B 328     -15.818   0.461  -6.940  1.00  0.39           H  \n",
      "ATOM    871  HD1 PHE B 328     -14.349   3.483  -8.458  1.00  0.43           H  \n",
      "ATOM    872  HD2 PHE B 328     -16.962   0.123  -9.073  1.00  0.50           H  \n",
      "ATOM    873  HE1 PHE B 328     -14.321   3.908 -10.904  1.00  0.54           H  \n",
      "ATOM    874  HE2 PHE B 328     -16.934   0.548 -11.519  1.00  0.60           H  \n",
      "ATOM    875  HZ  PHE B 328     -15.613   2.441 -12.435  1.00  0.59           H  \n",
      "ATOM    876  N   THR B 329     -16.606   0.893  -4.508  1.00  0.31           N  \n",
      "ATOM    877  CA  THR B 329     -16.561   0.642  -3.039  1.00  0.31           C  \n",
      "ATOM    878  C   THR B 329     -15.755  -0.628  -2.764  1.00  0.32           C  \n",
      "ATOM    879  O   THR B 329     -15.445  -1.384  -3.664  1.00  0.36           O  \n",
      "ATOM    880  CB  THR B 329     -17.985   0.471  -2.507  1.00  0.34           C  \n",
      "ATOM    881  OG1 THR B 329     -18.721  -0.367  -3.387  1.00  0.37           O  \n",
      "ATOM    882  CG2 THR B 329     -18.664   1.839  -2.414  1.00  0.38           C  \n",
      "ATOM    883  H   THR B 329     -16.508   0.146  -5.136  1.00  0.32           H  \n",
      "ATOM    884  HA  THR B 329     -16.091   1.480  -2.546  1.00  0.32           H  \n",
      "ATOM    885  HB  THR B 329     -17.952   0.023  -1.526  1.00  0.36           H  \n",
      "ATOM    886  HG1 THR B 329     -19.272  -0.945  -2.855  1.00  0.92           H  \n",
      "ATOM    887 HG21 THR B 329     -18.052   2.505  -1.824  1.00  1.08           H  \n",
      "ATOM    888 HG22 THR B 329     -18.787   2.248  -3.407  1.00  1.10           H  \n",
      "ATOM    889 HG23 THR B 329     -19.631   1.730  -1.948  1.00  1.09           H  \n",
      "ATOM    890  N   LEU B 330     -15.407  -0.869  -1.529  1.00  0.28           N  \n",
      "ATOM    891  CA  LEU B 330     -14.617  -2.090  -1.208  1.00  0.30           C  \n",
      "ATOM    892  C   LEU B 330     -14.924  -2.544   0.221  1.00  0.28           C  \n",
      "ATOM    893  O   LEU B 330     -14.886  -1.765   1.153  1.00  0.27           O  \n",
      "ATOM    894  CB  LEU B 330     -13.126  -1.771  -1.332  1.00  0.30           C  \n",
      "ATOM    895  CG  LEU B 330     -12.305  -3.010  -0.977  1.00  0.31           C  \n",
      "ATOM    896  CD1 LEU B 330     -12.424  -4.039  -2.102  1.00  0.36           C  \n",
      "ATOM    897  CD2 LEU B 330     -10.838  -2.612  -0.801  1.00  0.34           C  \n",
      "ATOM    898  H   LEU B 330     -15.663  -0.247  -0.816  1.00  0.26           H  \n",
      "ATOM    899  HA  LEU B 330     -14.875  -2.878  -1.899  1.00  0.33           H  \n",
      "ATOM    900  HB2 LEU B 330     -12.906  -1.471  -2.346  1.00  0.35           H  \n",
      "ATOM    901  HB3 LEU B 330     -12.872  -0.968  -0.656  1.00  0.30           H  \n",
      "ATOM    902  HG  LEU B 330     -12.677  -3.437  -0.057  1.00  0.32           H  \n",
      "ATOM    903 HD11 LEU B 330     -12.202  -3.567  -3.047  1.00  1.06           H  \n",
      "ATOM    904 HD12 LEU B 330     -11.727  -4.845  -1.930  1.00  1.04           H  \n",
      "ATOM    905 HD13 LEU B 330     -13.430  -4.432  -2.125  1.00  1.10           H  \n",
      "ATOM    906 HD21 LEU B 330     -10.615  -1.768  -1.438  1.00  1.06           H  \n",
      "ATOM    907 HD22 LEU B 330     -10.659  -2.342   0.229  1.00  1.05           H  \n",
      "ATOM    908 HD23 LEU B 330     -10.204  -3.444  -1.071  1.00  1.10           H  \n",
      "ATOM    909  N   GLN B 331     -15.220  -3.803   0.400  1.00  0.31           N  \n",
      "ATOM    910  CA  GLN B 331     -15.521  -4.313   1.767  1.00  0.32           C  \n",
      "ATOM    911  C   GLN B 331     -14.218  -4.750   2.442  1.00  0.32           C  \n",
      "ATOM    912  O   GLN B 331     -13.587  -5.704   2.032  1.00  0.34           O  \n",
      "ATOM    913  CB  GLN B 331     -16.472  -5.508   1.666  1.00  0.36           C  \n",
      "ATOM    914  CG  GLN B 331     -16.911  -5.933   3.069  1.00  0.42           C  \n",
      "ATOM    915  CD  GLN B 331     -18.418  -5.716   3.219  1.00  0.62           C  \n",
      "ATOM    916  OE1 GLN B 331     -19.202  -6.300   2.499  1.00  1.36           O  \n",
      "ATOM    917  NE2 GLN B 331     -18.858  -4.894   4.132  1.00  0.59           N  \n",
      "ATOM    918  H   GLN B 331     -15.239  -4.414  -0.365  1.00  0.33           H  \n",
      "ATOM    919  HA  GLN B 331     -15.985  -3.531   2.351  1.00  0.31           H  \n",
      "ATOM    920  HB2 GLN B 331     -17.339  -5.229   1.086  1.00  0.43           H  \n",
      "ATOM    921  HB3 GLN B 331     -15.965  -6.332   1.185  1.00  0.40           H  \n",
      "ATOM    922  HG2 GLN B 331     -16.680  -6.978   3.217  1.00  0.52           H  \n",
      "ATOM    923  HG3 GLN B 331     -16.388  -5.340   3.805  1.00  0.61           H  \n",
      "ATOM    924 HE21 GLN B 331     -18.226  -4.423   4.713  1.00  1.00           H  \n",
      "ATOM    925 HE22 GLN B 331     -19.822  -4.748   4.237  1.00  0.71           H  \n",
      "ATOM    926  N   ILE B 332     -13.806  -4.058   3.469  1.00  0.32           N  \n",
      "ATOM    927  CA  ILE B 332     -12.540  -4.437   4.159  1.00  0.34           C  \n",
      "ATOM    928  C   ILE B 332     -12.859  -5.156   5.472  1.00  0.35           C  \n",
      "ATOM    929  O   ILE B 332     -13.299  -4.552   6.431  1.00  0.37           O  \n",
      "ATOM    930  CB  ILE B 332     -11.726  -3.176   4.454  1.00  0.36           C  \n",
      "ATOM    931  CG1 ILE B 332     -11.533  -2.380   3.161  1.00  0.34           C  \n",
      "ATOM    932  CG2 ILE B 332     -10.359  -3.569   5.017  1.00  0.49           C  \n",
      "ATOM    933  CD1 ILE B 332     -11.045  -0.970   3.496  1.00  0.39           C  \n",
      "ATOM    934  H   ILE B 332     -14.326  -3.289   3.784  1.00  0.32           H  \n",
      "ATOM    935  HA  ILE B 332     -11.966  -5.093   3.522  1.00  0.36           H  \n",
      "ATOM    936  HB  ILE B 332     -12.252  -2.569   5.177  1.00  0.40           H  \n",
      "ATOM    937 HG12 ILE B 332     -10.803  -2.877   2.538  1.00  0.42           H  \n",
      "ATOM    938 HG13 ILE B 332     -12.473  -2.318   2.633  1.00  0.34           H  \n",
      "ATOM    939 HG21 ILE B 332      -9.884  -4.273   4.350  1.00  1.14           H  \n",
      "ATOM    940 HG22 ILE B 332      -9.742  -2.688   5.111  1.00  1.17           H  \n",
      "ATOM    941 HG23 ILE B 332     -10.488  -4.024   5.989  1.00  1.05           H  \n",
      "ATOM    942 HD11 ILE B 332     -10.107  -1.032   4.028  1.00  1.07           H  \n",
      "ATOM    943 HD12 ILE B 332     -10.904  -0.411   2.583  1.00  1.13           H  \n",
      "ATOM    944 HD13 ILE B 332     -11.777  -0.473   4.114  1.00  1.05           H  \n",
      "ATOM    945  N   ARG B 333     -12.636  -6.441   5.525  1.00  0.36           N  \n",
      "ATOM    946  CA  ARG B 333     -12.921  -7.197   6.777  1.00  0.40           C  \n",
      "ATOM    947  C   ARG B 333     -11.867  -6.855   7.832  1.00  0.41           C  \n",
      "ATOM    948  O   ARG B 333     -10.755  -6.483   7.515  1.00  0.43           O  \n",
      "ATOM    949  CB  ARG B 333     -12.878  -8.698   6.486  1.00  0.44           C  \n",
      "ATOM    950  CG  ARG B 333     -13.736  -9.006   5.258  1.00  0.46           C  \n",
      "ATOM    951  CD  ARG B 333     -13.499 -10.452   4.823  1.00  0.93           C  \n",
      "ATOM    952  NE  ARG B 333     -14.370 -10.771   3.657  1.00  1.07           N  \n",
      "ATOM    953  CZ  ARG B 333     -14.134 -11.834   2.935  1.00  1.50           C  \n",
      "ATOM    954  NH1 ARG B 333     -13.136 -12.623   3.233  1.00  2.10           N  \n",
      "ATOM    955  NH2 ARG B 333     -14.897 -12.109   1.913  1.00  2.10           N  \n",
      "ATOM    956  H   ARG B 333     -12.278  -6.908   4.742  1.00  0.37           H  \n",
      "ATOM    957  HA  ARG B 333     -13.900  -6.930   7.145  1.00  0.41           H  \n",
      "ATOM    958  HB2 ARG B 333     -11.857  -8.999   6.298  1.00  0.49           H  \n",
      "ATOM    959  HB3 ARG B 333     -13.262  -9.241   7.336  1.00  0.51           H  \n",
      "ATOM    960  HG2 ARG B 333     -14.778  -8.868   5.504  1.00  0.80           H  \n",
      "ATOM    961  HG3 ARG B 333     -13.464  -8.341   4.453  1.00  0.88           H  \n",
      "ATOM    962  HD2 ARG B 333     -12.464 -10.579   4.544  1.00  1.68           H  \n",
      "ATOM    963  HD3 ARG B 333     -13.735 -11.117   5.641  1.00  1.57           H  \n",
      "ATOM    964  HE  ARG B 333     -15.120 -10.182   3.431  1.00  1.64           H  \n",
      "ATOM    965 HH11 ARG B 333     -12.549 -12.417   4.015  1.00  2.21           H  \n",
      "ATOM    966 HH12 ARG B 333     -12.960 -13.435   2.677  1.00  2.79           H  \n",
      "ATOM    967 HH21 ARG B 333     -15.662 -11.507   1.684  1.00  2.41           H  \n",
      "ATOM    968 HH22 ARG B 333     -14.718 -12.922   1.360  1.00  2.59           H  \n",
      "ATOM    969  N   GLY B 334     -12.206  -6.984   9.085  1.00  0.42           N  \n",
      "ATOM    970  CA  GLY B 334     -11.223  -6.673  10.161  1.00  0.44           C  \n",
      "ATOM    971  C   GLY B 334     -11.089  -5.158  10.319  1.00  0.41           C  \n",
      "ATOM    972  O   GLY B 334     -10.822  -4.445   9.372  1.00  0.39           O  \n",
      "ATOM    973  H   GLY B 334     -13.108  -7.290   9.319  1.00  0.44           H  \n",
      "ATOM    974  HA2 GLY B 334     -11.561  -7.104  11.092  1.00  0.48           H  \n",
      "ATOM    975  HA3 GLY B 334     -10.261  -7.090   9.900  1.00  0.46           H  \n",
      "ATOM    976  N   ARG B 335     -11.269  -4.661  11.512  1.00  0.45           N  \n",
      "ATOM    977  CA  ARG B 335     -11.149  -3.193  11.734  1.00  0.46           C  \n",
      "ATOM    978  C   ARG B 335      -9.672  -2.793  11.690  1.00  0.45           C  \n",
      "ATOM    979  O   ARG B 335      -9.322  -1.721  11.238  1.00  0.43           O  \n",
      "ATOM    980  CB  ARG B 335     -11.733  -2.836  13.103  1.00  0.52           C  \n",
      "ATOM    981  CG  ARG B 335     -11.512  -1.348  13.384  1.00  0.57           C  \n",
      "ATOM    982  CD  ARG B 335     -11.429  -1.121  14.893  1.00  0.93           C  \n",
      "ATOM    983  NE  ARG B 335     -11.844   0.274  15.211  1.00  1.39           N  \n",
      "ATOM    984  CZ  ARG B 335     -12.140   0.606  16.439  1.00  1.89           C  \n",
      "ATOM    985  NH1 ARG B 335     -12.077  -0.283  17.393  1.00  2.25           N  \n",
      "ATOM    986  NH2 ARG B 335     -12.501   1.830  16.713  1.00  2.72           N  \n",
      "ATOM    987  H   ARG B 335     -11.481  -5.254  12.263  1.00  0.49           H  \n",
      "ATOM    988  HA  ARG B 335     -11.689  -2.665  10.963  1.00  0.45           H  \n",
      "ATOM    989  HB2 ARG B 335     -12.791  -3.051  13.109  1.00  0.54           H  \n",
      "ATOM    990  HB3 ARG B 335     -11.242  -3.420  13.867  1.00  0.60           H  \n",
      "ATOM    991  HG2 ARG B 335     -10.591  -1.025  12.920  1.00  0.78           H  \n",
      "ATOM    992  HG3 ARG B 335     -12.336  -0.779  12.981  1.00  0.81           H  \n",
      "ATOM    993  HD2 ARG B 335     -12.085  -1.815  15.398  1.00  1.59           H  \n",
      "ATOM    994  HD3 ARG B 335     -10.414  -1.279  15.226  1.00  1.50           H  \n",
      "ATOM    995  HE  ARG B 335     -11.894   0.944  14.497  1.00  2.00           H  \n",
      "ATOM    996 HH11 ARG B 335     -11.802  -1.222  17.188  1.00  2.24           H  \n",
      "ATOM    997 HH12 ARG B 335     -12.305  -0.025  18.332  1.00  2.96           H  \n",
      "ATOM    998 HH21 ARG B 335     -12.551   2.512  15.984  1.00  3.13           H  \n",
      "ATOM    999 HH22 ARG B 335     -12.728   2.085  17.653  1.00  3.22           H  \n",
      "ATOM   1000  N   GLU B 336      -8.804  -3.648  12.157  1.00  0.49           N  \n",
      "ATOM   1001  CA  GLU B 336      -7.351  -3.319  12.145  1.00  0.52           C  \n",
      "ATOM   1002  C   GLU B 336      -6.899  -3.042  10.710  1.00  0.46           C  \n",
      "ATOM   1003  O   GLU B 336      -6.423  -1.970  10.394  1.00  0.43           O  \n",
      "ATOM   1004  CB  GLU B 336      -6.559  -4.502  12.707  1.00  0.60           C  \n",
      "ATOM   1005  CG  GLU B 336      -5.881  -4.091  14.015  1.00  1.16           C  \n",
      "ATOM   1006  CD  GLU B 336      -5.265  -5.324  14.679  1.00  1.55           C  \n",
      "ATOM   1007  OE1 GLU B 336      -4.548  -6.042  14.002  1.00  2.21           O  \n",
      "ATOM   1008  OE2 GLU B 336      -5.521  -5.529  15.854  1.00  2.01           O  \n",
      "ATOM   1009  H   GLU B 336      -9.109  -4.506  12.518  1.00  0.53           H  \n",
      "ATOM   1010  HA  GLU B 336      -7.173  -2.445  12.754  1.00  0.55           H  \n",
      "ATOM   1011  HB2 GLU B 336      -7.231  -5.328  12.893  1.00  1.02           H  \n",
      "ATOM   1012  HB3 GLU B 336      -5.808  -4.804  11.993  1.00  0.99           H  \n",
      "ATOM   1013  HG2 GLU B 336      -5.105  -3.368  13.806  1.00  1.70           H  \n",
      "ATOM   1014  HG3 GLU B 336      -6.612  -3.654  14.678  1.00  1.70           H  \n",
      "ATOM   1015  N   ARG B 337      -7.043  -4.003   9.839  1.00  0.46           N  \n",
      "ATOM   1016  CA  ARG B 337      -6.618  -3.799   8.426  1.00  0.43           C  \n",
      "ATOM   1017  C   ARG B 337      -7.412  -2.646   7.813  1.00  0.37           C  \n",
      "ATOM   1018  O   ARG B 337      -6.933  -1.940   6.948  1.00  0.34           O  \n",
      "ATOM   1019  CB  ARG B 337      -6.876  -5.080   7.630  1.00  0.49           C  \n",
      "ATOM   1020  CG  ARG B 337      -5.957  -6.187   8.148  1.00  0.62           C  \n",
      "ATOM   1021  CD  ARG B 337      -4.592  -6.069   7.469  1.00  1.13           C  \n",
      "ATOM   1022  NE  ARG B 337      -4.034  -7.435   7.215  1.00  1.43           N  \n",
      "ATOM   1023  CZ  ARG B 337      -3.955  -8.324   8.170  1.00  2.16           C  \n",
      "ATOM   1024  NH1 ARG B 337      -4.176  -7.988   9.412  1.00  2.75           N  \n",
      "ATOM   1025  NH2 ARG B 337      -3.591  -9.545   7.888  1.00  2.79           N  \n",
      "ATOM   1026  H   ARG B 337      -7.427  -4.860  10.116  1.00  0.50           H  \n",
      "ATOM   1027  HA  ARG B 337      -5.565  -3.565   8.397  1.00  0.44           H  \n",
      "ATOM   1028  HB2 ARG B 337      -7.908  -5.379   7.751  1.00  0.53           H  \n",
      "ATOM   1029  HB3 ARG B 337      -6.671  -4.903   6.585  1.00  0.50           H  \n",
      "ATOM   1030  HG2 ARG B 337      -5.844  -6.081   9.216  1.00  1.26           H  \n",
      "ATOM   1031  HG3 ARG B 337      -6.389  -7.150   7.922  1.00  1.07           H  \n",
      "ATOM   1032  HD2 ARG B 337      -4.709  -5.576   6.521  1.00  1.68           H  \n",
      "ATOM   1033  HD3 ARG B 337      -3.928  -5.481   8.096  1.00  1.86           H  \n",
      "ATOM   1034  HE  ARG B 337      -3.771  -7.684   6.304  1.00  1.69           H  \n",
      "ATOM   1035 HH11 ARG B 337      -4.401  -7.045   9.644  1.00  2.63           H  \n",
      "ATOM   1036 HH12 ARG B 337      -4.121  -8.678  10.133  1.00  3.55           H  \n",
      "ATOM   1037 HH21 ARG B 337      -3.373  -9.797   6.945  1.00  2.80           H  \n",
      "ATOM   1038 HH22 ARG B 337      -3.528 -10.229   8.615  1.00  3.49           H  \n",
      "ATOM   1039  N   PHE B 338      -8.623  -2.448   8.252  1.00  0.37           N  \n",
      "ATOM   1040  CA  PHE B 338      -9.443  -1.340   7.692  1.00  0.34           C  \n",
      "ATOM   1041  C   PHE B 338      -8.687  -0.018   7.837  1.00  0.32           C  \n",
      "ATOM   1042  O   PHE B 338      -8.439   0.673   6.872  1.00  0.29           O  \n",
      "ATOM   1043  CB  PHE B 338     -10.769  -1.252   8.448  1.00  0.38           C  \n",
      "ATOM   1044  CG  PHE B 338     -11.516  -0.019   8.004  1.00  0.37           C  \n",
      "ATOM   1045  CD1 PHE B 338     -12.107   0.022   6.732  1.00  0.38           C  \n",
      "ATOM   1046  CD2 PHE B 338     -11.618   1.088   8.860  1.00  0.42           C  \n",
      "ATOM   1047  CE1 PHE B 338     -12.799   1.169   6.316  1.00  0.40           C  \n",
      "ATOM   1048  CE2 PHE B 338     -12.310   2.235   8.445  1.00  0.45           C  \n",
      "ATOM   1049  CZ  PHE B 338     -12.901   2.275   7.173  1.00  0.42           C  \n",
      "ATOM   1050  H   PHE B 338      -8.993  -3.028   8.950  1.00  0.40           H  \n",
      "ATOM   1051  HA  PHE B 338      -9.637  -1.530   6.647  1.00  0.34           H  \n",
      "ATOM   1052  HB2 PHE B 338     -11.363  -2.130   8.238  1.00  0.40           H  \n",
      "ATOM   1053  HB3 PHE B 338     -10.576  -1.192   9.509  1.00  0.42           H  \n",
      "ATOM   1054  HD1 PHE B 338     -12.028  -0.830   6.073  1.00  0.42           H  \n",
      "ATOM   1055  HD2 PHE B 338     -11.163   1.056   9.839  1.00  0.49           H  \n",
      "ATOM   1056  HE1 PHE B 338     -13.253   1.200   5.337  1.00  0.45           H  \n",
      "ATOM   1057  HE2 PHE B 338     -12.389   3.087   9.104  1.00  0.52           H  \n",
      "ATOM   1058  HZ  PHE B 338     -13.434   3.158   6.852  1.00  0.46           H  \n",
      "ATOM   1059  N   GLU B 339      -8.327   0.342   9.037  1.00  0.37           N  \n",
      "ATOM   1060  CA  GLU B 339      -7.595   1.625   9.248  1.00  0.39           C  \n",
      "ATOM   1061  C   GLU B 339      -6.375   1.696   8.322  1.00  0.34           C  \n",
      "ATOM   1062  O   GLU B 339      -6.006   2.754   7.850  1.00  0.33           O  \n",
      "ATOM   1063  CB  GLU B 339      -7.132   1.711  10.704  1.00  0.46           C  \n",
      "ATOM   1064  CG  GLU B 339      -8.351   1.741  11.628  1.00  0.60           C  \n",
      "ATOM   1065  CD  GLU B 339      -7.945   2.302  12.992  1.00  1.00           C  \n",
      "ATOM   1066  OE1 GLU B 339      -7.732   3.500  13.078  1.00  1.68           O  \n",
      "ATOM   1067  OE2 GLU B 339      -7.853   1.524  13.927  1.00  1.56           O  \n",
      "ATOM   1068  H   GLU B 339      -8.543  -0.228   9.804  1.00  0.41           H  \n",
      "ATOM   1069  HA  GLU B 339      -8.255   2.452   9.033  1.00  0.40           H  \n",
      "ATOM   1070  HB2 GLU B 339      -6.522   0.851  10.939  1.00  0.47           H  \n",
      "ATOM   1071  HB3 GLU B 339      -6.554   2.612  10.845  1.00  0.51           H  \n",
      "ATOM   1072  HG2 GLU B 339      -9.117   2.367  11.194  1.00  0.75           H  \n",
      "ATOM   1073  HG3 GLU B 339      -8.733   0.739  11.753  1.00  0.83           H  \n",
      "ATOM   1074  N   MET B 340      -5.737   0.586   8.070  1.00  0.33           N  \n",
      "ATOM   1075  CA  MET B 340      -4.532   0.601   7.189  1.00  0.31           C  \n",
      "ATOM   1076  C   MET B 340      -4.916   1.044   5.774  1.00  0.27           C  \n",
      "ATOM   1077  O   MET B 340      -4.306   1.927   5.205  1.00  0.27           O  \n",
      "ATOM   1078  CB  MET B 340      -3.924  -0.803   7.135  1.00  0.34           C  \n",
      "ATOM   1079  CG  MET B 340      -2.548  -0.739   6.469  1.00  0.34           C  \n",
      "ATOM   1080  SD  MET B 340      -1.902  -2.416   6.254  1.00  0.43           S  \n",
      "ATOM   1081  CE  MET B 340      -2.501  -2.675   4.566  1.00  0.43           C  \n",
      "ATOM   1082  H   MET B 340      -6.042  -0.256   8.469  1.00  0.35           H  \n",
      "ATOM   1083  HA  MET B 340      -3.804   1.289   7.592  1.00  0.33           H  \n",
      "ATOM   1084  HB2 MET B 340      -3.821  -1.189   8.139  1.00  0.41           H  \n",
      "ATOM   1085  HB3 MET B 340      -4.568  -1.454   6.563  1.00  0.34           H  \n",
      "ATOM   1086  HG2 MET B 340      -2.637  -0.261   5.505  1.00  0.34           H  \n",
      "ATOM   1087  HG3 MET B 340      -1.873  -0.170   7.092  1.00  0.43           H  \n",
      "ATOM   1088  HE1 MET B 340      -3.563  -2.479   4.528  1.00  1.11           H  \n",
      "ATOM   1089  HE2 MET B 340      -1.990  -2.003   3.895  1.00  1.15           H  \n",
      "ATOM   1090  HE3 MET B 340      -2.306  -3.696   4.269  1.00  1.07           H  \n",
      "ATOM   1091  N   PHE B 341      -5.914   0.434   5.197  1.00  0.25           N  \n",
      "ATOM   1092  CA  PHE B 341      -6.324   0.817   3.815  1.00  0.22           C  \n",
      "ATOM   1093  C   PHE B 341      -6.785   2.277   3.794  1.00  0.22           C  \n",
      "ATOM   1094  O   PHE B 341      -6.397   3.047   2.939  1.00  0.22           O  \n",
      "ATOM   1095  CB  PHE B 341      -7.471  -0.087   3.359  1.00  0.22           C  \n",
      "ATOM   1096  CG  PHE B 341      -6.915  -1.421   2.920  1.00  0.22           C  \n",
      "ATOM   1097  CD1 PHE B 341      -6.181  -1.517   1.729  1.00  0.26           C  \n",
      "ATOM   1098  CD2 PHE B 341      -7.133  -2.564   3.705  1.00  0.25           C  \n",
      "ATOM   1099  CE1 PHE B 341      -5.665  -2.757   1.322  1.00  0.28           C  \n",
      "ATOM   1100  CE2 PHE B 341      -6.616  -3.803   3.297  1.00  0.28           C  \n",
      "ATOM   1101  CZ  PHE B 341      -5.882  -3.899   2.106  1.00  0.28           C  \n",
      "ATOM   1102  H   PHE B 341      -6.390  -0.280   5.669  1.00  0.26           H  \n",
      "ATOM   1103  HA  PHE B 341      -5.485   0.697   3.146  1.00  0.22           H  \n",
      "ATOM   1104  HB2 PHE B 341      -8.160  -0.235   4.178  1.00  0.23           H  \n",
      "ATOM   1105  HB3 PHE B 341      -7.988   0.377   2.532  1.00  0.22           H  \n",
      "ATOM   1106  HD1 PHE B 341      -6.013  -0.638   1.125  1.00  0.30           H  \n",
      "ATOM   1107  HD2 PHE B 341      -7.698  -2.490   4.622  1.00  0.28           H  \n",
      "ATOM   1108  HE1 PHE B 341      -5.099  -2.831   0.404  1.00  0.33           H  \n",
      "ATOM   1109  HE2 PHE B 341      -6.784  -4.683   3.901  1.00  0.33           H  \n",
      "ATOM   1110  HZ  PHE B 341      -5.485  -4.853   1.792  1.00  0.31           H  \n",
      "ATOM   1111  N   ARG B 342      -7.615   2.661   4.723  1.00  0.25           N  \n",
      "ATOM   1112  CA  ARG B 342      -8.110   4.067   4.757  1.00  0.28           C  \n",
      "ATOM   1113  C   ARG B 342      -6.932   5.038   4.656  1.00  0.27           C  \n",
      "ATOM   1114  O   ARG B 342      -6.997   6.036   3.965  1.00  0.27           O  \n",
      "ATOM   1115  CB  ARG B 342      -8.857   4.308   6.070  1.00  0.34           C  \n",
      "ATOM   1116  CG  ARG B 342      -9.492   5.699   6.050  1.00  0.42           C  \n",
      "ATOM   1117  CD  ARG B 342      -9.887   6.099   7.472  1.00  0.91           C  \n",
      "ATOM   1118  NE  ARG B 342     -10.499   7.458   7.455  1.00  1.33           N  \n",
      "ATOM   1119  CZ  ARG B 342     -11.177   7.886   8.486  1.00  1.80           C  \n",
      "ATOM   1120  NH1 ARG B 342     -11.320   7.128   9.541  1.00  2.22           N  \n",
      "ATOM   1121  NH2 ARG B 342     -11.713   9.075   8.463  1.00  2.52           N  \n",
      "ATOM   1122  H   ARG B 342      -7.918   2.023   5.398  1.00  0.27           H  \n",
      "ATOM   1123  HA  ARG B 342      -8.782   4.231   3.928  1.00  0.29           H  \n",
      "ATOM   1124  HB2 ARG B 342      -9.629   3.560   6.187  1.00  0.38           H  \n",
      "ATOM   1125  HB3 ARG B 342      -8.164   4.242   6.896  1.00  0.38           H  \n",
      "ATOM   1126  HG2 ARG B 342      -8.782   6.413   5.658  1.00  0.80           H  \n",
      "ATOM   1127  HG3 ARG B 342     -10.371   5.685   5.425  1.00  0.74           H  \n",
      "ATOM   1128  HD2 ARG B 342     -10.602   5.388   7.861  1.00  1.52           H  \n",
      "ATOM   1129  HD3 ARG B 342      -9.010   6.108   8.101  1.00  1.44           H  \n",
      "ATOM   1130  HE  ARG B 342     -10.394   8.031   6.667  1.00  1.92           H  \n",
      "ATOM   1131 HH11 ARG B 342     -10.911   6.216   9.563  1.00  2.21           H  \n",
      "ATOM   1132 HH12 ARG B 342     -11.840   7.461  10.327  1.00  2.93           H  \n",
      "ATOM   1133 HH21 ARG B 342     -11.605   9.658   7.658  1.00  2.86           H  \n",
      "ATOM   1134 HH22 ARG B 342     -12.233   9.405   9.252  1.00  3.02           H  \n",
      "ATOM   1135  N   GLU B 343      -5.857   4.762   5.342  1.00  0.27           N  \n",
      "ATOM   1136  CA  GLU B 343      -4.683   5.678   5.287  1.00  0.27           C  \n",
      "ATOM   1137  C   GLU B 343      -4.204   5.815   3.840  1.00  0.23           C  \n",
      "ATOM   1138  O   GLU B 343      -3.953   6.903   3.360  1.00  0.24           O  \n",
      "ATOM   1139  CB  GLU B 343      -3.552   5.111   6.147  1.00  0.31           C  \n",
      "ATOM   1140  CG  GLU B 343      -2.321   6.013   6.030  1.00  0.35           C  \n",
      "ATOM   1141  CD  GLU B 343      -1.720   6.240   7.418  1.00  1.03           C  \n",
      "ATOM   1142  OE1 GLU B 343      -1.763   5.321   8.219  1.00  1.75           O  \n",
      "ATOM   1143  OE2 GLU B 343      -1.226   7.330   7.657  1.00  1.71           O  \n",
      "ATOM   1144  H   GLU B 343      -5.824   3.955   5.898  1.00  0.28           H  \n",
      "ATOM   1145  HA  GLU B 343      -4.968   6.649   5.663  1.00  0.30           H  \n",
      "ATOM   1146  HB2 GLU B 343      -3.871   5.068   7.178  1.00  0.36           H  \n",
      "ATOM   1147  HB3 GLU B 343      -3.301   4.118   5.805  1.00  0.35           H  \n",
      "ATOM   1148  HG2 GLU B 343      -1.589   5.541   5.392  1.00  0.70           H  \n",
      "ATOM   1149  HG3 GLU B 343      -2.610   6.963   5.606  1.00  0.68           H  \n",
      "ATOM   1150  N   LEU B 344      -4.072   4.722   3.141  1.00  0.21           N  \n",
      "ATOM   1151  CA  LEU B 344      -3.606   4.795   1.727  1.00  0.20           C  \n",
      "ATOM   1152  C   LEU B 344      -4.613   5.589   0.894  1.00  0.21           C  \n",
      "ATOM   1153  O   LEU B 344      -4.250   6.320  -0.004  1.00  0.23           O  \n",
      "ATOM   1154  CB  LEU B 344      -3.482   3.380   1.158  1.00  0.22           C  \n",
      "ATOM   1155  CG  LEU B 344      -2.361   2.635   1.885  1.00  0.25           C  \n",
      "ATOM   1156  CD1 LEU B 344      -2.488   1.134   1.618  1.00  0.31           C  \n",
      "ATOM   1157  CD2 LEU B 344      -1.007   3.131   1.374  1.00  0.30           C  \n",
      "ATOM   1158  H   LEU B 344      -4.277   3.854   3.545  1.00  0.22           H  \n",
      "ATOM   1159  HA  LEU B 344      -2.644   5.283   1.689  1.00  0.21           H  \n",
      "ATOM   1160  HB2 LEU B 344      -4.415   2.853   1.297  1.00  0.23           H  \n",
      "ATOM   1161  HB3 LEU B 344      -3.251   3.433   0.105  1.00  0.24           H  \n",
      "ATOM   1162  HG  LEU B 344      -2.437   2.819   2.947  1.00  0.28           H  \n",
      "ATOM   1163 HD11 LEU B 344      -2.977   0.977   0.668  1.00  1.01           H  \n",
      "ATOM   1164 HD12 LEU B 344      -1.505   0.688   1.595  1.00  1.02           H  \n",
      "ATOM   1165 HD13 LEU B 344      -3.072   0.677   2.403  1.00  1.04           H  \n",
      "ATOM   1166 HD21 LEU B 344      -1.118   4.125   0.966  1.00  1.07           H  \n",
      "ATOM   1167 HD22 LEU B 344      -0.300   3.153   2.191  1.00  1.01           H  \n",
      "ATOM   1168 HD23 LEU B 344      -0.646   2.464   0.604  1.00  1.06           H  \n",
      "ATOM   1169  N   ASN B 345      -5.877   5.451   1.184  1.00  0.26           N  \n",
      "ATOM   1170  CA  ASN B 345      -6.905   6.195   0.405  1.00  0.30           C  \n",
      "ATOM   1171  C   ASN B 345      -6.623   7.697   0.482  1.00  0.26           C  \n",
      "ATOM   1172  O   ASN B 345      -6.521   8.370  -0.523  1.00  0.25           O  \n",
      "ATOM   1173  CB  ASN B 345      -8.291   5.908   0.984  1.00  0.38           C  \n",
      "ATOM   1174  CG  ASN B 345      -9.345   6.674   0.182  1.00  0.48           C  \n",
      "ATOM   1175  OD1 ASN B 345      -9.070   7.727  -0.358  1.00  1.21           O  \n",
      "ATOM   1176  ND2 ASN B 345     -10.550   6.186   0.079  1.00  0.58           N  \n",
      "ATOM   1177  H   ASN B 345      -6.150   4.856   1.911  1.00  0.30           H  \n",
      "ATOM   1178  HA  ASN B 345      -6.874   5.876  -0.626  1.00  0.33           H  \n",
      "ATOM   1179  HB2 ASN B 345      -8.492   4.848   0.927  1.00  0.42           H  \n",
      "ATOM   1180  HB3 ASN B 345      -8.324   6.227   2.015  1.00  0.42           H  \n",
      "ATOM   1181 HD21 ASN B 345     -10.772   5.337   0.514  1.00  1.14           H  \n",
      "ATOM   1182 HD22 ASN B 345     -11.232   6.668  -0.433  1.00  0.58           H  \n",
      "ATOM   1183  N   GLU B 346      -6.504   8.228   1.668  1.00  0.28           N  \n",
      "ATOM   1184  CA  GLU B 346      -6.237   9.687   1.808  1.00  0.29           C  \n",
      "ATOM   1185  C   GLU B 346      -4.907  10.035   1.136  1.00  0.25           C  \n",
      "ATOM   1186  O   GLU B 346      -4.718  11.129   0.646  1.00  0.26           O  \n",
      "ATOM   1187  CB  GLU B 346      -6.168  10.051   3.293  1.00  0.37           C  \n",
      "ATOM   1188  CG  GLU B 346      -5.955  11.559   3.442  1.00  0.43           C  \n",
      "ATOM   1189  CD  GLU B 346      -6.615  12.044   4.734  1.00  0.95           C  \n",
      "ATOM   1190  OE1 GLU B 346      -7.766  11.704   4.952  1.00  1.63           O  \n",
      "ATOM   1191  OE2 GLU B 346      -5.958  12.747   5.485  1.00  1.69           O  \n",
      "ATOM   1192  H   GLU B 346      -6.596   7.667   2.467  1.00  0.33           H  \n",
      "ATOM   1193  HA  GLU B 346      -7.034  10.245   1.338  1.00  0.32           H  \n",
      "ATOM   1194  HB2 GLU B 346      -7.092   9.768   3.776  1.00  0.43           H  \n",
      "ATOM   1195  HB3 GLU B 346      -5.345   9.526   3.753  1.00  0.39           H  \n",
      "ATOM   1196  HG2 GLU B 346      -4.895  11.771   3.477  1.00  0.83           H  \n",
      "ATOM   1197  HG3 GLU B 346      -6.397  12.071   2.601  1.00  0.86           H  \n",
      "ATOM   1198  N   ALA B 347      -3.982   9.116   1.117  1.00  0.23           N  \n",
      "ATOM   1199  CA  ALA B 347      -2.661   9.395   0.482  1.00  0.24           C  \n",
      "ATOM   1200  C   ALA B 347      -2.853   9.719  -1.002  1.00  0.23           C  \n",
      "ATOM   1201  O   ALA B 347      -2.456  10.765  -1.476  1.00  0.25           O  \n",
      "ATOM   1202  CB  ALA B 347      -1.760   8.166   0.622  1.00  0.27           C  \n",
      "ATOM   1203  H   ALA B 347      -4.154   8.241   1.523  1.00  0.24           H  \n",
      "ATOM   1204  HA  ALA B 347      -2.198  10.237   0.975  1.00  0.27           H  \n",
      "ATOM   1205  HB1 ALA B 347      -2.347   7.327   0.966  1.00  1.02           H  \n",
      "ATOM   1206  HB2 ALA B 347      -1.322   7.930  -0.336  1.00  1.01           H  \n",
      "ATOM   1207  HB3 ALA B 347      -0.977   8.374   1.336  1.00  1.06           H  \n",
      "ATOM   1208  N   LEU B 348      -3.452   8.825  -1.741  1.00  0.22           N  \n",
      "ATOM   1209  CA  LEU B 348      -3.661   9.075  -3.196  1.00  0.24           C  \n",
      "ATOM   1210  C   LEU B 348      -4.544  10.311  -3.388  1.00  0.28           C  \n",
      "ATOM   1211  O   LEU B 348      -4.273  11.156  -4.218  1.00  0.30           O  \n",
      "ATOM   1212  CB  LEU B 348      -4.350   7.860  -3.825  1.00  0.25           C  \n",
      "ATOM   1213  CG  LEU B 348      -3.327   6.745  -4.053  1.00  0.25           C  \n",
      "ATOM   1214  CD1 LEU B 348      -2.150   7.286  -4.865  1.00  0.30           C  \n",
      "ATOM   1215  CD2 LEU B 348      -2.822   6.231  -2.704  1.00  0.26           C  \n",
      "ATOM   1216  H   LEU B 348      -3.758   7.985  -1.340  1.00  0.21           H  \n",
      "ATOM   1217  HA  LEU B 348      -2.707   9.237  -3.675  1.00  0.25           H  \n",
      "ATOM   1218  HB2 LEU B 348      -5.127   7.506  -3.163  1.00  0.25           H  \n",
      "ATOM   1219  HB3 LEU B 348      -4.786   8.144  -4.771  1.00  0.29           H  \n",
      "ATOM   1220  HG  LEU B 348      -3.795   5.935  -4.595  1.00  0.28           H  \n",
      "ATOM   1221 HD11 LEU B 348      -2.521   7.918  -5.658  1.00  1.08           H  \n",
      "ATOM   1222 HD12 LEU B 348      -1.502   7.860  -4.220  1.00  1.08           H  \n",
      "ATOM   1223 HD13 LEU B 348      -1.597   6.462  -5.291  1.00  1.02           H  \n",
      "ATOM   1224 HD21 LEU B 348      -3.646   6.174  -2.010  1.00  1.01           H  \n",
      "ATOM   1225 HD22 LEU B 348      -2.390   5.249  -2.831  1.00  1.06           H  \n",
      "ATOM   1226 HD23 LEU B 348      -2.073   6.907  -2.318  1.00  1.03           H  \n",
      "ATOM   1227  N   GLU B 349      -5.603  10.419  -2.634  1.00  0.31           N  \n",
      "ATOM   1228  CA  GLU B 349      -6.509  11.595  -2.780  1.00  0.36           C  \n",
      "ATOM   1229  C   GLU B 349      -5.722  12.890  -2.558  1.00  0.33           C  \n",
      "ATOM   1230  O   GLU B 349      -6.001  13.905  -3.164  1.00  0.35           O  \n",
      "ATOM   1231  CB  GLU B 349      -7.637  11.498  -1.751  1.00  0.41           C  \n",
      "ATOM   1232  CG  GLU B 349      -8.471  10.244  -2.025  1.00  0.49           C  \n",
      "ATOM   1233  CD  GLU B 349      -9.958  10.578  -1.890  1.00  1.14           C  \n",
      "ATOM   1234  OE1 GLU B 349     -10.281  11.450  -1.100  1.00  1.93           O  \n",
      "ATOM   1235  OE2 GLU B 349     -10.750   9.955  -2.578  1.00  1.68           O  \n",
      "ATOM   1236  H   GLU B 349      -5.807   9.724  -1.975  1.00  0.32           H  \n",
      "ATOM   1237  HA  GLU B 349      -6.931  11.599  -3.774  1.00  0.40           H  \n",
      "ATOM   1238  HB2 GLU B 349      -7.215  11.440  -0.758  1.00  0.40           H  \n",
      "ATOM   1239  HB3 GLU B 349      -8.268  12.371  -1.824  1.00  0.48           H  \n",
      "ATOM   1240  HG2 GLU B 349      -8.271   9.891  -3.027  1.00  0.91           H  \n",
      "ATOM   1241  HG3 GLU B 349      -8.211   9.475  -1.314  1.00  0.80           H  \n",
      "ATOM   1242  N   LEU B 350      -4.741  12.865  -1.698  1.00  0.30           N  \n",
      "ATOM   1243  CA  LEU B 350      -3.942  14.098  -1.448  1.00  0.29           C  \n",
      "ATOM   1244  C   LEU B 350      -3.218  14.494  -2.734  1.00  0.28           C  \n",
      "ATOM   1245  O   LEU B 350      -3.192  15.647  -3.117  1.00  0.32           O  \n",
      "ATOM   1246  CB  LEU B 350      -2.916  13.830  -0.345  1.00  0.28           C  \n",
      "ATOM   1247  CG  LEU B 350      -2.443  15.159   0.249  1.00  0.32           C  \n",
      "ATOM   1248  CD1 LEU B 350      -1.875  14.920   1.649  1.00  0.40           C  \n",
      "ATOM   1249  CD2 LEU B 350      -1.356  15.761  -0.644  1.00  0.35           C  \n",
      "ATOM   1250  H   LEU B 350      -4.527  12.037  -1.222  1.00  0.30           H  \n",
      "ATOM   1251  HA  LEU B 350      -4.600  14.899  -1.143  1.00  0.32           H  \n",
      "ATOM   1252  HB2 LEU B 350      -3.370  13.230   0.430  1.00  0.33           H  \n",
      "ATOM   1253  HB3 LEU B 350      -2.070  13.303  -0.760  1.00  0.27           H  \n",
      "ATOM   1254  HG  LEU B 350      -3.278  15.841   0.313  1.00  0.48           H  \n",
      "ATOM   1255 HD11 LEU B 350      -1.474  13.919   1.709  1.00  1.09           H  \n",
      "ATOM   1256 HD12 LEU B 350      -1.089  15.635   1.846  1.00  1.13           H  \n",
      "ATOM   1257 HD13 LEU B 350      -2.659  15.037   2.382  1.00  1.06           H  \n",
      "ATOM   1258 HD21 LEU B 350      -0.550  15.052  -0.759  1.00  1.14           H  \n",
      "ATOM   1259 HD22 LEU B 350      -1.772  15.992  -1.612  1.00  1.08           H  \n",
      "ATOM   1260 HD23 LEU B 350      -0.979  16.666  -0.191  1.00  1.03           H  \n",
      "ATOM   1261  N   LYS B 351      -2.634  13.543  -3.408  1.00  0.27           N  \n",
      "ATOM   1262  CA  LYS B 351      -1.915  13.858  -4.672  1.00  0.31           C  \n",
      "ATOM   1263  C   LYS B 351      -2.896  14.480  -5.666  1.00  0.37           C  \n",
      "ATOM   1264  O   LYS B 351      -2.580  15.428  -6.358  1.00  0.42           O  \n",
      "ATOM   1265  CB  LYS B 351      -1.338  12.568  -5.258  1.00  0.34           C  \n",
      "ATOM   1266  CG  LYS B 351      -0.375  12.905  -6.398  1.00  0.44           C  \n",
      "ATOM   1267  CD  LYS B 351       0.358  11.634  -6.833  1.00  0.50           C  \n",
      "ATOM   1268  CE  LYS B 351       1.631  12.010  -7.592  1.00  0.81           C  \n",
      "ATOM   1269  NZ  LYS B 351       1.974  10.924  -8.553  1.00  1.59           N  \n",
      "ATOM   1270  H   LYS B 351      -2.672  12.619  -3.083  1.00  0.27           H  \n",
      "ATOM   1271  HA  LYS B 351      -1.114  14.554  -4.469  1.00  0.31           H  \n",
      "ATOM   1272  HB2 LYS B 351      -0.808  12.029  -4.487  1.00  0.37           H  \n",
      "ATOM   1273  HB3 LYS B 351      -2.141  11.955  -5.638  1.00  0.38           H  \n",
      "ATOM   1274  HG2 LYS B 351      -0.932  13.305  -7.233  1.00  0.57           H  \n",
      "ATOM   1275  HG3 LYS B 351       0.344  13.636  -6.059  1.00  0.55           H  \n",
      "ATOM   1276  HD2 LYS B 351       0.617  11.053  -5.959  1.00  0.57           H  \n",
      "ATOM   1277  HD3 LYS B 351      -0.284  11.051  -7.476  1.00  0.66           H  \n",
      "ATOM   1278  HE2 LYS B 351       1.470  12.932  -8.132  1.00  1.34           H  \n",
      "ATOM   1279  HE3 LYS B 351       2.443  12.140  -6.892  1.00  1.15           H  \n",
      "ATOM   1280  HZ1 LYS B 351       1.570  10.026  -8.220  1.00  2.19           H  \n",
      "ATOM   1281  HZ2 LYS B 351       1.582  11.155  -9.490  1.00  2.05           H  \n",
      "ATOM   1282  HZ3 LYS B 351       3.007  10.833  -8.621  1.00  2.02           H  \n",
      "ATOM   1283  N   ASP B 352      -4.088  13.953  -5.741  1.00  0.42           N  \n",
      "ATOM   1284  CA  ASP B 352      -5.095  14.511  -6.686  1.00  0.50           C  \n",
      "ATOM   1285  C   ASP B 352      -5.295  16.001  -6.401  1.00  0.50           C  \n",
      "ATOM   1286  O   ASP B 352      -5.405  16.807  -7.303  1.00  0.58           O  \n",
      "ATOM   1287  CB  ASP B 352      -6.425  13.776  -6.504  1.00  0.59           C  \n",
      "ATOM   1288  CG  ASP B 352      -6.322  12.373  -7.105  1.00  0.68           C  \n",
      "ATOM   1289  OD1 ASP B 352      -5.374  11.679  -6.780  1.00  1.15           O  \n",
      "ATOM   1290  OD2 ASP B 352      -7.195  12.017  -7.880  1.00  1.43           O  \n",
      "ATOM   1291  H   ASP B 352      -4.322  13.190  -5.172  1.00  0.43           H  \n",
      "ATOM   1292  HA  ASP B 352      -4.748  14.382  -7.700  1.00  0.55           H  \n",
      "ATOM   1293  HB2 ASP B 352      -6.654  13.702  -5.451  1.00  0.57           H  \n",
      "ATOM   1294  HB3 ASP B 352      -7.209  14.324  -7.005  1.00  0.69           H  \n",
      "ATOM   1295  N   ALA B 353      -5.348  16.373  -5.151  1.00  0.49           N  \n",
      "ATOM   1296  CA  ALA B 353      -5.547  17.810  -4.808  1.00  0.56           C  \n",
      "ATOM   1297  C   ALA B 353      -4.325  18.617  -5.248  1.00  0.54           C  \n",
      "ATOM   1298  O   ALA B 353      -4.420  19.791  -5.547  1.00  0.65           O  \n",
      "ATOM   1299  CB  ALA B 353      -5.734  17.951  -3.296  1.00  0.64           C  \n",
      "ATOM   1300  H   ALA B 353      -5.261  15.707  -4.437  1.00  0.50           H  \n",
      "ATOM   1301  HA  ALA B 353      -6.425  18.182  -5.315  1.00  0.64           H  \n",
      "ATOM   1302  HB1 ALA B 353      -6.380  17.163  -2.938  1.00  1.31           H  \n",
      "ATOM   1303  HB2 ALA B 353      -4.774  17.879  -2.807  1.00  1.12           H  \n",
      "ATOM   1304  HB3 ALA B 353      -6.179  18.910  -3.076  1.00  1.22           H  \n",
      "ATOM   1305  N   GLN B 354      -3.176  18.000  -5.294  1.00  0.51           N  \n",
      "ATOM   1306  CA  GLN B 354      -1.953  18.736  -5.719  1.00  0.59           C  \n",
      "ATOM   1307  C   GLN B 354      -1.864  18.738  -7.247  1.00  0.66           C  \n",
      "ATOM   1308  O   GLN B 354      -1.085  19.462  -7.833  1.00  0.86           O  \n",
      "ATOM   1309  CB  GLN B 354      -0.717  18.048  -5.138  1.00  0.61           C  \n",
      "ATOM   1310  CG  GLN B 354      -0.013  18.996  -4.166  1.00  0.94           C  \n",
      "ATOM   1311  CD  GLN B 354       1.349  18.413  -3.782  1.00  0.83           C  \n",
      "ATOM   1312  OE1 GLN B 354       2.266  19.142  -3.462  1.00  1.19           O  \n",
      "ATOM   1313  NE2 GLN B 354       1.519  17.119  -3.801  1.00  0.67           N  \n",
      "ATOM   1314  H   GLN B 354      -3.119  17.052  -5.052  1.00  0.49           H  \n",
      "ATOM   1315  HA  GLN B 354      -2.001  19.753  -5.360  1.00  0.68           H  \n",
      "ATOM   1316  HB2 GLN B 354      -1.017  17.151  -4.615  1.00  0.85           H  \n",
      "ATOM   1317  HB3 GLN B 354      -0.040  17.788  -5.938  1.00  0.87           H  \n",
      "ATOM   1318  HG2 GLN B 354       0.126  19.957  -4.638  1.00  1.36           H  \n",
      "ATOM   1319  HG3 GLN B 354      -0.615  19.114  -3.277  1.00  1.40           H  \n",
      "ATOM   1320 HE21 GLN B 354       0.779  16.531  -4.059  1.00  0.70           H  \n",
      "ATOM   1321 HE22 GLN B 354       2.387  16.735  -3.557  1.00  0.81           H  \n",
      "ATOM   1322  N   ALA B 355      -2.658  17.930  -7.895  1.00  0.67           N  \n",
      "ATOM   1323  CA  ALA B 355      -2.623  17.883  -9.383  1.00  0.82           C  \n",
      "ATOM   1324  C   ALA B 355      -3.277  19.145  -9.952  1.00  0.89           C  \n",
      "ATOM   1325  O   ALA B 355      -2.767  19.764 -10.865  1.00  1.22           O  \n",
      "ATOM   1326  CB  ALA B 355      -3.382  16.646  -9.868  1.00  1.02           C  \n",
      "ATOM   1327  H   ALA B 355      -3.279  17.354  -7.403  1.00  0.72           H  \n",
      "ATOM   1328  HA  ALA B 355      -1.597  17.828  -9.717  1.00  0.93           H  \n",
      "ATOM   1329  HB1 ALA B 355      -3.802  16.126  -9.018  1.00  1.52           H  \n",
      "ATOM   1330  HB2 ALA B 355      -4.177  16.948 -10.533  1.00  1.57           H  \n",
      "ATOM   1331  HB3 ALA B 355      -2.704  15.989 -10.391  1.00  1.32           H  \n",
      "ATOM   1332  N   GLY B 356      -4.406  19.529  -9.420  1.00  1.01           N  \n",
      "ATOM   1333  CA  GLY B 356      -5.101  20.749  -9.925  1.00  1.22           C  \n",
      "ATOM   1334  C   GLY B 356      -4.478  22.009  -9.313  1.00  1.17           C  \n",
      "ATOM   1335  O   GLY B 356      -5.049  23.080  -9.373  1.00  1.53           O  \n",
      "ATOM   1336  H   GLY B 356      -4.800  19.013  -8.686  1.00  1.23           H  \n",
      "ATOM   1337  HA2 GLY B 356      -5.009  20.792 -11.001  1.00  1.43           H  \n",
      "ATOM   1338  HA3 GLY B 356      -6.145  20.704  -9.657  1.00  1.52           H  \n",
      "ATOM   1339  N   LYS B 357      -3.320  21.898  -8.719  1.00  1.30           N  \n",
      "ATOM   1340  CA  LYS B 357      -2.685  23.098  -8.106  1.00  1.51           C  \n",
      "ATOM   1341  C   LYS B 357      -1.904  23.864  -9.173  1.00  1.94           C  \n",
      "ATOM   1342  O   LYS B 357      -0.914  23.390  -9.696  1.00  2.47           O  \n",
      "ATOM   1343  CB  LYS B 357      -1.734  22.658  -6.990  1.00  1.85           C  \n",
      "ATOM   1344  CG  LYS B 357      -1.736  23.706  -5.875  1.00  2.23           C  \n",
      "ATOM   1345  CD  LYS B 357      -0.319  23.868  -5.321  1.00  2.95           C  \n",
      "ATOM   1346  CE  LYS B 357      -0.330  24.891  -4.184  1.00  3.51           C  \n",
      "ATOM   1347  NZ  LYS B 357       1.061  25.110  -3.697  1.00  4.21           N  \n",
      "ATOM   1348  H   LYS B 357      -2.872  21.030  -8.672  1.00  1.59           H  \n",
      "ATOM   1349  HA  LYS B 357      -3.451  23.738  -7.693  1.00  1.71           H  \n",
      "ATOM   1350  HB2 LYS B 357      -2.061  21.708  -6.594  1.00  2.20           H  \n",
      "ATOM   1351  HB3 LYS B 357      -0.735  22.558  -7.387  1.00  2.11           H  \n",
      "ATOM   1352  HG2 LYS B 357      -2.080  24.651  -6.271  1.00  2.38           H  \n",
      "ATOM   1353  HG3 LYS B 357      -2.396  23.386  -5.083  1.00  2.53           H  \n",
      "ATOM   1354  HD2 LYS B 357       0.032  22.917  -4.948  1.00  3.42           H  \n",
      "ATOM   1355  HD3 LYS B 357       0.337  24.212  -6.106  1.00  3.15           H  \n",
      "ATOM   1356  HE2 LYS B 357      -0.736  25.825  -4.544  1.00  3.68           H  \n",
      "ATOM   1357  HE3 LYS B 357      -0.941  24.522  -3.374  1.00  3.78           H  \n",
      "ATOM   1358  HZ1 LYS B 357       1.734  24.699  -4.374  1.00  4.49           H  \n",
      "ATOM   1359  HZ2 LYS B 357       1.242  26.130  -3.609  1.00  4.58           H  \n",
      "ATOM   1360  HZ3 LYS B 357       1.179  24.654  -2.769  1.00  4.46           H  \n",
      "ATOM   1361  N   GLU B 358      -2.343  25.048  -9.503  1.00  2.39           N  \n",
      "ATOM   1362  CA  GLU B 358      -1.632  25.848 -10.538  1.00  3.15           C  \n",
      "ATOM   1363  C   GLU B 358      -0.132  25.879 -10.214  1.00  3.39           C  \n",
      "ATOM   1364  O   GLU B 358       0.246  26.124  -9.086  1.00  3.42           O  \n",
      "ATOM   1365  CB  GLU B 358      -2.182  27.277 -10.541  1.00  3.87           C  \n",
      "ATOM   1366  CG  GLU B 358      -2.354  27.759 -11.983  1.00  4.49           C  \n",
      "ATOM   1367  CD  GLU B 358      -2.696  29.250 -11.985  1.00  5.22           C  \n",
      "ATOM   1368  OE1 GLU B 358      -2.030  29.991 -11.280  1.00  5.62           O  \n",
      "ATOM   1369  OE2 GLU B 358      -3.618  29.625 -12.690  1.00  5.68           O  \n",
      "ATOM   1370  H   GLU B 358      -3.144  25.411  -9.070  1.00  2.56           H  \n",
      "ATOM   1371  HA  GLU B 358      -1.790  25.399 -11.505  1.00  3.45           H  \n",
      "ATOM   1372  HB2 GLU B 358      -3.139  27.294 -10.039  1.00  4.19           H  \n",
      "ATOM   1373  HB3 GLU B 358      -1.493  27.929 -10.025  1.00  4.12           H  \n",
      "ATOM   1374  HG2 GLU B 358      -1.435  27.601 -12.527  1.00  4.60           H  \n",
      "ATOM   1375  HG3 GLU B 358      -3.152  27.208 -12.455  1.00  4.74           H  \n",
      "ATOM   1376  N   PRO B 359       0.682  25.632 -11.213  1.00  4.03           N  \n",
      "ATOM   1377  CA  PRO B 359       2.145  25.631 -11.045  1.00  4.70           C  \n",
      "ATOM   1378  C   PRO B 359       2.621  27.016 -10.599  1.00  4.87           C  \n",
      "ATOM   1379  O   PRO B 359       1.860  27.962 -10.561  1.00  4.90           O  \n",
      "ATOM   1380  CB  PRO B 359       2.702  25.286 -12.433  1.00  5.57           C  \n",
      "ATOM   1381  CG  PRO B 359       1.498  25.096 -13.390  1.00  5.54           C  \n",
      "ATOM   1382  CD  PRO B 359       0.216  25.338 -12.581  1.00  4.57           C  \n",
      "ATOM   1383  HA  PRO B 359       2.443  24.880 -10.331  1.00  4.80           H  \n",
      "ATOM   1384  HB2 PRO B 359       3.330  26.092 -12.786  1.00  5.99           H  \n",
      "ATOM   1385  HB3 PRO B 359       3.271  24.371 -12.383  1.00  6.05           H  \n",
      "ATOM   1386  HG2 PRO B 359       1.561  25.806 -14.202  1.00  6.00           H  \n",
      "ATOM   1387  HG3 PRO B 359       1.495  24.090 -13.780  1.00  5.99           H  \n",
      "ATOM   1388  HD2 PRO B 359      -0.325  26.181 -12.984  1.00  4.68           H  \n",
      "ATOM   1389  HD3 PRO B 359      -0.403  24.455 -12.580  1.00  4.50           H  \n",
      "ATOM   1390  N   GLY B 360       3.874  27.142 -10.260  1.00  5.39           N  \n",
      "ATOM   1391  CA  GLY B 360       4.396  28.465  -9.817  1.00  5.90           C  \n",
      "ATOM   1392  C   GLY B 360       5.886  28.563 -10.146  1.00  6.72           C  \n",
      "ATOM   1393  O   GLY B 360       6.291  27.987 -11.143  1.00  7.20           O  \n",
      "ATOM   1394  OXT GLY B 360       6.598  29.211  -9.397  1.00  7.11           O  \n",
      "ATOM   1395  H   GLY B 360       4.471  26.366 -10.297  1.00  5.67           H  \n",
      "ATOM   1396  HA2 GLY B 360       3.860  29.253 -10.328  1.00  5.94           H  \n",
      "ATOM   1397  HA3 GLY B 360       4.259  28.569  -8.751  1.00  5.99           H  \n",
      "TER    1398      GLY B 360                                                      \n",
      "ATOM   1399  N   LYS C 319      13.974 -25.061 -10.197  1.00  6.27           N  \n",
      "ATOM   1400  CA  LYS C 319      14.002 -24.415  -8.852  1.00  5.90           C  \n",
      "ATOM   1401  C   LYS C 319      15.140 -23.371  -8.789  1.00  5.22           C  \n",
      "ATOM   1402  O   LYS C 319      16.287 -23.720  -8.617  1.00  5.00           O  \n",
      "ATOM   1403  CB  LYS C 319      14.229 -25.494  -7.776  1.00  6.41           C  \n",
      "ATOM   1404  CG  LYS C 319      13.101 -26.543  -7.814  1.00  7.00           C  \n",
      "ATOM   1405  CD  LYS C 319      13.580 -27.821  -8.524  1.00  7.69           C  \n",
      "ATOM   1406  CE  LYS C 319      13.431 -29.029  -7.585  1.00  8.33           C  \n",
      "ATOM   1407  NZ  LYS C 319      12.707 -30.135  -8.285  1.00  8.96           N  \n",
      "ATOM   1408  H1  LYS C 319      14.750 -24.689 -10.780  1.00  6.53           H  \n",
      "ATOM   1409  H2  LYS C 319      14.084 -26.090 -10.088  1.00  6.51           H  \n",
      "ATOM   1410  H3  LYS C 319      13.066 -24.857 -10.661  1.00  6.39           H  \n",
      "ATOM   1411  HA  LYS C 319      13.057 -23.924  -8.673  1.00  6.14           H  \n",
      "ATOM   1412  HB2 LYS C 319      15.177 -25.980  -7.951  1.00  6.65           H  \n",
      "ATOM   1413  HB3 LYS C 319      14.243 -25.025  -6.803  1.00  6.48           H  \n",
      "ATOM   1414  HG2 LYS C 319      12.809 -26.788  -6.802  1.00  7.03           H  \n",
      "ATOM   1415  HG3 LYS C 319      12.251 -26.139  -8.343  1.00  7.21           H  \n",
      "ATOM   1416  HD2 LYS C 319      12.986 -27.982  -9.412  1.00  7.83           H  \n",
      "ATOM   1417  HD3 LYS C 319      14.618 -27.712  -8.801  1.00  7.86           H  \n",
      "ATOM   1418  HE2 LYS C 319      14.409 -29.377  -7.289  1.00  8.50           H  \n",
      "ATOM   1419  HE3 LYS C 319      12.874 -28.737  -6.707  1.00  8.41           H  \n",
      "ATOM   1420  HZ1 LYS C 319      12.305 -29.779  -9.176  1.00  9.22           H  \n",
      "ATOM   1421  HZ2 LYS C 319      13.370 -30.910  -8.488  1.00  9.18           H  \n",
      "ATOM   1422  HZ3 LYS C 319      11.942 -30.489  -7.676  1.00  9.15           H  \n",
      "ATOM   1423  N   LYS C 320      14.818 -22.095  -8.926  1.00  5.24           N  \n",
      "ATOM   1424  CA  LYS C 320      15.860 -21.002  -8.873  1.00  4.93           C  \n",
      "ATOM   1425  C   LYS C 320      16.927 -21.250  -9.942  1.00  4.31           C  \n",
      "ATOM   1426  O   LYS C 320      17.616 -22.248  -9.930  1.00  4.51           O  \n",
      "ATOM   1427  CB  LYS C 320      16.490 -20.933  -7.465  1.00  5.31           C  \n",
      "ATOM   1428  CG  LYS C 320      17.775 -21.775  -7.384  1.00  6.01           C  \n",
      "ATOM   1429  CD  LYS C 320      17.877 -22.450  -6.019  1.00  6.69           C  \n",
      "ATOM   1430  CE  LYS C 320      16.774 -23.502  -5.887  1.00  7.46           C  \n",
      "ATOM   1431  NZ  LYS C 320      17.330 -24.851  -6.197  1.00  8.15           N  \n",
      "ATOM   1432  H   LYS C 320      13.880 -21.852  -9.066  1.00  5.72           H  \n",
      "ATOM   1433  HA  LYS C 320      15.378 -20.057  -9.086  1.00  5.29           H  \n",
      "ATOM   1434  HB2 LYS C 320      16.729 -19.904  -7.237  1.00  5.50           H  \n",
      "ATOM   1435  HB3 LYS C 320      15.777 -21.299  -6.741  1.00  5.30           H  \n",
      "ATOM   1436  HG2 LYS C 320      17.761 -22.531  -8.139  1.00  6.15           H  \n",
      "ATOM   1437  HG3 LYS C 320      18.632 -21.137  -7.531  1.00  6.27           H  \n",
      "ATOM   1438  HD2 LYS C 320      18.842 -22.928  -5.932  1.00  6.87           H  \n",
      "ATOM   1439  HD3 LYS C 320      17.768 -21.712  -5.240  1.00  6.79           H  \n",
      "ATOM   1440  HE2 LYS C 320      16.389 -23.495  -4.879  1.00  7.62           H  \n",
      "ATOM   1441  HE3 LYS C 320      15.975 -23.275  -6.578  1.00  7.64           H  \n",
      "ATOM   1442  HZ1 LYS C 320      17.701 -24.858  -7.170  1.00  8.51           H  \n",
      "ATOM   1443  HZ2 LYS C 320      18.098 -25.071  -5.532  1.00  8.39           H  \n",
      "ATOM   1444  HZ3 LYS C 320      16.581 -25.565  -6.106  1.00  8.28           H  \n",
      "ATOM   1445  N   LYS C 321      17.063 -20.383 -10.900  1.00  3.98           N  \n",
      "ATOM   1446  CA  LYS C 321      18.065 -20.682 -11.938  1.00  3.74           C  \n",
      "ATOM   1447  C   LYS C 321      18.463 -19.382 -12.734  1.00  3.33           C  \n",
      "ATOM   1448  O   LYS C 321      18.856 -18.440 -12.082  1.00  3.47           O  \n",
      "ATOM   1449  CB  LYS C 321      17.479 -21.879 -12.718  1.00  4.29           C  \n",
      "ATOM   1450  CG  LYS C 321      15.930 -21.806 -12.774  1.00  4.88           C  \n",
      "ATOM   1451  CD  LYS C 321      15.435 -20.378 -13.070  1.00  5.72           C  \n",
      "ATOM   1452  CE  LYS C 321      14.483 -19.898 -11.961  1.00  6.54           C  \n",
      "ATOM   1453  NZ  LYS C 321      14.766 -18.463 -11.647  1.00  7.15           N  \n",
      "ATOM   1454  H   LYS C 321      16.502 -19.586 -10.952  1.00  4.23           H  \n",
      "ATOM   1455  HA  LYS C 321      18.957 -21.026 -11.431  1.00  4.04           H  \n",
      "ATOM   1456  HB2 LYS C 321      17.890 -21.948 -13.694  1.00  4.61           H  \n",
      "ATOM   1457  HB3 LYS C 321      17.746 -22.779 -12.181  1.00  4.47           H  \n",
      "ATOM   1458  HG2 LYS C 321      15.573 -22.469 -13.547  1.00  4.96           H  \n",
      "ATOM   1459  HG3 LYS C 321      15.529 -22.128 -11.823  1.00  5.09           H  \n",
      "ATOM   1460  HD2 LYS C 321      16.273 -19.709 -13.134  1.00  5.99           H  \n",
      "ATOM   1461  HD3 LYS C 321      14.906 -20.376 -14.011  1.00  5.83           H  \n",
      "ATOM   1462  HE2 LYS C 321      13.461 -19.996 -12.295  1.00  6.76           H  \n",
      "ATOM   1463  HE3 LYS C 321      14.630 -20.495 -11.073  1.00  6.82           H  \n",
      "ATOM   1464  HZ1 LYS C 321      15.529 -18.117 -12.262  1.00  7.31           H  \n",
      "ATOM   1465  HZ2 LYS C 321      13.907 -17.896 -11.805  1.00  7.42           H  \n",
      "ATOM   1466  HZ3 LYS C 321      15.060 -18.376 -10.653  1.00  7.42           H  \n",
      "ATOM   1467  N   PRO C 322      18.423 -19.326 -14.072  1.00  3.32           N  \n",
      "ATOM   1468  CA  PRO C 322      18.850 -18.109 -14.830  1.00  3.43           C  \n",
      "ATOM   1469  C   PRO C 322      18.428 -16.739 -14.231  1.00  2.69           C  \n",
      "ATOM   1470  O   PRO C 322      18.591 -16.468 -13.060  1.00  2.47           O  \n",
      "ATOM   1471  CB  PRO C 322      18.261 -18.308 -16.231  1.00  4.12           C  \n",
      "ATOM   1472  CG  PRO C 322      17.899 -19.798 -16.366  1.00  4.35           C  \n",
      "ATOM   1473  CD  PRO C 322      17.976 -20.418 -14.967  1.00  3.79           C  \n",
      "ATOM   1474  HA  PRO C 322      19.925 -18.123 -14.919  1.00  3.97           H  \n",
      "ATOM   1475  HB2 PRO C 322      17.371 -17.706 -16.344  1.00  4.05           H  \n",
      "ATOM   1476  HB3 PRO C 322      18.989 -18.041 -16.981  1.00  4.85           H  \n",
      "ATOM   1477  HG2 PRO C 322      16.897 -19.898 -16.760  1.00  4.53           H  \n",
      "ATOM   1478  HG3 PRO C 322      18.603 -20.289 -17.021  1.00  5.05           H  \n",
      "ATOM   1479  HD2 PRO C 322      17.007 -20.784 -14.678  1.00  3.73           H  \n",
      "ATOM   1480  HD3 PRO C 322      18.705 -21.217 -14.965  1.00  4.20           H  \n",
      "ATOM   1481  N   LEU C 323      17.996 -15.835 -15.085  1.00  2.71           N  \n",
      "ATOM   1482  CA  LEU C 323      17.679 -14.429 -14.662  1.00  2.32           C  \n",
      "ATOM   1483  C   LEU C 323      16.385 -14.292 -13.852  1.00  1.85           C  \n",
      "ATOM   1484  O   LEU C 323      15.324 -14.717 -14.264  1.00  2.13           O  \n",
      "ATOM   1485  CB  LEU C 323      17.498 -13.579 -15.923  1.00  3.06           C  \n",
      "ATOM   1486  CG  LEU C 323      18.510 -13.984 -16.995  1.00  3.73           C  \n",
      "ATOM   1487  CD1 LEU C 323      18.517 -12.929 -18.101  1.00  4.63           C  \n",
      "ATOM   1488  CD2 LEU C 323      19.904 -14.095 -16.375  1.00  3.78           C  \n",
      "ATOM   1489  H   LEU C 323      17.964 -16.057 -16.033  1.00  3.24           H  \n",
      "ATOM   1490  HA  LEU C 323      18.502 -14.031 -14.093  1.00  2.30           H  \n",
      "ATOM   1491  HB2 LEU C 323      16.498 -13.722 -16.306  1.00  3.39           H  \n",
      "ATOM   1492  HB3 LEU C 323      17.638 -12.537 -15.675  1.00  3.08           H  \n",
      "ATOM   1493  HG  LEU C 323      18.224 -14.936 -17.418  1.00  3.82           H  \n",
      "ATOM   1494 HD11 LEU C 323      18.750 -11.964 -17.677  1.00  4.83           H  \n",
      "ATOM   1495 HD12 LEU C 323      19.259 -13.188 -18.842  1.00  5.14           H  \n",
      "ATOM   1496 HD13 LEU C 323      17.542 -12.892 -18.566  1.00  4.89           H  \n",
      "ATOM   1497 HD21 LEU C 323      19.924 -13.561 -15.437  1.00  3.74           H  \n",
      "ATOM   1498 HD22 LEU C 323      20.140 -15.135 -16.203  1.00  3.95           H  \n",
      "ATOM   1499 HD23 LEU C 323      20.633 -13.668 -17.049  1.00  4.22           H  \n",
      "ATOM   1500  N   ASP C 324      16.471 -13.613 -12.733  1.00  1.52           N  \n",
      "ATOM   1501  CA  ASP C 324      15.260 -13.334 -11.907  1.00  1.53           C  \n",
      "ATOM   1502  C   ASP C 324      14.668 -11.994 -12.375  1.00  1.21           C  \n",
      "ATOM   1503  O   ASP C 324      14.992 -11.507 -13.440  1.00  1.15           O  \n",
      "ATOM   1504  CB  ASP C 324      15.655 -13.224 -10.432  1.00  1.97           C  \n",
      "ATOM   1505  CG  ASP C 324      16.329 -14.520  -9.979  1.00  2.39           C  \n",
      "ATOM   1506  OD1 ASP C 324      17.250 -14.952 -10.651  1.00  2.70           O  \n",
      "ATOM   1507  OD2 ASP C 324      15.914 -15.056  -8.964  1.00  2.85           O  \n",
      "ATOM   1508  H   ASP C 324      17.334 -13.233 -12.465  1.00  1.65           H  \n",
      "ATOM   1509  HA  ASP C 324      14.534 -14.124 -12.038  1.00  1.89           H  \n",
      "ATOM   1510  HB2 ASP C 324      16.341 -12.398 -10.306  1.00  2.01           H  \n",
      "ATOM   1511  HB3 ASP C 324      14.772 -13.051  -9.835  1.00  2.31           H  \n",
      "ATOM   1512  N   GLY C 325      13.829 -11.377 -11.583  1.00  1.12           N  \n",
      "ATOM   1513  CA  GLY C 325      13.256 -10.053 -11.986  1.00  0.94           C  \n",
      "ATOM   1514  C   GLY C 325      14.332  -8.967 -11.831  1.00  0.75           C  \n",
      "ATOM   1515  O   GLY C 325      15.390  -9.212 -11.287  1.00  0.72           O  \n",
      "ATOM   1516  H   GLY C 325      13.592 -11.770 -10.717  1.00  1.27           H  \n",
      "ATOM   1517  HA2 GLY C 325      12.931 -10.098 -13.016  1.00  0.98           H  \n",
      "ATOM   1518  HA3 GLY C 325      12.415  -9.815 -11.352  1.00  1.05           H  \n",
      "ATOM   1519  N   GLU C 326      14.074  -7.766 -12.291  1.00  0.70           N  \n",
      "ATOM   1520  CA  GLU C 326      15.096  -6.683 -12.145  1.00  0.60           C  \n",
      "ATOM   1521  C   GLU C 326      15.422  -6.506 -10.659  1.00  0.51           C  \n",
      "ATOM   1522  O   GLU C 326      14.967  -7.265  -9.827  1.00  0.50           O  \n",
      "ATOM   1523  CB  GLU C 326      14.551  -5.367 -12.711  1.00  0.71           C  \n",
      "ATOM   1524  CG  GLU C 326      14.626  -5.395 -14.241  1.00  0.88           C  \n",
      "ATOM   1525  CD  GLU C 326      15.694  -4.410 -14.722  1.00  1.40           C  \n",
      "ATOM   1526  OE1 GLU C 326      15.772  -3.330 -14.160  1.00  2.11           O  \n",
      "ATOM   1527  OE2 GLU C 326      16.414  -4.752 -15.646  1.00  2.03           O  \n",
      "ATOM   1528  H   GLU C 326      13.214  -7.577 -12.720  1.00  0.79           H  \n",
      "ATOM   1529  HA  GLU C 326      15.994  -6.962 -12.678  1.00  0.62           H  \n",
      "ATOM   1530  HB2 GLU C 326      13.524  -5.239 -12.402  1.00  0.78           H  \n",
      "ATOM   1531  HB3 GLU C 326      15.144  -4.544 -12.341  1.00  0.71           H  \n",
      "ATOM   1532  HG2 GLU C 326      14.881  -6.392 -14.570  1.00  1.37           H  \n",
      "ATOM   1533  HG3 GLU C 326      13.668  -5.114 -14.653  1.00  1.29           H  \n",
      "ATOM   1534  N   TYR C 327      16.208  -5.521 -10.308  1.00  0.47           N  \n",
      "ATOM   1535  CA  TYR C 327      16.545  -5.336  -8.867  1.00  0.41           C  \n",
      "ATOM   1536  C   TYR C 327      16.817  -3.861  -8.558  1.00  0.38           C  \n",
      "ATOM   1537  O   TYR C 327      17.869  -3.334  -8.862  1.00  0.43           O  \n",
      "ATOM   1538  CB  TYR C 327      17.786  -6.163  -8.529  1.00  0.43           C  \n",
      "ATOM   1539  CG  TYR C 327      17.389  -7.612  -8.384  1.00  0.47           C  \n",
      "ATOM   1540  CD1 TYR C 327      16.907  -8.086  -7.156  1.00  0.47           C  \n",
      "ATOM   1541  CD2 TYR C 327      17.498  -8.482  -9.479  1.00  0.53           C  \n",
      "ATOM   1542  CE1 TYR C 327      16.535  -9.432  -7.021  1.00  0.52           C  \n",
      "ATOM   1543  CE2 TYR C 327      17.125  -9.827  -9.345  1.00  0.58           C  \n",
      "ATOM   1544  CZ  TYR C 327      16.643 -10.302  -8.116  1.00  0.57           C  \n",
      "ATOM   1545  OH  TYR C 327      16.276 -11.626  -7.985  1.00  0.64           O  \n",
      "ATOM   1546  H   TYR C 327      16.576  -4.916 -10.985  1.00  0.50           H  \n",
      "ATOM   1547  HA  TYR C 327      15.718  -5.677  -8.263  1.00  0.40           H  \n",
      "ATOM   1548  HB2 TYR C 327      18.513  -6.064  -9.322  1.00  0.47           H  \n",
      "ATOM   1549  HB3 TYR C 327      18.213  -5.813  -7.602  1.00  0.43           H  \n",
      "ATOM   1550  HD1 TYR C 327      16.824  -7.416  -6.313  1.00  0.45           H  \n",
      "ATOM   1551  HD2 TYR C 327      17.868  -8.115 -10.426  1.00  0.57           H  \n",
      "ATOM   1552  HE1 TYR C 327      16.163  -9.797  -6.075  1.00  0.56           H  \n",
      "ATOM   1553  HE2 TYR C 327      17.208 -10.497 -10.188  1.00  0.64           H  \n",
      "ATOM   1554  HH  TYR C 327      16.640 -11.954  -7.159  1.00  1.01           H  \n",
      "ATOM   1555  N   PHE C 328      15.879  -3.197  -7.937  1.00  0.35           N  \n",
      "ATOM   1556  CA  PHE C 328      16.080  -1.763  -7.584  1.00  0.34           C  \n",
      "ATOM   1557  C   PHE C 328      16.149  -1.629  -6.059  1.00  0.31           C  \n",
      "ATOM   1558  O   PHE C 328      16.351  -2.596  -5.353  1.00  0.32           O  \n",
      "ATOM   1559  CB  PHE C 328      14.910  -0.934  -8.118  1.00  0.36           C  \n",
      "ATOM   1560  CG  PHE C 328      14.813  -1.109  -9.615  1.00  0.37           C  \n",
      "ATOM   1561  CD1 PHE C 328      15.513  -0.244 -10.469  1.00  0.43           C  \n",
      "ATOM   1562  CD2 PHE C 328      14.023  -2.136 -10.151  1.00  0.43           C  \n",
      "ATOM   1563  CE1 PHE C 328      15.422  -0.407 -11.860  1.00  0.49           C  \n",
      "ATOM   1564  CE2 PHE C 328      13.932  -2.299 -11.541  1.00  0.48           C  \n",
      "ATOM   1565  CZ  PHE C 328      14.632  -1.434 -12.396  1.00  0.48           C  \n",
      "ATOM   1566  H   PHE C 328      15.045  -3.648  -7.691  1.00  0.36           H  \n",
      "ATOM   1567  HA  PHE C 328      17.003  -1.409  -8.020  1.00  0.36           H  \n",
      "ATOM   1568  HB2 PHE C 328      13.993  -1.267  -7.656  1.00  0.38           H  \n",
      "ATOM   1569  HB3 PHE C 328      15.072   0.108  -7.888  1.00  0.39           H  \n",
      "ATOM   1570  HD1 PHE C 328      16.122   0.547 -10.057  1.00  0.50           H  \n",
      "ATOM   1571  HD2 PHE C 328      13.484  -2.802  -9.492  1.00  0.50           H  \n",
      "ATOM   1572  HE1 PHE C 328      15.961   0.259 -12.518  1.00  0.58           H  \n",
      "ATOM   1573  HE2 PHE C 328      13.323  -3.090 -11.953  1.00  0.57           H  \n",
      "ATOM   1574  HZ  PHE C 328      14.562  -1.560 -13.466  1.00  0.54           H  \n",
      "ATOM   1575  N   THR C 329      15.979  -0.441  -5.544  1.00  0.31           N  \n",
      "ATOM   1576  CA  THR C 329      16.034  -0.257  -4.065  1.00  0.32           C  \n",
      "ATOM   1577  C   THR C 329      15.225   0.981  -3.674  1.00  0.32           C  \n",
      "ATOM   1578  O   THR C 329      14.830   1.765  -4.514  1.00  0.38           O  \n",
      "ATOM   1579  CB  THR C 329      17.488  -0.079  -3.625  1.00  0.35           C  \n",
      "ATOM   1580  OG1 THR C 329      18.116   0.892  -4.451  1.00  0.47           O  \n",
      "ATOM   1581  CG2 THR C 329      18.226  -1.413  -3.749  1.00  0.38           C  \n",
      "ATOM   1582  H   THR C 329      15.814   0.328  -6.128  1.00  0.32           H  \n",
      "ATOM   1583  HA  THR C 329      15.616  -1.127  -3.580  1.00  0.33           H  \n",
      "ATOM   1584  HB  THR C 329      17.517   0.249  -2.597  1.00  0.44           H  \n",
      "ATOM   1585  HG1 THR C 329      18.299   0.484  -5.301  1.00  1.16           H  \n",
      "ATOM   1586 HG21 THR C 329      17.545  -2.222  -3.530  1.00  1.15           H  \n",
      "ATOM   1587 HG22 THR C 329      18.605  -1.522  -4.754  1.00  1.11           H  \n",
      "ATOM   1588 HG23 THR C 329      19.048  -1.436  -3.049  1.00  1.03           H  \n",
      "ATOM   1589  N   LEU C 330      14.972   1.165  -2.406  1.00  0.29           N  \n",
      "ATOM   1590  CA  LEU C 330      14.185   2.355  -1.973  1.00  0.29           C  \n",
      "ATOM   1591  C   LEU C 330      14.584   2.754  -0.550  1.00  0.28           C  \n",
      "ATOM   1592  O   LEU C 330      14.596   1.942   0.354  1.00  0.27           O  \n",
      "ATOM   1593  CB  LEU C 330      12.694   2.012  -2.007  1.00  0.30           C  \n",
      "ATOM   1594  CG  LEU C 330      11.878   3.228  -1.567  1.00  0.33           C  \n",
      "ATOM   1595  CD1 LEU C 330      11.900   4.285  -2.672  1.00  0.42           C  \n",
      "ATOM   1596  CD2 LEU C 330      10.432   2.800  -1.301  1.00  0.37           C  \n",
      "ATOM   1597  H   LEU C 330      15.296   0.522  -1.741  1.00  0.27           H  \n",
      "ATOM   1598  HA  LEU C 330      14.379   3.178  -2.645  1.00  0.33           H  \n",
      "ATOM   1599  HB2 LEU C 330      12.412   1.733  -3.012  1.00  0.36           H  \n",
      "ATOM   1600  HB3 LEU C 330      12.499   1.189  -1.336  1.00  0.30           H  \n",
      "ATOM   1601  HG  LEU C 330      12.305   3.641  -0.664  1.00  0.36           H  \n",
      "ATOM   1602 HD11 LEU C 330      12.889   4.335  -3.103  1.00  1.07           H  \n",
      "ATOM   1603 HD12 LEU C 330      11.187   4.020  -3.438  1.00  1.04           H  \n",
      "ATOM   1604 HD13 LEU C 330      11.640   5.247  -2.256  1.00  1.17           H  \n",
      "ATOM   1605 HD21 LEU C 330      10.421   1.793  -0.912  1.00  1.02           H  \n",
      "ATOM   1606 HD22 LEU C 330       9.985   3.469  -0.581  1.00  1.09           H  \n",
      "ATOM   1607 HD23 LEU C 330       9.871   2.837  -2.223  1.00  1.15           H  \n",
      "ATOM   1608  N   GLN C 331      14.900   4.004  -0.344  1.00  0.32           N  \n",
      "ATOM   1609  CA  GLN C 331      15.287   4.463   1.021  1.00  0.33           C  \n",
      "ATOM   1610  C   GLN C 331      14.029   4.882   1.785  1.00  0.31           C  \n",
      "ATOM   1611  O   GLN C 331      13.374   5.844   1.438  1.00  0.33           O  \n",
      "ATOM   1612  CB  GLN C 331      16.237   5.657   0.908  1.00  0.42           C  \n",
      "ATOM   1613  CG  GLN C 331      16.730   6.053   2.301  1.00  0.50           C  \n",
      "ATOM   1614  CD  GLN C 331      18.241   5.835   2.390  1.00  0.86           C  \n",
      "ATOM   1615  OE1 GLN C 331      19.008   6.567   1.797  1.00  1.81           O  \n",
      "ATOM   1616  NE2 GLN C 331      18.705   4.851   3.111  1.00  0.86           N  \n",
      "ATOM   1617  H   GLN C 331      14.876   4.644  -1.086  1.00  0.35           H  \n",
      "ATOM   1618  HA  GLN C 331      15.778   3.658   1.547  1.00  0.34           H  \n",
      "ATOM   1619  HB2 GLN C 331      17.082   5.387   0.290  1.00  0.50           H  \n",
      "ATOM   1620  HB3 GLN C 331      15.717   6.491   0.461  1.00  0.49           H  \n",
      "ATOM   1621  HG2 GLN C 331      16.504   7.094   2.480  1.00  0.67           H  \n",
      "ATOM   1622  HG3 GLN C 331      16.236   5.444   3.044  1.00  0.86           H  \n",
      "ATOM   1623 HE21 GLN C 331      18.086   4.260   3.589  1.00  1.28           H  \n",
      "ATOM   1624 HE22 GLN C 331      19.671   4.703   3.175  1.00  1.14           H  \n",
      "ATOM   1625  N   ILE C 332      13.682   4.166   2.820  1.00  0.29           N  \n",
      "ATOM   1626  CA  ILE C 332      12.461   4.529   3.594  1.00  0.29           C  \n",
      "ATOM   1627  C   ILE C 332      12.858   5.217   4.902  1.00  0.30           C  \n",
      "ATOM   1628  O   ILE C 332      13.400   4.603   5.799  1.00  0.31           O  \n",
      "ATOM   1629  CB  ILE C 332      11.663   3.262   3.909  1.00  0.32           C  \n",
      "ATOM   1630  CG1 ILE C 332      11.187   2.624   2.603  1.00  0.33           C  \n",
      "ATOM   1631  CG2 ILE C 332      10.451   3.621   4.770  1.00  0.42           C  \n",
      "ATOM   1632  CD1 ILE C 332      10.392   1.355   2.916  1.00  0.29           C  \n",
      "ATOM   1633  H   ILE C 332      14.219   3.391   3.084  1.00  0.31           H  \n",
      "ATOM   1634  HA  ILE C 332      11.850   5.199   3.008  1.00  0.31           H  \n",
      "ATOM   1635  HB  ILE C 332      12.291   2.565   4.444  1.00  0.37           H  \n",
      "ATOM   1636 HG12 ILE C 332      10.557   3.322   2.070  1.00  0.40           H  \n",
      "ATOM   1637 HG13 ILE C 332      12.040   2.370   1.993  1.00  0.43           H  \n",
      "ATOM   1638 HG21 ILE C 332       9.972   4.502   4.368  1.00  1.10           H  \n",
      "ATOM   1639 HG22 ILE C 332       9.751   2.798   4.767  1.00  1.13           H  \n",
      "ATOM   1640 HG23 ILE C 332      10.773   3.816   5.782  1.00  1.11           H  \n",
      "ATOM   1641 HD11 ILE C 332      10.532   1.090   3.954  1.00  1.00           H  \n",
      "ATOM   1642 HD12 ILE C 332       9.343   1.531   2.728  1.00  1.07           H  \n",
      "ATOM   1643 HD13 ILE C 332      10.740   0.548   2.289  1.00  1.06           H  \n",
      "ATOM   1644  N   ARG C 333      12.583   6.488   5.020  1.00  0.33           N  \n",
      "ATOM   1645  CA  ARG C 333      12.934   7.214   6.272  1.00  0.37           C  \n",
      "ATOM   1646  C   ARG C 333      11.972   6.795   7.386  1.00  0.37           C  \n",
      "ATOM   1647  O   ARG C 333      10.810   6.533   7.148  1.00  0.40           O  \n",
      "ATOM   1648  CB  ARG C 333      12.815   8.721   6.039  1.00  0.43           C  \n",
      "ATOM   1649  CG  ARG C 333      13.792   9.147   4.942  1.00  0.46           C  \n",
      "ATOM   1650  CD  ARG C 333      13.118  10.167   4.024  1.00  0.91           C  \n",
      "ATOM   1651  NE  ARG C 333      14.125  11.167   3.569  1.00  1.25           N  \n",
      "ATOM   1652  CZ  ARG C 333      13.861  11.951   2.558  1.00  1.59           C  \n",
      "ATOM   1653  NH1 ARG C 333      12.714  11.864   1.940  1.00  2.00           N  \n",
      "ATOM   1654  NH2 ARG C 333      14.746  12.825   2.164  1.00  2.35           N  \n",
      "ATOM   1655  H   ARG C 333      12.140   6.963   4.287  1.00  0.36           H  \n",
      "ATOM   1656  HA  ARG C 333      13.947   6.971   6.559  1.00  0.38           H  \n",
      "ATOM   1657  HB2 ARG C 333      11.806   8.960   5.736  1.00  0.50           H  \n",
      "ATOM   1658  HB3 ARG C 333      13.051   9.247   6.951  1.00  0.56           H  \n",
      "ATOM   1659  HG2 ARG C 333      14.664   9.593   5.394  1.00  0.85           H  \n",
      "ATOM   1660  HG3 ARG C 333      14.085   8.283   4.365  1.00  0.81           H  \n",
      "ATOM   1661  HD2 ARG C 333      12.702   9.659   3.166  1.00  1.66           H  \n",
      "ATOM   1662  HD3 ARG C 333      12.329  10.670   4.562  1.00  1.54           H  \n",
      "ATOM   1663  HE  ARG C 333      14.988  11.236   4.029  1.00  1.93           H  \n",
      "ATOM   1664 HH11 ARG C 333      12.033  11.196   2.238  1.00  2.11           H  \n",
      "ATOM   1665 HH12 ARG C 333      12.516  12.466   1.167  1.00  2.64           H  \n",
      "ATOM   1666 HH21 ARG C 333      15.625  12.894   2.635  1.00  2.82           H  \n",
      "ATOM   1667 HH22 ARG C 333      14.545  13.426   1.390  1.00  2.75           H  \n",
      "ATOM   1668  N   GLY C 334      12.443   6.732   8.601  1.00  0.40           N  \n",
      "ATOM   1669  CA  GLY C 334      11.549   6.333   9.726  1.00  0.41           C  \n",
      "ATOM   1670  C   GLY C 334      11.454   4.808   9.793  1.00  0.37           C  \n",
      "ATOM   1671  O   GLY C 334      11.176   4.149   8.811  1.00  0.35           O  \n",
      "ATOM   1672  H   GLY C 334      13.383   6.949   8.775  1.00  0.45           H  \n",
      "ATOM   1673  HA2 GLY C 334      11.950   6.712  10.655  1.00  0.45           H  \n",
      "ATOM   1674  HA3 GLY C 334      10.564   6.745   9.565  1.00  0.43           H  \n",
      "ATOM   1675  N   ARG C 335      11.683   4.242  10.946  1.00  0.41           N  \n",
      "ATOM   1676  CA  ARG C 335      11.605   2.760  11.080  1.00  0.42           C  \n",
      "ATOM   1677  C   ARG C 335      10.137   2.331  11.160  1.00  0.40           C  \n",
      "ATOM   1678  O   ARG C 335       9.768   1.262  10.717  1.00  0.39           O  \n",
      "ATOM   1679  CB  ARG C 335      12.335   2.326  12.353  1.00  0.50           C  \n",
      "ATOM   1680  CG  ARG C 335      12.218   0.810  12.520  1.00  0.62           C  \n",
      "ATOM   1681  CD  ARG C 335      12.390   0.446  13.996  1.00  1.14           C  \n",
      "ATOM   1682  NE  ARG C 335      12.381  -1.037  14.150  1.00  1.44           N  \n",
      "ATOM   1683  CZ  ARG C 335      12.804  -1.584  15.259  1.00  1.88           C  \n",
      "ATOM   1684  NH1 ARG C 335      13.231  -0.835  16.241  1.00  2.34           N  \n",
      "ATOM   1685  NH2 ARG C 335      12.798  -2.882  15.387  1.00  2.59           N  \n",
      "ATOM   1686  H   ARG C 335      11.905   4.793  11.726  1.00  0.46           H  \n",
      "ATOM   1687  HA  ARG C 335      12.069   2.296  10.222  1.00  0.41           H  \n",
      "ATOM   1688  HB2 ARG C 335      13.377   2.602  12.281  1.00  0.55           H  \n",
      "ATOM   1689  HB3 ARG C 335      11.891   2.815  13.207  1.00  0.52           H  \n",
      "ATOM   1690  HG2 ARG C 335      11.247   0.483  12.178  1.00  0.99           H  \n",
      "ATOM   1691  HG3 ARG C 335      12.987   0.323  11.940  1.00  1.03           H  \n",
      "ATOM   1692  HD2 ARG C 335      13.329   0.839  14.357  1.00  1.84           H  \n",
      "ATOM   1693  HD3 ARG C 335      11.579   0.872  14.568  1.00  1.77           H  \n",
      "ATOM   1694  HE  ARG C 335      12.060  -1.602  13.417  1.00  2.02           H  \n",
      "ATOM   1695 HH11 ARG C 335      13.236   0.160  16.148  1.00  2.38           H  \n",
      "ATOM   1696 HH12 ARG C 335      13.553  -1.258  17.087  1.00  3.03           H  \n",
      "ATOM   1697 HH21 ARG C 335      12.471  -3.458  14.638  1.00  2.96           H  \n",
      "ATOM   1698 HH22 ARG C 335      13.121  -3.302  16.236  1.00  3.08           H  \n",
      "ATOM   1699  N   GLU C 336       9.299   3.156  11.727  1.00  0.44           N  \n",
      "ATOM   1700  CA  GLU C 336       7.857   2.795  11.842  1.00  0.46           C  \n",
      "ATOM   1701  C   GLU C 336       7.275   2.524  10.452  1.00  0.40           C  \n",
      "ATOM   1702  O   GLU C 336       6.893   1.415  10.134  1.00  0.38           O  \n",
      "ATOM   1703  CB  GLU C 336       7.094   3.949  12.497  1.00  0.53           C  \n",
      "ATOM   1704  CG  GLU C 336       6.792   3.598  13.955  1.00  1.19           C  \n",
      "ATOM   1705  CD  GLU C 336       6.133   4.794  14.645  1.00  1.86           C  \n",
      "ATOM   1706  OE1 GLU C 336       6.772   5.830  14.729  1.00  2.39           O  \n",
      "ATOM   1707  OE2 GLU C 336       5.001   4.653  15.077  1.00  2.54           O  \n",
      "ATOM   1708  H   GLU C 336       9.619   4.012  12.081  1.00  0.47           H  \n",
      "ATOM   1709  HA  GLU C 336       7.756   1.909  12.451  1.00  0.49           H  \n",
      "ATOM   1710  HB2 GLU C 336       7.696   4.846  12.459  1.00  0.86           H  \n",
      "ATOM   1711  HB3 GLU C 336       6.167   4.115  11.969  1.00  0.98           H  \n",
      "ATOM   1712  HG2 GLU C 336       6.125   2.749  13.990  1.00  1.68           H  \n",
      "ATOM   1713  HG3 GLU C 336       7.712   3.354  14.465  1.00  1.72           H  \n",
      "ATOM   1714  N   ARG C 337       7.198   3.528   9.621  1.00  0.41           N  \n",
      "ATOM   1715  CA  ARG C 337       6.633   3.322   8.258  1.00  0.39           C  \n",
      "ATOM   1716  C   ARG C 337       7.458   2.281   7.502  1.00  0.33           C  \n",
      "ATOM   1717  O   ARG C 337       6.967   1.620   6.610  1.00  0.30           O  \n",
      "ATOM   1718  CB  ARG C 337       6.638   4.647   7.492  1.00  0.47           C  \n",
      "ATOM   1719  CG  ARG C 337       5.426   5.475   7.920  1.00  0.67           C  \n",
      "ATOM   1720  CD  ARG C 337       5.820   6.946   8.059  1.00  1.24           C  \n",
      "ATOM   1721  NE  ARG C 337       4.638   7.803   7.761  1.00  1.75           N  \n",
      "ATOM   1722  CZ  ARG C 337       4.623   9.057   8.125  1.00  2.56           C  \n",
      "ATOM   1723  NH1 ARG C 337       5.634   9.567   8.775  1.00  2.94           N  \n",
      "ATOM   1724  NH2 ARG C 337       3.589   9.801   7.843  1.00  3.39           N  \n",
      "ATOM   1725  H   ARG C 337       7.506   4.417   9.896  1.00  0.46           H  \n",
      "ATOM   1726  HA  ARG C 337       5.615   2.969   8.347  1.00  0.41           H  \n",
      "ATOM   1727  HB2 ARG C 337       7.546   5.189   7.715  1.00  0.53           H  \n",
      "ATOM   1728  HB3 ARG C 337       6.584   4.452   6.432  1.00  0.52           H  \n",
      "ATOM   1729  HG2 ARG C 337       4.648   5.381   7.177  1.00  1.41           H  \n",
      "ATOM   1730  HG3 ARG C 337       5.061   5.112   8.869  1.00  1.27           H  \n",
      "ATOM   1731  HD2 ARG C 337       6.159   7.133   9.069  1.00  1.87           H  \n",
      "ATOM   1732  HD3 ARG C 337       6.612   7.176   7.363  1.00  1.97           H  \n",
      "ATOM   1733  HE  ARG C 337       3.870   7.425   7.283  1.00  1.98           H  \n",
      "ATOM   1734 HH11 ARG C 337       6.426   9.000   8.997  1.00  2.70           H  \n",
      "ATOM   1735 HH12 ARG C 337       5.615  10.527   9.054  1.00  3.74           H  \n",
      "ATOM   1736 HH21 ARG C 337       2.811   9.411   7.350  1.00  3.50           H  \n",
      "ATOM   1737 HH22 ARG C 337       3.574  10.762   8.119  1.00  4.11           H  \n",
      "ATOM   1738  N   PHE C 338       8.704   2.117   7.852  1.00  0.33           N  \n",
      "ATOM   1739  CA  PHE C 338       9.538   1.105   7.147  1.00  0.31           C  \n",
      "ATOM   1740  C   PHE C 338       8.839  -0.253   7.222  1.00  0.29           C  \n",
      "ATOM   1741  O   PHE C 338       8.559  -0.874   6.218  1.00  0.28           O  \n",
      "ATOM   1742  CB  PHE C 338      10.911   1.009   7.816  1.00  0.36           C  \n",
      "ATOM   1743  CG  PHE C 338      11.659  -0.179   7.255  1.00  0.35           C  \n",
      "ATOM   1744  CD1 PHE C 338      12.165  -0.135   5.948  1.00  0.36           C  \n",
      "ATOM   1745  CD2 PHE C 338      11.843  -1.327   8.040  1.00  0.42           C  \n",
      "ATOM   1746  CE1 PHE C 338      12.855  -1.238   5.425  1.00  0.41           C  \n",
      "ATOM   1747  CE2 PHE C 338      12.534  -2.431   7.518  1.00  0.47           C  \n",
      "ATOM   1748  CZ  PHE C 338      13.040  -2.386   6.210  1.00  0.45           C  \n",
      "ATOM   1749  H   PHE C 338       9.086   2.652   8.578  1.00  0.37           H  \n",
      "ATOM   1750  HA  PHE C 338       9.658   1.392   6.113  1.00  0.30           H  \n",
      "ATOM   1751  HB2 PHE C 338      11.471   1.912   7.621  1.00  0.40           H  \n",
      "ATOM   1752  HB3 PHE C 338      10.786   0.884   8.881  1.00  0.41           H  \n",
      "ATOM   1753  HD1 PHE C 338      12.023   0.749   5.343  1.00  0.40           H  \n",
      "ATOM   1754  HD2 PHE C 338      11.454  -1.361   9.047  1.00  0.49           H  \n",
      "ATOM   1755  HE1 PHE C 338      13.245  -1.204   4.418  1.00  0.47           H  \n",
      "ATOM   1756  HE2 PHE C 338      12.676  -3.315   8.122  1.00  0.57           H  \n",
      "ATOM   1757  HZ  PHE C 338      13.572  -3.236   5.808  1.00  0.52           H  \n",
      "ATOM   1758  N   GLU C 339       8.556  -0.717   8.408  1.00  0.32           N  \n",
      "ATOM   1759  CA  GLU C 339       7.874  -2.033   8.555  1.00  0.34           C  \n",
      "ATOM   1760  C   GLU C 339       6.588  -2.045   7.724  1.00  0.31           C  \n",
      "ATOM   1761  O   GLU C 339       6.186  -3.065   7.200  1.00  0.32           O  \n",
      "ATOM   1762  CB  GLU C 339       7.528  -2.266  10.027  1.00  0.40           C  \n",
      "ATOM   1763  CG  GLU C 339       8.795  -2.148  10.877  1.00  0.53           C  \n",
      "ATOM   1764  CD  GLU C 339       8.521  -2.700  12.278  1.00  0.96           C  \n",
      "ATOM   1765  OE1 GLU C 339       7.529  -3.392  12.437  1.00  1.65           O  \n",
      "ATOM   1766  OE2 GLU C 339       9.308  -2.420  13.167  1.00  1.66           O  \n",
      "ATOM   1767  H   GLU C 339       8.791  -0.197   9.205  1.00  0.34           H  \n",
      "ATOM   1768  HA  GLU C 339       8.531  -2.819   8.212  1.00  0.36           H  \n",
      "ATOM   1769  HB2 GLU C 339       6.809  -1.526  10.348  1.00  0.40           H  \n",
      "ATOM   1770  HB3 GLU C 339       7.107  -3.253  10.147  1.00  0.47           H  \n",
      "ATOM   1771  HG2 GLU C 339       9.592  -2.714  10.416  1.00  1.01           H  \n",
      "ATOM   1772  HG3 GLU C 339       9.085  -1.111  10.951  1.00  0.83           H  \n",
      "ATOM   1773  N   MET C 340       5.933  -0.921   7.607  1.00  0.29           N  \n",
      "ATOM   1774  CA  MET C 340       4.668  -0.874   6.819  1.00  0.28           C  \n",
      "ATOM   1775  C   MET C 340       4.954  -1.222   5.357  1.00  0.26           C  \n",
      "ATOM   1776  O   MET C 340       4.301  -2.062   4.770  1.00  0.27           O  \n",
      "ATOM   1777  CB  MET C 340       4.065   0.531   6.898  1.00  0.29           C  \n",
      "ATOM   1778  CG  MET C 340       2.672   0.526   6.265  1.00  0.34           C  \n",
      "ATOM   1779  SD  MET C 340       2.044   2.221   6.171  1.00  0.47           S  \n",
      "ATOM   1780  CE  MET C 340       2.659   2.600   4.511  1.00  0.61           C  \n",
      "ATOM   1781  H   MET C 340       6.270  -0.110   8.043  1.00  0.29           H  \n",
      "ATOM   1782  HA  MET C 340       3.967  -1.588   7.225  1.00  0.31           H  \n",
      "ATOM   1783  HB2 MET C 340       3.989   0.832   7.934  1.00  0.38           H  \n",
      "ATOM   1784  HB3 MET C 340       4.697   1.227   6.366  1.00  0.31           H  \n",
      "ATOM   1785  HG2 MET C 340       2.731   0.108   5.271  1.00  0.42           H  \n",
      "ATOM   1786  HG3 MET C 340       2.005  -0.072   6.868  1.00  0.48           H  \n",
      "ATOM   1787  HE1 MET C 340       2.291   1.866   3.813  1.00  1.16           H  \n",
      "ATOM   1788  HE2 MET C 340       2.313   3.581   4.216  1.00  1.29           H  \n",
      "ATOM   1789  HE3 MET C 340       3.740   2.580   4.516  1.00  1.19           H  \n",
      "ATOM   1790  N   PHE C 341       5.922  -0.584   4.761  1.00  0.23           N  \n",
      "ATOM   1791  CA  PHE C 341       6.242  -0.880   3.337  1.00  0.23           C  \n",
      "ATOM   1792  C   PHE C 341       6.680  -2.339   3.200  1.00  0.23           C  \n",
      "ATOM   1793  O   PHE C 341       6.244  -3.045   2.314  1.00  0.25           O  \n",
      "ATOM   1794  CB  PHE C 341       7.371   0.038   2.866  1.00  0.22           C  \n",
      "ATOM   1795  CG  PHE C 341       6.805   1.395   2.524  1.00  0.23           C  \n",
      "ATOM   1796  CD1 PHE C 341       5.979   1.548   1.400  1.00  0.30           C  \n",
      "ATOM   1797  CD2 PHE C 341       7.105   2.504   3.329  1.00  0.22           C  \n",
      "ATOM   1798  CE1 PHE C 341       5.453   2.808   1.082  1.00  0.33           C  \n",
      "ATOM   1799  CE2 PHE C 341       6.579   3.765   3.012  1.00  0.26           C  \n",
      "ATOM   1800  CZ  PHE C 341       5.753   3.917   1.888  1.00  0.30           C  \n",
      "ATOM   1801  H   PHE C 341       6.437   0.092   5.249  1.00  0.23           H  \n",
      "ATOM   1802  HA  PHE C 341       5.365  -0.710   2.730  1.00  0.24           H  \n",
      "ATOM   1803  HB2 PHE C 341       8.104   0.141   3.653  1.00  0.22           H  \n",
      "ATOM   1804  HB3 PHE C 341       7.839  -0.386   1.991  1.00  0.25           H  \n",
      "ATOM   1805  HD1 PHE C 341       5.748   0.694   0.780  1.00  0.35           H  \n",
      "ATOM   1806  HD2 PHE C 341       7.740   2.387   4.194  1.00  0.24           H  \n",
      "ATOM   1807  HE1 PHE C 341       4.817   2.926   0.217  1.00  0.40           H  \n",
      "ATOM   1808  HE2 PHE C 341       6.810   4.618   3.632  1.00  0.29           H  \n",
      "ATOM   1809  HZ  PHE C 341       5.348   4.888   1.644  1.00  0.34           H  \n",
      "ATOM   1810  N   ARG C 342       7.544  -2.794   4.065  1.00  0.27           N  \n",
      "ATOM   1811  CA  ARG C 342       8.016  -4.205   3.982  1.00  0.30           C  \n",
      "ATOM   1812  C   ARG C 342       6.816  -5.153   3.909  1.00  0.28           C  \n",
      "ATOM   1813  O   ARG C 342       6.822  -6.118   3.171  1.00  0.28           O  \n",
      "ATOM   1814  CB  ARG C 342       8.852  -4.538   5.219  1.00  0.36           C  \n",
      "ATOM   1815  CG  ARG C 342       9.214  -6.024   5.203  1.00  0.42           C  \n",
      "ATOM   1816  CD  ARG C 342      10.115  -6.344   6.396  1.00  1.13           C  \n",
      "ATOM   1817  NE  ARG C 342      10.728  -7.690   6.211  1.00  1.77           N  \n",
      "ATOM   1818  CZ  ARG C 342      11.749  -8.053   6.940  1.00  2.41           C  \n",
      "ATOM   1819  NH1 ARG C 342      12.238  -7.240   7.838  1.00  2.75           N  \n",
      "ATOM   1820  NH2 ARG C 342      12.283  -9.232   6.772  1.00  3.28           N  \n",
      "ATOM   1821  H   ARG C 342       7.885  -2.206   4.768  1.00  0.31           H  \n",
      "ATOM   1822  HA  ARG C 342       8.622  -4.327   3.097  1.00  0.32           H  \n",
      "ATOM   1823  HB2 ARG C 342       9.756  -3.946   5.212  1.00  0.43           H  \n",
      "ATOM   1824  HB3 ARG C 342       8.283  -4.317   6.109  1.00  0.40           H  \n",
      "ATOM   1825  HG2 ARG C 342       8.311  -6.615   5.264  1.00  1.08           H  \n",
      "ATOM   1826  HG3 ARG C 342       9.736  -6.258   4.287  1.00  0.95           H  \n",
      "ATOM   1827  HD2 ARG C 342      10.896  -5.601   6.467  1.00  1.76           H  \n",
      "ATOM   1828  HD3 ARG C 342       9.529  -6.337   7.303  1.00  1.77           H  \n",
      "ATOM   1829  HE  ARG C 342      10.363  -8.305   5.540  1.00  2.27           H  \n",
      "ATOM   1830 HH11 ARG C 342      11.833  -6.336   7.971  1.00  2.63           H  \n",
      "ATOM   1831 HH12 ARG C 342      13.020  -7.522   8.394  1.00  3.50           H  \n",
      "ATOM   1832 HH21 ARG C 342      11.910  -9.857   6.086  1.00  3.61           H  \n",
      "ATOM   1833 HH22 ARG C 342      13.065  -9.511   7.330  1.00  3.85           H  \n",
      "ATOM   1834  N   GLU C 343       5.791  -4.893   4.673  1.00  0.28           N  \n",
      "ATOM   1835  CA  GLU C 343       4.600  -5.790   4.648  1.00  0.30           C  \n",
      "ATOM   1836  C   GLU C 343       4.046  -5.868   3.224  1.00  0.25           C  \n",
      "ATOM   1837  O   GLU C 343       3.800  -6.938   2.702  1.00  0.26           O  \n",
      "ATOM   1838  CB  GLU C 343       3.524  -5.235   5.583  1.00  0.34           C  \n",
      "ATOM   1839  CG  GLU C 343       2.294  -6.144   5.541  1.00  0.38           C  \n",
      "ATOM   1840  CD  GLU C 343       1.942  -6.592   6.960  1.00  1.05           C  \n",
      "ATOM   1841  OE1 GLU C 343       2.404  -5.956   7.893  1.00  1.78           O  \n",
      "ATOM   1842  OE2 GLU C 343       1.216  -7.564   7.090  1.00  1.70           O  \n",
      "ATOM   1843  H   GLU C 343       5.805  -4.113   5.266  1.00  0.30           H  \n",
      "ATOM   1844  HA  GLU C 343       4.887  -6.778   4.976  1.00  0.33           H  \n",
      "ATOM   1845  HB2 GLU C 343       3.910  -5.196   6.592  1.00  0.41           H  \n",
      "ATOM   1846  HB3 GLU C 343       3.246  -4.242   5.264  1.00  0.39           H  \n",
      "ATOM   1847  HG2 GLU C 343       1.461  -5.602   5.116  1.00  0.80           H  \n",
      "ATOM   1848  HG3 GLU C 343       2.508  -7.011   4.934  1.00  0.72           H  \n",
      "ATOM   1849  N   LEU C 344       3.847  -4.746   2.591  1.00  0.23           N  \n",
      "ATOM   1850  CA  LEU C 344       3.310  -4.761   1.201  1.00  0.24           C  \n",
      "ATOM   1851  C   LEU C 344       4.295  -5.489   0.286  1.00  0.23           C  \n",
      "ATOM   1852  O   LEU C 344       3.912  -6.144  -0.661  1.00  0.25           O  \n",
      "ATOM   1853  CB  LEU C 344       3.126  -3.324   0.710  1.00  0.26           C  \n",
      "ATOM   1854  CG  LEU C 344       2.042  -2.637   1.540  1.00  0.30           C  \n",
      "ATOM   1855  CD1 LEU C 344       2.096  -1.127   1.302  1.00  0.36           C  \n",
      "ATOM   1856  CD2 LEU C 344       0.670  -3.169   1.122  1.00  0.38           C  \n",
      "ATOM   1857  H   LEU C 344       4.052  -3.893   3.028  1.00  0.25           H  \n",
      "ATOM   1858  HA  LEU C 344       2.358  -5.272   1.187  1.00  0.26           H  \n",
      "ATOM   1859  HB2 LEU C 344       4.057  -2.786   0.815  1.00  0.27           H  \n",
      "ATOM   1860  HB3 LEU C 344       2.830  -3.333  -0.328  1.00  0.31           H  \n",
      "ATOM   1861  HG  LEU C 344       2.207  -2.842   2.588  1.00  0.32           H  \n",
      "ATOM   1862 HD11 LEU C 344       2.593  -0.928   0.364  1.00  1.00           H  \n",
      "ATOM   1863 HD12 LEU C 344       1.092  -0.731   1.268  1.00  1.06           H  \n",
      "ATOM   1864 HD13 LEU C 344       2.642  -0.655   2.106  1.00  1.17           H  \n",
      "ATOM   1865 HD21 LEU C 344       0.564  -3.090   0.051  1.00  1.06           H  \n",
      "ATOM   1866 HD22 LEU C 344       0.581  -4.204   1.417  1.00  1.12           H  \n",
      "ATOM   1867 HD23 LEU C 344      -0.103  -2.590   1.605  1.00  1.11           H  \n",
      "ATOM   1868  N   ASN C 345       5.563  -5.376   0.565  1.00  0.28           N  \n",
      "ATOM   1869  CA  ASN C 345       6.580  -6.057  -0.282  1.00  0.31           C  \n",
      "ATOM   1870  C   ASN C 345       6.323  -7.564  -0.284  1.00  0.26           C  \n",
      "ATOM   1871  O   ASN C 345       6.153  -8.172  -1.321  1.00  0.25           O  \n",
      "ATOM   1872  CB  ASN C 345       7.974  -5.783   0.283  1.00  0.41           C  \n",
      "ATOM   1873  CG  ASN C 345       9.004  -6.617  -0.480  1.00  0.50           C  \n",
      "ATOM   1874  OD1 ASN C 345       8.942  -7.830  -0.479  1.00  1.20           O  \n",
      "ATOM   1875  ND2 ASN C 345       9.954  -6.012  -1.137  1.00  0.55           N  \n",
      "ATOM   1876  H   ASN C 345       5.848  -4.843   1.334  1.00  0.34           H  \n",
      "ATOM   1877  HA  ASN C 345       6.520  -5.679  -1.292  1.00  0.34           H  \n",
      "ATOM   1878  HB2 ASN C 345       8.207  -4.734   0.175  1.00  0.47           H  \n",
      "ATOM   1879  HB3 ASN C 345       7.999  -6.053   1.328  1.00  0.45           H  \n",
      "ATOM   1880 HD21 ASN C 345      10.002  -5.034  -1.137  1.00  1.09           H  \n",
      "ATOM   1881 HD22 ASN C 345      10.620  -6.535  -1.630  1.00  0.55           H  \n",
      "ATOM   1882  N   GLU C 346       6.301  -8.170   0.870  1.00  0.27           N  \n",
      "ATOM   1883  CA  GLU C 346       6.062  -9.639   0.937  1.00  0.28           C  \n",
      "ATOM   1884  C   GLU C 346       4.684  -9.967   0.360  1.00  0.24           C  \n",
      "ATOM   1885  O   GLU C 346       4.466 -11.031  -0.183  1.00  0.26           O  \n",
      "ATOM   1886  CB  GLU C 346       6.123 -10.099   2.395  1.00  0.35           C  \n",
      "ATOM   1887  CG  GLU C 346       5.863 -11.605   2.466  1.00  0.45           C  \n",
      "ATOM   1888  CD  GLU C 346       6.480 -12.171   3.746  1.00  1.36           C  \n",
      "ATOM   1889  OE1 GLU C 346       7.663 -11.958   3.955  1.00  2.12           O  \n",
      "ATOM   1890  OE2 GLU C 346       5.758 -12.808   4.496  1.00  2.10           O  \n",
      "ATOM   1891  H   GLU C 346       6.446  -7.659   1.694  1.00  0.32           H  \n",
      "ATOM   1892  HA  GLU C 346       6.822 -10.153   0.367  1.00  0.31           H  \n",
      "ATOM   1893  HB2 GLU C 346       7.101  -9.881   2.801  1.00  0.43           H  \n",
      "ATOM   1894  HB3 GLU C 346       5.371  -9.579   2.968  1.00  0.39           H  \n",
      "ATOM   1895  HG2 GLU C 346       4.798 -11.786   2.467  1.00  1.03           H  \n",
      "ATOM   1896  HG3 GLU C 346       6.310 -12.088   1.610  1.00  1.05           H  \n",
      "ATOM   1897  N   ALA C 347       3.749  -9.065   0.481  1.00  0.22           N  \n",
      "ATOM   1898  CA  ALA C 347       2.382  -9.327  -0.056  1.00  0.23           C  \n",
      "ATOM   1899  C   ALA C 347       2.444  -9.518  -1.574  1.00  0.24           C  \n",
      "ATOM   1900  O   ALA C 347       1.928 -10.479  -2.110  1.00  0.26           O  \n",
      "ATOM   1901  CB  ALA C 347       1.471  -8.143   0.273  1.00  0.26           C  \n",
      "ATOM   1902  H   ALA C 347       3.944  -8.216   0.928  1.00  0.23           H  \n",
      "ATOM   1903  HA  ALA C 347       1.985 -10.222   0.400  1.00  0.26           H  \n",
      "ATOM   1904  HB1 ALA C 347       1.755  -7.726   1.228  1.00  1.07           H  \n",
      "ATOM   1905  HB2 ALA C 347       1.569  -7.389  -0.493  1.00  1.03           H  \n",
      "ATOM   1906  HB3 ALA C 347       0.446  -8.480   0.318  1.00  1.05           H  \n",
      "ATOM   1907  N   LEU C 348       3.071  -8.611  -2.272  1.00  0.24           N  \n",
      "ATOM   1908  CA  LEU C 348       3.165  -8.744  -3.755  1.00  0.26           C  \n",
      "ATOM   1909  C   LEU C 348       4.021  -9.963  -4.114  1.00  0.29           C  \n",
      "ATOM   1910  O   LEU C 348       3.663 -10.755  -4.962  1.00  0.31           O  \n",
      "ATOM   1911  CB  LEU C 348       3.793  -7.477  -4.360  1.00  0.29           C  \n",
      "ATOM   1912  CG  LEU C 348       2.735  -6.378  -4.553  1.00  0.30           C  \n",
      "ATOM   1913  CD1 LEU C 348       1.488  -6.968  -5.213  1.00  0.40           C  \n",
      "ATOM   1914  CD2 LEU C 348       2.349  -5.759  -3.206  1.00  0.32           C  \n",
      "ATOM   1915  H   LEU C 348       3.485  -7.848  -1.821  1.00  0.24           H  \n",
      "ATOM   1916  HA  LEU C 348       2.175  -8.882  -4.159  1.00  0.28           H  \n",
      "ATOM   1917  HB2 LEU C 348       4.564  -7.113  -3.696  1.00  0.28           H  \n",
      "ATOM   1918  HB3 LEU C 348       4.233  -7.719  -5.315  1.00  0.32           H  \n",
      "ATOM   1919  HG  LEU C 348       3.141  -5.609  -5.195  1.00  0.33           H  \n",
      "ATOM   1920 HD11 LEU C 348       1.784  -7.652  -5.995  1.00  1.08           H  \n",
      "ATOM   1921 HD12 LEU C 348       0.906  -7.497  -4.474  1.00  1.14           H  \n",
      "ATOM   1922 HD13 LEU C 348       0.894  -6.172  -5.637  1.00  1.08           H  \n",
      "ATOM   1923 HD21 LEU C 348       2.488  -6.482  -2.421  1.00  1.09           H  \n",
      "ATOM   1924 HD22 LEU C 348       2.971  -4.896  -3.016  1.00  1.04           H  \n",
      "ATOM   1925 HD23 LEU C 348       1.313  -5.456  -3.235  1.00  1.05           H  \n",
      "ATOM   1926  N   GLU C 349       5.149 -10.115  -3.478  1.00  0.34           N  \n",
      "ATOM   1927  CA  GLU C 349       6.032 -11.277  -3.785  1.00  0.39           C  \n",
      "ATOM   1928  C   GLU C 349       5.268 -12.582  -3.559  1.00  0.34           C  \n",
      "ATOM   1929  O   GLU C 349       5.502 -13.572  -4.224  1.00  0.35           O  \n",
      "ATOM   1930  CB  GLU C 349       7.254 -11.238  -2.869  1.00  0.47           C  \n",
      "ATOM   1931  CG  GLU C 349       7.968  -9.896  -3.033  1.00  0.59           C  \n",
      "ATOM   1932  CD  GLU C 349       9.398 -10.136  -3.521  1.00  1.29           C  \n",
      "ATOM   1933  OE1 GLU C 349       9.959 -11.159  -3.167  1.00  1.91           O  \n",
      "ATOM   1934  OE2 GLU C 349       9.906  -9.292  -4.241  1.00  2.09           O  \n",
      "ATOM   1935  H   GLU C 349       5.420  -9.463  -2.801  1.00  0.37           H  \n",
      "ATOM   1936  HA  GLU C 349       6.353 -11.223  -4.815  1.00  0.43           H  \n",
      "ATOM   1937  HB2 GLU C 349       6.938 -11.357  -1.842  1.00  0.47           H  \n",
      "ATOM   1938  HB3 GLU C 349       7.929 -12.038  -3.133  1.00  0.54           H  \n",
      "ATOM   1939  HG2 GLU C 349       7.437  -9.293  -3.755  1.00  1.06           H  \n",
      "ATOM   1940  HG3 GLU C 349       7.995  -9.383  -2.084  1.00  1.19           H  \n",
      "ATOM   1941  N   LEU C 350       4.358 -12.594  -2.627  1.00  0.32           N  \n",
      "ATOM   1942  CA  LEU C 350       3.583 -13.837  -2.365  1.00  0.32           C  \n",
      "ATOM   1943  C   LEU C 350       2.747 -14.172  -3.599  1.00  0.30           C  \n",
      "ATOM   1944  O   LEU C 350       2.711 -15.299  -4.054  1.00  0.33           O  \n",
      "ATOM   1945  CB  LEU C 350       2.662 -13.622  -1.162  1.00  0.35           C  \n",
      "ATOM   1946  CG  LEU C 350       2.118 -14.970  -0.688  1.00  0.44           C  \n",
      "ATOM   1947  CD1 LEU C 350       1.607 -14.840   0.748  1.00  0.75           C  \n",
      "ATOM   1948  CD2 LEU C 350       0.967 -15.404  -1.598  1.00  0.72           C  \n",
      "ATOM   1949  H   LEU C 350       4.181 -11.787  -2.103  1.00  0.34           H  \n",
      "ATOM   1950  HA  LEU C 350       4.264 -14.650  -2.159  1.00  0.35           H  \n",
      "ATOM   1951  HB2 LEU C 350       3.220 -13.157  -0.362  1.00  0.39           H  \n",
      "ATOM   1952  HB3 LEU C 350       1.840 -12.983  -1.447  1.00  0.47           H  \n",
      "ATOM   1953  HG  LEU C 350       2.905 -15.709  -0.724  1.00  0.80           H  \n",
      "ATOM   1954 HD11 LEU C 350       1.047 -13.923   0.849  1.00  1.30           H  \n",
      "ATOM   1955 HD12 LEU C 350       0.968 -15.680   0.980  1.00  1.38           H  \n",
      "ATOM   1956 HD13 LEU C 350       2.446 -14.827   1.428  1.00  1.32           H  \n",
      "ATOM   1957 HD21 LEU C 350       0.311 -14.565  -1.775  1.00  1.27           H  \n",
      "ATOM   1958 HD22 LEU C 350       1.362 -15.757  -2.537  1.00  1.31           H  \n",
      "ATOM   1959 HD23 LEU C 350       0.416 -16.199  -1.122  1.00  1.30           H  \n",
      "ATOM   1960  N   LYS C 351       2.080 -13.197  -4.149  1.00  0.31           N  \n",
      "ATOM   1961  CA  LYS C 351       1.247 -13.449  -5.356  1.00  0.35           C  \n",
      "ATOM   1962  C   LYS C 351       2.128 -14.028  -6.464  1.00  0.39           C  \n",
      "ATOM   1963  O   LYS C 351       1.745 -14.947  -7.160  1.00  0.44           O  \n",
      "ATOM   1964  CB  LYS C 351       0.633 -12.129  -5.825  1.00  0.40           C  \n",
      "ATOM   1965  CG  LYS C 351      -0.375 -12.396  -6.943  1.00  0.51           C  \n",
      "ATOM   1966  CD  LYS C 351      -1.160 -11.115  -7.231  1.00  0.82           C  \n",
      "ATOM   1967  CE  LYS C 351      -2.496 -11.469  -7.886  1.00  1.34           C  \n",
      "ATOM   1968  NZ  LYS C 351      -3.038 -10.270  -8.587  1.00  2.19           N  \n",
      "ATOM   1969  H   LYS C 351       2.129 -12.295  -3.768  1.00  0.33           H  \n",
      "ATOM   1970  HA  LYS C 351       0.461 -14.149  -5.114  1.00  0.37           H  \n",
      "ATOM   1971  HB2 LYS C 351       0.133 -11.651  -4.995  1.00  0.42           H  \n",
      "ATOM   1972  HB3 LYS C 351       1.414 -11.481  -6.195  1.00  0.45           H  \n",
      "ATOM   1973  HG2 LYS C 351       0.149 -12.708  -7.834  1.00  0.71           H  \n",
      "ATOM   1974  HG3 LYS C 351      -1.058 -13.173  -6.634  1.00  0.64           H  \n",
      "ATOM   1975  HD2 LYS C 351      -1.341 -10.588  -6.305  1.00  1.00           H  \n",
      "ATOM   1976  HD3 LYS C 351      -0.590 -10.486  -7.898  1.00  1.12           H  \n",
      "ATOM   1977  HE2 LYS C 351      -2.348 -12.266  -8.599  1.00  1.66           H  \n",
      "ATOM   1978  HE3 LYS C 351      -3.195 -11.789  -7.128  1.00  1.73           H  \n",
      "ATOM   1979  HZ1 LYS C 351      -2.466  -9.438  -8.343  1.00  2.61           H  \n",
      "ATOM   1980  HZ2 LYS C 351      -3.005 -10.426  -9.615  1.00  2.62           H  \n",
      "ATOM   1981  HZ3 LYS C 351      -4.022 -10.111  -8.292  1.00  2.68           H  \n",
      "ATOM   1982  N   ASP C 352       3.309 -13.498  -6.629  1.00  0.42           N  \n",
      "ATOM   1983  CA  ASP C 352       4.218 -14.019  -7.687  1.00  0.49           C  \n",
      "ATOM   1984  C   ASP C 352       4.439 -15.518  -7.475  1.00  0.50           C  \n",
      "ATOM   1985  O   ASP C 352       4.482 -16.288  -8.415  1.00  0.57           O  \n",
      "ATOM   1986  CB  ASP C 352       5.560 -13.288  -7.609  1.00  0.58           C  \n",
      "ATOM   1987  CG  ASP C 352       5.409 -11.881  -8.191  1.00  0.69           C  \n",
      "ATOM   1988  OD1 ASP C 352       5.292 -11.771  -9.400  1.00  1.27           O  \n",
      "ATOM   1989  OD2 ASP C 352       5.414 -10.938  -7.417  1.00  1.34           O  \n",
      "ATOM   1990  H   ASP C 352       3.599 -12.760  -6.054  1.00  0.43           H  \n",
      "ATOM   1991  HA  ASP C 352       3.774 -13.854  -8.657  1.00  0.53           H  \n",
      "ATOM   1992  HB2 ASP C 352       5.874 -13.220  -6.577  1.00  0.58           H  \n",
      "ATOM   1993  HB3 ASP C 352       6.300 -13.832  -8.176  1.00  0.64           H  \n",
      "ATOM   1994  N   ALA C 353       4.582 -15.939  -6.248  1.00  0.52           N  \n",
      "ATOM   1995  CA  ALA C 353       4.803 -17.388  -5.977  1.00  0.60           C  \n",
      "ATOM   1996  C   ALA C 353       3.545 -18.178  -6.342  1.00  0.60           C  \n",
      "ATOM   1997  O   ALA C 353       3.610 -19.342  -6.684  1.00  0.72           O  \n",
      "ATOM   1998  CB  ALA C 353       5.115 -17.586  -4.492  1.00  0.71           C  \n",
      "ATOM   1999  H   ALA C 353       4.546 -15.301  -5.505  1.00  0.55           H  \n",
      "ATOM   2000  HA  ALA C 353       5.634 -17.742  -6.569  1.00  0.67           H  \n",
      "ATOM   2001  HB1 ALA C 353       5.040 -16.639  -3.979  1.00  1.24           H  \n",
      "ATOM   2002  HB2 ALA C 353       4.410 -18.284  -4.066  1.00  1.30           H  \n",
      "ATOM   2003  HB3 ALA C 353       6.117 -17.975  -4.384  1.00  1.25           H  \n",
      "ATOM   2004  N   GLN C 354       2.400 -17.557  -6.273  1.00  0.58           N  \n",
      "ATOM   2005  CA  GLN C 354       1.143 -18.277  -6.619  1.00  0.70           C  \n",
      "ATOM   2006  C   GLN C 354       0.935 -18.245  -8.136  1.00  0.73           C  \n",
      "ATOM   2007  O   GLN C 354       0.074 -18.918  -8.669  1.00  0.96           O  \n",
      "ATOM   2008  CB  GLN C 354      -0.038 -17.590  -5.932  1.00  0.79           C  \n",
      "ATOM   2009  CG  GLN C 354      -0.605 -18.504  -4.845  1.00  1.13           C  \n",
      "ATOM   2010  CD  GLN C 354      -1.946 -17.945  -4.366  1.00  1.20           C  \n",
      "ATOM   2011  OE1 GLN C 354      -2.783 -18.676  -3.875  1.00  1.80           O  \n",
      "ATOM   2012  NE2 GLN C 354      -2.186 -16.668  -4.489  1.00  1.12           N  \n",
      "ATOM   2013  H   GLN C 354       2.367 -16.617  -5.996  1.00  0.57           H  \n",
      "ATOM   2014  HA  GLN C 354       1.209 -19.301  -6.285  1.00  0.82           H  \n",
      "ATOM   2015  HB2 GLN C 354       0.296 -16.664  -5.485  1.00  1.09           H  \n",
      "ATOM   2016  HB3 GLN C 354      -0.806 -17.381  -6.660  1.00  1.27           H  \n",
      "ATOM   2017  HG2 GLN C 354      -0.750 -19.497  -5.247  1.00  1.67           H  \n",
      "ATOM   2018  HG3 GLN C 354       0.082 -18.547  -4.014  1.00  1.60           H  \n",
      "ATOM   2019 HE21 GLN C 354      -1.511 -16.078  -4.885  1.00  1.18           H  \n",
      "ATOM   2020 HE22 GLN C 354      -3.042 -16.299  -4.188  1.00  1.41           H  \n",
      "ATOM   2021  N   ALA C 355       1.710 -17.460  -8.834  1.00  0.66           N  \n",
      "ATOM   2022  CA  ALA C 355       1.556 -17.372 -10.314  1.00  0.81           C  \n",
      "ATOM   2023  C   ALA C 355       2.125 -18.626 -10.986  1.00  0.96           C  \n",
      "ATOM   2024  O   ALA C 355       1.438 -19.320 -11.708  1.00  1.39           O  \n",
      "ATOM   2025  CB  ALA C 355       2.301 -16.138 -10.825  1.00  0.92           C  \n",
      "ATOM   2026  H   ALA C 355       2.393 -16.921  -8.382  1.00  0.66           H  \n",
      "ATOM   2027  HA  ALA C 355       0.508 -17.281 -10.558  1.00  0.94           H  \n",
      "ATOM   2028  HB1 ALA C 355       3.244 -16.046 -10.306  1.00  1.38           H  \n",
      "ATOM   2029  HB2 ALA C 355       2.482 -16.239 -11.885  1.00  1.47           H  \n",
      "ATOM   2030  HB3 ALA C 355       1.704 -15.256 -10.644  1.00  1.34           H  \n",
      "ATOM   2031  N   GLY C 356       3.382 -18.910 -10.770  1.00  1.01           N  \n",
      "ATOM   2032  CA  GLY C 356       4.005 -20.105 -11.415  1.00  1.28           C  \n",
      "ATOM   2033  C   GLY C 356       3.510 -21.398 -10.757  1.00  1.31           C  \n",
      "ATOM   2034  O   GLY C 356       3.935 -22.480 -11.112  1.00  1.66           O  \n",
      "ATOM   2035  H   GLY C 356       3.923 -18.328 -10.196  1.00  1.18           H  \n",
      "ATOM   2036  HA2 GLY C 356       3.747 -20.117 -12.464  1.00  1.58           H  \n",
      "ATOM   2037  HA3 GLY C 356       5.079 -20.045 -11.314  1.00  1.55           H  \n",
      "ATOM   2038  N   LYS C 357       2.619 -21.307  -9.809  1.00  1.57           N  \n",
      "ATOM   2039  CA  LYS C 357       2.111 -22.544  -9.148  1.00  1.88           C  \n",
      "ATOM   2040  C   LYS C 357       0.958 -23.117  -9.981  1.00  2.17           C  \n",
      "ATOM   2041  O   LYS C 357       0.099 -22.396 -10.447  1.00  2.57           O  \n",
      "ATOM   2042  CB  LYS C 357       1.634 -22.196  -7.725  1.00  2.39           C  \n",
      "ATOM   2043  CG  LYS C 357       0.463 -23.097  -7.306  1.00  2.87           C  \n",
      "ATOM   2044  CD  LYS C 357       0.373 -23.149  -5.778  1.00  3.33           C  \n",
      "ATOM   2045  CE  LYS C 357       0.187 -24.600  -5.328  1.00  4.10           C  \n",
      "ATOM   2046  NZ  LYS C 357      -1.072 -24.721  -4.539  1.00  4.45           N  \n",
      "ATOM   2047  H   LYS C 357       2.283 -20.431  -9.531  1.00  1.90           H  \n",
      "ATOM   2048  HA  LYS C 357       2.907 -23.272  -9.092  1.00  2.10           H  \n",
      "ATOM   2049  HB2 LYS C 357       2.452 -22.332  -7.034  1.00  2.75           H  \n",
      "ATOM   2050  HB3 LYS C 357       1.315 -21.164  -7.699  1.00  2.75           H  \n",
      "ATOM   2051  HG2 LYS C 357      -0.457 -22.698  -7.707  1.00  3.10           H  \n",
      "ATOM   2052  HG3 LYS C 357       0.621 -24.094  -7.689  1.00  3.33           H  \n",
      "ATOM   2053  HD2 LYS C 357       1.281 -22.751  -5.350  1.00  3.64           H  \n",
      "ATOM   2054  HD3 LYS C 357      -0.470 -22.560  -5.447  1.00  3.33           H  \n",
      "ATOM   2055  HE2 LYS C 357       0.130 -25.240  -6.195  1.00  4.48           H  \n",
      "ATOM   2056  HE3 LYS C 357       1.025 -24.896  -4.716  1.00  4.48           H  \n",
      "ATOM   2057  HZ1 LYS C 357      -1.394 -23.774  -4.250  1.00  4.57           H  \n",
      "ATOM   2058  HZ2 LYS C 357      -1.805 -25.174  -5.121  1.00  4.68           H  \n",
      "ATOM   2059  HZ3 LYS C 357      -0.898 -25.301  -3.694  1.00  4.78           H  \n",
      "ATOM   2060  N   GLU C 358       0.932 -24.409 -10.167  1.00  2.67           N  \n",
      "ATOM   2061  CA  GLU C 358      -0.167 -25.021 -10.963  1.00  3.46           C  \n",
      "ATOM   2062  C   GLU C 358      -1.509 -24.501 -10.432  1.00  3.57           C  \n",
      "ATOM   2063  O   GLU C 358      -1.706 -24.429  -9.235  1.00  3.59           O  \n",
      "ATOM   2064  CB  GLU C 358      -0.116 -26.544 -10.821  1.00  4.28           C  \n",
      "ATOM   2065  CG  GLU C 358       0.634 -27.147 -12.011  1.00  4.94           C  \n",
      "ATOM   2066  CD  GLU C 358       0.305 -28.638 -12.118  1.00  5.73           C  \n",
      "ATOM   2067  OE1 GLU C 358      -0.766 -29.020 -11.675  1.00  6.15           O  \n",
      "ATOM   2068  OE2 GLU C 358       1.129 -29.371 -12.639  1.00  6.19           O  \n",
      "ATOM   2069  H   GLU C 358       1.633 -24.975  -9.781  1.00  2.86           H  \n",
      "ATOM   2070  HA  GLU C 358      -0.052 -24.749 -12.001  1.00  3.75           H  \n",
      "ATOM   2071  HB2 GLU C 358       0.395 -26.802  -9.905  1.00  4.61           H  \n",
      "ATOM   2072  HB3 GLU C 358      -1.121 -26.937 -10.796  1.00  4.52           H  \n",
      "ATOM   2073  HG2 GLU C 358       0.334 -26.645 -12.919  1.00  4.98           H  \n",
      "ATOM   2074  HG3 GLU C 358       1.697 -27.025 -11.866  1.00  5.25           H  \n",
      "ATOM   2075  N   PRO C 359      -2.394 -24.147 -11.332  1.00  4.12           N  \n",
      "ATOM   2076  CA  PRO C 359      -3.720 -23.625 -10.957  1.00  4.64           C  \n",
      "ATOM   2077  C   PRO C 359      -4.500 -24.673 -10.159  1.00  4.93           C  \n",
      "ATOM   2078  O   PRO C 359      -5.129 -24.368  -9.165  1.00  5.28           O  \n",
      "ATOM   2079  CB  PRO C 359      -4.418 -23.328 -12.290  1.00  5.52           C  \n",
      "ATOM   2080  CG  PRO C 359      -3.437 -23.702 -13.430  1.00  5.58           C  \n",
      "ATOM   2081  CD  PRO C 359      -2.150 -24.236 -12.784  1.00  4.67           C  \n",
      "ATOM   2082  HA  PRO C 359      -3.619 -22.716 -10.386  1.00  4.60           H  \n",
      "ATOM   2083  HB2 PRO C 359      -5.320 -23.920 -12.370  1.00  6.03           H  \n",
      "ATOM   2084  HB3 PRO C 359      -4.661 -22.279 -12.352  1.00  5.85           H  \n",
      "ATOM   2085  HG2 PRO C 359      -3.878 -24.464 -14.057  1.00  6.18           H  \n",
      "ATOM   2086  HG3 PRO C 359      -3.210 -22.827 -14.020  1.00  5.90           H  \n",
      "ATOM   2087  HD2 PRO C 359      -1.980 -25.262 -13.077  1.00  4.93           H  \n",
      "ATOM   2088  HD3 PRO C 359      -1.307 -23.619 -13.057  1.00  4.50           H  \n",
      "ATOM   2089  N   GLY C 360      -4.464 -25.906 -10.584  1.00  5.18           N  \n",
      "ATOM   2090  CA  GLY C 360      -5.205 -26.968  -9.847  1.00  5.78           C  \n",
      "ATOM   2091  C   GLY C 360      -6.697 -26.633  -9.830  1.00  6.47           C  \n",
      "ATOM   2092  O   GLY C 360      -7.265 -26.597  -8.751  1.00  6.85           O  \n",
      "ATOM   2093  OXT GLY C 360      -7.248 -26.418 -10.898  1.00  6.90           O  \n",
      "ATOM   2094  H   GLY C 360      -3.951 -26.133 -11.387  1.00  5.24           H  \n",
      "ATOM   2095  HA2 GLY C 360      -5.053 -27.919 -10.337  1.00  5.91           H  \n",
      "ATOM   2096  HA3 GLY C 360      -4.841 -27.023  -8.832  1.00  5.95           H  \n",
      "TER    2097      GLY C 360                                                      \n",
      "ATOM   2098  N   LYS D 319     -18.336 -25.670  10.765  1.00  4.83           N  \n",
      "ATOM   2099  CA  LYS D 319     -17.721 -25.526   9.414  1.00  4.34           C  \n",
      "ATOM   2100  C   LYS D 319     -17.903 -24.089   8.923  1.00  3.74           C  \n",
      "ATOM   2101  O   LYS D 319     -19.008 -23.636   8.723  1.00  3.68           O  \n",
      "ATOM   2102  CB  LYS D 319     -18.397 -26.489   8.433  1.00  4.69           C  \n",
      "ATOM   2103  CG  LYS D 319     -17.916 -26.190   7.010  1.00  5.18           C  \n",
      "ATOM   2104  CD  LYS D 319     -16.385 -26.220   6.964  1.00  5.93           C  \n",
      "ATOM   2105  CE  LYS D 319     -15.873 -27.521   7.586  1.00  6.53           C  \n",
      "ATOM   2106  NZ  LYS D 319     -16.278 -28.673   6.731  1.00  7.36           N  \n",
      "ATOM   2107  H1  LYS D 319     -18.198 -24.791  11.303  1.00  5.12           H  \n",
      "ATOM   2108  H2  LYS D 319     -19.354 -25.860  10.665  1.00  5.20           H  \n",
      "ATOM   2109  H3  LYS D 319     -17.884 -26.457  11.271  1.00  4.97           H  \n",
      "ATOM   2110  HA  LYS D 319     -16.667 -25.755   9.473  1.00  4.69           H  \n",
      "ATOM   2111  HB2 LYS D 319     -18.147 -27.506   8.694  1.00  4.98           H  \n",
      "ATOM   2112  HB3 LYS D 319     -19.468 -26.359   8.482  1.00  4.82           H  \n",
      "ATOM   2113  HG2 LYS D 319     -18.311 -26.933   6.333  1.00  5.24           H  \n",
      "ATOM   2114  HG3 LYS D 319     -18.262 -25.212   6.712  1.00  5.36           H  \n",
      "ATOM   2115  HD2 LYS D 319     -16.056 -26.160   5.936  1.00  6.23           H  \n",
      "ATOM   2116  HD3 LYS D 319     -15.991 -25.381   7.517  1.00  6.10           H  \n",
      "ATOM   2117  HE2 LYS D 319     -14.796 -27.487   7.656  1.00  6.70           H  \n",
      "ATOM   2118  HE3 LYS D 319     -16.294 -27.640   8.573  1.00  6.51           H  \n",
      "ATOM   2119  HZ1 LYS D 319     -17.304 -28.638   6.564  1.00  7.62           H  \n",
      "ATOM   2120  HZ2 LYS D 319     -15.780 -28.621   5.820  1.00  7.59           H  \n",
      "ATOM   2121  HZ3 LYS D 319     -16.034 -29.563   7.211  1.00  7.69           H  \n",
      "ATOM   2122  N   LYS D 320     -16.815 -23.384   8.731  1.00  3.83           N  \n",
      "ATOM   2123  CA  LYS D 320     -16.877 -21.966   8.250  1.00  3.79           C  \n",
      "ATOM   2124  C   LYS D 320     -17.448 -21.052   9.345  1.00  3.36           C  \n",
      "ATOM   2125  O   LYS D 320     -16.724 -20.314   9.977  1.00  3.69           O  \n",
      "ATOM   2126  CB  LYS D 320     -17.732 -21.876   6.981  1.00  4.16           C  \n",
      "ATOM   2127  CG  LYS D 320     -16.889 -22.305   5.776  1.00  4.94           C  \n",
      "ATOM   2128  CD  LYS D 320     -17.763 -22.346   4.521  1.00  5.75           C  \n",
      "ATOM   2129  CE  LYS D 320     -18.544 -23.661   4.486  1.00  6.48           C  \n",
      "ATOM   2130  NZ  LYS D 320     -17.658 -24.752   3.988  1.00  7.15           N  \n",
      "ATOM   2131  H   LYS D 320     -15.943 -23.793   8.906  1.00  4.29           H  \n",
      "ATOM   2132  HA  LYS D 320     -15.874 -21.638   8.018  1.00  4.32           H  \n",
      "ATOM   2133  HB2 LYS D 320     -18.589 -22.524   7.070  1.00  4.21           H  \n",
      "ATOM   2134  HB3 LYS D 320     -18.063 -20.858   6.841  1.00  4.34           H  \n",
      "ATOM   2135  HG2 LYS D 320     -16.084 -21.598   5.631  1.00  5.21           H  \n",
      "ATOM   2136  HG3 LYS D 320     -16.476 -23.286   5.957  1.00  5.08           H  \n",
      "ATOM   2137  HD2 LYS D 320     -18.454 -21.516   4.536  1.00  5.83           H  \n",
      "ATOM   2138  HD3 LYS D 320     -17.137 -22.278   3.644  1.00  6.07           H  \n",
      "ATOM   2139  HE2 LYS D 320     -18.888 -23.903   5.480  1.00  6.78           H  \n",
      "ATOM   2140  HE3 LYS D 320     -19.393 -23.558   3.826  1.00  6.57           H  \n",
      "ATOM   2141  HZ1 LYS D 320     -16.680 -24.568   4.290  1.00  7.36           H  \n",
      "ATOM   2142  HZ2 LYS D 320     -17.976 -25.662   4.377  1.00  7.40           H  \n",
      "ATOM   2143  HZ3 LYS D 320     -17.701 -24.788   2.950  1.00  7.45           H  \n",
      "ATOM   2144  N   LYS D 321     -18.732 -21.088   9.574  1.00  3.02           N  \n",
      "ATOM   2145  CA  LYS D 321     -19.335 -20.220  10.631  1.00  2.81           C  \n",
      "ATOM   2146  C   LYS D 321     -19.248 -18.747  10.200  1.00  2.47           C  \n",
      "ATOM   2147  O   LYS D 321     -18.220 -18.308   9.722  1.00  2.60           O  \n",
      "ATOM   2148  CB  LYS D 321     -18.587 -20.387  11.966  1.00  3.35           C  \n",
      "ATOM   2149  CG  LYS D 321     -17.916 -21.764  12.062  1.00  4.03           C  \n",
      "ATOM   2150  CD  LYS D 321     -16.402 -21.580  12.224  1.00  4.79           C  \n",
      "ATOM   2151  CE  LYS D 321     -16.014 -21.753  13.693  1.00  5.71           C  \n",
      "ATOM   2152  NZ  LYS D 321     -16.370 -20.519  14.448  1.00  6.50           N  \n",
      "ATOM   2153  H   LYS D 321     -19.306 -21.680   9.051  1.00  3.22           H  \n",
      "ATOM   2154  HA  LYS D 321     -20.368 -20.498  10.765  1.00  3.16           H  \n",
      "ATOM   2155  HB2 LYS D 321     -17.832 -19.619  12.047  1.00  3.53           H  \n",
      "ATOM   2156  HB3 LYS D 321     -19.290 -20.279  12.779  1.00  3.66           H  \n",
      "ATOM   2157  HG2 LYS D 321     -18.305 -22.291  12.921  1.00  4.37           H  \n",
      "ATOM   2158  HG3 LYS D 321     -18.115 -22.338  11.173  1.00  4.12           H  \n",
      "ATOM   2159  HD2 LYS D 321     -15.884 -22.316  11.627  1.00  4.85           H  \n",
      "ATOM   2160  HD3 LYS D 321     -16.122 -20.591  11.899  1.00  5.06           H  \n",
      "ATOM   2161  HE2 LYS D 321     -16.542 -22.596  14.110  1.00  5.94           H  \n",
      "ATOM   2162  HE3 LYS D 321     -14.950 -21.922  13.764  1.00  5.91           H  \n",
      "ATOM   2163  HZ1 LYS D 321     -16.476 -19.726  13.786  1.00  6.75           H  \n",
      "ATOM   2164  HZ2 LYS D 321     -17.267 -20.668  14.953  1.00  6.76           H  \n",
      "ATOM   2165  HZ3 LYS D 321     -15.617 -20.303  15.133  1.00  6.84           H  \n",
      "ATOM   2166  N   PRO D 322     -20.329 -18.023  10.388  1.00  2.84           N  \n",
      "ATOM   2167  CA  PRO D 322     -20.396 -16.590  10.030  1.00  3.26           C  \n",
      "ATOM   2168  C   PRO D 322     -19.627 -15.725  11.044  1.00  2.75           C  \n",
      "ATOM   2169  O   PRO D 322     -19.788 -14.522  11.082  1.00  2.71           O  \n",
      "ATOM   2170  CB  PRO D 322     -21.891 -16.252  10.093  1.00  4.29           C  \n",
      "ATOM   2171  CG  PRO D 322     -22.613 -17.449  10.762  1.00  4.50           C  \n",
      "ATOM   2172  CD  PRO D 322     -21.578 -18.569  10.951  1.00  3.58           C  \n",
      "ATOM   2173  HA  PRO D 322     -20.024 -16.429   9.032  1.00  3.66           H  \n",
      "ATOM   2174  HB2 PRO D 322     -22.041 -15.355  10.677  1.00  4.48           H  \n",
      "ATOM   2175  HB3 PRO D 322     -22.277 -16.110   9.095  1.00  4.98           H  \n",
      "ATOM   2176  HG2 PRO D 322     -23.009 -17.147  11.721  1.00  4.95           H  \n",
      "ATOM   2177  HG3 PRO D 322     -23.413 -17.798  10.127  1.00  5.12           H  \n",
      "ATOM   2178  HD2 PRO D 322     -21.456 -18.794  12.002  1.00  3.75           H  \n",
      "ATOM   2179  HD3 PRO D 322     -21.876 -19.452  10.407  1.00  3.71           H  \n",
      "ATOM   2180  N   LEU D 323     -18.809 -16.319  11.876  1.00  2.66           N  \n",
      "ATOM   2181  CA  LEU D 323     -18.056 -15.518  12.884  1.00  2.35           C  \n",
      "ATOM   2182  C   LEU D 323     -16.783 -14.935  12.262  1.00  1.75           C  \n",
      "ATOM   2183  O   LEU D 323     -15.694 -15.136  12.761  1.00  1.85           O  \n",
      "ATOM   2184  CB  LEU D 323     -17.676 -16.417  14.065  1.00  2.89           C  \n",
      "ATOM   2185  CG  LEU D 323     -18.934 -16.796  14.850  1.00  3.62           C  \n",
      "ATOM   2186  CD1 LEU D 323     -18.534 -17.441  16.179  1.00  4.32           C  \n",
      "ATOM   2187  CD2 LEU D 323     -19.762 -15.539  15.126  1.00  3.81           C  \n",
      "ATOM   2188  H   LEU D 323     -18.698 -17.289  11.848  1.00  3.01           H  \n",
      "ATOM   2189  HA  LEU D 323     -18.679 -14.711  13.237  1.00  2.52           H  \n",
      "ATOM   2190  HB2 LEU D 323     -17.199 -17.313  13.694  1.00  3.05           H  \n",
      "ATOM   2191  HB3 LEU D 323     -16.993 -15.889  14.715  1.00  2.85           H  \n",
      "ATOM   2192  HG  LEU D 323     -19.520 -17.497  14.273  1.00  3.73           H  \n",
      "ATOM   2193 HD11 LEU D 323     -17.690 -16.911  16.597  1.00  4.66           H  \n",
      "ATOM   2194 HD12 LEU D 323     -19.365 -17.393  16.867  1.00  4.60           H  \n",
      "ATOM   2195 HD13 LEU D 323     -18.264 -18.473  16.012  1.00  4.56           H  \n",
      "ATOM   2196 HD21 LEU D 323     -19.106 -14.735  15.427  1.00  4.00           H  \n",
      "ATOM   2197 HD22 LEU D 323     -20.291 -15.254  14.228  1.00  3.90           H  \n",
      "ATOM   2198 HD23 LEU D 323     -20.472 -15.739  15.914  1.00  4.14           H  \n",
      "ATOM   2199  N   ASP D 324     -16.905 -14.205  11.187  1.00  1.48           N  \n",
      "ATOM   2200  CA  ASP D 324     -15.694 -13.606  10.558  1.00  1.30           C  \n",
      "ATOM   2201  C   ASP D 324     -15.310 -12.333  11.317  1.00  1.05           C  \n",
      "ATOM   2202  O   ASP D 324     -15.950 -11.958  12.279  1.00  1.00           O  \n",
      "ATOM   2203  CB  ASP D 324     -15.989 -13.263   9.096  1.00  1.75           C  \n",
      "ATOM   2204  CG  ASP D 324     -16.558 -14.492   8.385  1.00  2.06           C  \n",
      "ATOM   2205  OD1 ASP D 324     -16.033 -15.573   8.599  1.00  2.40           O  \n",
      "ATOM   2206  OD2 ASP D 324     -17.509 -14.332   7.637  1.00  2.46           O  \n",
      "ATOM   2207  H   ASP D 324     -17.789 -14.042  10.800  1.00  1.75           H  \n",
      "ATOM   2208  HA  ASP D 324     -14.879 -14.313  10.605  1.00  1.49           H  \n",
      "ATOM   2209  HB2 ASP D 324     -16.706 -12.457   9.053  1.00  1.91           H  \n",
      "ATOM   2210  HB3 ASP D 324     -15.075 -12.959   8.607  1.00  2.04           H  \n",
      "ATOM   2211  N   GLY D 325     -14.271 -11.665  10.894  1.00  0.97           N  \n",
      "ATOM   2212  CA  GLY D 325     -13.853 -10.417  11.596  1.00  0.84           C  \n",
      "ATOM   2213  C   GLY D 325     -14.897  -9.322  11.359  1.00  0.68           C  \n",
      "ATOM   2214  O   GLY D 325     -15.931  -9.557  10.767  1.00  0.66           O  \n",
      "ATOM   2215  H   GLY D 325     -13.767 -11.982  10.116  1.00  1.08           H  \n",
      "ATOM   2216  HA2 GLY D 325     -13.768 -10.612  12.656  1.00  0.87           H  \n",
      "ATOM   2217  HA3 GLY D 325     -12.899 -10.089  11.212  1.00  0.92           H  \n",
      "ATOM   2218  N   GLU D 326     -14.635  -8.126  11.815  1.00  0.64           N  \n",
      "ATOM   2219  CA  GLU D 326     -15.615  -7.021  11.611  1.00  0.57           C  \n",
      "ATOM   2220  C   GLU D 326     -15.860  -6.832  10.113  1.00  0.51           C  \n",
      "ATOM   2221  O   GLU D 326     -15.367  -7.584   9.296  1.00  0.51           O  \n",
      "ATOM   2222  CB  GLU D 326     -15.060  -5.726  12.208  1.00  0.65           C  \n",
      "ATOM   2223  CG  GLU D 326     -15.088  -5.814  13.735  1.00  0.76           C  \n",
      "ATOM   2224  CD  GLU D 326     -16.176  -4.888  14.282  1.00  1.06           C  \n",
      "ATOM   2225  OE1 GLU D 326     -17.332  -5.277  14.242  1.00  1.66           O  \n",
      "ATOM   2226  OE2 GLU D 326     -15.836  -3.807  14.733  1.00  1.67           O  \n",
      "ATOM   2227  H   GLU D 326     -13.795  -7.955  12.289  1.00  0.71           H  \n",
      "ATOM   2228  HA  GLU D 326     -16.546  -7.272  12.099  1.00  0.59           H  \n",
      "ATOM   2229  HB2 GLU D 326     -14.043  -5.581  11.872  1.00  0.69           H  \n",
      "ATOM   2230  HB3 GLU D 326     -15.667  -4.892  11.887  1.00  0.66           H  \n",
      "ATOM   2231  HG2 GLU D 326     -15.296  -6.831  14.033  1.00  0.96           H  \n",
      "ATOM   2232  HG3 GLU D 326     -14.130  -5.512  14.131  1.00  0.96           H  \n",
      "ATOM   2233  N   TYR D 327     -16.618  -5.837   9.743  1.00  0.49           N  \n",
      "ATOM   2234  CA  TYR D 327     -16.890  -5.612   8.296  1.00  0.45           C  \n",
      "ATOM   2235  C   TYR D 327     -17.120  -4.122   8.034  1.00  0.43           C  \n",
      "ATOM   2236  O   TYR D 327     -18.154  -3.575   8.364  1.00  0.48           O  \n",
      "ATOM   2237  CB  TYR D 327     -18.133  -6.401   7.884  1.00  0.48           C  \n",
      "ATOM   2238  CG  TYR D 327     -17.755  -7.844   7.654  1.00  0.50           C  \n",
      "ATOM   2239  CD1 TYR D 327     -17.297  -8.256   6.393  1.00  0.49           C  \n",
      "ATOM   2240  CD2 TYR D 327     -17.860  -8.773   8.700  1.00  0.55           C  \n",
      "ATOM   2241  CE1 TYR D 327     -16.944  -9.597   6.179  1.00  0.52           C  \n",
      "ATOM   2242  CE2 TYR D 327     -17.508 -10.113   8.486  1.00  0.58           C  \n",
      "ATOM   2243  CZ  TYR D 327     -17.050 -10.525   7.225  1.00  0.56           C  \n",
      "ATOM   2244  OH  TYR D 327     -16.704 -11.845   7.014  1.00  0.61           O  \n",
      "ATOM   2245  H   TYR D 327     -17.010  -5.241  10.415  1.00  0.51           H  \n",
      "ATOM   2246  HA  TYR D 327     -16.043  -5.950   7.717  1.00  0.43           H  \n",
      "ATOM   2247  HB2 TYR D 327     -18.873  -6.343   8.668  1.00  0.52           H  \n",
      "ATOM   2248  HB3 TYR D 327     -18.539  -5.987   6.973  1.00  0.48           H  \n",
      "ATOM   2249  HD1 TYR D 327     -17.216  -7.541   5.588  1.00  0.47           H  \n",
      "ATOM   2250  HD2 TYR D 327     -18.211  -8.455   9.671  1.00  0.57           H  \n",
      "ATOM   2251  HE1 TYR D 327     -16.591  -9.913   5.208  1.00  0.53           H  \n",
      "ATOM   2252  HE2 TYR D 327     -17.589 -10.827   9.292  1.00  0.63           H  \n",
      "ATOM   2253  HH  TYR D 327     -17.095 -12.124   6.183  1.00  1.01           H  \n",
      "ATOM   2254  N   PHE D 328     -16.166  -3.463   7.436  1.00  0.39           N  \n",
      "ATOM   2255  CA  PHE D 328     -16.328  -2.011   7.143  1.00  0.39           C  \n",
      "ATOM   2256  C   PHE D 328     -16.404  -1.809   5.628  1.00  0.36           C  \n",
      "ATOM   2257  O   PHE D 328     -16.596  -2.745   4.878  1.00  0.37           O  \n",
      "ATOM   2258  CB  PHE D 328     -15.131  -1.239   7.702  1.00  0.39           C  \n",
      "ATOM   2259  CG  PHE D 328     -15.004  -1.502   9.183  1.00  0.43           C  \n",
      "ATOM   2260  CD1 PHE D 328     -14.233  -2.582   9.640  1.00  0.46           C  \n",
      "ATOM   2261  CD2 PHE D 328     -15.657  -0.667  10.103  1.00  0.49           C  \n",
      "ATOM   2262  CE1 PHE D 328     -14.114  -2.826  11.016  1.00  0.53           C  \n",
      "ATOM   2263  CE2 PHE D 328     -15.538  -0.912  11.479  1.00  0.56           C  \n",
      "ATOM   2264  CZ  PHE D 328     -14.767  -1.991  11.936  1.00  0.56           C  \n",
      "ATOM   2265  H   PHE D 328     -15.342  -3.926   7.174  1.00  0.38           H  \n",
      "ATOM   2266  HA  PHE D 328     -17.237  -1.650   7.602  1.00  0.42           H  \n",
      "ATOM   2267  HB2 PHE D 328     -14.230  -1.561   7.200  1.00  0.39           H  \n",
      "ATOM   2268  HB3 PHE D 328     -15.276  -0.181   7.536  1.00  0.42           H  \n",
      "ATOM   2269  HD1 PHE D 328     -13.731  -3.224   8.932  1.00  0.47           H  \n",
      "ATOM   2270  HD2 PHE D 328     -16.250   0.164   9.751  1.00  0.53           H  \n",
      "ATOM   2271  HE1 PHE D 328     -13.520  -3.657  11.368  1.00  0.58           H  \n",
      "ATOM   2272  HE2 PHE D 328     -16.040  -0.269  12.187  1.00  0.63           H  \n",
      "ATOM   2273  HZ  PHE D 328     -14.675  -2.179  12.995  1.00  0.63           H  \n",
      "ATOM   2274  N   THR D 329     -16.255  -0.596   5.171  1.00  0.34           N  \n",
      "ATOM   2275  CA  THR D 329     -16.319  -0.342   3.704  1.00  0.34           C  \n",
      "ATOM   2276  C   THR D 329     -15.522   0.920   3.371  1.00  0.33           C  \n",
      "ATOM   2277  O   THR D 329     -15.137   1.670   4.245  1.00  0.37           O  \n",
      "ATOM   2278  CB  THR D 329     -17.778  -0.153   3.282  1.00  0.37           C  \n",
      "ATOM   2279  OG1 THR D 329     -18.434   0.691   4.218  1.00  0.41           O  \n",
      "ATOM   2280  CG2 THR D 329     -18.477  -1.512   3.239  1.00  0.43           C  \n",
      "ATOM   2281  H   THR D 329     -16.100   0.147   5.791  1.00  0.35           H  \n",
      "ATOM   2282  HA  THR D 329     -15.898  -1.184   3.174  1.00  0.35           H  \n",
      "ATOM   2283  HB  THR D 329     -17.815   0.298   2.302  1.00  0.39           H  \n",
      "ATOM   2284  HG1 THR D 329     -19.017   1.278   3.731  1.00  0.97           H  \n",
      "ATOM   2285 HG21 THR D 329     -17.920  -2.183   2.602  1.00  1.07           H  \n",
      "ATOM   2286 HG22 THR D 329     -18.529  -1.922   4.236  1.00  1.13           H  \n",
      "ATOM   2287 HG23 THR D 329     -19.476  -1.391   2.847  1.00  1.13           H  \n",
      "ATOM   2288  N   LEU D 330     -15.266   1.159   2.113  1.00  0.29           N  \n",
      "ATOM   2289  CA  LEU D 330     -14.489   2.372   1.735  1.00  0.29           C  \n",
      "ATOM   2290  C   LEU D 330     -14.898   2.834   0.334  1.00  0.27           C  \n",
      "ATOM   2291  O   LEU D 330     -14.940   2.058  -0.600  1.00  0.26           O  \n",
      "ATOM   2292  CB  LEU D 330     -12.996   2.036   1.744  1.00  0.29           C  \n",
      "ATOM   2293  CG  LEU D 330     -12.191   3.267   1.328  1.00  0.29           C  \n",
      "ATOM   2294  CD1 LEU D 330     -12.210   4.296   2.460  1.00  0.35           C  \n",
      "ATOM   2295  CD2 LEU D 330     -10.746   2.854   1.039  1.00  0.31           C  \n",
      "ATOM   2296  H   LEU D 330     -15.582   0.542   1.421  1.00  0.28           H  \n",
      "ATOM   2297  HA  LEU D 330     -14.684   3.162   2.445  1.00  0.32           H  \n",
      "ATOM   2298  HB2 LEU D 330     -12.703   1.732   2.738  1.00  0.34           H  \n",
      "ATOM   2299  HB3 LEU D 330     -12.804   1.232   1.050  1.00  0.29           H  \n",
      "ATOM   2300  HG  LEU D 330     -12.628   3.700   0.440  1.00  0.31           H  \n",
      "ATOM   2301 HD11 LEU D 330     -11.920   3.819   3.384  1.00  1.05           H  \n",
      "ATOM   2302 HD12 LEU D 330     -11.519   5.094   2.233  1.00  1.03           H  \n",
      "ATOM   2303 HD13 LEU D 330     -13.207   4.700   2.561  1.00  1.10           H  \n",
      "ATOM   2304 HD21 LEU D 330     -10.483   2.008   1.658  1.00  1.06           H  \n",
      "ATOM   2305 HD22 LEU D 330     -10.651   2.581  -0.001  1.00  1.05           H  \n",
      "ATOM   2306 HD23 LEU D 330     -10.085   3.679   1.257  1.00  1.09           H  \n",
      "ATOM   2307  N   GLN D 331     -15.192   4.097   0.180  1.00  0.29           N  \n",
      "ATOM   2308  CA  GLN D 331     -15.591   4.615  -1.159  1.00  0.29           C  \n",
      "ATOM   2309  C   GLN D 331     -14.338   5.038  -1.930  1.00  0.29           C  \n",
      "ATOM   2310  O   GLN D 331     -13.666   5.983  -1.568  1.00  0.31           O  \n",
      "ATOM   2311  CB  GLN D 331     -16.516   5.820  -0.983  1.00  0.34           C  \n",
      "ATOM   2312  CG  GLN D 331     -17.057   6.254  -2.347  1.00  0.38           C  \n",
      "ATOM   2313  CD  GLN D 331     -18.573   6.055  -2.382  1.00  0.63           C  \n",
      "ATOM   2314  OE1 GLN D 331     -19.293   6.646  -1.602  1.00  1.37           O  \n",
      "ATOM   2315  NE2 GLN D 331     -19.092   5.241  -3.260  1.00  0.61           N  \n",
      "ATOM   2316  H   GLN D 331     -15.146   4.706   0.947  1.00  0.32           H  \n",
      "ATOM   2317  HA  GLN D 331     -16.107   3.840  -1.706  1.00  0.29           H  \n",
      "ATOM   2318  HB2 GLN D 331     -17.340   5.550  -0.338  1.00  0.41           H  \n",
      "ATOM   2319  HB3 GLN D 331     -15.965   6.636  -0.541  1.00  0.39           H  \n",
      "ATOM   2320  HG2 GLN D 331     -16.825   7.296  -2.511  1.00  0.48           H  \n",
      "ATOM   2321  HG3 GLN D 331     -16.599   5.657  -3.122  1.00  0.58           H  \n",
      "ATOM   2322 HE21 GLN D 331     -18.512   4.763  -3.889  1.00  1.01           H  \n",
      "ATOM   2323 HE22 GLN D 331     -20.063   5.107  -3.290  1.00  0.76           H  \n",
      "ATOM   2324  N   ILE D 332     -14.014   4.344  -2.987  1.00  0.29           N  \n",
      "ATOM   2325  CA  ILE D 332     -12.800   4.709  -3.772  1.00  0.32           C  \n",
      "ATOM   2326  C   ILE D 332     -13.210   5.436  -5.055  1.00  0.32           C  \n",
      "ATOM   2327  O   ILE D 332     -13.727   4.840  -5.979  1.00  0.33           O  \n",
      "ATOM   2328  CB  ILE D 332     -12.027   3.439  -4.131  1.00  0.35           C  \n",
      "ATOM   2329  CG1 ILE D 332     -11.746   2.638  -2.857  1.00  0.34           C  \n",
      "ATOM   2330  CG2 ILE D 332     -10.703   3.817  -4.796  1.00  0.48           C  \n",
      "ATOM   2331  CD1 ILE D 332     -11.300   1.223  -3.232  1.00  0.40           C  \n",
      "ATOM   2332  H   ILE D 332     -14.566   3.583  -3.263  1.00  0.30           H  \n",
      "ATOM   2333  HA  ILE D 332     -12.171   5.357  -3.179  1.00  0.33           H  \n",
      "ATOM   2334  HB  ILE D 332     -12.614   2.841  -4.812  1.00  0.39           H  \n",
      "ATOM   2335 HG12 ILE D 332     -10.964   3.125  -2.291  1.00  0.41           H  \n",
      "ATOM   2336 HG13 ILE D 332     -12.643   2.584  -2.260  1.00  0.34           H  \n",
      "ATOM   2337 HG21 ILE D 332     -10.169   4.514  -4.167  1.00  1.15           H  \n",
      "ATOM   2338 HG22 ILE D 332     -10.105   2.929  -4.938  1.00  1.15           H  \n",
      "ATOM   2339 HG23 ILE D 332     -10.899   4.275  -5.754  1.00  1.04           H  \n",
      "ATOM   2340 HD11 ILE D 332     -10.405   1.276  -3.834  1.00  1.11           H  \n",
      "ATOM   2341 HD12 ILE D 332     -11.097   0.660  -2.333  1.00  1.12           H  \n",
      "ATOM   2342 HD13 ILE D 332     -12.084   0.736  -3.793  1.00  1.05           H  \n",
      "ATOM   2343  N   ARG D 333     -12.979   6.719  -5.122  1.00  0.33           N  \n",
      "ATOM   2344  CA  ARG D 333     -13.350   7.482  -6.347  1.00  0.36           C  \n",
      "ATOM   2345  C   ARG D 333     -12.383   7.131  -7.480  1.00  0.37           C  \n",
      "ATOM   2346  O   ARG D 333     -11.254   6.747  -7.250  1.00  0.40           O  \n",
      "ATOM   2347  CB  ARG D 333     -13.268   8.982  -6.058  1.00  0.39           C  \n",
      "ATOM   2348  CG  ARG D 333     -14.027   9.296  -4.769  1.00  0.41           C  \n",
      "ATOM   2349  CD  ARG D 333     -13.738  10.738  -4.348  1.00  0.89           C  \n",
      "ATOM   2350  NE  ARG D 333     -14.512  11.063  -3.118  1.00  1.04           N  \n",
      "ATOM   2351  CZ  ARG D 333     -14.209  12.121  -2.414  1.00  1.47           C  \n",
      "ATOM   2352  NH1 ARG D 333     -13.228  12.900  -2.786  1.00  2.08           N  \n",
      "ATOM   2353  NH2 ARG D 333     -14.888  12.402  -1.336  1.00  2.08           N  \n",
      "ATOM   2354  H   ARG D 333     -12.558   7.180  -4.367  1.00  0.34           H  \n",
      "ATOM   2355  HA  ARG D 333     -14.357   7.226  -6.640  1.00  0.36           H  \n",
      "ATOM   2356  HB2 ARG D 333     -12.232   9.270  -5.947  1.00  0.45           H  \n",
      "ATOM   2357  HB3 ARG D 333     -13.709   9.531  -6.876  1.00  0.47           H  \n",
      "ATOM   2358  HG2 ARG D 333     -15.087   9.174  -4.937  1.00  0.78           H  \n",
      "ATOM   2359  HG3 ARG D 333     -13.706   8.624  -3.988  1.00  0.87           H  \n",
      "ATOM   2360  HD2 ARG D 333     -12.682  10.849  -4.148  1.00  1.67           H  \n",
      "ATOM   2361  HD3 ARG D 333     -14.026  11.409  -5.144  1.00  1.53           H  \n",
      "ATOM   2362  HE  ARG D 333     -15.249  10.483  -2.835  1.00  1.62           H  \n",
      "ATOM   2363 HH11 ARG D 333     -12.705  12.688  -3.612  1.00  2.19           H  \n",
      "ATOM   2364 HH12 ARG D 333     -12.999  13.708  -2.244  1.00  2.77           H  \n",
      "ATOM   2365 HH21 ARG D 333     -15.639  11.808  -1.049  1.00  2.39           H  \n",
      "ATOM   2366 HH22 ARG D 333     -14.656  13.212  -0.796  1.00  2.57           H  \n",
      "ATOM   2367  N   GLY D 334     -12.816   7.266  -8.704  1.00  0.37           N  \n",
      "ATOM   2368  CA  GLY D 334     -11.921   6.947  -9.852  1.00  0.40           C  \n",
      "ATOM   2369  C   GLY D 334     -11.817   5.431 -10.023  1.00  0.38           C  \n",
      "ATOM   2370  O   GLY D 334     -11.486   4.712  -9.100  1.00  0.37           O  \n",
      "ATOM   2371  H   GLY D 334     -13.730   7.582  -8.868  1.00  0.39           H  \n",
      "ATOM   2372  HA2 GLY D 334     -12.324   7.384 -10.754  1.00  0.43           H  \n",
      "ATOM   2373  HA3 GLY D 334     -10.938   7.352  -9.664  1.00  0.41           H  \n",
      "ATOM   2374  N   ARG D 335     -12.094   4.939 -11.199  1.00  0.41           N  \n",
      "ATOM   2375  CA  ARG D 335     -12.007   3.470 -11.434  1.00  0.42           C  \n",
      "ATOM   2376  C   ARG D 335     -10.537   3.054 -11.503  1.00  0.42           C  \n",
      "ATOM   2377  O   ARG D 335     -10.166   1.976 -11.082  1.00  0.42           O  \n",
      "ATOM   2378  CB  ARG D 335     -12.698   3.124 -12.755  1.00  0.48           C  \n",
      "ATOM   2379  CG  ARG D 335     -12.518   1.634 -13.054  1.00  0.56           C  \n",
      "ATOM   2380  CD  ARG D 335     -12.553   1.411 -14.566  1.00  0.93           C  \n",
      "ATOM   2381  NE  ARG D 335     -13.009   0.022 -14.853  1.00  1.34           N  \n",
      "ATOM   2382  CZ  ARG D 335     -13.402  -0.303 -16.056  1.00  1.80           C  \n",
      "ATOM   2383  NH1 ARG D 335     -13.401   0.588 -17.011  1.00  2.16           N  \n",
      "ATOM   2384  NH2 ARG D 335     -13.798  -1.521 -16.305  1.00  2.64           N  \n",
      "ATOM   2385  H   ARG D 335     -12.356   5.537 -11.930  1.00  0.44           H  \n",
      "ATOM   2386  HA  ARG D 335     -12.493   2.946 -10.624  1.00  0.42           H  \n",
      "ATOM   2387  HB2 ARG D 335     -13.752   3.351 -12.680  1.00  0.48           H  \n",
      "ATOM   2388  HB3 ARG D 335     -12.261   3.705 -13.553  1.00  0.56           H  \n",
      "ATOM   2389  HG2 ARG D 335     -11.568   1.300 -12.663  1.00  0.79           H  \n",
      "ATOM   2390  HG3 ARG D 335     -13.315   1.074 -12.590  1.00  0.78           H  \n",
      "ATOM   2391  HD2 ARG D 335     -13.237   2.114 -15.017  1.00  1.64           H  \n",
      "ATOM   2392  HD3 ARG D 335     -11.565   1.557 -14.975  1.00  1.50           H  \n",
      "ATOM   2393  HE  ARG D 335     -13.012  -0.650 -14.139  1.00  1.98           H  \n",
      "ATOM   2394 HH11 ARG D 335     -13.099   1.523 -16.825  1.00  2.17           H  \n",
      "ATOM   2395 HH12 ARG D 335     -13.703   0.336 -17.930  1.00  2.86           H  \n",
      "ATOM   2396 HH21 ARG D 335     -13.800  -2.205 -15.575  1.00  3.07           H  \n",
      "ATOM   2397 HH22 ARG D 335     -14.099  -1.771 -17.225  1.00  3.12           H  \n",
      "ATOM   2398  N   GLU D 336      -9.697   3.900 -12.034  1.00  0.46           N  \n",
      "ATOM   2399  CA  GLU D 336      -8.251   3.555 -12.134  1.00  0.49           C  \n",
      "ATOM   2400  C   GLU D 336      -7.694   3.268 -10.738  1.00  0.44           C  \n",
      "ATOM   2401  O   GLU D 336      -7.207   2.189 -10.461  1.00  0.40           O  \n",
      "ATOM   2402  CB  GLU D 336      -7.491   4.730 -12.752  1.00  0.56           C  \n",
      "ATOM   2403  CG  GLU D 336      -6.919   4.315 -14.109  1.00  1.17           C  \n",
      "ATOM   2404  CD  GLU D 336      -6.344   5.543 -14.817  1.00  1.55           C  \n",
      "ATOM   2405  OE1 GLU D 336      -5.596   6.270 -14.185  1.00  2.22           O  \n",
      "ATOM   2406  OE2 GLU D 336      -6.662   5.735 -15.979  1.00  1.98           O  \n",
      "ATOM   2407  H   GLU D 336     -10.018   4.763 -12.369  1.00  0.48           H  \n",
      "ATOM   2408  HA  GLU D 336      -8.131   2.681 -12.756  1.00  0.52           H  \n",
      "ATOM   2409  HB2 GLU D 336      -8.165   5.564 -12.884  1.00  0.98           H  \n",
      "ATOM   2410  HB3 GLU D 336      -6.683   5.021 -12.097  1.00  0.96           H  \n",
      "ATOM   2411  HG2 GLU D 336      -6.138   3.584 -13.961  1.00  1.71           H  \n",
      "ATOM   2412  HG3 GLU D 336      -7.704   3.887 -14.714  1.00  1.72           H  \n",
      "ATOM   2413  N   ARG D 337      -7.759   4.228  -9.857  1.00  0.45           N  \n",
      "ATOM   2414  CA  ARG D 337      -7.230   4.015  -8.480  1.00  0.43           C  \n",
      "ATOM   2415  C   ARG D 337      -7.988   2.869  -7.810  1.00  0.37           C  \n",
      "ATOM   2416  O   ARG D 337      -7.453   2.155  -6.986  1.00  0.35           O  \n",
      "ATOM   2417  CB  ARG D 337      -7.410   5.296  -7.664  1.00  0.49           C  \n",
      "ATOM   2418  CG  ARG D 337      -6.520   6.393  -8.248  1.00  0.60           C  \n",
      "ATOM   2419  CD  ARG D 337      -5.108   6.256  -7.677  1.00  1.13           C  \n",
      "ATOM   2420  NE  ARG D 337      -4.514   7.614  -7.464  1.00  1.40           N  \n",
      "ATOM   2421  CZ  ARG D 337      -4.500   8.506  -8.420  1.00  2.12           C  \n",
      "ATOM   2422  NH1 ARG D 337      -4.821   8.177  -9.642  1.00  2.70           N  \n",
      "ATOM   2423  NH2 ARG D 337      -4.101   9.722  -8.164  1.00  2.74           N  \n",
      "ATOM   2424  H   ARG D 337      -8.153   5.091 -10.101  1.00  0.49           H  \n",
      "ATOM   2425  HA  ARG D 337      -6.180   3.767  -8.533  1.00  0.44           H  \n",
      "ATOM   2426  HB2 ARG D 337      -8.444   5.608  -7.705  1.00  0.52           H  \n",
      "ATOM   2427  HB3 ARG D 337      -7.128   5.114  -6.638  1.00  0.51           H  \n",
      "ATOM   2428  HG2 ARG D 337      -6.491   6.290  -9.322  1.00  1.22           H  \n",
      "ATOM   2429  HG3 ARG D 337      -6.920   7.361  -7.988  1.00  1.08           H  \n",
      "ATOM   2430  HD2 ARG D 337      -5.157   5.761  -6.724  1.00  1.71           H  \n",
      "ATOM   2431  HD3 ARG D 337      -4.502   5.662  -8.355  1.00  1.85           H  \n",
      "ATOM   2432  HE  ARG D 337      -4.177   7.856  -6.576  1.00  1.67           H  \n",
      "ATOM   2433 HH11 ARG D 337      -5.073   7.237  -9.859  1.00  2.59           H  \n",
      "ATOM   2434 HH12 ARG D 337      -4.814   8.870 -10.363  1.00  3.50           H  \n",
      "ATOM   2435 HH21 ARG D 337      -3.807   9.967  -7.240  1.00  2.76           H  \n",
      "ATOM   2436 HH22 ARG D 337      -4.087  10.408  -8.891  1.00  3.44           H  \n",
      "ATOM   2437  N   PHE D 338      -9.231   2.687  -8.156  1.00  0.37           N  \n",
      "ATOM   2438  CA  PHE D 338     -10.019   1.587  -7.537  1.00  0.35           C  \n",
      "ATOM   2439  C   PHE D 338      -9.292   0.256  -7.742  1.00  0.33           C  \n",
      "ATOM   2440  O   PHE D 338      -8.980  -0.440  -6.800  1.00  0.29           O  \n",
      "ATOM   2441  CB  PHE D 338     -11.400   1.517  -8.190  1.00  0.38           C  \n",
      "ATOM   2442  CG  PHE D 338     -12.126   0.291  -7.692  1.00  0.36           C  \n",
      "ATOM   2443  CD1 PHE D 338     -12.618   0.255  -6.379  1.00  0.37           C  \n",
      "ATOM   2444  CD2 PHE D 338     -12.306  -0.812  -8.540  1.00  0.41           C  \n",
      "ATOM   2445  CE1 PHE D 338     -13.291  -0.885  -5.914  1.00  0.39           C  \n",
      "ATOM   2446  CE2 PHE D 338     -12.978  -1.951  -8.075  1.00  0.43           C  \n",
      "ATOM   2447  CZ  PHE D 338     -13.471  -1.988  -6.762  1.00  0.40           C  \n",
      "ATOM   2448  H   PHE D 338      -9.647   3.273  -8.823  1.00  0.40           H  \n",
      "ATOM   2449  HA  PHE D 338     -10.131   1.776  -6.480  1.00  0.35           H  \n",
      "ATOM   2450  HB2 PHE D 338     -11.966   2.401  -7.934  1.00  0.41           H  \n",
      "ATOM   2451  HB3 PHE D 338     -11.289   1.458  -9.263  1.00  0.43           H  \n",
      "ATOM   2452  HD1 PHE D 338     -12.480   1.104  -5.726  1.00  0.41           H  \n",
      "ATOM   2453  HD2 PHE D 338     -11.926  -0.783  -9.551  1.00  0.48           H  \n",
      "ATOM   2454  HE1 PHE D 338     -13.669  -0.913  -4.903  1.00  0.44           H  \n",
      "ATOM   2455  HE2 PHE D 338     -13.117  -2.801  -8.728  1.00  0.50           H  \n",
      "ATOM   2456  HZ  PHE D 338     -13.988  -2.866  -6.403  1.00  0.43           H  \n",
      "ATOM   2457  N   GLU D 339      -9.030  -0.105  -8.967  1.00  0.37           N  \n",
      "ATOM   2458  CA  GLU D 339      -8.331  -1.395  -9.235  1.00  0.37           C  \n",
      "ATOM   2459  C   GLU D 339      -7.045  -1.484  -8.406  1.00  0.32           C  \n",
      "ATOM   2460  O   GLU D 339      -6.654  -2.546  -7.966  1.00  0.30           O  \n",
      "ATOM   2461  CB  GLU D 339      -7.982  -1.483 -10.723  1.00  0.44           C  \n",
      "ATOM   2462  CG  GLU D 339      -9.269  -1.495 -11.551  1.00  0.58           C  \n",
      "ATOM   2463  CD  GLU D 339      -8.975  -2.057 -12.943  1.00  0.97           C  \n",
      "ATOM   2464  OE1 GLU D 339      -8.784  -3.257 -13.048  1.00  1.69           O  \n",
      "ATOM   2465  OE2 GLU D 339      -8.946  -1.277 -13.881  1.00  1.50           O  \n",
      "ATOM   2466  H   GLU D 339      -9.297   0.471  -9.714  1.00  0.41           H  \n",
      "ATOM   2467  HA  GLU D 339      -8.983  -2.215  -8.972  1.00  0.38           H  \n",
      "ATOM   2468  HB2 GLU D 339      -7.382  -0.629 -11.002  1.00  0.46           H  \n",
      "ATOM   2469  HB3 GLU D 339      -7.428  -2.390 -10.910  1.00  0.47           H  \n",
      "ATOM   2470  HG2 GLU D 339     -10.007  -2.114 -11.061  1.00  0.74           H  \n",
      "ATOM   2471  HG3 GLU D 339      -9.647  -0.488 -11.644  1.00  0.81           H  \n",
      "ATOM   2472  N   MET D 340      -6.377  -0.382  -8.202  1.00  0.32           N  \n",
      "ATOM   2473  CA  MET D 340      -5.108  -0.413  -7.415  1.00  0.29           C  \n",
      "ATOM   2474  C   MET D 340      -5.388  -0.855  -5.976  1.00  0.25           C  \n",
      "ATOM   2475  O   MET D 340      -4.746  -1.747  -5.457  1.00  0.25           O  \n",
      "ATOM   2476  CB  MET D 340      -4.481   0.984  -7.405  1.00  0.32           C  \n",
      "ATOM   2477  CG  MET D 340      -3.059   0.901  -6.847  1.00  0.31           C  \n",
      "ATOM   2478  SD  MET D 340      -2.378   2.570  -6.678  1.00  0.42           S  \n",
      "ATOM   2479  CE  MET D 340      -2.843   2.831  -4.949  1.00  0.43           C  \n",
      "ATOM   2480  H   MET D 340      -6.702   0.465  -8.574  1.00  0.34           H  \n",
      "ATOM   2481  HA  MET D 340      -4.422  -1.108  -7.874  1.00  0.29           H  \n",
      "ATOM   2482  HB2 MET D 340      -4.450   1.371  -8.414  1.00  0.39           H  \n",
      "ATOM   2483  HB3 MET D 340      -5.071   1.641  -6.785  1.00  0.32           H  \n",
      "ATOM   2484  HG2 MET D 340      -3.079   0.421  -5.879  1.00  0.32           H  \n",
      "ATOM   2485  HG3 MET D 340      -2.440   0.326  -7.520  1.00  0.40           H  \n",
      "ATOM   2486  HE1 MET D 340      -3.902   2.649  -4.831  1.00  1.12           H  \n",
      "ATOM   2487  HE2 MET D 340      -2.291   2.152  -4.320  1.00  1.14           H  \n",
      "ATOM   2488  HE3 MET D 340      -2.613   3.849  -4.665  1.00  1.07           H  \n",
      "ATOM   2489  N   PHE D 341      -6.332  -0.236  -5.323  1.00  0.24           N  \n",
      "ATOM   2490  CA  PHE D 341      -6.639  -0.618  -3.915  1.00  0.22           C  \n",
      "ATOM   2491  C   PHE D 341      -7.114  -2.072  -3.860  1.00  0.23           C  \n",
      "ATOM   2492  O   PHE D 341      -6.671  -2.848  -3.039  1.00  0.23           O  \n",
      "ATOM   2493  CB  PHE D 341      -7.736   0.298  -3.370  1.00  0.23           C  \n",
      "ATOM   2494  CG  PHE D 341      -7.133   1.625  -2.973  1.00  0.23           C  \n",
      "ATOM   2495  CD1 PHE D 341      -6.308   1.709  -1.841  1.00  0.26           C  \n",
      "ATOM   2496  CD2 PHE D 341      -7.397   2.772  -3.735  1.00  0.25           C  \n",
      "ATOM   2497  CE1 PHE D 341      -5.747   2.941  -1.472  1.00  0.27           C  \n",
      "ATOM   2498  CE2 PHE D 341      -6.836   4.004  -3.366  1.00  0.26           C  \n",
      "ATOM   2499  CZ  PHE D 341      -6.012   4.089  -2.235  1.00  0.26           C  \n",
      "ATOM   2500  H   PHE D 341      -6.834   0.486  -5.756  1.00  0.25           H  \n",
      "ATOM   2501  HA  PHE D 341      -5.750  -0.509  -3.312  1.00  0.22           H  \n",
      "ATOM   2502  HB2 PHE D 341      -8.484   0.457  -4.133  1.00  0.24           H  \n",
      "ATOM   2503  HB3 PHE D 341      -8.194  -0.161  -2.506  1.00  0.24           H  \n",
      "ATOM   2504  HD1 PHE D 341      -6.104   0.826  -1.254  1.00  0.30           H  \n",
      "ATOM   2505  HD2 PHE D 341      -8.031   2.708  -4.607  1.00  0.28           H  \n",
      "ATOM   2506  HE1 PHE D 341      -5.112   3.006  -0.601  1.00  0.32           H  \n",
      "ATOM   2507  HE2 PHE D 341      -7.040   4.888  -3.953  1.00  0.31           H  \n",
      "ATOM   2508  HZ  PHE D 341      -5.580   5.037  -1.951  1.00  0.28           H  \n",
      "ATOM   2509  N   ARG D 342      -8.017  -2.444  -4.724  1.00  0.26           N  \n",
      "ATOM   2510  CA  ARG D 342      -8.530  -3.843  -4.722  1.00  0.28           C  \n",
      "ATOM   2511  C   ARG D 342      -7.359  -4.829  -4.713  1.00  0.26           C  \n",
      "ATOM   2512  O   ARG D 342      -7.383  -5.827  -4.022  1.00  0.27           O  \n",
      "ATOM   2513  CB  ARG D 342      -9.378  -4.072  -5.975  1.00  0.34           C  \n",
      "ATOM   2514  CG  ARG D 342     -10.026  -5.456  -5.909  1.00  0.43           C  \n",
      "ATOM   2515  CD  ARG D 342     -10.534  -5.847  -7.298  1.00  0.88           C  \n",
      "ATOM   2516  NE  ARG D 342     -11.159  -7.198  -7.236  1.00  1.26           N  \n",
      "ATOM   2517  CZ  ARG D 342     -11.919  -7.616  -8.213  1.00  1.74           C  \n",
      "ATOM   2518  NH1 ARG D 342     -12.133  -6.853  -9.252  1.00  2.21           N  \n",
      "ATOM   2519  NH2 ARG D 342     -12.466  -8.799  -8.152  1.00  2.45           N  \n",
      "ATOM   2520  H   ARG D 342      -8.363  -1.800  -5.373  1.00  0.28           H  \n",
      "ATOM   2521  HA  ARG D 342      -9.139  -4.002  -3.844  1.00  0.30           H  \n",
      "ATOM   2522  HB2 ARG D 342     -10.148  -3.316  -6.031  1.00  0.38           H  \n",
      "ATOM   2523  HB3 ARG D 342      -8.750  -4.012  -6.851  1.00  0.38           H  \n",
      "ATOM   2524  HG2 ARG D 342      -9.296  -6.179  -5.575  1.00  0.80           H  \n",
      "ATOM   2525  HG3 ARG D 342     -10.854  -5.434  -5.218  1.00  0.77           H  \n",
      "ATOM   2526  HD2 ARG D 342     -11.267  -5.127  -7.629  1.00  1.52           H  \n",
      "ATOM   2527  HD3 ARG D 342      -9.707  -5.864  -7.992  1.00  1.40           H  \n",
      "ATOM   2528  HE  ARG D 342     -11.000  -7.775  -6.460  1.00  1.84           H  \n",
      "ATOM   2529 HH11 ARG D 342     -11.716  -5.946  -9.304  1.00  2.21           H  \n",
      "ATOM   2530 HH12 ARG D 342     -12.716  -7.177  -9.997  1.00  2.92           H  \n",
      "ATOM   2531 HH21 ARG D 342     -12.303  -9.385  -7.358  1.00  2.78           H  \n",
      "ATOM   2532 HH22 ARG D 342     -13.048  -9.120  -8.899  1.00  2.96           H  \n",
      "ATOM   2533  N   GLU D 343      -6.337  -4.563  -5.480  1.00  0.26           N  \n",
      "ATOM   2534  CA  GLU D 343      -5.173  -5.493  -5.516  1.00  0.26           C  \n",
      "ATOM   2535  C   GLU D 343      -4.586  -5.640  -4.110  1.00  0.23           C  \n",
      "ATOM   2536  O   GLU D 343      -4.312  -6.732  -3.653  1.00  0.24           O  \n",
      "ATOM   2537  CB  GLU D 343      -4.104  -4.937  -6.459  1.00  0.29           C  \n",
      "ATOM   2538  CG  GLU D 343      -2.879  -5.854  -6.439  1.00  0.35           C  \n",
      "ATOM   2539  CD  GLU D 343      -2.389  -6.084  -7.870  1.00  1.06           C  \n",
      "ATOM   2540  OE1 GLU D 343      -2.482  -5.162  -8.663  1.00  1.75           O  \n",
      "ATOM   2541  OE2 GLU D 343      -1.927  -7.179  -8.148  1.00  1.75           O  \n",
      "ATOM   2542  H   GLU D 343      -6.337  -3.755  -6.035  1.00  0.27           H  \n",
      "ATOM   2543  HA  GLU D 343      -5.498  -6.460  -5.871  1.00  0.29           H  \n",
      "ATOM   2544  HB2 GLU D 343      -4.501  -4.887  -7.463  1.00  0.35           H  \n",
      "ATOM   2545  HB3 GLU D 343      -3.817  -3.949  -6.136  1.00  0.32           H  \n",
      "ATOM   2546  HG2 GLU D 343      -2.094  -5.393  -5.858  1.00  0.71           H  \n",
      "ATOM   2547  HG3 GLU D 343      -3.146  -6.802  -5.996  1.00  0.66           H  \n",
      "ATOM   2548  N   LEU D 344      -4.388  -4.550  -3.421  1.00  0.21           N  \n",
      "ATOM   2549  CA  LEU D 344      -3.816  -4.631  -2.047  1.00  0.21           C  \n",
      "ATOM   2550  C   LEU D 344      -4.766  -5.417  -1.142  1.00  0.22           C  \n",
      "ATOM   2551  O   LEU D 344      -4.344  -6.153  -0.274  1.00  0.25           O  \n",
      "ATOM   2552  CB  LEU D 344      -3.632  -3.220  -1.487  1.00  0.22           C  \n",
      "ATOM   2553  CG  LEU D 344      -2.562  -2.486  -2.297  1.00  0.26           C  \n",
      "ATOM   2554  CD1 LEU D 344      -2.649  -0.984  -2.017  1.00  0.31           C  \n",
      "ATOM   2555  CD2 LEU D 344      -1.178  -2.999  -1.894  1.00  0.31           C  \n",
      "ATOM   2556  H   LEU D 344      -4.613  -3.678  -3.808  1.00  0.22           H  \n",
      "ATOM   2557  HA  LEU D 344      -2.860  -5.131  -2.084  1.00  0.22           H  \n",
      "ATOM   2558  HB2 LEU D 344      -4.566  -2.682  -1.553  1.00  0.23           H  \n",
      "ATOM   2559  HB3 LEU D 344      -3.321  -3.279  -0.455  1.00  0.24           H  \n",
      "ATOM   2560  HG  LEU D 344      -2.723  -2.665  -3.350  1.00  0.28           H  \n",
      "ATOM   2561 HD11 LEU D 344      -3.063  -0.825  -1.032  1.00  1.04           H  \n",
      "ATOM   2562 HD12 LEU D 344      -1.662  -0.551  -2.068  1.00  1.02           H  \n",
      "ATOM   2563 HD13 LEU D 344      -3.286  -0.517  -2.754  1.00  1.01           H  \n",
      "ATOM   2564 HD21 LEU D 344      -1.269  -3.993  -1.481  1.00  1.08           H  \n",
      "ATOM   2565 HD22 LEU D 344      -0.538  -3.028  -2.764  1.00  1.05           H  \n",
      "ATOM   2566 HD23 LEU D 344      -0.751  -2.339  -1.154  1.00  1.02           H  \n",
      "ATOM   2567  N   ASN D 345      -6.046  -5.263  -1.334  1.00  0.25           N  \n",
      "ATOM   2568  CA  ASN D 345      -7.020  -5.998  -0.480  1.00  0.30           C  \n",
      "ATOM   2569  C   ASN D 345      -6.763  -7.502  -0.581  1.00  0.26           C  \n",
      "ATOM   2570  O   ASN D 345      -6.591  -8.179   0.412  1.00  0.26           O  \n",
      "ATOM   2571  CB  ASN D 345      -8.443  -5.694  -0.951  1.00  0.38           C  \n",
      "ATOM   2572  CG  ASN D 345      -9.442  -6.450  -0.072  1.00  0.49           C  \n",
      "ATOM   2573  OD1 ASN D 345      -9.136  -7.505   0.447  1.00  1.22           O  \n",
      "ATOM   2574  ND2 ASN D 345     -10.631  -5.951   0.120  1.00  0.56           N  \n",
      "ATOM   2575  H   ASN D 345      -6.367  -4.664  -2.038  1.00  0.29           H  \n",
      "ATOM   2576  HA  ASN D 345      -6.908  -5.682   0.546  1.00  0.33           H  \n",
      "ATOM   2577  HB2 ASN D 345      -8.628  -4.632  -0.878  1.00  0.42           H  \n",
      "ATOM   2578  HB3 ASN D 345      -8.559  -6.010  -1.977  1.00  0.42           H  \n",
      "ATOM   2579 HD21 ASN D 345     -10.877  -5.100  -0.296  1.00  1.13           H  \n",
      "ATOM   2580 HD22 ASN D 345     -11.278  -6.429   0.680  1.00  0.57           H  \n",
      "ATOM   2581  N   GLU D 346      -6.742  -8.032  -1.773  1.00  0.28           N  \n",
      "ATOM   2582  CA  GLU D 346      -6.503  -9.494  -1.936  1.00  0.29           C  \n",
      "ATOM   2583  C   GLU D 346      -5.129  -9.859  -1.369  1.00  0.25           C  \n",
      "ATOM   2584  O   GLU D 346      -4.916 -10.956  -0.897  1.00  0.26           O  \n",
      "ATOM   2585  CB  GLU D 346      -6.552  -9.854  -3.423  1.00  0.36           C  \n",
      "ATOM   2586  CG  GLU D 346      -6.369 -11.364  -3.591  1.00  0.43           C  \n",
      "ATOM   2587  CD  GLU D 346      -7.132 -11.837  -4.830  1.00  1.02           C  \n",
      "ATOM   2588  OE1 GLU D 346      -8.292 -11.480  -4.959  1.00  1.71           O  \n",
      "ATOM   2589  OE2 GLU D 346      -6.544 -12.547  -5.629  1.00  1.73           O  \n",
      "ATOM   2590  H   GLU D 346      -6.889  -7.468  -2.562  1.00  0.32           H  \n",
      "ATOM   2591  HA  GLU D 346      -7.268 -10.043  -1.408  1.00  0.32           H  \n",
      "ATOM   2592  HB2 GLU D 346      -7.506  -9.558  -3.833  1.00  0.41           H  \n",
      "ATOM   2593  HB3 GLU D 346      -5.760  -9.338  -3.944  1.00  0.38           H  \n",
      "ATOM   2594  HG2 GLU D 346      -5.318 -11.588  -3.708  1.00  0.84           H  \n",
      "ATOM   2595  HG3 GLU D 346      -6.751 -11.873  -2.719  1.00  0.82           H  \n",
      "ATOM   2596  N   ALA D 347      -4.195  -8.950  -1.418  1.00  0.24           N  \n",
      "ATOM   2597  CA  ALA D 347      -2.833  -9.245  -0.887  1.00  0.25           C  \n",
      "ATOM   2598  C   ALA D 347      -2.914  -9.571   0.607  1.00  0.24           C  \n",
      "ATOM   2599  O   ALA D 347      -2.492 -10.622   1.046  1.00  0.25           O  \n",
      "ATOM   2600  CB  ALA D 347      -1.932  -8.025  -1.094  1.00  0.28           C  \n",
      "ATOM   2601  H   ALA D 347      -4.388  -8.072  -1.809  1.00  0.25           H  \n",
      "ATOM   2602  HA  ALA D 347      -2.418 -10.090  -1.416  1.00  0.27           H  \n",
      "ATOM   2603  HB1 ALA D 347      -2.534  -7.180  -1.394  1.00  0.99           H  \n",
      "ATOM   2604  HB2 ALA D 347      -1.422  -7.794  -0.171  1.00  1.04           H  \n",
      "ATOM   2605  HB3 ALA D 347      -1.206  -8.241  -1.863  1.00  1.07           H  \n",
      "ATOM   2606  N   LEU D 348      -3.445  -8.674   1.391  1.00  0.24           N  \n",
      "ATOM   2607  CA  LEU D 348      -3.545  -8.926   2.858  1.00  0.26           C  \n",
      "ATOM   2608  C   LEU D 348      -4.423 -10.153   3.115  1.00  0.28           C  \n",
      "ATOM   2609  O   LEU D 348      -4.099 -11.002   3.921  1.00  0.31           O  \n",
      "ATOM   2610  CB  LEU D 348      -4.170  -7.706   3.541  1.00  0.26           C  \n",
      "ATOM   2611  CG  LEU D 348      -3.121  -6.602   3.692  1.00  0.26           C  \n",
      "ATOM   2612  CD1 LEU D 348      -1.892  -7.157   4.412  1.00  0.29           C  \n",
      "ATOM   2613  CD2 LEU D 348      -2.713  -6.092   2.308  1.00  0.26           C  \n",
      "ATOM   2614  H   LEU D 348      -3.773  -7.830   1.017  1.00  0.24           H  \n",
      "ATOM   2615  HA  LEU D 348      -2.558  -9.099   3.261  1.00  0.26           H  \n",
      "ATOM   2616  HB2 LEU D 348      -4.992  -7.342   2.942  1.00  0.27           H  \n",
      "ATOM   2617  HB3 LEU D 348      -4.534  -7.989   4.517  1.00  0.29           H  \n",
      "ATOM   2618  HG  LEU D 348      -3.537  -5.788   4.268  1.00  0.28           H  \n",
      "ATOM   2619 HD11 LEU D 348      -2.209  -7.785   5.232  1.00  1.11           H  \n",
      "ATOM   2620 HD12 LEU D 348      -1.301  -7.739   3.720  1.00  1.04           H  \n",
      "ATOM   2621 HD13 LEU D 348      -1.299  -6.340   4.794  1.00  1.01           H  \n",
      "ATOM   2622 HD21 LEU D 348      -3.587  -6.024   1.679  1.00  1.02           H  \n",
      "ATOM   2623 HD22 LEU D 348      -2.261  -5.116   2.403  1.00  1.04           H  \n",
      "ATOM   2624 HD23 LEU D 348      -2.004  -6.776   1.866  1.00  1.04           H  \n",
      "ATOM   2625  N   GLU D 349      -5.537 -10.249   2.442  1.00  0.31           N  \n",
      "ATOM   2626  CA  GLU D 349      -6.441 -11.415   2.655  1.00  0.35           C  \n",
      "ATOM   2627  C   GLU D 349      -5.689 -12.716   2.367  1.00  0.33           C  \n",
      "ATOM   2628  O   GLU D 349      -5.932 -13.733   2.987  1.00  0.35           O  \n",
      "ATOM   2629  CB  GLU D 349      -7.645 -11.303   1.717  1.00  0.40           C  \n",
      "ATOM   2630  CG  GLU D 349      -8.433 -10.035   2.051  1.00  0.47           C  \n",
      "ATOM   2631  CD  GLU D 349      -9.932 -10.342   2.033  1.00  1.12           C  \n",
      "ATOM   2632  OE1 GLU D 349     -10.329 -11.212   1.275  1.00  1.91           O  \n",
      "ATOM   2633  OE2 GLU D 349     -10.657  -9.703   2.777  1.00  1.66           O  \n",
      "ATOM   2634  H   GLU D 349      -5.783  -9.550   1.801  1.00  0.32           H  \n",
      "ATOM   2635  HA  GLU D 349      -6.785 -11.420   3.678  1.00  0.39           H  \n",
      "ATOM   2636  HB2 GLU D 349      -7.302 -11.256   0.694  1.00  0.39           H  \n",
      "ATOM   2637  HB3 GLU D 349      -8.283 -12.165   1.845  1.00  0.47           H  \n",
      "ATOM   2638  HG2 GLU D 349      -8.149  -9.683   3.032  1.00  0.89           H  \n",
      "ATOM   2639  HG3 GLU D 349      -8.216  -9.272   1.318  1.00  0.78           H  \n",
      "ATOM   2640  N   LEU D 350      -4.777 -12.697   1.436  1.00  0.30           N  \n",
      "ATOM   2641  CA  LEU D 350      -4.015 -13.937   1.121  1.00  0.30           C  \n",
      "ATOM   2642  C   LEU D 350      -3.200 -14.346   2.347  1.00  0.29           C  \n",
      "ATOM   2643  O   LEU D 350      -3.158 -15.501   2.723  1.00  0.33           O  \n",
      "ATOM   2644  CB  LEU D 350      -3.074 -13.676  -0.058  1.00  0.30           C  \n",
      "ATOM   2645  CG  LEU D 350      -2.663 -15.009  -0.687  1.00  0.36           C  \n",
      "ATOM   2646  CD1 LEU D 350      -2.201 -14.776  -2.126  1.00  0.43           C  \n",
      "ATOM   2647  CD2 LEU D 350      -1.517 -15.624   0.120  1.00  0.40           C  \n",
      "ATOM   2648  H   LEU D 350      -4.590 -11.869   0.949  1.00  0.31           H  \n",
      "ATOM   2649  HA  LEU D 350      -4.704 -14.728   0.866  1.00  0.33           H  \n",
      "ATOM   2650  HB2 LEU D 350      -3.582 -13.071  -0.796  1.00  0.34           H  \n",
      "ATOM   2651  HB3 LEU D 350      -2.194 -13.157   0.290  1.00  0.30           H  \n",
      "ATOM   2652  HG  LEU D 350      -3.507 -15.683  -0.686  1.00  0.51           H  \n",
      "ATOM   2653 HD11 LEU D 350      -1.809 -13.774  -2.221  1.00  1.09           H  \n",
      "ATOM   2654 HD12 LEU D 350      -1.430 -15.489  -2.377  1.00  1.14           H  \n",
      "ATOM   2655 HD13 LEU D 350      -3.038 -14.900  -2.798  1.00  1.08           H  \n",
      "ATOM   2656 HD21 LEU D 350      -0.691 -14.929   0.160  1.00  1.17           H  \n",
      "ATOM   2657 HD22 LEU D 350      -1.855 -15.838   1.121  1.00  1.03           H  \n",
      "ATOM   2658 HD23 LEU D 350      -1.197 -16.539  -0.352  1.00  1.11           H  \n",
      "ATOM   2659  N   LYS D 351      -2.557 -13.404   2.979  1.00  0.29           N  \n",
      "ATOM   2660  CA  LYS D 351      -1.749 -13.733   4.185  1.00  0.32           C  \n",
      "ATOM   2661  C   LYS D 351      -2.660 -14.349   5.249  1.00  0.38           C  \n",
      "ATOM   2662  O   LYS D 351      -2.304 -15.303   5.911  1.00  0.43           O  \n",
      "ATOM   2663  CB  LYS D 351      -1.115 -12.453   4.731  1.00  0.35           C  \n",
      "ATOM   2664  CG  LYS D 351      -0.071 -12.804   5.792  1.00  0.44           C  \n",
      "ATOM   2665  CD  LYS D 351       0.709 -11.544   6.170  1.00  0.48           C  \n",
      "ATOM   2666  CE  LYS D 351       2.033 -11.936   6.829  1.00  0.81           C  \n",
      "ATOM   2667  NZ  LYS D 351       2.460 -10.856   7.763  1.00  1.57           N  \n",
      "ATOM   2668  H   LYS D 351      -2.610 -12.478   2.663  1.00  0.28           H  \n",
      "ATOM   2669  HA  LYS D 351      -0.974 -14.437   3.920  1.00  0.33           H  \n",
      "ATOM   2670  HB2 LYS D 351      -0.640 -11.915   3.923  1.00  0.38           H  \n",
      "ATOM   2671  HB3 LYS D 351      -1.881 -11.833   5.174  1.00  0.39           H  \n",
      "ATOM   2672  HG2 LYS D 351      -0.566 -13.200   6.668  1.00  0.55           H  \n",
      "ATOM   2673  HG3 LYS D 351       0.610 -13.543   5.398  1.00  0.55           H  \n",
      "ATOM   2674  HD2 LYS D 351       0.906 -10.964   5.280  1.00  0.56           H  \n",
      "ATOM   2675  HD3 LYS D 351       0.126 -10.954   6.862  1.00  0.63           H  \n",
      "ATOM   2676  HE2 LYS D 351       1.904 -12.857   7.377  1.00  1.33           H  \n",
      "ATOM   2677  HE3 LYS D 351       2.787 -12.073   6.068  1.00  1.19           H  \n",
      "ATOM   2678  HZ1 LYS D 351       2.042  -9.953   7.464  1.00  2.15           H  \n",
      "ATOM   2679  HZ2 LYS D 351       2.140 -11.085   8.726  1.00  2.04           H  \n",
      "ATOM   2680  HZ3 LYS D 351       3.496 -10.777   7.751  1.00  1.99           H  \n",
      "ATOM   2681  N   ASP D 352      -3.837 -13.809   5.414  1.00  0.43           N  \n",
      "ATOM   2682  CA  ASP D 352      -4.777 -14.359   6.430  1.00  0.51           C  \n",
      "ATOM   2683  C   ASP D 352      -5.013 -15.847   6.159  1.00  0.51           C  \n",
      "ATOM   2684  O   ASP D 352      -5.063 -16.653   7.067  1.00  0.58           O  \n",
      "ATOM   2685  CB  ASP D 352      -6.110 -13.611   6.349  1.00  0.59           C  \n",
      "ATOM   2686  CG  ASP D 352      -5.949 -12.211   6.944  1.00  0.67           C  \n",
      "ATOM   2687  OD1 ASP D 352      -5.015 -11.528   6.556  1.00  1.12           O  \n",
      "ATOM   2688  OD2 ASP D 352      -6.763 -11.845   7.776  1.00  1.44           O  \n",
      "ATOM   2689  H   ASP D 352      -4.104 -13.041   4.866  1.00  0.44           H  \n",
      "ATOM   2690  HA  ASP D 352      -4.354 -14.235   7.416  1.00  0.55           H  \n",
      "ATOM   2691  HB2 ASP D 352      -6.415 -13.532   5.316  1.00  0.56           H  \n",
      "ATOM   2692  HB3 ASP D 352      -6.860 -14.152   6.906  1.00  0.69           H  \n",
      "ATOM   2693  N   ALA D 353      -5.164 -16.217   4.917  1.00  0.49           N  \n",
      "ATOM   2694  CA  ALA D 353      -5.404 -17.651   4.590  1.00  0.55           C  \n",
      "ATOM   2695  C   ALA D 353      -4.159 -18.472   4.935  1.00  0.54           C  \n",
      "ATOM   2696  O   ALA D 353      -4.244 -19.645   5.238  1.00  0.64           O  \n",
      "ATOM   2697  CB  ALA D 353      -5.707 -17.788   3.096  1.00  0.63           C  \n",
      "ATOM   2698  H   ALA D 353      -5.125 -15.550   4.199  1.00  0.49           H  \n",
      "ATOM   2699  HA  ALA D 353      -6.244 -18.014   5.162  1.00  0.63           H  \n",
      "ATOM   2700  HB1 ALA D 353      -6.368 -16.992   2.788  1.00  1.30           H  \n",
      "ATOM   2701  HB2 ALA D 353      -4.786 -17.728   2.535  1.00  1.11           H  \n",
      "ATOM   2702  HB3 ALA D 353      -6.179 -18.741   2.911  1.00  1.22           H  \n",
      "ATOM   2703  N   GLN D 354      -3.004 -17.866   4.894  1.00  0.51           N  \n",
      "ATOM   2704  CA  GLN D 354      -1.760 -18.617   5.224  1.00  0.59           C  \n",
      "ATOM   2705  C   GLN D 354      -1.556 -18.623   6.740  1.00  0.66           C  \n",
      "ATOM   2706  O   GLN D 354      -0.743 -19.359   7.264  1.00  0.85           O  \n",
      "ATOM   2707  CB  GLN D 354      -0.564 -17.939   4.552  1.00  0.62           C  \n",
      "ATOM   2708  CG  GLN D 354       0.052 -18.889   3.524  1.00  0.96           C  \n",
      "ATOM   2709  CD  GLN D 354       1.383 -18.317   3.034  1.00  0.86           C  \n",
      "ATOM   2710  OE1 GLN D 354       2.268 -19.053   2.645  1.00  1.21           O  \n",
      "ATOM   2711  NE2 GLN D 354       1.563 -17.024   3.038  1.00  0.71           N  \n",
      "ATOM   2712  H   GLN D 354      -2.956 -16.918   4.649  1.00  0.49           H  \n",
      "ATOM   2713  HA  GLN D 354      -1.846 -19.632   4.867  1.00  0.68           H  \n",
      "ATOM   2714  HB2 GLN D 354      -0.894 -17.037   4.058  1.00  0.83           H  \n",
      "ATOM   2715  HB3 GLN D 354       0.175 -17.692   5.299  1.00  0.90           H  \n",
      "ATOM   2716  HG2 GLN D 354       0.219 -19.854   3.981  1.00  1.38           H  \n",
      "ATOM   2717  HG3 GLN D 354      -0.621 -18.999   2.686  1.00  1.40           H  \n",
      "ATOM   2718 HE21 GLN D 354       0.850 -16.431   3.352  1.00  0.74           H  \n",
      "ATOM   2719 HE22 GLN D 354       2.413 -16.647   2.726  1.00  0.84           H  \n",
      "ATOM   2720  N   ALA D 355      -2.289 -17.808   7.449  1.00  0.68           N  \n",
      "ATOM   2721  CA  ALA D 355      -2.140 -17.766   8.930  1.00  0.82           C  \n",
      "ATOM   2722  C   ALA D 355      -2.763 -19.022   9.543  1.00  0.87           C  \n",
      "ATOM   2723  O   ALA D 355      -2.191 -19.651  10.412  1.00  1.22           O  \n",
      "ATOM   2724  CB  ALA D 355      -2.847 -16.522   9.473  1.00  1.01           C  \n",
      "ATOM   2725  H   ALA D 355      -2.938 -17.224   7.006  1.00  0.72           H  \n",
      "ATOM   2726  HA  ALA D 355      -1.092 -17.723   9.185  1.00  0.94           H  \n",
      "ATOM   2727  HB1 ALA D 355      -3.324 -15.996   8.659  1.00  1.50           H  \n",
      "ATOM   2728  HB2 ALA D 355      -3.592 -16.818  10.197  1.00  1.60           H  \n",
      "ATOM   2729  HB3 ALA D 355      -2.123 -15.874   9.944  1.00  1.28           H  \n",
      "ATOM   2730  N   GLY D 356      -3.935 -19.390   9.099  1.00  0.98           N  \n",
      "ATOM   2731  CA  GLY D 356      -4.603 -20.604   9.653  1.00  1.18           C  \n",
      "ATOM   2732  C   GLY D 356      -4.046 -21.869   8.991  1.00  1.14           C  \n",
      "ATOM   2733  O   GLY D 356      -4.625 -22.932   9.090  1.00  1.50           O  \n",
      "ATOM   2734  H   GLY D 356      -4.378 -18.866   8.399  1.00  1.20           H  \n",
      "ATOM   2735  HA2 GLY D 356      -4.429 -20.651  10.719  1.00  1.39           H  \n",
      "ATOM   2736  HA3 GLY D 356      -5.665 -20.544   9.467  1.00  1.48           H  \n",
      "ATOM   2737  N   LYS D 357      -2.934 -21.770   8.313  1.00  1.30           N  \n",
      "ATOM   2738  CA  LYS D 357      -2.363 -22.976   7.649  1.00  1.52           C  \n",
      "ATOM   2739  C   LYS D 357      -1.512 -23.756   8.653  1.00  1.98           C  \n",
      "ATOM   2740  O   LYS D 357      -0.481 -23.294   9.100  1.00  2.51           O  \n",
      "ATOM   2741  CB  LYS D 357      -1.495 -22.544   6.466  1.00  1.87           C  \n",
      "ATOM   2742  CG  LYS D 357      -1.594 -23.590   5.353  1.00  2.24           C  \n",
      "ATOM   2743  CD  LYS D 357      -0.226 -23.765   4.690  1.00  2.96           C  \n",
      "ATOM   2744  CE  LYS D 357      -0.337 -24.784   3.554  1.00  3.50           C  \n",
      "ATOM   2745  NZ  LYS D 357       1.010 -25.018   2.962  1.00  4.18           N  \n",
      "ATOM   2746  H   LYS D 357      -2.479 -20.908   8.236  1.00  1.59           H  \n",
      "ATOM   2747  HA  LYS D 357      -3.166 -23.605   7.295  1.00  1.70           H  \n",
      "ATOM   2748  HB2 LYS D 357      -1.840 -21.590   6.096  1.00  2.21           H  \n",
      "ATOM   2749  HB3 LYS D 357      -0.467 -22.457   6.785  1.00  2.13           H  \n",
      "ATOM   2750  HG2 LYS D 357      -1.915 -24.532   5.773  1.00  2.39           H  \n",
      "ATOM   2751  HG3 LYS D 357      -2.310 -23.262   4.614  1.00  2.52           H  \n",
      "ATOM   2752  HD2 LYS D 357       0.105 -22.816   4.293  1.00  3.42           H  \n",
      "ATOM   2753  HD3 LYS D 357       0.486 -24.119   5.420  1.00  3.18           H  \n",
      "ATOM   2754  HE2 LYS D 357      -0.726 -25.714   3.942  1.00  3.68           H  \n",
      "ATOM   2755  HE3 LYS D 357      -1.004 -24.405   2.794  1.00  3.76           H  \n",
      "ATOM   2756  HZ1 LYS D 357       1.738 -24.618   3.586  1.00  4.47           H  \n",
      "ATOM   2757  HZ2 LYS D 357       1.170 -26.040   2.857  1.00  4.55           H  \n",
      "ATOM   2758  HZ3 LYS D 357       1.061 -24.561   2.028  1.00  4.44           H  \n",
      "ATOM   2759  N   GLU D 358      -1.940 -24.935   9.011  1.00  2.42           N  \n",
      "ATOM   2760  CA  GLU D 358      -1.161 -25.747   9.987  1.00  3.19           C  \n",
      "ATOM   2761  C   GLU D 358       0.309 -25.794   9.551  1.00  3.44           C  \n",
      "ATOM   2762  O   GLU D 358       0.598 -26.042   8.397  1.00  3.47           O  \n",
      "ATOM   2763  CB  GLU D 358      -1.727 -27.168  10.029  1.00  3.90           C  \n",
      "ATOM   2764  CG  GLU D 358      -1.794 -27.653  11.480  1.00  4.50           C  \n",
      "ATOM   2765  CD  GLU D 358      -2.154 -29.139  11.504  1.00  5.22           C  \n",
      "ATOM   2766  OE1 GLU D 358      -1.554 -29.886  10.749  1.00  5.62           O  \n",
      "ATOM   2767  OE2 GLU D 358      -3.023 -29.505  12.278  1.00  5.67           O  \n",
      "ATOM   2768  H   GLU D 358      -2.775 -25.287   8.639  1.00  2.58           H  \n",
      "ATOM   2769  HA  GLU D 358      -1.241 -25.299  10.965  1.00  3.49           H  \n",
      "ATOM   2770  HB2 GLU D 358      -2.719 -27.173   9.602  1.00  4.21           H  \n",
      "ATOM   2771  HB3 GLU D 358      -1.088 -27.827   9.461  1.00  4.17           H  \n",
      "ATOM   2772  HG2 GLU D 358      -0.834 -27.507  11.952  1.00  4.62           H  \n",
      "ATOM   2773  HG3 GLU D 358      -2.547 -27.093  12.013  1.00  4.75           H  \n",
      "ATOM   2774  N   PRO D 359       1.199 -25.558  10.485  1.00  4.10           N  \n",
      "ATOM   2775  CA  PRO D 359       2.645 -25.573  10.206  1.00  4.76           C  \n",
      "ATOM   2776  C   PRO D 359       3.071 -26.962   9.725  1.00  4.95           C  \n",
      "ATOM   2777  O   PRO D 359       2.300 -27.900   9.747  1.00  4.97           O  \n",
      "ATOM   2778  CB  PRO D 359       3.310 -25.235  11.548  1.00  5.65           C  \n",
      "ATOM   2779  CG  PRO D 359       2.185 -25.033  12.594  1.00  5.61           C  \n",
      "ATOM   2780  CD  PRO D 359       0.842 -25.260  11.884  1.00  4.63           C  \n",
      "ATOM   2781  HA  PRO D 359       2.896 -24.824   9.472  1.00  4.86           H  \n",
      "ATOM   2782  HB2 PRO D 359       3.955 -26.048  11.852  1.00  6.07           H  \n",
      "ATOM   2783  HB3 PRO D 359       3.884 -24.326  11.455  1.00  6.12           H  \n",
      "ATOM   2784  HG2 PRO D 359       2.301 -25.744  13.399  1.00  6.07           H  \n",
      "ATOM   2785  HG3 PRO D 359       2.222 -24.027  12.983  1.00  6.05           H  \n",
      "ATOM   2786  HD2 PRO D 359       0.324 -26.097  12.328  1.00  4.74           H  \n",
      "ATOM   2787  HD3 PRO D 359       0.234 -24.370  11.932  1.00  4.55           H  \n",
      "ATOM   2788  N   GLY D 360       4.292 -27.100   9.291  1.00  5.47           N  \n",
      "ATOM   2789  CA  GLY D 360       4.765 -28.428   8.809  1.00  5.98           C  \n",
      "ATOM   2790  C   GLY D 360       6.275 -28.542   9.023  1.00  6.80           C  \n",
      "ATOM   2791  O   GLY D 360       6.761 -27.972   9.986  1.00  7.27           O  \n",
      "ATOM   2792  OXT GLY D 360       6.919 -29.196   8.219  1.00  7.20           O  \n",
      "ATOM   2793  H   GLY D 360       4.898 -26.331   9.281  1.00  5.75           H  \n",
      "ATOM   2794  HA2 GLY D 360       4.261 -29.210   9.360  1.00  6.02           H  \n",
      "ATOM   2795  HA3 GLY D 360       4.545 -28.529   7.757  1.00  6.08           H  \n",
      "TER    2796      GLY D 360                                                      \n",
      "MASTER      120    0    0    4    4    0    0    6 2792    4    0   16          \n",
      "END                                                                             \n"
     ]
    }
   ],
   "source": [
    "# code to preview the whole file here\n",
    "!type data\\1OLG.pdb\n",
    "\n",
    "# code to preview the whole file in a its original program (here pymol is used)\n",
    "#!>cat data\\1OLG.pdb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8c56cb5e",
   "metadata": {},
   "source": [
    "We want to save only the 21st column with the entry for the chain 'A' in the lines that describe 'ATOM'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a73c9d24",
   "metadata": {},
   "outputs": [],
   "source": [
    "# extract from 1OLG all ATOM lines describing chain A\n",
    "with open('data/1OLG.pdb', 'r') as f, open('atoms_chain_A.txt', 'w') as f_out:\n",
    "    # Put the ATOM lines from chain A in new file\n",
    "    for line in f:\n",
    "        if len(line) > 21 and line[:4] == 'ATOM' and line[21] == 'A':\n",
    "            f_out.write(line)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9451a02",
   "metadata": {},
   "source": [
    "preview the first 10 or the last 10 lines of a file (conda install posix)\n",
    "\n",
    "-> doesnt work!\n",
    "\n",
    "neither `!head -10 atoms_chain_A.txt` nor `!tail -10 atoms_chain_A.txt`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "87ee9e8d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ATOM      1  N   LYS A 319      18.634  25.437  10.685  1.00  4.81           N  \n",
      "ATOM      2  CA  LYS A 319      17.984  25.295   9.354  1.00  4.32           C  \n",
      "ATOM      3  C   LYS A 319      18.160  23.876   8.818  1.00  3.74           C  \n",
      "ATOM      4  O   LYS A 319      19.259  23.441   8.537  1.00  3.67           O  \n",
      "ATOM      5  CB  LYS A 319      18.609  26.282   8.371  1.00  4.67           C  \n",
      "ATOM      6  CG  LYS A 319      18.003  26.056   6.986  1.00  5.15           C  \n",
      "ATOM      7  CD  LYS A 319      16.476  26.057   7.091  1.00  5.90           C  \n",
      "ATOM      8  CE  LYS A 319      16.014  27.341   7.784  1.00  6.51           C  \n",
      "ATOM      9  NZ  LYS A 319      16.388  28.518   6.952  1.00  7.33           N  \n",
      "ATOM     10  H1  LYS A 319      18.414  24.606  11.281  1.00  5.09           H  \n",
      "ATOM     11  H2  LYS A 319      19.658  25.516  10.555  1.00  5.18           H  \n",
      "ATOM     12  H3  LYS A 319      18.282  26.296  11.150  1.00  4.96           H  \n",
      "ATOM     13  HA  LYS A 319      16.934  25.509   9.452  1.00  4.66           H  \n",
      "ATOM     14  HB2 LYS A 319      18.399  27.288   8.697  1.00  4.96           H  \n",
      "ATOM     15  HB3 LYS A 319      19.679  26.132   8.326  1.00  4.81           H  \n",
      "ATOM     16  HG2 LYS A 319      18.320  26.846   6.324  1.00  5.22           H  \n",
      "ATOM     17  HG3 LYS A 319      18.334  25.104   6.598  1.00  5.32           H  \n",
      "ATOM     18  HD2 LYS A 319      16.046  26.005   6.102  1.00  6.19           H  \n",
      "ATOM     19  HD3 LYS A 319      16.154  25.204   7.670  1.00  6.08           H  \n",
      "ATOM     20  HE2 LYS A 319      14.941  27.316   7.911  1.00  6.70           H  \n",
      "ATOM     21  HE3 LYS A 319      16.488  27.419   8.752  1.00  6.49           H  \n",
      "ATOM     22  HZ1 LYS A 319      17.395  28.460   6.701  1.00  7.60           H  \n",
      "ATOM     23  HZ2 LYS A 319      15.816  28.526   6.085  1.00  7.57           H  \n",
      "ATOM     24  HZ3 LYS A 319      16.214  29.392   7.491  1.00  7.66           H  \n",
      "ATOM     25  N   LYS A 320      17.063  23.177   8.658  1.00  3.85           N  \n",
      "ATOM     26  CA  LYS A 320      17.085  21.778   8.123  1.00  3.82           C  \n",
      "ATOM     27  C   LYS A 320      17.727  20.814   9.132  1.00  3.38           C  \n",
      "ATOM     28  O   LYS A 320      17.048  20.031   9.762  1.00  3.69           O  \n",
      "ATOM     29  CB  LYS A 320      17.844  21.740   6.797  1.00  4.21           C  \n",
      "ATOM     30  CG  LYS A 320      16.917  22.223   5.676  1.00  5.00           C  \n",
      "ATOM     31  CD  LYS A 320      17.701  22.331   4.365  1.00  5.81           C  \n",
      "ATOM     32  CE  LYS A 320      18.474  23.652   4.334  1.00  6.54           C  \n",
      "ATOM     33  NZ  LYS A 320      17.555  24.758   3.943  1.00  7.21           N  \n",
      "ATOM     34  H   LYS A 320      16.203  23.590   8.879  1.00  4.30           H  \n",
      "ATOM     35  HA  LYS A 320      16.067  21.461   7.949  1.00  4.36           H  \n",
      "ATOM     36  HB2 LYS A 320      18.707  22.383   6.855  1.00  4.26           H  \n",
      "ATOM     37  HB3 LYS A 320      18.159  20.728   6.592  1.00  4.38           H  \n",
      "ATOM     38  HG2 LYS A 320      16.104  21.520   5.553  1.00  5.26           H  \n",
      "ATOM     39  HG3 LYS A 320      16.516  23.192   5.934  1.00  5.13           H  \n",
      "ATOM     40  HD2 LYS A 320      18.394  21.505   4.293  1.00  5.89           H  \n",
      "ATOM     41  HD3 LYS A 320      17.015  22.299   3.532  1.00  6.15           H  \n",
      "ATOM     42  HE2 LYS A 320      18.883  23.852   5.313  1.00  6.83           H  \n",
      "ATOM     43  HE3 LYS A 320      19.278  23.581   3.617  1.00  6.61           H  \n",
      "ATOM     44  HZ1 LYS A 320      16.599  24.558   4.300  1.00  7.43           H  \n",
      "ATOM     45  HZ2 LYS A 320      17.897  25.653   4.348  1.00  7.46           H  \n",
      "ATOM     46  HZ3 LYS A 320      17.529  24.838   2.906  1.00  7.51           H  \n",
      "ATOM     47  N   LYS A 321      19.020  20.847   9.285  1.00  3.05           N  \n",
      "ATOM     48  CA  LYS A 321      19.688  19.926  10.251  1.00  2.81           C  \n",
      "ATOM     49  C   LYS A 321      19.570  18.479   9.751  1.00  2.50           C  \n",
      "ATOM     50  O   LYS A 321      18.512  18.062   9.322  1.00  2.62           O  \n",
      "ATOM     51  CB  LYS A 321      19.029  20.015  11.637  1.00  3.34           C  \n",
      "ATOM     52  CG  LYS A 321      18.339  21.373  11.842  1.00  4.02           C  \n",
      "ATOM     53  CD  LYS A 321      16.841  21.152  12.059  1.00  4.77           C  \n",
      "ATOM     54  CE  LYS A 321      16.501  21.307  13.544  1.00  5.69           C  \n",
      "ATOM     55  NZ  LYS A 321      17.082  20.148  14.319  1.00  6.47           N  \n",
      "ATOM     56  H   LYS A 321      19.556  21.470   8.761  1.00  3.26           H  \n",
      "ATOM     57  HA  LYS A 321      20.727  20.200  10.333  1.00  3.16           H  \n",
      "ATOM     58  HB2 LYS A 321      18.300  19.225  11.731  1.00  3.51           H  \n",
      "ATOM     59  HB3 LYS A 321      19.789  19.886  12.396  1.00  3.66           H  \n",
      "ATOM     60  HG2 LYS A 321      18.756  21.853  12.711  1.00  4.36           H  \n",
      "ATOM     61  HG3 LYS A 321      18.488  21.998  10.978  1.00  4.11           H  \n",
      "ATOM     62  HD2 LYS A 321      16.288  21.876  11.488  1.00  4.81           H  \n",
      "ATOM     63  HD3 LYS A 321      16.572  20.152  11.738  1.00  5.02           H  \n",
      "ATOM     64  HE2 LYS A 321      16.904  22.244  13.913  1.00  5.93           H  \n",
      "ATOM     65  HE3 LYS A 321      15.432  21.313  13.656  1.00  5.90           H  \n",
      "ATOM     66  HZ1 LYS A 321      17.162  19.350  13.688  1.00  6.74           H  \n",
      "ATOM     67  HZ2 LYS A 321      18.033  20.394  14.685  1.00  6.71           H  \n",
      "ATOM     68  HZ3 LYS A 321      16.460  19.881  15.120  1.00  6.81           H  \n",
      "ATOM     69  N   PRO A 322      20.660  17.750   9.832  1.00  2.87           N  \n",
      "ATOM     70  CA  PRO A 322      20.701  16.337   9.404  1.00  3.30           C  \n",
      "ATOM     71  C   PRO A 322      20.002  15.426  10.428  1.00  2.78           C  \n",
      "ATOM     72  O   PRO A 322      20.165  14.222  10.401  1.00  2.72           O  \n",
      "ATOM     73  CB  PRO A 322      22.198  16.001   9.352  1.00  4.33           C  \n",
      "ATOM     74  CG  PRO A 322      22.963  17.169  10.023  1.00  4.53           C  \n",
      "ATOM     75  CD  PRO A 322      21.944  18.275  10.333  1.00  3.61           C  \n",
      "ATOM     76  HA  PRO A 322      20.264  16.220   8.427  1.00  3.71           H  \n",
      "ATOM     77  HB2 PRO A 322      22.386  15.079   9.886  1.00  4.51           H  \n",
      "ATOM     78  HB3 PRO A 322      22.516  15.903   8.326  1.00  5.02           H  \n",
      "ATOM     79  HG2 PRO A 322      23.425  16.826  10.939  1.00  4.97           H  \n",
      "ATOM     80  HG3 PRO A 322      23.718  17.549   9.351  1.00  5.15           H  \n",
      "ATOM     81  HD2 PRO A 322      21.893  18.452  11.398  1.00  3.76           H  \n",
      "ATOM     82  HD3 PRO A 322      22.204  19.182   9.810  1.00  3.74           H  \n",
      "ATOM     83  N   LEU A 323      19.244  15.979  11.341  1.00  2.70           N  \n",
      "ATOM     84  CA  LEU A 323      18.560  15.130  12.361  1.00  2.39           C  \n",
      "ATOM     85  C   LEU A 323      17.248  14.572  11.801  1.00  1.78           C  \n",
      "ATOM     86  O   LEU A 323      16.196  14.744  12.384  1.00  1.89           O  \n",
      "ATOM     87  CB  LEU A 323      18.262  15.973  13.605  1.00  2.93           C  \n",
      "ATOM     88  CG  LEU A 323      19.571  16.320  14.320  1.00  3.65           C  \n",
      "ATOM     89  CD1 LEU A 323      19.263  16.901  15.702  1.00  4.35           C  \n",
      "ATOM     90  CD2 LEU A 323      20.417  15.054  14.480  1.00  3.83           C  \n",
      "ATOM     91  H   LEU A 323      19.132  16.949  11.365  1.00  3.05           H  \n",
      "ATOM     92  HA  LEU A 323      19.207  14.310  12.632  1.00  2.54           H  \n",
      "ATOM     93  HB2 LEU A 323      17.761  16.884  13.310  1.00  3.10           H  \n",
      "ATOM     94  HB3 LEU A 323      17.626  15.414  14.275  1.00  2.89           H  \n",
      "ATOM     95  HG  LEU A 323      20.116  17.048  13.738  1.00  3.76           H  \n",
      "ATOM     96 HD11 LEU A 323      18.450  16.350  16.151  1.00  4.70           H  \n",
      "ATOM     97 HD12 LEU A 323      20.139  16.824  16.328  1.00  4.63           H  \n",
      "ATOM     98 HD13 LEU A 323      18.982  17.939  15.601  1.00  4.59           H  \n",
      "ATOM     99 HD21 LEU A 323      19.783  14.235  14.789  1.00  4.03           H  \n",
      "ATOM    100 HD22 LEU A 323      20.881  14.811  13.535  1.00  3.92           H  \n",
      "ATOM    101 HD23 LEU A 323      21.180  15.220  15.225  1.00  4.16           H  \n",
      "ATOM    102  N   ASP A 324      17.296  13.893  10.687  1.00  1.51           N  \n",
      "ATOM    103  CA  ASP A 324      16.045  13.320  10.116  1.00  1.33           C  \n",
      "ATOM    104  C   ASP A 324      15.714  12.012  10.841  1.00  1.07           C  \n",
      "ATOM    105  O   ASP A 324      16.418  11.596  11.741  1.00  1.04           O  \n",
      "ATOM    106  CB  ASP A 324      16.239  13.043   8.624  1.00  1.76           C  \n",
      "ATOM    107  CG  ASP A 324      16.759  14.305   7.931  1.00  2.08           C  \n",
      "ATOM    108  OD1 ASP A 324      16.250  15.373   8.229  1.00  2.42           O  \n",
      "ATOM    109  OD2 ASP A 324      17.656  14.180   7.114  1.00  2.48           O  \n",
      "ATOM    110  H   ASP A 324      18.151  13.753  10.233  1.00  1.76           H  \n",
      "ATOM    111  HA  ASP A 324      15.234  14.021  10.251  1.00  1.53           H  \n",
      "ATOM    112  HB2 ASP A 324      16.951  12.243   8.496  1.00  1.90           H  \n",
      "ATOM    113  HB3 ASP A 324      15.294  12.758   8.185  1.00  2.04           H  \n",
      "ATOM    114  N   GLY A 325      14.651  11.359  10.459  1.00  0.98           N  \n",
      "ATOM    115  CA  GLY A 325      14.281  10.078  11.129  1.00  0.84           C  \n",
      "ATOM    116  C   GLY A 325      15.307   8.999  10.773  1.00  0.70           C  \n",
      "ATOM    117  O   GLY A 325      16.298   9.264  10.122  1.00  0.68           O  \n",
      "ATOM    118  H   GLY A 325      14.095  11.709   9.731  1.00  1.08           H  \n",
      "ATOM    119  HA2 GLY A 325      14.268  10.224  12.200  1.00  0.89           H  \n",
      "ATOM    120  HA3 GLY A 325      13.304   9.765  10.796  1.00  0.91           H  \n",
      "ATOM    121  N   GLU A 326      15.076   7.783  11.190  1.00  0.65           N  \n",
      "ATOM    122  CA  GLU A 326      16.040   6.691  10.869  1.00  0.58           C  \n",
      "ATOM    123  C   GLU A 326      16.181   6.571   9.351  1.00  0.52           C  \n",
      "ATOM    124  O   GLU A 326      15.633   7.358   8.605  1.00  0.51           O  \n",
      "ATOM    125  CB  GLU A 326      15.526   5.368  11.443  1.00  0.64           C  \n",
      "ATOM    126  CG  GLU A 326      15.658   5.387  12.967  1.00  0.75           C  \n",
      "ATOM    127  CD  GLU A 326      16.782   4.441  13.396  1.00  1.06           C  \n",
      "ATOM    128  OE1 GLU A 326      17.932   4.836  13.295  1.00  1.70           O  \n",
      "ATOM    129  OE2 GLU A 326      16.473   3.339  13.817  1.00  1.64           O  \n",
      "ATOM    130  H   GLU A 326      14.270   7.588  11.712  1.00  0.71           H  \n",
      "ATOM    131  HA  GLU A 326      17.002   6.922  11.303  1.00  0.59           H  \n",
      "ATOM    132  HB2 GLU A 326      14.489   5.237  11.171  1.00  0.68           H  \n",
      "ATOM    133  HB3 GLU A 326      16.109   4.552  11.043  1.00  0.65           H  \n",
      "ATOM    134  HG2 GLU A 326      15.887   6.391  13.296  1.00  0.92           H  \n",
      "ATOM    135  HG3 GLU A 326      14.730   5.065  13.414  1.00  0.97           H  \n",
      "ATOM    136  N   TYR A 327      16.912   5.596   8.884  1.00  0.49           N  \n",
      "ATOM    137  CA  TYR A 327      17.083   5.439   7.413  1.00  0.45           C  \n",
      "ATOM    138  C   TYR A 327      17.296   3.964   7.067  1.00  0.43           C  \n",
      "ATOM    139  O   TYR A 327      18.352   3.408   7.296  1.00  0.48           O  \n",
      "ATOM    140  CB  TYR A 327      18.294   6.252   6.953  1.00  0.47           C  \n",
      "ATOM    141  CG  TYR A 327      17.900   7.702   6.817  1.00  0.49           C  \n",
      "ATOM    142  CD1 TYR A 327      17.353   8.169   5.612  1.00  0.49           C  \n",
      "ATOM    143  CD2 TYR A 327      18.079   8.582   7.894  1.00  0.54           C  \n",
      "ATOM    144  CE1 TYR A 327      16.986   9.516   5.485  1.00  0.52           C  \n",
      "ATOM    145  CE2 TYR A 327      17.711   9.930   7.767  1.00  0.58           C  \n",
      "ATOM    146  CZ  TYR A 327      17.164  10.397   6.562  1.00  0.56           C  \n",
      "ATOM    147  OH  TYR A 327      16.801  11.723   6.437  1.00  0.61           O  \n",
      "ATOM    148  H   TYR A 327      17.349   4.971   9.500  1.00  0.51           H  \n",
      "ATOM    149  HA  TYR A 327      16.198   5.800   6.909  1.00  0.44           H  \n",
      "ATOM    150  HB2 TYR A 327      19.087   6.160   7.681  1.00  0.51           H  \n",
      "ATOM    151  HB3 TYR A 327      18.636   5.881   5.998  1.00  0.46           H  \n",
      "ATOM    152  HD1 TYR A 327      17.215   7.490   4.783  1.00  0.47           H  \n",
      "ATOM    153  HD2 TYR A 327      18.499   8.223   8.822  1.00  0.57           H  \n",
      "ATOM    154  HE1 TYR A 327      16.563   9.875   4.558  1.00  0.53           H  \n",
      "ATOM    155  HE2 TYR A 327      17.849  10.607   8.596  1.00  0.62           H  \n",
      "ATOM    156  HH  TYR A 327      17.140  12.045   5.598  1.00  1.04           H  \n",
      "ATOM    157  N   PHE A 328      16.303   3.329   6.508  1.00  0.39           N  \n",
      "ATOM    158  CA  PHE A 328      16.446   1.893   6.139  1.00  0.39           C  \n",
      "ATOM    159  C   PHE A 328      16.419   1.760   4.614  1.00  0.36           C  \n",
      "ATOM    160  O   PHE A 328      16.563   2.730   3.896  1.00  0.37           O  \n",
      "ATOM    161  CB  PHE A 328      15.290   1.090   6.741  1.00  0.40           C  \n",
      "ATOM    162  CG  PHE A 328      15.266   1.286   8.238  1.00  0.44           C  \n",
      "ATOM    163  CD1 PHE A 328      14.528   2.342   8.795  1.00  0.47           C  \n",
      "ATOM    164  CD2 PHE A 328      15.979   0.413   9.072  1.00  0.51           C  \n",
      "ATOM    165  CE1 PHE A 328      14.504   2.524  10.186  1.00  0.54           C  \n",
      "ATOM    166  CE2 PHE A 328      15.956   0.594  10.463  1.00  0.57           C  \n",
      "ATOM    167  CZ  PHE A 328      15.218   1.650  11.020  1.00  0.58           C  \n",
      "ATOM    168  H   PHE A 328      15.462   3.799   6.328  1.00  0.39           H  \n",
      "ATOM    169  HA  PHE A 328      17.384   1.514   6.518  1.00  0.42           H  \n",
      "ATOM    170  HB2 PHE A 328      14.357   1.431   6.317  1.00  0.40           H  \n",
      "ATOM    171  HB3 PHE A 328      15.425   0.042   6.518  1.00  0.43           H  \n",
      "ATOM    172  HD1 PHE A 328      13.978   3.014   8.152  1.00  0.48           H  \n",
      "ATOM    173  HD2 PHE A 328      16.547  -0.400   8.643  1.00  0.54           H  \n",
      "ATOM    174  HE1 PHE A 328      13.936   3.336  10.614  1.00  0.59           H  \n",
      "ATOM    175  HE2 PHE A 328      16.505  -0.078  11.105  1.00  0.65           H  \n",
      "ATOM    176  HZ  PHE A 328      15.199   1.790  12.090  1.00  0.64           H  \n",
      "ATOM    177  N   THR A 329      16.238   0.570   4.113  1.00  0.35           N  \n",
      "ATOM    178  CA  THR A 329      16.204   0.382   2.635  1.00  0.34           C  \n",
      "ATOM    179  C   THR A 329      15.386  -0.866   2.299  1.00  0.33           C  \n",
      "ATOM    180  O   THR A 329      15.063  -1.657   3.162  1.00  0.37           O  \n",
      "ATOM    181  CB  THR A 329      17.631   0.217   2.108  1.00  0.37           C  \n",
      "ATOM    182  OG1 THR A 329      18.349  -0.666   2.958  1.00  0.39           O  \n",
      "ATOM    183  CG2 THR A 329      18.326   1.580   2.079  1.00  0.42           C  \n",
      "ATOM    184  H   THR A 329      16.125  -0.201   4.707  1.00  0.35           H  \n",
      "ATOM    185  HA  THR A 329      15.748   1.246   2.174  1.00  0.35           H  \n",
      "ATOM    186  HB  THR A 329      17.602  -0.188   1.108  1.00  0.39           H  \n",
      "ATOM    187  HG1 THR A 329      18.897  -1.229   2.406  1.00  0.97           H  \n",
      "ATOM    188 HG21 THR A 329      17.726   2.277   1.513  1.00  1.08           H  \n",
      "ATOM    189 HG22 THR A 329      18.445   1.944   3.088  1.00  1.14           H  \n",
      "ATOM    190 HG23 THR A 329      19.296   1.480   1.615  1.00  1.09           H  \n",
      "ATOM    191  N   LEU A 330      15.043  -1.048   1.053  1.00  0.29           N  \n",
      "ATOM    192  CA  LEU A 330      14.242  -2.245   0.673  1.00  0.29           C  \n",
      "ATOM    193  C   LEU A 330      14.555  -2.641  -0.772  1.00  0.27           C  \n",
      "ATOM    194  O   LEU A 330      14.531  -1.823  -1.670  1.00  0.25           O  \n",
      "ATOM    195  CB  LEU A 330      12.754  -1.915   0.800  1.00  0.29           C  \n",
      "ATOM    196  CG  LEU A 330      11.923  -3.128   0.385  1.00  0.30           C  \n",
      "ATOM    197  CD1 LEU A 330      12.020  -4.207   1.465  1.00  0.36           C  \n",
      "ATOM    198  CD2 LEU A 330      10.461  -2.707   0.214  1.00  0.32           C  \n",
      "ATOM    199  H   LEU A 330      15.310  -0.398   0.369  1.00  0.27           H  \n",
      "ATOM    200  HA  LEU A 330      14.486  -3.065   1.332  1.00  0.32           H  \n",
      "ATOM    201  HB2 LEU A 330      12.529  -1.657   1.825  1.00  0.35           H  \n",
      "ATOM    202  HB3 LEU A 330      12.515  -1.080   0.158  1.00  0.29           H  \n",
      "ATOM    203  HG  LEU A 330      12.298  -3.519  -0.550  1.00  0.31           H  \n",
      "ATOM    204 HD11 LEU A 330      11.794  -3.773   2.428  1.00  1.08           H  \n",
      "ATOM    205 HD12 LEU A 330      11.315  -4.996   1.251  1.00  1.03           H  \n",
      "ATOM    206 HD13 LEU A 330      13.021  -4.612   1.479  1.00  1.09           H  \n",
      "ATOM    207 HD21 LEU A 330      10.240  -1.893   0.889  1.00  1.08           H  \n",
      "ATOM    208 HD22 LEU A 330      10.295  -2.387  -0.804  1.00  1.05           H  \n",
      "ATOM    209 HD23 LEU A 330       9.817  -3.545   0.438  1.00  1.08           H  \n",
      "ATOM    210  N   GLN A 331      14.840  -3.894  -1.002  1.00  0.29           N  \n",
      "ATOM    211  CA  GLN A 331      15.146  -4.349  -2.388  1.00  0.29           C  \n",
      "ATOM    212  C   GLN A 331      13.844  -4.741  -3.090  1.00  0.28           C  \n",
      "ATOM    213  O   GLN A 331      13.198  -5.703  -2.726  1.00  0.31           O  \n",
      "ATOM    214  CB  GLN A 331      16.081  -5.559  -2.331  1.00  0.34           C  \n",
      "ATOM    215  CG  GLN A 331      16.527  -5.928  -3.747  1.00  0.40           C  \n",
      "ATOM    216  CD  GLN A 331      18.038  -5.723  -3.876  1.00  0.64           C  \n",
      "ATOM    217  OE1 GLN A 331      18.809  -6.347  -3.175  1.00  1.37           O  \n",
      "ATOM    218  NE2 GLN A 331      18.495  -4.868  -4.749  1.00  0.62           N  \n",
      "ATOM    219  H   GLN A 331      14.848  -4.538  -0.263  1.00  0.32           H  \n",
      "ATOM    220  HA  GLN A 331      15.624  -3.549  -2.933  1.00  0.28           H  \n",
      "ATOM    221  HB2 GLN A 331      16.947  -5.316  -1.732  1.00  0.41           H  \n",
      "ATOM    222  HB3 GLN A 331      15.561  -6.396  -1.890  1.00  0.37           H  \n",
      "ATOM    223  HG2 GLN A 331      16.285  -6.963  -3.942  1.00  0.48           H  \n",
      "ATOM    224  HG3 GLN A 331      16.018  -5.298  -4.461  1.00  0.61           H  \n",
      "ATOM    225 HE21 GLN A 331      17.873  -4.364  -5.315  1.00  1.01           H  \n",
      "ATOM    226 HE22 GLN A 331      19.461  -4.729  -4.840  1.00  0.76           H  \n",
      "ATOM    227  N   ILE A 332      13.449  -4.001  -4.090  1.00  0.29           N  \n",
      "ATOM    228  CA  ILE A 332      12.184  -4.335  -4.806  1.00  0.31           C  \n",
      "ATOM    229  C   ILE A 332      12.506  -5.000  -6.146  1.00  0.32           C  \n",
      "ATOM    230  O   ILE A 332      12.959  -4.362  -7.075  1.00  0.33           O  \n",
      "ATOM    231  CB  ILE A 332      11.387  -3.052  -5.053  1.00  0.35           C  \n",
      "ATOM    232  CG1 ILE A 332      11.193  -2.311  -3.728  1.00  0.34           C  \n",
      "ATOM    233  CG2 ILE A 332      10.021  -3.405  -5.643  1.00  0.47           C  \n",
      "ATOM    234  CD1 ILE A 332      10.723  -0.882  -4.007  1.00  0.39           C  \n",
      "ATOM    235  H   ILE A 332      13.981  -3.226  -4.368  1.00  0.30           H  \n",
      "ATOM    236  HA  ILE A 332      11.597  -5.010  -4.202  1.00  0.33           H  \n",
      "ATOM    237  HB  ILE A 332      11.926  -2.422  -5.745  1.00  0.38           H  \n",
      "ATOM    238 HG12 ILE A 332      10.452  -2.825  -3.133  1.00  0.41           H  \n",
      "ATOM    239 HG13 ILE A 332      12.129  -2.282  -3.191  1.00  0.35           H  \n",
      "ATOM    240 HG21 ILE A 332       9.532  -4.131  -5.011  1.00  1.15           H  \n",
      "ATOM    241 HG22 ILE A 332       9.414  -2.513  -5.703  1.00  1.15           H  \n",
      "ATOM    242 HG23 ILE A 332      10.152  -3.819  -6.632  1.00  1.04           H  \n",
      "ATOM    243 HD11 ILE A 332       9.789  -0.909  -4.548  1.00  1.09           H  \n",
      "ATOM    244 HD12 ILE A 332      10.582  -0.361  -3.072  1.00  1.12           H  \n",
      "ATOM    245 HD13 ILE A 332      11.467  -0.367  -4.597  1.00  1.05           H  \n",
      "ATOM    246  N   ARG A 333      12.269  -6.279  -6.256  1.00  0.33           N  \n",
      "ATOM    247  CA  ARG A 333      12.555  -6.984  -7.537  1.00  0.36           C  \n",
      "ATOM    248  C   ARG A 333      11.514  -6.586  -8.584  1.00  0.38           C  \n",
      "ATOM    249  O   ARG A 333      10.403  -6.215  -8.260  1.00  0.42           O  \n",
      "ATOM    250  CB  ARG A 333      12.494  -8.496  -7.312  1.00  0.40           C  \n",
      "ATOM    251  CG  ARG A 333      13.339  -8.867  -6.092  1.00  0.43           C  \n",
      "ATOM    252  CD  ARG A 333      13.079 -10.327  -5.720  1.00  0.91           C  \n",
      "ATOM    253  NE  ARG A 333      13.936 -10.706  -4.561  1.00  1.05           N  \n",
      "ATOM    254  CZ  ARG A 333      13.682 -11.796  -3.888  1.00  1.50           C  \n",
      "ATOM    255  NH1 ARG A 333      12.677 -12.560  -4.228  1.00  2.10           N  \n",
      "ATOM    256  NH2 ARG A 333      14.433 -12.124  -2.873  1.00  2.10           N  \n",
      "ATOM    257  H   ARG A 333      11.899  -6.776  -5.496  1.00  0.34           H  \n",
      "ATOM    258  HA  ARG A 333      13.541  -6.712  -7.886  1.00  0.37           H  \n",
      "ATOM    259  HB2 ARG A 333      11.468  -8.793  -7.144  1.00  0.45           H  \n",
      "ATOM    260  HB3 ARG A 333      12.878  -9.006  -8.182  1.00  0.48           H  \n",
      "ATOM    261  HG2 ARG A 333      14.385  -8.733  -6.325  1.00  0.78           H  \n",
      "ATOM    262  HG3 ARG A 333      13.070  -8.232  -5.261  1.00  0.88           H  \n",
      "ATOM    263  HD2 ARG A 333      12.040 -10.451  -5.454  1.00  1.66           H  \n",
      "ATOM    264  HD3 ARG A 333      13.313 -10.960  -6.563  1.00  1.56           H  \n",
      "ATOM    265  HE  ARG A 333      14.690 -10.136  -4.304  1.00  1.63           H  \n",
      "ATOM    266 HH11 ARG A 333      12.099 -12.313  -5.005  1.00  2.21           H  \n",
      "ATOM    267 HH12 ARG A 333      12.487 -13.393  -3.709  1.00  2.79           H  \n",
      "ATOM    268 HH21 ARG A 333      15.202 -11.541  -2.611  1.00  2.40           H  \n",
      "ATOM    269 HH22 ARG A 333      14.239 -12.959  -2.357  1.00  2.59           H  \n",
      "ATOM    270  N   GLY A 334      11.862  -6.664  -9.839  1.00  0.39           N  \n",
      "ATOM    271  CA  GLY A 334      10.891  -6.296 -10.908  1.00  0.42           C  \n",
      "ATOM    272  C   GLY A 334      10.776  -4.774 -11.002  1.00  0.40           C  \n",
      "ATOM    273  O   GLY A 334      10.508  -4.099 -10.027  1.00  0.39           O  \n",
      "ATOM    274  H   GLY A 334      12.762  -6.970 -10.079  1.00  0.40           H  \n",
      "ATOM    275  HA2 GLY A 334      11.232  -6.691 -11.854  1.00  0.45           H  \n",
      "ATOM    276  HA3 GLY A 334       9.923  -6.712 -10.673  1.00  0.44           H  \n",
      "ATOM    277  N   ARG A 335      10.972  -4.228 -12.171  1.00  0.43           N  \n",
      "ATOM    278  CA  ARG A 335      10.869  -2.751 -12.332  1.00  0.45           C  \n",
      "ATOM    279  C   ARG A 335       9.397  -2.337 -12.280  1.00  0.45           C  \n",
      "ATOM    280  O   ARG A 335       9.056  -1.281 -11.786  1.00  0.44           O  \n",
      "ATOM    281  CB  ARG A 335      11.468  -2.343 -13.680  1.00  0.50           C  \n",
      "ATOM    282  CG  ARG A 335      11.267  -0.841 -13.897  1.00  0.58           C  \n",
      "ATOM    283  CD  ARG A 335      11.199  -0.548 -15.396  1.00  0.91           C  \n",
      "ATOM    284  NE  ARG A 335      11.634   0.854 -15.650  1.00  1.36           N  \n",
      "ATOM    285  CZ  ARG A 335      11.944   1.235 -16.860  1.00  1.83           C  \n",
      "ATOM    286  NH1 ARG A 335      11.879   0.388 -17.853  1.00  2.18           N  \n",
      "ATOM    287  NH2 ARG A 335      12.322   2.464 -17.079  1.00  2.68           N  \n",
      "ATOM    288  H   ARG A 335      11.183  -4.791 -12.945  1.00  0.47           H  \n",
      "ATOM    289  HA  ARG A 335      11.410  -2.263 -11.534  1.00  0.44           H  \n",
      "ATOM    290  HB2 ARG A 335      12.524  -2.570 -13.688  1.00  0.51           H  \n",
      "ATOM    291  HB3 ARG A 335      10.976  -2.888 -14.472  1.00  0.56           H  \n",
      "ATOM    292  HG2 ARG A 335      10.346  -0.528 -13.427  1.00  0.81           H  \n",
      "ATOM    293  HG3 ARG A 335      12.095  -0.300 -13.464  1.00  0.83           H  \n",
      "ATOM    294  HD2 ARG A 335      11.851  -1.228 -15.925  1.00  1.56           H  \n",
      "ATOM    295  HD3 ARG A 335      10.185  -0.679 -15.743  1.00  1.51           H  \n",
      "ATOM    296  HE  ARG A 335      11.685   1.492 -14.908  1.00  2.00           H  \n",
      "ATOM    297 HH11 ARG A 335      11.591  -0.556 -17.690  1.00  2.17           H  \n",
      "ATOM    298 HH12 ARG A 335      12.118   0.683 -18.777  1.00  2.88           H  \n",
      "ATOM    299 HH21 ARG A 335      12.373   3.114 -16.321  1.00  3.10           H  \n",
      "ATOM    300 HH22 ARG A 335      12.560   2.757 -18.005  1.00  3.16           H  \n",
      "ATOM    301  N   GLU A 336       8.524  -3.160 -12.791  1.00  0.50           N  \n",
      "ATOM    302  CA  GLU A 336       7.074  -2.816 -12.776  1.00  0.54           C  \n",
      "ATOM    303  C   GLU A 336       6.613  -2.596 -11.333  1.00  0.48           C  \n",
      "ATOM    304  O   GLU A 336       6.146  -1.533 -10.975  1.00  0.44           O  \n",
      "ATOM    305  CB  GLU A 336       6.274  -3.964 -13.395  1.00  0.62           C  \n",
      "ATOM    306  CG  GLU A 336       5.611  -3.488 -14.689  1.00  1.19           C  \n",
      "ATOM    307  CD  GLU A 336       4.989  -4.684 -15.413  1.00  1.58           C  \n",
      "ATOM    308  OE1 GLU A 336       4.286  -5.443 -14.766  1.00  2.22           O  \n",
      "ATOM    309  OE2 GLU A 336       5.225  -4.820 -16.602  1.00  2.04           O  \n",
      "ATOM    310  H   GLU A 336       8.821  -4.006 -13.187  1.00  0.53           H  \n",
      "ATOM    311  HA  GLU A 336       6.911  -1.915 -13.348  1.00  0.57           H  \n",
      "ATOM    312  HB2 GLU A 336       6.938  -4.788 -13.612  1.00  1.01           H  \n",
      "ATOM    313  HB3 GLU A 336       5.513  -4.288 -12.701  1.00  1.01           H  \n",
      "ATOM    314  HG2 GLU A 336       4.841  -2.767 -14.455  1.00  1.72           H  \n",
      "ATOM    315  HG3 GLU A 336       6.352  -3.030 -15.326  1.00  1.72           H  \n",
      "ATOM    316  N   ARG A 337       6.738  -3.595 -10.504  1.00  0.48           N  \n",
      "ATOM    317  CA  ARG A 337       6.303  -3.446  -9.086  1.00  0.46           C  \n",
      "ATOM    318  C   ARG A 337       7.106  -2.330  -8.418  1.00  0.39           C  \n",
      "ATOM    319  O   ARG A 337       6.628  -1.658  -7.526  1.00  0.36           O  \n",
      "ATOM    320  CB  ARG A 337       6.537  -4.762  -8.342  1.00  0.51           C  \n",
      "ATOM    321  CG  ARG A 337       5.607  -5.833  -8.912  1.00  0.63           C  \n",
      "ATOM    322  CD  ARG A 337       4.238  -5.725  -8.239  1.00  1.16           C  \n",
      "ATOM    323  NE  ARG A 337       3.657  -7.092  -8.051  1.00  1.43           N  \n",
      "ATOM    324  CZ  ARG A 337       3.584  -7.941  -9.044  1.00  2.15           C  \n",
      "ATOM    325  NH1 ARG A 337       3.827  -7.556 -10.268  1.00  2.74           N  \n",
      "ATOM    326  NH2 ARG A 337       3.203  -9.168  -8.818  1.00  2.77           N  \n",
      "ATOM    327  H   ARG A 337       7.115  -4.444 -10.814  1.00  0.53           H  \n",
      "ATOM    328  HA  ARG A 337       5.253  -3.199  -9.057  1.00  0.46           H  \n",
      "ATOM    329  HB2 ARG A 337       7.565  -5.071  -8.469  1.00  0.55           H  \n",
      "ATOM    330  HB3 ARG A 337       6.327  -4.626  -7.292  1.00  0.53           H  \n",
      "ATOM    331  HG2 ARG A 337       5.504  -5.682  -9.976  1.00  1.23           H  \n",
      "ATOM    332  HG3 ARG A 337       6.024  -6.811  -8.725  1.00  1.10           H  \n",
      "ATOM    333  HD2 ARG A 337       4.353  -5.276  -7.269  1.00  1.72           H  \n",
      "ATOM    334  HD3 ARG A 337       3.587  -5.101  -8.845  1.00  1.86           H  \n",
      "ATOM    335  HE  ARG A 337       3.376  -7.375  -7.156  1.00  1.69           H  \n",
      "ATOM    336 HH11 ARG A 337       4.062  -6.607 -10.459  1.00  2.62           H  \n",
      "ATOM    337 HH12 ARG A 337       3.777  -8.217 -11.017  1.00  3.54           H  \n",
      "ATOM    338 HH21 ARG A 337       2.968  -9.456  -7.890  1.00  2.78           H  \n",
      "ATOM    339 HH22 ARG A 337       3.145  -9.821  -9.573  1.00  3.47           H  \n",
      "ATOM    340  N   PHE A 338       8.322  -2.126  -8.840  1.00  0.39           N  \n",
      "ATOM    341  CA  PHE A 338       9.150  -1.053  -8.226  1.00  0.35           C  \n",
      "ATOM    342  C   PHE A 338       8.411   0.283  -8.320  1.00  0.33           C  \n",
      "ATOM    343  O   PHE A 338       8.162   0.935  -7.328  1.00  0.29           O  \n",
      "ATOM    344  CB  PHE A 338      10.483  -0.949  -8.967  1.00  0.39           C  \n",
      "ATOM    345  CG  PHE A 338      11.242   0.256  -8.464  1.00  0.37           C  \n",
      "ATOM    346  CD1 PHE A 338      11.823   0.234  -7.188  1.00  0.38           C  \n",
      "ATOM    347  CD2 PHE A 338      11.363   1.396  -9.272  1.00  0.42           C  \n",
      "ATOM    348  CE1 PHE A 338      12.526   1.354  -6.718  1.00  0.40           C  \n",
      "ATOM    349  CE2 PHE A 338      12.067   2.516  -8.803  1.00  0.43           C  \n",
      "ATOM    350  CZ  PHE A 338      12.648   2.494  -7.526  1.00  0.41           C  \n",
      "ATOM    351  H   PHE A 338       8.690  -2.679  -9.561  1.00  0.42           H  \n",
      "ATOM    352  HA  PHE A 338       9.333  -1.290  -7.188  1.00  0.35           H  \n",
      "ATOM    353  HB2 PHE A 338      11.065  -1.842  -8.790  1.00  0.42           H  \n",
      "ATOM    354  HB3 PHE A 338      10.300  -0.842 -10.026  1.00  0.43           H  \n",
      "ATOM    355  HD1 PHE A 338      11.729  -0.644  -6.566  1.00  0.42           H  \n",
      "ATOM    356  HD2 PHE A 338      10.915   1.412 -10.254  1.00  0.48           H  \n",
      "ATOM    357  HE1 PHE A 338      12.973   1.337  -5.735  1.00  0.45           H  \n",
      "ATOM    358  HE2 PHE A 338      12.160   3.394  -9.424  1.00  0.50           H  \n",
      "ATOM    359  HZ  PHE A 338      13.189   3.356  -7.164  1.00  0.44           H  \n",
      "ATOM    360  N   GLU A 339       8.065   0.699  -9.507  1.00  0.37           N  \n",
      "ATOM    361  CA  GLU A 339       7.351   1.998  -9.667  1.00  0.38           C  \n",
      "ATOM    362  C   GLU A 339       6.124   2.044  -8.750  1.00  0.33           C  \n",
      "ATOM    363  O   GLU A 339       5.763   3.085  -8.236  1.00  0.30           O  \n",
      "ATOM    364  CB  GLU A 339       6.901   2.152 -11.122  1.00  0.45           C  \n",
      "ATOM    365  CG  GLU A 339       8.128   2.207 -12.035  1.00  0.58           C  \n",
      "ATOM    366  CD  GLU A 339       7.740   2.830 -13.377  1.00  0.98           C  \n",
      "ATOM    367  OE1 GLU A 339       7.542   4.034 -13.413  1.00  1.66           O  \n",
      "ATOM    368  OE2 GLU A 339       7.646   2.094 -14.345  1.00  1.53           O  \n",
      "ATOM    369  H   GLU A 339       8.282   0.159 -10.295  1.00  0.41           H  \n",
      "ATOM    370  HA  GLU A 339       8.018   2.807  -9.412  1.00  0.38           H  \n",
      "ATOM    371  HB2 GLU A 339       6.283   1.310 -11.399  1.00  0.46           H  \n",
      "ATOM    372  HB3 GLU A 339       6.335   3.065 -11.229  1.00  0.48           H  \n",
      "ATOM    373  HG2 GLU A 339       8.898   2.805 -11.568  1.00  0.74           H  \n",
      "ATOM    374  HG3 GLU A 339       8.500   1.206 -12.199  1.00  0.81           H  \n",
      "ATOM    375  N   MET A 340       5.471   0.932  -8.550  1.00  0.33           N  \n",
      "ATOM    376  CA  MET A 340       4.260   0.923  -7.679  1.00  0.30           C  \n",
      "ATOM    377  C   MET A 340       4.637   1.300  -6.243  1.00  0.26           C  \n",
      "ATOM    378  O   MET A 340       4.032   2.165  -5.641  1.00  0.25           O  \n",
      "ATOM    379  CB  MET A 340       3.635  -0.475  -7.690  1.00  0.34           C  \n",
      "ATOM    380  CG  MET A 340       2.254  -0.422  -7.033  1.00  0.34           C  \n",
      "ATOM    381  SD  MET A 340       1.587  -2.100  -6.895  1.00  0.43           S  \n",
      "ATOM    382  CE  MET A 340       2.169  -2.437  -5.215  1.00  0.44           C  \n",
      "ATOM    383  H   MET A 340       5.771   0.104  -8.982  1.00  0.36           H  \n",
      "ATOM    384  HA  MET A 340       3.543   1.636  -8.057  1.00  0.31           H  \n",
      "ATOM    385  HB2 MET A 340       3.536  -0.816  -8.711  1.00  0.41           H  \n",
      "ATOM    386  HB3 MET A 340       4.266  -1.158  -7.142  1.00  0.34           H  \n",
      "ATOM    387  HG2 MET A 340       2.341   0.013  -6.048  1.00  0.35           H  \n",
      "ATOM    388  HG3 MET A 340       1.591   0.180  -7.636  1.00  0.44           H  \n",
      "ATOM    389  HE1 MET A 340       3.233  -2.257  -5.160  1.00  1.12           H  \n",
      "ATOM    390  HE2 MET A 340       1.661  -1.789  -4.519  1.00  1.15           H  \n",
      "ATOM    391  HE3 MET A 340       1.959  -3.468  -4.963  1.00  1.08           H  \n",
      "ATOM    392  N   PHE A 341       5.623   0.655  -5.685  1.00  0.24           N  \n",
      "ATOM    393  CA  PHE A 341       6.026   0.973  -4.285  1.00  0.22           C  \n",
      "ATOM    394  C   PHE A 341       6.504   2.425  -4.197  1.00  0.22           C  \n",
      "ATOM    395  O   PHE A 341       6.117   3.162  -3.313  1.00  0.22           O  \n",
      "ATOM    396  CB  PHE A 341       7.158   0.037  -3.859  1.00  0.22           C  \n",
      "ATOM    397  CG  PHE A 341       6.583  -1.308  -3.482  1.00  0.23           C  \n",
      "ATOM    398  CD1 PHE A 341       5.838  -1.447  -2.302  1.00  0.26           C  \n",
      "ATOM    399  CD2 PHE A 341       6.794  -2.419  -4.313  1.00  0.26           C  \n",
      "ATOM    400  CE1 PHE A 341       5.304  -2.696  -1.952  1.00  0.29           C  \n",
      "ATOM    401  CE2 PHE A 341       6.260  -3.668  -3.963  1.00  0.29           C  \n",
      "ATOM    402  CZ  PHE A 341       5.515  -3.806  -2.782  1.00  0.29           C  \n",
      "ATOM    403  H   PHE A 341       6.095  -0.045  -6.184  1.00  0.26           H  \n",
      "ATOM    404  HA  PHE A 341       5.180   0.834  -3.628  1.00  0.22           H  \n",
      "ATOM    405  HB2 PHE A 341       7.852  -0.084  -4.678  1.00  0.24           H  \n",
      "ATOM    406  HB3 PHE A 341       7.674   0.458  -3.009  1.00  0.23           H  \n",
      "ATOM    407  HD1 PHE A 341       5.675  -0.591  -1.662  1.00  0.30           H  \n",
      "ATOM    408  HD2 PHE A 341       7.368  -2.312  -5.222  1.00  0.30           H  \n",
      "ATOM    409  HE1 PHE A 341       4.730  -2.802  -1.043  1.00  0.33           H  \n",
      "ATOM    410  HE2 PHE A 341       6.423  -4.523  -4.603  1.00  0.34           H  \n",
      "ATOM    411  HZ  PHE A 341       5.104  -4.768  -2.513  1.00  0.32           H  \n",
      "ATOM    412  N   ARG A 342       7.346   2.839  -5.102  1.00  0.25           N  \n",
      "ATOM    413  CA  ARG A 342       7.857   4.239  -5.071  1.00  0.27           C  \n",
      "ATOM    414  C   ARG A 342       6.689   5.219  -4.938  1.00  0.25           C  \n",
      "ATOM    415  O   ARG A 342       6.760   6.185  -4.205  1.00  0.26           O  \n",
      "ATOM    416  CB  ARG A 342       8.617   4.527  -6.367  1.00  0.33           C  \n",
      "ATOM    417  CG  ARG A 342       9.268   5.909  -6.283  1.00  0.43           C  \n",
      "ATOM    418  CD  ARG A 342       9.680   6.364  -7.684  1.00  0.88           C  \n",
      "ATOM    419  NE  ARG A 342      10.307   7.714  -7.603  1.00  1.25           N  \n",
      "ATOM    420  CZ  ARG A 342      10.999   8.177  -8.610  1.00  1.72           C  \n",
      "ATOM    421  NH1 ARG A 342      11.142   7.464  -9.694  1.00  2.19           N  \n",
      "ATOM    422  NH2 ARG A 342      11.549   9.359  -8.532  1.00  2.44           N  \n",
      "ATOM    423  H   ARG A 342       7.647   2.227  -5.802  1.00  0.28           H  \n",
      "ATOM    424  HA  ARG A 342       8.524   4.359  -4.231  1.00  0.29           H  \n",
      "ATOM    425  HB2 ARG A 342       9.382   3.776  -6.510  1.00  0.37           H  \n",
      "ATOM    426  HB3 ARG A 342       7.931   4.505  -7.200  1.00  0.37           H  \n",
      "ATOM    427  HG2 ARG A 342       8.563   6.613  -5.867  1.00  0.80           H  \n",
      "ATOM    428  HG3 ARG A 342      10.142   5.857  -5.652  1.00  0.77           H  \n",
      "ATOM    429  HD2 ARG A 342      10.389   5.662  -8.096  1.00  1.51           H  \n",
      "ATOM    430  HD3 ARG A 342       8.808   6.410  -8.319  1.00  1.42           H  \n",
      "ATOM    431  HE  ARG A 342      10.202   8.254  -6.792  1.00  1.84           H  \n",
      "ATOM    432 HH11 ARG A 342      10.723   6.559  -9.759  1.00  2.21           H  \n",
      "ATOM    433 HH12 ARG A 342      11.673   7.823 -10.462  1.00  2.91           H  \n",
      "ATOM    434 HH21 ARG A 342      11.440   9.907  -7.703  1.00  2.77           H  \n",
      "ATOM    435 HH22 ARG A 342      12.079   9.715  -9.301  1.00  2.94           H  \n",
      "ATOM    436  N   GLU A 343       5.617   4.985  -5.644  1.00  0.25           N  \n",
      "ATOM    437  CA  GLU A 343       4.453   5.912  -5.558  1.00  0.25           C  \n",
      "ATOM    438  C   GLU A 343       3.963   5.993  -4.111  1.00  0.21           C  \n",
      "ATOM    439  O   GLU A 343       3.722   7.063  -3.587  1.00  0.22           O  \n",
      "ATOM    440  CB  GLU A 343       3.323   5.397  -6.451  1.00  0.28           C  \n",
      "ATOM    441  CG  GLU A 343       2.102   6.307  -6.305  1.00  0.33           C  \n",
      "ATOM    442  CD  GLU A 343       1.514   6.601  -7.687  1.00  1.05           C  \n",
      "ATOM    443  OE1 GLU A 343       1.553   5.716  -8.526  1.00  1.74           O  \n",
      "ATOM    444  OE2 GLU A 343       1.034   7.705  -7.882  1.00  1.74           O  \n",
      "ATOM    445  H   GLU A 343       5.579   4.203  -6.234  1.00  0.27           H  \n",
      "ATOM    446  HA  GLU A 343       4.753   6.895  -5.890  1.00  0.28           H  \n",
      "ATOM    447  HB2 GLU A 343       3.650   5.394  -7.481  1.00  0.35           H  \n",
      "ATOM    448  HB3 GLU A 343       3.058   4.393  -6.154  1.00  0.32           H  \n",
      "ATOM    449  HG2 GLU A 343       1.359   5.817  -5.693  1.00  0.69           H  \n",
      "ATOM    450  HG3 GLU A 343       2.398   7.234  -5.838  1.00  0.64           H  \n",
      "ATOM    451  N   LEU A 344       3.812   4.873  -3.460  1.00  0.20           N  \n",
      "ATOM    452  CA  LEU A 344       3.335   4.890  -2.048  1.00  0.19           C  \n",
      "ATOM    453  C   LEU A 344       4.345   5.636  -1.175  1.00  0.20           C  \n",
      "ATOM    454  O   LEU A 344       3.984   6.333  -0.248  1.00  0.23           O  \n",
      "ATOM    455  CB  LEU A 344       3.191   3.454  -1.542  1.00  0.20           C  \n",
      "ATOM    456  CG  LEU A 344       2.070   2.753  -2.309  1.00  0.24           C  \n",
      "ATOM    457  CD1 LEU A 344       2.177   1.241  -2.105  1.00  0.29           C  \n",
      "ATOM    458  CD2 LEU A 344       0.716   3.242  -1.790  1.00  0.30           C  \n",
      "ATOM    459  H   LEU A 344       4.010   4.020  -3.900  1.00  0.21           H  \n",
      "ATOM    460  HA  LEU A 344       2.378   5.387  -1.997  1.00  0.20           H  \n",
      "ATOM    461  HB2 LEU A 344       4.120   2.923  -1.695  1.00  0.22           H  \n",
      "ATOM    462  HB3 LEU A 344       2.951   3.465  -0.489  1.00  0.23           H  \n",
      "ATOM    463  HG  LEU A 344       2.158   2.981  -3.362  1.00  0.27           H  \n",
      "ATOM    464 HD11 LEU A 344       2.657   1.039  -1.158  1.00  1.02           H  \n",
      "ATOM    465 HD12 LEU A 344       1.189   0.806  -2.107  1.00  1.03           H  \n",
      "ATOM    466 HD13 LEU A 344       2.762   0.810  -2.904  1.00  1.00           H  \n",
      "ATOM    467 HD21 LEU A 344       0.835   4.216  -1.338  1.00  1.07           H  \n",
      "ATOM    468 HD22 LEU A 344       0.018   3.310  -2.612  1.00  1.04           H  \n",
      "ATOM    469 HD23 LEU A 344       0.340   2.547  -1.054  1.00  1.04           H  \n",
      "ATOM    470  N   ASN A 345       5.609   5.493  -1.460  1.00  0.24           N  \n",
      "ATOM    471  CA  ASN A 345       6.640   6.190  -0.643  1.00  0.29           C  \n",
      "ATOM    472  C   ASN A 345       6.376   7.697  -0.657  1.00  0.25           C  \n",
      "ATOM    473  O   ASN A 345       6.272   8.327   0.376  1.00  0.25           O  \n",
      "ATOM    474  CB  ASN A 345       8.026   5.913  -1.227  1.00  0.38           C  \n",
      "ATOM    475  CG  ASN A 345       9.084   6.627  -0.385  1.00  0.49           C  \n",
      "ATOM    476  OD1 ASN A 345       8.821   7.662   0.195  1.00  1.22           O  \n",
      "ATOM    477  ND2 ASN A 345      10.280   6.116  -0.292  1.00  0.58           N  \n",
      "ATOM    478  H   ASN A 345       5.880   4.926  -2.210  1.00  0.28           H  \n",
      "ATOM    479  HA  ASN A 345       6.599   5.827   0.373  1.00  0.32           H  \n",
      "ATOM    480  HB2 ASN A 345       8.214   4.849  -1.217  1.00  0.41           H  \n",
      "ATOM    481  HB3 ASN A 345       8.069   6.277  -2.242  1.00  0.41           H  \n",
      "ATOM    482 HD21 ASN A 345      10.493   5.282  -0.759  1.00  1.14           H  \n",
      "ATOM    483 HD22 ASN A 345      10.965   6.566   0.246  1.00  0.58           H  \n",
      "ATOM    484  N   GLU A 346       6.274   8.280  -1.819  1.00  0.27           N  \n",
      "ATOM    485  CA  GLU A 346       6.025   9.747  -1.898  1.00  0.28           C  \n",
      "ATOM    486  C   GLU A 346       4.693  10.081  -1.223  1.00  0.23           C  \n",
      "ATOM    487  O   GLU A 346       4.513  11.155  -0.687  1.00  0.25           O  \n",
      "ATOM    488  CB  GLU A 346       5.972  10.176  -3.367  1.00  0.34           C  \n",
      "ATOM    489  CG  GLU A 346       5.777  11.691  -3.452  1.00  0.40           C  \n",
      "ATOM    490  CD  GLU A 346       6.454  12.223  -4.717  1.00  1.00           C  \n",
      "ATOM    491  OE1 GLU A 346       7.602  11.878  -4.941  1.00  1.70           O  \n",
      "ATOM    492  OE2 GLU A 346       5.812  12.968  -5.440  1.00  1.72           O  \n",
      "ATOM    493  H   GLU A 346       6.367   7.754  -2.641  1.00  0.30           H  \n",
      "ATOM    494  HA  GLU A 346       6.824  10.274  -1.399  1.00  0.31           H  \n",
      "ATOM    495  HB2 GLU A 346       6.897   9.903  -3.854  1.00  0.40           H  \n",
      "ATOM    496  HB3 GLU A 346       5.147   9.681  -3.856  1.00  0.37           H  \n",
      "ATOM    497  HG2 GLU A 346       4.721  11.916  -3.486  1.00  0.82           H  \n",
      "ATOM    498  HG3 GLU A 346       6.218  12.161  -2.586  1.00  0.80           H  \n",
      "ATOM    499  N   ALA A 347       3.757   9.173  -1.250  1.00  0.22           N  \n",
      "ATOM    500  CA  ALA A 347       2.435   9.441  -0.614  1.00  0.23           C  \n",
      "ATOM    501  C   ALA A 347       2.619   9.699   0.884  1.00  0.22           C  \n",
      "ATOM    502  O   ALA A 347       2.233  10.730   1.398  1.00  0.24           O  \n",
      "ATOM    503  CB  ALA A 347       1.520   8.231  -0.814  1.00  0.27           C  \n",
      "ATOM    504  H   ALA A 347       3.921   8.315  -1.693  1.00  0.23           H  \n",
      "ATOM    505  HA  ALA A 347       1.986  10.309  -1.073  1.00  0.26           H  \n",
      "ATOM    506  HB1 ALA A 347       2.097   7.402  -1.194  1.00  1.00           H  \n",
      "ATOM    507  HB2 ALA A 347       1.074   7.957   0.131  1.00  1.01           H  \n",
      "ATOM    508  HB3 ALA A 347       0.741   8.481  -1.520  1.00  1.07           H  \n",
      "ATOM    509  N   LEU A 348       3.199   8.767   1.589  1.00  0.22           N  \n",
      "ATOM    510  CA  LEU A 348       3.399   8.953   3.055  1.00  0.24           C  \n",
      "ATOM    511  C   LEU A 348       4.295  10.169   3.306  1.00  0.27           C  \n",
      "ATOM    512  O   LEU A 348       4.025  10.983   4.166  1.00  0.29           O  \n",
      "ATOM    513  CB  LEU A 348       4.067   7.705   3.639  1.00  0.25           C  \n",
      "ATOM    514  CG  LEU A 348       3.030   6.592   3.810  1.00  0.24           C  \n",
      "ATOM    515  CD1 LEU A 348       1.852   7.111   4.634  1.00  0.29           C  \n",
      "ATOM    516  CD2 LEU A 348       2.532   6.142   2.435  1.00  0.24           C  \n",
      "ATOM    517  H   LEU A 348       3.497   7.940   1.156  1.00  0.22           H  \n",
      "ATOM    518  HA  LEU A 348       2.442   9.106   3.532  1.00  0.25           H  \n",
      "ATOM    519  HB2 LEU A 348       4.847   7.369   2.971  1.00  0.25           H  \n",
      "ATOM    520  HB3 LEU A 348       4.496   7.944   4.601  1.00  0.28           H  \n",
      "ATOM    521  HG  LEU A 348       3.484   5.755   4.321  1.00  0.27           H  \n",
      "ATOM    522 HD11 LEU A 348       2.222   7.696   5.463  1.00  1.08           H  \n",
      "ATOM    523 HD12 LEU A 348       1.221   7.727   4.012  1.00  1.06           H  \n",
      "ATOM    524 HD13 LEU A 348       1.280   6.275   5.011  1.00  1.03           H  \n",
      "ATOM    525 HD21 LEU A 348       3.363   6.103   1.748  1.00  1.01           H  \n",
      "ATOM    526 HD22 LEU A 348       2.086   5.162   2.517  1.00  1.05           H  \n",
      "ATOM    527 HD23 LEU A 348       1.796   6.843   2.072  1.00  1.03           H  \n",
      "ATOM    528  N   GLU A 349       5.363  10.296   2.567  1.00  0.30           N  \n",
      "ATOM    529  CA  GLU A 349       6.280  11.455   2.771  1.00  0.35           C  \n",
      "ATOM    530  C   GLU A 349       5.509  12.766   2.597  1.00  0.32           C  \n",
      "ATOM    531  O   GLU A 349       5.793  13.752   3.248  1.00  0.34           O  \n",
      "ATOM    532  CB  GLU A 349       7.418  11.392   1.749  1.00  0.40           C  \n",
      "ATOM    533  CG  GLU A 349       8.235  10.116   1.971  1.00  0.48           C  \n",
      "ATOM    534  CD  GLU A 349       9.727  10.437   1.863  1.00  1.13           C  \n",
      "ATOM    535  OE1 GLU A 349      10.067  11.338   1.115  1.00  1.91           O  \n",
      "ATOM    536  OE2 GLU A 349      10.503   9.773   2.531  1.00  1.68           O  \n",
      "ATOM    537  H   GLU A 349       5.567   9.626   1.883  1.00  0.31           H  \n",
      "ATOM    538  HA  GLU A 349       6.693  11.412   3.768  1.00  0.39           H  \n",
      "ATOM    539  HB2 GLU A 349       7.005  11.387   0.750  1.00  0.38           H  \n",
      "ATOM    540  HB3 GLU A 349       8.059  12.252   1.870  1.00  0.46           H  \n",
      "ATOM    541  HG2 GLU A 349       8.021   9.719   2.952  1.00  0.90           H  \n",
      "ATOM    542  HG3 GLU A 349       7.971   9.385   1.221  1.00  0.78           H  \n",
      "ATOM    543  N   LEU A 350       4.534  12.789   1.730  1.00  0.29           N  \n",
      "ATOM    544  CA  LEU A 350       3.751  14.041   1.526  1.00  0.29           C  \n",
      "ATOM    545  C   LEU A 350       3.021  14.391   2.822  1.00  0.28           C  \n",
      "ATOM    546  O   LEU A 350       3.004  15.527   3.252  1.00  0.32           O  \n",
      "ATOM    547  CB  LEU A 350       2.730  13.833   0.405  1.00  0.29           C  \n",
      "ATOM    548  CG  LEU A 350       2.280  15.193  -0.135  1.00  0.34           C  \n",
      "ATOM    549  CD1 LEU A 350       1.720  15.022  -1.548  1.00  0.41           C  \n",
      "ATOM    550  CD2 LEU A 350       1.192  15.767   0.775  1.00  0.38           C  \n",
      "ATOM    551  H   LEU A 350       4.316  11.985   1.216  1.00  0.30           H  \n",
      "ATOM    552  HA  LEU A 350       4.421  14.846   1.261  1.00  0.32           H  \n",
      "ATOM    553  HB2 LEU A 350       3.182  13.261  -0.392  1.00  0.33           H  \n",
      "ATOM    554  HB3 LEU A 350       1.875  13.300   0.791  1.00  0.28           H  \n",
      "ATOM    555  HG  LEU A 350       3.123  15.867  -0.161  1.00  0.49           H  \n",
      "ATOM    556 HD11 LEU A 350       1.317  14.026  -1.658  1.00  1.08           H  \n",
      "ATOM    557 HD12 LEU A 350       0.937  15.748  -1.715  1.00  1.15           H  \n",
      "ATOM    558 HD13 LEU A 350       2.510  15.173  -2.269  1.00  1.06           H  \n",
      "ATOM    559 HD21 LEU A 350       0.387  15.054   0.865  1.00  1.15           H  \n",
      "ATOM    560 HD22 LEU A 350       1.606  15.968   1.750  1.00  1.06           H  \n",
      "ATOM    561 HD23 LEU A 350       0.815  16.685   0.349  1.00  1.07           H  \n",
      "ATOM    562  N   LYS A 351       2.423  13.419   3.451  1.00  0.27           N  \n",
      "ATOM    563  CA  LYS A 351       1.698  13.688   4.722  1.00  0.31           C  \n",
      "ATOM    564  C   LYS A 351       2.680  14.257   5.748  1.00  0.38           C  \n",
      "ATOM    565  O   LYS A 351       2.368  15.175   6.480  1.00  0.44           O  \n",
      "ATOM    566  CB  LYS A 351       1.104  12.381   5.251  1.00  0.33           C  \n",
      "ATOM    567  CG  LYS A 351       0.134  12.681   6.394  1.00  0.40           C  \n",
      "ATOM    568  CD  LYS A 351      -0.618  11.402   6.767  1.00  0.45           C  \n",
      "ATOM    569  CE  LYS A 351      -1.894  11.761   7.531  1.00  0.77           C  \n",
      "ATOM    570  NZ  LYS A 351      -2.257  10.639   8.442  1.00  1.56           N  \n",
      "ATOM    571  H   LYS A 351       2.455  12.509   3.089  1.00  0.26           H  \n",
      "ATOM    572  HA  LYS A 351       0.907  14.401   4.544  1.00  0.31           H  \n",
      "ATOM    573  HB2 LYS A 351       0.576  11.879   4.453  1.00  0.35           H  \n",
      "ATOM    574  HB3 LYS A 351       1.898  11.745   5.612  1.00  0.38           H  \n",
      "ATOM    575  HG2 LYS A 351       0.687  13.038   7.251  1.00  0.53           H  \n",
      "ATOM    576  HG3 LYS A 351      -0.573  13.434   6.080  1.00  0.52           H  \n",
      "ATOM    577  HD2 LYS A 351      -0.876  10.862   5.867  1.00  0.55           H  \n",
      "ATOM    578  HD3 LYS A 351       0.010  10.783   7.389  1.00  0.63           H  \n",
      "ATOM    579  HE2 LYS A 351      -1.727  12.656   8.111  1.00  1.30           H  \n",
      "ATOM    580  HE3 LYS A 351      -2.698  11.930   6.830  1.00  1.15           H  \n",
      "ATOM    581  HZ1 LYS A 351      -1.861   9.751   8.074  1.00  2.14           H  \n",
      "ATOM    582  HZ2 LYS A 351      -1.872  10.825   9.391  1.00  2.02           H  \n",
      "ATOM    583  HZ3 LYS A 351      -3.292  10.557   8.497  1.00  2.00           H  \n",
      "ATOM    584  N   ASP A 352       3.866  13.716   5.804  1.00  0.42           N  \n",
      "ATOM    585  CA  ASP A 352       4.874  14.221   6.778  1.00  0.51           C  \n",
      "ATOM    586  C   ASP A 352       5.093  15.720   6.560  1.00  0.51           C  \n",
      "ATOM    587  O   ASP A 352       5.209  16.483   7.498  1.00  0.59           O  \n",
      "ATOM    588  CB  ASP A 352       6.196  13.479   6.573  1.00  0.58           C  \n",
      "ATOM    589  CG  ASP A 352       6.074  12.056   7.119  1.00  0.67           C  \n",
      "ATOM    590  OD1 ASP A 352       5.104  11.396   6.785  1.00  1.14           O  \n",
      "ATOM    591  OD2 ASP A 352       6.952  11.650   7.862  1.00  1.43           O  \n",
      "ATOM    592  H   ASP A 352       4.095  12.977   5.202  1.00  0.42           H  \n",
      "ATOM    593  HA  ASP A 352       4.518  14.052   7.783  1.00  0.56           H  \n",
      "ATOM    594  HB2 ASP A 352       6.427  13.441   5.518  1.00  0.56           H  \n",
      "ATOM    595  HB3 ASP A 352       6.985  13.997   7.096  1.00  0.68           H  \n",
      "ATOM    596  N   ALA A 353       5.155  16.147   5.328  1.00  0.49           N  \n",
      "ATOM    597  CA  ALA A 353       5.372  17.596   5.053  1.00  0.56           C  \n",
      "ATOM    598  C   ALA A 353       4.154  18.396   5.520  1.00  0.56           C  \n",
      "ATOM    599  O   ALA A 353       4.260  19.554   5.872  1.00  0.66           O  \n",
      "ATOM    600  CB  ALA A 353       5.574  17.805   3.550  1.00  0.64           C  \n",
      "ATOM    601  H   ALA A 353       5.062  15.516   4.585  1.00  0.49           H  \n",
      "ATOM    602  HA  ALA A 353       6.250  17.935   5.583  1.00  0.63           H  \n",
      "ATOM    603  HB1 ALA A 353       6.214  17.027   3.161  1.00  1.30           H  \n",
      "ATOM    604  HB2 ALA A 353       4.617  17.767   3.050  1.00  1.11           H  \n",
      "ATOM    605  HB3 ALA A 353       6.032  18.767   3.379  1.00  1.23           H  \n",
      "ATOM    606  N   GLN A 354       2.999  17.789   5.528  1.00  0.53           N  \n",
      "ATOM    607  CA  GLN A 354       1.780  18.519   5.976  1.00  0.62           C  \n",
      "ATOM    608  C   GLN A 354       1.680  18.457   7.502  1.00  0.68           C  \n",
      "ATOM    609  O   GLN A 354       0.906  19.166   8.113  1.00  0.88           O  \n",
      "ATOM    610  CB  GLN A 354       0.541  17.869   5.358  1.00  0.64           C  \n",
      "ATOM    611  CG  GLN A 354      -0.144  18.864   4.419  1.00  0.97           C  \n",
      "ATOM    612  CD  GLN A 354      -1.506  18.311   3.996  1.00  0.84           C  \n",
      "ATOM    613  OE1 GLN A 354      -2.415  19.062   3.702  1.00  1.18           O  \n",
      "ATOM    614  NE2 GLN A 354      -1.687  17.019   3.953  1.00  0.68           N  \n",
      "ATOM    615  H   GLN A 354       2.934  16.854   5.242  1.00  0.51           H  \n",
      "ATOM    616  HA  GLN A 354       1.841  19.550   5.661  1.00  0.71           H  \n",
      "ATOM    617  HB2 GLN A 354       0.836  16.991   4.801  1.00  0.85           H  \n",
      "ATOM    618  HB3 GLN A 354      -0.145  17.585   6.141  1.00  0.89           H  \n",
      "ATOM    619  HG2 GLN A 354      -0.280  19.806   4.931  1.00  1.39           H  \n",
      "ATOM    620  HG3 GLN A 354       0.469  19.014   3.544  1.00  1.42           H  \n",
      "ATOM    621 HE21 GLN A 354      -0.954  16.414   4.190  1.00  0.73           H  \n",
      "ATOM    622 HE22 GLN A 354      -2.556  16.655   3.683  1.00  0.79           H  \n",
      "ATOM    623  N   ALA A 355       2.458  17.612   8.121  1.00  0.70           N  \n",
      "ATOM    624  CA  ALA A 355       2.410  17.502   9.606  1.00  0.83           C  \n",
      "ATOM    625  C   ALA A 355       3.074  18.731  10.233  1.00  0.91           C  \n",
      "ATOM    626  O   ALA A 355       2.564  19.316  11.167  1.00  1.23           O  \n",
      "ATOM    627  CB  ALA A 355       3.151  16.237  10.044  1.00  1.02           C  \n",
      "ATOM    628  H   ALA A 355       3.075  17.049   7.609  1.00  0.74           H  \n",
      "ATOM    629  HA  ALA A 355       1.381  17.446   9.929  1.00  0.95           H  \n",
      "ATOM    630  HB1 ALA A 355       3.572  15.749   9.177  1.00  1.52           H  \n",
      "ATOM    631  HB2 ALA A 355       3.943  16.501  10.728  1.00  1.58           H  \n",
      "ATOM    632  HB3 ALA A 355       2.459  15.567  10.533  1.00  1.31           H  \n",
      "ATOM    633  N   GLY A 356       4.212  19.124   9.726  1.00  1.03           N  \n",
      "ATOM    634  CA  GLY A 356       4.917  20.314  10.289  1.00  1.23           C  \n",
      "ATOM    635  C   GLY A 356       4.315  21.606   9.726  1.00  1.19           C  \n",
      "ATOM    636  O   GLY A 356       4.896  22.667   9.839  1.00  1.54           O  \n",
      "ATOM    637  H   GLY A 356       4.606  18.636   8.973  1.00  1.24           H  \n",
      "ATOM    638  HA2 GLY A 356       4.817  20.311  11.365  1.00  1.44           H  \n",
      "ATOM    639  HA3 GLY A 356       5.963  20.267  10.026  1.00  1.53           H  \n",
      "ATOM    640  N   LYS A 357       3.162  21.535   9.118  1.00  1.33           N  \n",
      "ATOM    641  CA  LYS A 357       2.546  22.768   8.552  1.00  1.54           C  \n",
      "ATOM    642  C   LYS A 357       1.765  23.498   9.646  1.00  1.98           C  \n",
      "ATOM    643  O   LYS A 357       0.766  23.013  10.140  1.00  2.51           O  \n",
      "ATOM    644  CB  LYS A 357       1.599  22.388   7.412  1.00  1.88           C  \n",
      "ATOM    645  CG  LYS A 357       1.623  23.481   6.342  1.00  2.25           C  \n",
      "ATOM    646  CD  LYS A 357       0.212  23.683   5.784  1.00  2.96           C  \n",
      "ATOM    647  CE  LYS A 357       0.244  24.754   4.692  1.00  3.50           C  \n",
      "ATOM    648  NZ  LYS A 357      -1.141  25.009   4.205  1.00  4.18           N  \n",
      "ATOM    649  H   LYS A 357       2.704  20.675   9.028  1.00  1.62           H  \n",
      "ATOM    650  HA  LYS A 357       3.323  23.416   8.173  1.00  1.72           H  \n",
      "ATOM    651  HB2 LYS A 357       1.918  21.451   6.978  1.00  2.23           H  \n",
      "ATOM    652  HB3 LYS A 357       0.596  22.284   7.796  1.00  2.14           H  \n",
      "ATOM    653  HG2 LYS A 357       1.973  24.405   6.780  1.00  2.39           H  \n",
      "ATOM    654  HG3 LYS A 357       2.285  23.188   5.542  1.00  2.54           H  \n",
      "ATOM    655  HD2 LYS A 357      -0.146  22.753   5.368  1.00  3.43           H  \n",
      "ATOM    656  HD3 LYS A 357      -0.447  24.001   6.578  1.00  3.17           H  \n",
      "ATOM    657  HE2 LYS A 357       0.658  25.666   5.094  1.00  3.67           H  \n",
      "ATOM    658  HE3 LYS A 357       0.857  24.412   3.871  1.00  3.76           H  \n",
      "ATOM    659  HZ1 LYS A 357      -1.824  24.577   4.859  1.00  4.46           H  \n",
      "ATOM    660  HZ2 LYS A 357      -1.310  26.034   4.159  1.00  4.55           H  \n",
      "ATOM    661  HZ3 LYS A 357      -1.257  24.594   3.258  1.00  4.44           H  \n",
      "ATOM    662  N   GLU A 358       2.216  24.661  10.029  1.00  2.43           N  \n",
      "ATOM    663  CA  GLU A 358       1.505  25.424  11.092  1.00  3.19           C  \n",
      "ATOM    664  C   GLU A 358       0.008  25.487  10.760  1.00  3.44           C  \n",
      "ATOM    665  O   GLU A 358      -0.359  25.786   9.641  1.00  3.47           O  \n",
      "ATOM    666  CB  GLU A 358       2.071  26.845  11.161  1.00  3.91           C  \n",
      "ATOM    667  CG  GLU A 358       2.237  27.262  12.624  1.00  4.53           C  \n",
      "ATOM    668  CD  GLU A 358       2.598  28.747  12.693  1.00  5.25           C  \n",
      "ATOM    669  OE1 GLU A 358       1.947  29.526  12.015  1.00  5.65           O  \n",
      "ATOM    670  OE2 GLU A 358       3.517  29.080  13.422  1.00  5.71           O  \n",
      "ATOM    671  H   GLU A 358       3.024  25.032   9.618  1.00  2.59           H  \n",
      "ATOM    672  HA  GLU A 358       1.651  24.933  12.041  1.00  3.49           H  \n",
      "ATOM    673  HB2 GLU A 358       3.031  26.873  10.667  1.00  4.22           H  \n",
      "ATOM    674  HB3 GLU A 358       1.393  27.527  10.670  1.00  4.17           H  \n",
      "ATOM    675  HG2 GLU A 358       1.312  27.092  13.154  1.00  4.65           H  \n",
      "ATOM    676  HG3 GLU A 358       3.026  26.681  13.077  1.00  4.78           H  \n",
      "ATOM    677  N   PRO A 359      -0.816  25.205  11.741  1.00  4.08           N  \n",
      "ATOM    678  CA  PRO A 359      -2.278  25.228  11.562  1.00  4.74           C  \n",
      "ATOM    679  C   PRO A 359      -2.736  26.637  11.176  1.00  4.92           C  \n",
      "ATOM    680  O   PRO A 359      -1.965  27.576  11.189  1.00  4.94           O  \n",
      "ATOM    681  CB  PRO A 359      -2.849  24.827  12.930  1.00  5.61           C  \n",
      "ATOM    682  CG  PRO A 359      -1.655  24.581  13.885  1.00  5.57           C  \n",
      "ATOM    683  CD  PRO A 359      -0.364  24.845  13.097  1.00  4.60           C  \n",
      "ATOM    684  HA  PRO A 359      -2.579  24.513  10.813  1.00  4.84           H  \n",
      "ATOM    685  HB2 PRO A 359      -3.472  25.623  13.314  1.00  6.02           H  \n",
      "ATOM    686  HB3 PRO A 359      -3.428  23.921  12.834  1.00  6.09           H  \n",
      "ATOM    687  HG2 PRO A 359      -1.716  25.254  14.729  1.00  6.03           H  \n",
      "ATOM    688  HG3 PRO A 359      -1.665  23.558  14.230  1.00  6.01           H  \n",
      "ATOM    689  HD2 PRO A 359       0.183  25.663  13.542  1.00  4.71           H  \n",
      "ATOM    690  HD3 PRO A 359       0.246  23.956  13.062  1.00  4.53           H  \n",
      "ATOM    691  N   GLY A 360      -3.984  26.791  10.832  1.00  5.45           N  \n",
      "ATOM    692  CA  GLY A 360      -4.489  28.138  10.445  1.00  5.95           C  \n",
      "ATOM    693  C   GLY A 360      -5.981  28.236  10.765  1.00  6.77           C  \n",
      "ATOM    694  O   GLY A 360      -6.401  27.621  11.732  1.00  7.24           O  \n",
      "ATOM    695  OXT GLY A 360      -6.679  28.924  10.039  1.00  7.15           O  \n",
      "ATOM    696  H   GLY A 360      -4.589  26.020  10.828  1.00  5.72           H  \n",
      "ATOM    697  HA2 GLY A 360      -3.950  28.896  10.995  1.00  5.99           H  \n",
      "ATOM    698  HA3 GLY A 360      -4.341  28.288   9.386  1.00  6.05           H  \n"
     ]
    }
   ],
   "source": [
    "!type atoms_chain_A.txt # shows the whole file"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "53851bb1",
   "metadata": {},
   "source": [
    "**Finding files with the glob package**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "83d9bced",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['data\\\\1FAG.pdb', 'data\\\\1J6Z.pdb', 'data\\\\1OLG.pdb', 'data\\\\2ERK.pdb']"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import glob\n",
    "\n",
    "file_list = glob.glob('data/*.pdb') # all files in data which contain a .pdb\n",
    "file_list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1fa9249f",
   "metadata": {},
   "source": [
    "We want to extract a set of data (sequence from each chain A) for all these files:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "edeef00f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dictionary to hold sequences\n",
    "seqs = {}\n",
    "\n",
    "# Loop through all matching files\n",
    "for file_name in file_list:\n",
    "    # Extract PDB ID\n",
    "    pdb_id = file_name[file_name.find('\\\\')+1:file_name.rfind('.')]\n",
    "\n",
    "    # Initialize sequence string, which we build as we go along\n",
    "    seq = ''\n",
    "    with open(file_name, 'r') as f:\n",
    "        for line in f:\n",
    "            if len(line) > 11 and line[:6] == 'SEQRES' and line[11] == 'A':\n",
    "                seq += line[19:].rstrip() + ' '\n",
    "\n",
    "    # Build sequence with dash-joined three letter codes\n",
    "    seq = '-'.join(seq.split())\n",
    "\n",
    "    # Store in the dictionary\n",
    "    seqs[pdb_id] = seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "8fb28a68",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "print dict with four entries first and then call 1OLG\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(None,\n",
       " {'1FAG': 'THR-ILE-LYS-GLU-MET-PRO-GLN-PRO-LYS-THR-PHE-GLY-GLU-LEU-LYS-ASN-LEU-PRO-LEU-LEU-ASN-THR-ASP-LYS-PRO-VAL-GLN-ALA-LEU-MET-LYS-ILE-ALA-ASP-GLU-LEU-GLY-GLU-ILE-PHE-LYS-PHE-GLU-ALA-PRO-GLY-ARG-VAL-THR-ARG-TYR-LEU-SER-SER-GLN-ARG-LEU-ILE-LYS-GLU-ALA-CYS-ASP-GLU-SER-ARG-PHE-ASP-LYS-ASN-LEU-SER-GLN-ALA-LEU-LYS-PHE-VAL-ARG-ASP-PHE-ALA-GLY-ASP-GLY-LEU-PHE-THR-SER-TRP-THR-HIS-GLU-LYS-ASN-TRP-LYS-LYS-ALA-HIS-ASN-ILE-LEU-LEU-PRO-SER-PHE-SER-GLN-GLN-ALA-MET-LYS-GLY-TYR-HIS-ALA-MET-MET-VAL-ASP-ILE-ALA-VAL-GLN-LEU-VAL-GLN-LYS-TRP-GLU-ARG-LEU-ASN-ALA-ASP-GLU-HIS-ILE-GLU-VAL-PRO-GLU-ASP-MET-THR-ARG-LEU-THR-LEU-ASP-THR-ILE-GLY-LEU-CYS-GLY-PHE-ASN-TYR-ARG-PHE-ASN-SER-PHE-TYR-ARG-ASP-GLN-PRO-HIS-PRO-PHE-ILE-THR-SER-MET-VAL-ARG-ALA-LEU-ASP-GLU-ALA-MET-ASN-LYS-LEU-GLN-ARG-ALA-ASN-PRO-ASP-ASP-PRO-ALA-TYR-ASP-GLU-ASN-LYS-ARG-GLN-PHE-GLN-GLU-ASP-ILE-LYS-VAL-MET-ASN-ASP-LEU-VAL-ASP-LYS-ILE-ILE-ALA-ASP-ARG-LYS-ALA-SER-GLY-GLU-GLN-SER-ASP-ASP-LEU-LEU-THR-HIS-MET-LEU-ASN-GLY-LYS-ASP-PRO-GLU-THR-GLY-GLU-PRO-LEU-ASP-ASP-GLU-ASN-ILE-ARG-TYR-GLN-ILE-ILE-THR-PHE-LEU-ILE-ALA-GLY-HIS-GLU-THR-THR-SER-GLY-LEU-LEU-SER-PHE-ALA-LEU-TYR-PHE-LEU-VAL-LYS-ASN-PRO-HIS-VAL-LEU-GLN-LYS-ALA-ALA-GLU-GLU-ALA-ALA-ARG-VAL-LEU-VAL-ASP-PRO-VAL-PRO-SER-TYR-LYS-GLN-VAL-LYS-GLN-LEU-LYS-TYR-VAL-GLY-MET-VAL-LEU-ASN-GLU-ALA-LEU-ARG-LEU-TRP-PRO-THR-ALA-PRO-ALA-PHE-SER-LEU-TYR-ALA-LYS-GLU-ASP-THR-VAL-LEU-GLY-GLY-GLU-TYR-PRO-LEU-GLU-LYS-GLY-ASP-GLU-LEU-MET-VAL-LEU-ILE-PRO-GLN-LEU-HIS-ARG-ASP-LYS-THR-ILE-TRP-GLY-ASP-ASP-VAL-GLU-GLU-PHE-ARG-PRO-GLU-ARG-PHE-GLU-ASN-PRO-SER-ALA-ILE-PRO-GLN-HIS-ALA-PHE-LYS-PRO-PHE-GLY-ASN-GLY-GLN-ARG-ALA-CYS-ILE-GLY-GLN-GLN-PHE-ALA-LEU-HIS-GLU-ALA-THR-LEU-VAL-LEU-GLY-MET-MET-LEU-LYS-HIS-PHE-ASP-PHE-GLU-ASP-HIS-THR-ASN-TYR-GLU-LEU-ASP-ILE-LYS-GLU-THR-LEU-THR-LEU-LYS-PRO-GLU-GLY-PHE-VAL-VAL-LYS-ALA-LYS-SER-LYS-LYS-ILE-PRO-LEU-GLY-GLY-ILE-PRO-SER-PRO-SER-THR-GLU-GLN-SER-ALA-LYS-LYS-VAL-ARG',\n",
       "  '1J6Z': 'ASP-GLU-ASP-GLU-THR-THR-ALA-LEU-VAL-CYS-ASP-ASN-GLY-SER-GLY-LEU-VAL-LYS-ALA-GLY-PHE-ALA-GLY-ASP-ASP-ALA-PRO-ARG-ALA-VAL-PHE-PRO-SER-ILE-VAL-GLY-ARG-PRO-ARG-HIS-GLN-GLY-VAL-MET-VAL-GLY-MET-GLY-GLN-LYS-ASP-SER-TYR-VAL-GLY-ASP-GLU-ALA-GLN-SER-LYS-ARG-GLY-ILE-LEU-THR-LEU-LYS-TYR-PRO-ILE-GLU-HIC-GLY-ILE-ILE-THR-ASN-TRP-ASP-ASP-MET-GLU-LYS-ILE-TRP-HIS-HIS-THR-PHE-TYR-ASN-GLU-LEU-ARG-VAL-ALA-PRO-GLU-GLU-HIS-PRO-THR-LEU-LEU-THR-GLU-ALA-PRO-LEU-ASN-PRO-LYS-ALA-ASN-ARG-GLU-LYS-MET-THR-GLN-ILE-MET-PHE-GLU-THR-PHE-ASN-VAL-PRO-ALA-MET-TYR-VAL-ALA-ILE-GLN-ALA-VAL-LEU-SER-LEU-TYR-ALA-SER-GLY-ARG-THR-THR-GLY-ILE-VAL-LEU-ASP-SER-GLY-ASP-GLY-VAL-THR-HIS-ASN-VAL-PRO-ILE-TYR-GLU-GLY-TYR-ALA-LEU-PRO-HIS-ALA-ILE-MET-ARG-LEU-ASP-LEU-ALA-GLY-ARG-ASP-LEU-THR-ASP-TYR-LEU-MET-LYS-ILE-LEU-THR-GLU-ARG-GLY-TYR-SER-PHE-VAL-THR-THR-ALA-GLU-ARG-GLU-ILE-VAL-ARG-ASP-ILE-LYS-GLU-LYS-LEU-CYS-TYR-VAL-ALA-LEU-ASP-PHE-GLU-ASN-GLU-MET-ALA-THR-ALA-ALA-SER-SER-SER-SER-LEU-GLU-LYS-SER-TYR-GLU-LEU-PRO-ASP-GLY-GLN-VAL-ILE-THR-ILE-GLY-ASN-GLU-ARG-PHE-ARG-CYS-PRO-GLU-THR-LEU-PHE-GLN-PRO-SER-PHE-ILE-GLY-MET-GLU-SER-ALA-GLY-ILE-HIS-GLU-THR-THR-TYR-ASN-SER-ILE-MET-LYS-CYS-ASP-ILE-ASP-ILE-ARG-LYS-ASP-LEU-TYR-ALA-ASN-ASN-VAL-MET-SER-GLY-GLY-THR-THR-MET-TYR-PRO-GLY-ILE-ALA-ASP-ARG-MET-GLN-LYS-GLU-ILE-THR-ALA-LEU-ALA-PRO-SER-THR-MET-LYS-ILE-LYS-ILE-ILE-ALA-PRO-PRO-GLU-ARG-LYS-TYR-SER-VAL-TRP-ILE-GLY-GLY-SER-ILE-LEU-ALA-SER-LEU-SER-THR-PHE-GLN-GLN-MET-TRP-ILE-THR-LYS-GLN-GLU-TYR-ASP-GLU-ALA-GLY-PRO-SER-ILE-VAL-HIS-ARG-LYS-CYS-PHE',\n",
       "  '1OLG': 'LYS-LYS-LYS-PRO-LEU-ASP-GLY-GLU-TYR-PHE-THR-LEU-GLN-ILE-ARG-GLY-ARG-GLU-ARG-PHE-GLU-MET-PHE-ARG-GLU-LEU-ASN-GLU-ALA-LEU-GLU-LEU-LYS-ASP-ALA-GLN-ALA-GLY-LYS-GLU-PRO-GLY',\n",
       "  '2ERK': 'ALA-HIS-HIS-HIS-HIS-HIS-HIS-MET-ALA-ALA-ALA-ALA-ALA-ALA-GLY-PRO-GLU-MET-VAL-ARG-GLY-GLN-VAL-PHE-ASP-VAL-GLY-PRO-ARG-TYR-THR-ASN-LEU-SER-TYR-ILE-GLY-GLU-GLY-ALA-TYR-GLY-MET-VAL-CYS-SER-ALA-TYR-ASP-ASN-LEU-ASN-LYS-VAL-ARG-VAL-ALA-ILE-LYS-LYS-ILE-SER-PRO-PHE-GLU-HIS-GLN-THR-TYR-CYS-GLN-ARG-THR-LEU-ARG-GLU-ILE-LYS-ILE-LEU-LEU-ARG-PHE-ARG-HIS-GLU-ASN-ILE-ILE-GLY-ILE-ASN-ASP-ILE-ILE-ARG-ALA-PRO-THR-ILE-GLU-GLN-MET-LYS-ASP-VAL-TYR-ILE-VAL-GLN-ASP-LEU-MET-GLU-THR-ASP-LEU-TYR-LYS-LEU-LEU-LYS-THR-GLN-HIS-LEU-SER-ASN-ASP-HIS-ILE-CYS-TYR-PHE-LEU-TYR-GLN-ILE-LEU-ARG-GLY-LEU-LYS-TYR-ILE-HIS-SER-ALA-ASN-VAL-LEU-HIS-ARG-ASP-LEU-LYS-PRO-SER-ASN-LEU-LEU-LEU-ASN-THR-THR-CYS-ASP-LEU-LYS-ILE-CYS-ASP-PHE-GLY-LEU-ALA-ARG-VAL-ALA-ASP-PRO-ASP-HIS-ASP-HIS-THR-GLY-PHE-LEU-TPO-GLU-PTR-VAL-ALA-THR-ARG-TRP-TYR-ARG-ALA-PRO-GLU-ILE-MET-LEU-ASN-SER-LYS-GLY-TYR-THR-LYS-SER-ILE-ASP-ILE-TRP-SER-VAL-GLY-CYS-ILE-LEU-ALA-GLU-MET-LEU-SER-ASN-ARG-PRO-ILE-PHE-PRO-GLY-LYS-HIS-TYR-LEU-ASP-GLN-LEU-ASN-HIS-ILE-LEU-GLY-ILE-LEU-GLY-SER-PRO-SER-GLN-GLU-ASP-LEU-ASN-CYS-ILE-ILE-ASN-LEU-LYS-ALA-ARG-ASN-TYR-LEU-LEU-SER-LEU-PRO-HIS-LYS-ASN-LYS-VAL-PRO-TRP-ASN-ARG-LEU-PHE-PRO-ASN-ALA-ASP-SER-LYS-ALA-LEU-ASP-LEU-LEU-ASP-LYS-MET-LEU-THR-PHE-ASN-PRO-HIS-LYS-ARG-ILE-GLU-VAL-GLU-GLN-ALA-LEU-ALA-HIS-PRO-TYR-LEU-GLU-GLN-TYR-TYR-ASP-PRO-SER-ASP-GLU-PRO-ILE-ALA-GLU-ALA-PRO-PHE-LYS-PHE-ASP-MET-GLU-LEU-ASP-ASP-LEU-PRO-LYS-GLU-LYS-LEU-LYS-GLU-LEU-ILE-PHE-GLU-GLU-THR-ALA-ARG-PHE-GLN-PRO-GLY-TYR-ARG-SER'},\n",
       " None,\n",
       " 'LYS-LYS-LYS-PRO-LEU-ASP-GLY-GLU-TYR-PHE-THR-LEU-GLN-ILE-ARG-GLY-ARG-GLU-ARG-PHE-GLU-MET-PHE-ARG-GLU-LEU-ASN-GLU-ALA-LEU-GLU-LEU-LYS-ASP-ALA-GLN-ALA-GLY-LYS-GLU-PRO-GLY')"
      ]
     },
     "execution_count": 89,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print('print dict with four entries first and then call 1OLG'), seqs, print(), seqs['1OLG']"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56030374",
   "metadata": {},
   "source": [
    "## Lesson 12: Version control with Git"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cc48cfb0",
   "metadata": {},
   "source": [
    "Clone repository o shakespearian insulter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a5a42b7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Cloning into 'insulter'...\n"
     ]
    }
   ],
   "source": [
    "!git clone https://gist.github.com/3165396.git insulter"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "293a94e3",
   "metadata": {},
   "source": [
    "run the insulter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "09b6e22c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Thou impertinent bunch-back'd bum-bailey!\n"
     ]
    }
   ],
   "source": [
    "!python insulter\\insulter.py"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f60d1527",
   "metadata": {},
   "source": [
    "Update external changes, first fetch and then merges the latest version of the repository with the local version (FETCH_HEAD is automatically created when running git fetch)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "e9501448",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Already up to date.\n"
     ]
    }
   ],
   "source": [
    "!git fetch \n",
    "!git merge FETCH_HEAD"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "23247fc6",
   "metadata": {},
   "source": [
    "OR combined `git pull origin main`, i.e. abbreviated as "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "f0818af8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Already up to date.\n"
     ]
    }
   ],
   "source": [
    "!git pull"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fa98bc74",
   "metadata": {},
   "source": [
    ">...\n",
    ">\n",
    "> good description of how to make an own package available for download, in the previous lesson there is a good description on how to create an own package from the command line\n",
    ">\n",
    ">..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "cbb19373",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting bootcamp_utils\n",
      "  Downloading bootcamp_utils-0.0.7-py3-none-any.whl (9.4 kB)\n",
      "Requirement already satisfied: bokeh>=2.0.0 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from bootcamp_utils) (3.2.1)\n",
      "Requirement already satisfied: Jinja2>=2.9 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from bokeh>=2.0.0->bootcamp_utils) (3.1.2)\n",
      "Requirement already satisfied: contourpy>=1 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from bokeh>=2.0.0->bootcamp_utils) (1.2.0)\n",
      "Requirement already satisfied: numpy>=1.16 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from bokeh>=2.0.0->bootcamp_utils) (1.24.3)\n",
      "Requirement already satisfied: packaging>=16.8 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from bokeh>=2.0.0->bootcamp_utils) (23.1)\n",
      "Requirement already satisfied: pandas>=1.2 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from bokeh>=2.0.0->bootcamp_utils) (2.1.1)\n",
      "Requirement already satisfied: pillow>=7.1.0 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from bokeh>=2.0.0->bootcamp_utils) (10.0.1)\n",
      "Requirement already satisfied: PyYAML>=3.10 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from bokeh>=2.0.0->bootcamp_utils) (6.0.1)\n",
      "Requirement already satisfied: tornado>=5.1 in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from bokeh>=2.0.0->bootcamp_utils) (6.3.3)\n",
      "Requirement already satisfied: xyzservices>=2021.09.1 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from bokeh>=2.0.0->bootcamp_utils) (2022.9.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from Jinja2>=2.9->bokeh>=2.0.0->bootcamp_utils) (2.1.1)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from pandas>=1.2->bokeh>=2.0.0->bootcamp_utils) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from pandas>=1.2->bokeh>=2.0.0->bootcamp_utils) (2023.3.post1)\n",
      "Requirement already satisfied: tzdata>=2022.1 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from pandas>=1.2->bokeh>=2.0.0->bootcamp_utils) (2023.3)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from python-dateutil>=2.8.2->pandas>=1.2->bokeh>=2.0.0->bootcamp_utils) (1.16.0)\n",
      "Installing collected packages: bootcamp_utils\n",
      "Successfully installed bootcamp_utils-0.0.7\n"
     ]
    }
   ],
   "source": [
    "# install a package for the bootcampl\n",
    "#!pip install bootcamp_utils"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a12250b",
   "metadata": {},
   "source": [
    "## Lesson 13: Errors and exception handling"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1089df51",
   "metadata": {},
   "source": [
    "In Python, an error detected during execution is called an exception, i.e. the interpreter “raised an exception.”\n",
    "\n",
    "[Built-in Exception](https://docs.python.org/3/library/exceptions.html#bltin-exceptions) (types of errors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "048df145",
   "metadata": {},
   "outputs": [],
   "source": [
    "import warnings\n",
    "import bootcamp_utils"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9bc49fa4",
   "metadata": {},
   "source": [
    "* Syntax errors: forgot a bracket e.g.\n",
    "* Runtimw errors: e.g. TypeError, or generally a RuntimeError (not-syntax)\n",
    "    * e.g. trying to add a string to an integer\n",
    "* Semantic errors: errors in the function of the program\n",
    "    * example below"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3a3297a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# A function to append a list onto itself, with the intention of\n",
    "# returning a new list, but leaving the input unaltered\n",
    "def double_list(in_list):\n",
    "    \"\"\"Append a list to itself.\"\"\"\n",
    "    in_list += in_list\n",
    "    return in_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "e1bd7a90",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([3, 2, 1], [3, 2, 1, 3, 2, 1])"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def double_list(in_list):\n",
    "    \"\"\"Append a list to itself.\"\"\"\n",
    "    #make a deep copy of in_list first\n",
    "    in_list = in_list.copy()\n",
    "    in_list += in_list\n",
    "    return in_list\n",
    "\n",
    "my_list = [3, 2, 1]\n",
    "\n",
    "# Double it\n",
    "my_list_double = double_list(my_list)\n",
    "my_list, my_list_double"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "272f35d2",
   "metadata": {},
   "source": [
    "or can modify like this:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "e73a7ed8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "([3, 2, 1], [3, 2, 1, 3, 2, 1])"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def double_list(in_list):\n",
    "    \"\"\"Create a new list by appending a copy of in_list to itself.\"\"\"\n",
    "    new_list = in_list + in_list  # Creating a new list without modifying the original\n",
    "    return new_list\n",
    "\n",
    "my_list_double = double_list(my_list) # but \n",
    "\n",
    "my_list, my_list_double"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fd889253",
   "metadata": {},
   "source": [
    "**Handle errors** \n",
    "* opening with the arguments `try:`, `except .. as e:` and `finally: .. pass`\n",
    "    * if the `try` argument works, the rest is skipped\n",
    "    * ImportError will be skipped\n",
    "* and closing with \n",
    "    * `if (True):` apply function\n",
    "    * `else:` manual calculus"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "fb3f4bd6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16\n"
     ]
    }
   ],
   "source": [
    "# Try to get the gc_content module\n",
    "try:\n",
    "    import gc_content\n",
    "    have_gc = True\n",
    "except ImportError as e: # only ImportError is a valid exception, i.e. if the module is not found\n",
    "    have_gc = False\n",
    "finally:\n",
    "    # Do whatever is necessary here, like close files\n",
    "    pass\n",
    "\n",
    "seq = 'ACGATCTACGATCAGCTGCGCGCATCG'\n",
    "\n",
    "if have_gc:\n",
    "    print(gc_content(seq))\n",
    "else:\n",
    "    print(seq.count('G') + seq.count('C'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3faad55d",
   "metadata": {},
   "source": [
    "Can also **give warnings**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "6eab3329",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "16\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\fulci\\AppData\\Local\\Temp\\ipykernel_27972\\165630899.py:8: UserWarning: Failed to load gc_content. Using custom function.\n",
      "  warnings.warn(\n"
     ]
    }
   ],
   "source": [
    "# Try to get the gc_content module\n",
    "try:\n",
    "    import gc_content\n",
    "\n",
    "    have_gc = True\n",
    "except ImportError as e:\n",
    "    have_gc = False\n",
    "    warnings.warn(\n",
    "        \"Failed to load gc_content. Using custom function.\", UserWarning\n",
    "    )\n",
    "finally:\n",
    "    pass\n",
    "\n",
    "seq = \"ACGATCTACGATCAGCTGCGCGCATCG\"\n",
    "\n",
    "if have_gc:\n",
    "    print(gc_content(seq))\n",
    "else:\n",
    "    print(seq.count(\"G\") + seq.count(\"C\"))\n",
    "\n",
    "# Normally, we would use an ImportWarning, but those are ignored by default, so we have used a UserWarning."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7cf08b62",
   "metadata": {},
   "source": [
    "**Check the input**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "18316fd2",
   "metadata": {},
   "outputs": [],
   "source": [
    "def one_to_three(seq):\n",
    "    \"\"\"\n",
    "    Converts a protein sequence using one-letter abbreviations\n",
    "    to one using three-letter abbreviations.\n",
    "    \"\"\"\n",
    "    # Convert seq to upper case\n",
    "    seq = seq.upper()\n",
    "\n",
    "    aa_list = []\n",
    "    for amino_acid in seq:\n",
    "        # Check if the `amino_acid` is in our dictionary `bootcamp_utils.aa`\n",
    "        if amino_acid not in bootcamp_utils.aa:\n",
    "            raise RuntimeError(f\"{amino_acid} is not a valid amino acid\")\n",
    "        # Add the `amino_acid` to our aa_list\n",
    "        aa_list.append(bootcamp_utils.aa[amino_acid])\n",
    "\n",
    "    # Return the amino acids, joined together, with a dash as a separator.\n",
    "    return \"-\".join(aa_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "fe6eb376",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1. Trp-Ala-Glu-Ile-Phe-Asn-Ser-Asp-Phe-Lys-Leu-Asn-Ser-Ala-Glu\n"
     ]
    },
    {
     "ename": "RuntimeError",
     "evalue": "Z is not a valid amino acid",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[45], line 3\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m1.\u001b[39m\u001b[38;5;124m'\u001b[39m, one_to_three(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mwaeifnsdfklnsae\u001b[39m\u001b[38;5;124m'\u001b[39m))\n\u001b[1;32m----> 3\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m2.\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[43mone_to_three\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mwaeifnsdfzklnsae\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m)\n",
      "Cell \u001b[1;32mIn[41], line 13\u001b[0m, in \u001b[0;36mone_to_three\u001b[1;34m(seq)\u001b[0m\n\u001b[0;32m     10\u001b[0m \u001b[38;5;28;01mfor\u001b[39;00m amino_acid \u001b[38;5;129;01min\u001b[39;00m seq:\n\u001b[0;32m     11\u001b[0m     \u001b[38;5;66;03m# Check if the `amino_acid` is in our dictionary `bootcamp_utils.aa`\u001b[39;00m\n\u001b[0;32m     12\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m amino_acid \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m bootcamp_utils\u001b[38;5;241m.\u001b[39maa:\n\u001b[1;32m---> 13\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mRuntimeError\u001b[39;00m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mamino_acid\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m is not a valid amino acid\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m     14\u001b[0m     \u001b[38;5;66;03m# Add the `amino_acid` to our aa_list\u001b[39;00m\n\u001b[0;32m     15\u001b[0m     aa_list\u001b[38;5;241m.\u001b[39mappend(bootcamp_utils\u001b[38;5;241m.\u001b[39maa[amino_acid])\n",
      "\u001b[1;31mRuntimeError\u001b[0m: Z is not a valid amino acid"
     ]
    }
   ],
   "source": [
    "print('1.', one_to_three('waeifnsdfklnsae'))\n",
    "\n",
    "print('2.', one_to_three('waeifnsdfzklnsae'))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "96a7f15e",
   "metadata": {},
   "source": [
    "Useful error description is very good to have when doing functions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cbec9e3f",
   "metadata": {},
   "source": [
    "## Lesson 14: Style"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "2f411412",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting blackcellmagic\n",
      "  Downloading blackcellmagic-0.0.3-py3-none-any.whl (4.2 kB)\n",
      "Collecting black<22.0,>=21.9b0 (from blackcellmagic)\n",
      "  Downloading black-21.12b0-py3-none-any.whl (156 kB)\n",
      "     ---------------------------------------- 0.0/156.7 kB ? eta -:--:--\n",
      "     -- ------------------------------------- 10.2/156.7 kB ? eta -:--:--\n",
      "     ------- ----------------------------- 30.7/156.7 kB 325.1 kB/s eta 0:00:01\n",
      "     --------------------- --------------- 92.2/156.7 kB 744.7 kB/s eta 0:00:01\n",
      "     -------------------------------------- 156.7/156.7 kB 1.0 MB/s eta 0:00:00\n",
      "Requirement already satisfied: jupyter<2.0.0,>=1.0.0 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from blackcellmagic) (1.0.0)\n",
      "Requirement already satisfied: click>=7.1.2 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from black<22.0,>=21.9b0->blackcellmagic) (8.1.7)\n",
      "Requirement already satisfied: platformdirs>=2 in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from black<22.0,>=21.9b0->blackcellmagic) (3.11.0)\n",
      "Collecting tomli<2.0.0,>=0.2.6 (from black<22.0,>=21.9b0->blackcellmagic)\n",
      "  Downloading tomli-1.2.3-py3-none-any.whl (12 kB)\n",
      "Requirement already satisfied: pathspec<1,>=0.9.0 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from black<22.0,>=21.9b0->blackcellmagic) (0.10.3)\n",
      "Requirement already satisfied: typing-extensions>=3.10.0.0 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from black<22.0,>=21.9b0->blackcellmagic) (4.7.1)\n",
      "Requirement already satisfied: mypy-extensions>=0.4.3 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from black<22.0,>=21.9b0->blackcellmagic) (1.0.0)\n",
      "Requirement already satisfied: notebook in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from jupyter<2.0.0,>=1.0.0->blackcellmagic) (6.5.4)\n",
      "Requirement already satisfied: qtconsole in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from jupyter<2.0.0,>=1.0.0->blackcellmagic) (5.4.2)\n",
      "Requirement already satisfied: jupyter-console in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from jupyter<2.0.0,>=1.0.0->blackcellmagic) (6.6.3)\n",
      "Requirement already satisfied: nbconvert in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from jupyter<2.0.0,>=1.0.0->blackcellmagic) (7.10.0)\n",
      "Requirement already satisfied: ipykernel in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from jupyter<2.0.0,>=1.0.0->blackcellmagic) (6.25.2)\n",
      "Requirement already satisfied: ipywidgets in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from jupyter<2.0.0,>=1.0.0->blackcellmagic) (8.0.4)\n",
      "Requirement already satisfied: colorama in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from click>=7.1.2->black<22.0,>=21.9b0->blackcellmagic) (0.4.6)\n",
      "Requirement already satisfied: comm>=0.1.1 in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (0.1.4)\n",
      "Requirement already satisfied: debugpy>=1.6.5 in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (1.8.0)\n",
      "Requirement already satisfied: ipython>=7.23.1 in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (8.16.1)\n",
      "Requirement already satisfied: jupyter-client>=6.1.12 in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (8.3.1)\n",
      "Requirement already satisfied: jupyter-core!=5.0.*,>=4.12 in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (5.3.2)\n",
      "Requirement already satisfied: matplotlib-inline>=0.1 in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (0.1.6)\n",
      "Requirement already satisfied: nest-asyncio in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (1.5.8)\n",
      "Requirement already satisfied: packaging in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (23.1)\n",
      "Requirement already satisfied: psutil in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (5.9.5)\n",
      "Requirement already satisfied: pyzmq>=20 in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (25.1.1)\n",
      "Requirement already satisfied: tornado>=6.1 in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (6.3.3)\n",
      "Requirement already satisfied: traitlets>=5.4.0 in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (5.11.2)\n",
      "Requirement already satisfied: widgetsnbextension~=4.0 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from ipywidgets->jupyter<2.0.0,>=1.0.0->blackcellmagic) (4.0.5)\n",
      "Requirement already satisfied: jupyterlab-widgets~=3.0 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from ipywidgets->jupyter<2.0.0,>=1.0.0->blackcellmagic) (3.0.9)\n",
      "Requirement already satisfied: prompt-toolkit>=3.0.30 in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from jupyter-console->jupyter<2.0.0,>=1.0.0->blackcellmagic) (3.0.39)\n",
      "Requirement already satisfied: pygments in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from jupyter-console->jupyter<2.0.0,>=1.0.0->blackcellmagic) (2.16.1)\n",
      "Requirement already satisfied: beautifulsoup4 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from nbconvert->jupyter<2.0.0,>=1.0.0->blackcellmagic) (4.12.2)\n",
      "Requirement already satisfied: bleach!=5.0.0 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from nbconvert->jupyter<2.0.0,>=1.0.0->blackcellmagic) (4.1.0)\n",
      "Requirement already satisfied: defusedxml in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from nbconvert->jupyter<2.0.0,>=1.0.0->blackcellmagic) (0.7.1)\n",
      "Requirement already satisfied: jinja2>=3.0 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from nbconvert->jupyter<2.0.0,>=1.0.0->blackcellmagic) (3.1.2)\n",
      "Requirement already satisfied: jupyterlab-pygments in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from nbconvert->jupyter<2.0.0,>=1.0.0->blackcellmagic) (0.2.2)\n",
      "Requirement already satisfied: markupsafe>=2.0 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from nbconvert->jupyter<2.0.0,>=1.0.0->blackcellmagic) (2.1.1)\n",
      "Requirement already satisfied: mistune<4,>=2.0.3 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from nbconvert->jupyter<2.0.0,>=1.0.0->blackcellmagic) (2.0.4)\n",
      "Requirement already satisfied: nbclient>=0.5.0 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from nbconvert->jupyter<2.0.0,>=1.0.0->blackcellmagic) (0.8.0)\n",
      "Requirement already satisfied: nbformat>=5.7 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from nbconvert->jupyter<2.0.0,>=1.0.0->blackcellmagic) (5.9.2)\n",
      "Requirement already satisfied: pandocfilters>=1.4.1 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from nbconvert->jupyter<2.0.0,>=1.0.0->blackcellmagic) (1.5.0)\n",
      "Requirement already satisfied: tinycss2 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from nbconvert->jupyter<2.0.0,>=1.0.0->blackcellmagic) (1.2.1)\n",
      "Requirement already satisfied: argon2-cffi in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from notebook->jupyter<2.0.0,>=1.0.0->blackcellmagic) (21.3.0)\n",
      "Requirement already satisfied: ipython-genutils in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from notebook->jupyter<2.0.0,>=1.0.0->blackcellmagic) (0.2.0)\n",
      "Requirement already satisfied: Send2Trash>=1.8.0 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from notebook->jupyter<2.0.0,>=1.0.0->blackcellmagic) (1.8.2)\n",
      "Requirement already satisfied: terminado>=0.8.3 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from notebook->jupyter<2.0.0,>=1.0.0->blackcellmagic) (0.17.1)\n",
      "Requirement already satisfied: prometheus-client in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from notebook->jupyter<2.0.0,>=1.0.0->blackcellmagic) (0.14.1)\n",
      "Requirement already satisfied: nbclassic>=0.4.7 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from notebook->jupyter<2.0.0,>=1.0.0->blackcellmagic) (1.0.0)\n",
      "Requirement already satisfied: qtpy>=2.0.1 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from qtconsole->jupyter<2.0.0,>=1.0.0->blackcellmagic) (2.4.1)\n",
      "Requirement already satisfied: six>=1.9.0 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from bleach!=5.0.0->nbconvert->jupyter<2.0.0,>=1.0.0->blackcellmagic) (1.16.0)\n",
      "Requirement already satisfied: webencodings in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from bleach!=5.0.0->nbconvert->jupyter<2.0.0,>=1.0.0->blackcellmagic) (0.5.1)\n",
      "Requirement already satisfied: backcall in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from ipython>=7.23.1->ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (0.2.0)\n",
      "Requirement already satisfied: decorator in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from ipython>=7.23.1->ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (5.1.1)\n",
      "Requirement already satisfied: jedi>=0.16 in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from ipython>=7.23.1->ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (0.19.1)\n",
      "Requirement already satisfied: pickleshare in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from ipython>=7.23.1->ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (0.7.5)\n",
      "Requirement already satisfied: stack-data in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from ipython>=7.23.1->ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (0.6.3)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from jupyter-client>=6.1.12->ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (2.8.2)\n",
      "Requirement already satisfied: pywin32>=300 in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from jupyter-core!=5.0.*,>=4.12->ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (306)\n",
      "Requirement already satisfied: jupyter-server>=1.8 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from nbclassic>=0.4.7->notebook->jupyter<2.0.0,>=1.0.0->blackcellmagic) (1.23.4)\n",
      "Requirement already satisfied: notebook-shim>=0.2.3 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from nbclassic>=0.4.7->notebook->jupyter<2.0.0,>=1.0.0->blackcellmagic) (0.2.3)\n",
      "Requirement already satisfied: fastjsonschema in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from nbformat>=5.7->nbconvert->jupyter<2.0.0,>=1.0.0->blackcellmagic) (2.16.2)\n",
      "Requirement already satisfied: jsonschema>=2.6 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from nbformat>=5.7->nbconvert->jupyter<2.0.0,>=1.0.0->blackcellmagic) (4.19.2)\n",
      "Requirement already satisfied: wcwidth in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from prompt-toolkit>=3.0.30->jupyter-console->jupyter<2.0.0,>=1.0.0->blackcellmagic) (0.2.8)\n",
      "Requirement already satisfied: pywinpty>=1.1.0 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from terminado>=0.8.3->notebook->jupyter<2.0.0,>=1.0.0->blackcellmagic) (2.0.10)\n",
      "Requirement already satisfied: argon2-cffi-bindings in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from argon2-cffi->notebook->jupyter<2.0.0,>=1.0.0->blackcellmagic) (21.2.0)\n",
      "Requirement already satisfied: soupsieve>1.2 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from beautifulsoup4->nbconvert->jupyter<2.0.0,>=1.0.0->blackcellmagic) (2.5)\n",
      "Requirement already satisfied: parso<0.9.0,>=0.8.3 in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from jedi>=0.16->ipython>=7.23.1->ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (0.8.3)\n",
      "Requirement already satisfied: attrs>=22.2.0 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from jsonschema>=2.6->nbformat>=5.7->nbconvert->jupyter<2.0.0,>=1.0.0->blackcellmagic) (23.1.0)\n",
      "Requirement already satisfied: jsonschema-specifications>=2023.03.6 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from jsonschema>=2.6->nbformat>=5.7->nbconvert->jupyter<2.0.0,>=1.0.0->blackcellmagic) (2023.7.1)\n",
      "Requirement already satisfied: referencing>=0.28.4 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from jsonschema>=2.6->nbformat>=5.7->nbconvert->jupyter<2.0.0,>=1.0.0->blackcellmagic) (0.30.2)\n",
      "Requirement already satisfied: rpds-py>=0.7.1 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from jsonschema>=2.6->nbformat>=5.7->nbconvert->jupyter<2.0.0,>=1.0.0->blackcellmagic) (0.10.6)\n",
      "Requirement already satisfied: anyio<4,>=3.1.0 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from jupyter-server>=1.8->nbclassic>=0.4.7->notebook->jupyter<2.0.0,>=1.0.0->blackcellmagic) (3.5.0)\n",
      "Requirement already satisfied: websocket-client in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from jupyter-server>=1.8->nbclassic>=0.4.7->notebook->jupyter<2.0.0,>=1.0.0->blackcellmagic) (0.58.0)\n",
      "Requirement already satisfied: cffi>=1.0.1 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from argon2-cffi-bindings->argon2-cffi->notebook->jupyter<2.0.0,>=1.0.0->blackcellmagic) (1.16.0)\n",
      "Requirement already satisfied: executing>=1.2.0 in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from stack-data->ipython>=7.23.1->ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (2.0.0)\n",
      "Requirement already satisfied: asttokens>=2.1.0 in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from stack-data->ipython>=7.23.1->ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (2.4.0)\n",
      "Requirement already satisfied: pure-eval in c:\\users\\fulci\\appdata\\roaming\\python\\python311\\site-packages (from stack-data->ipython>=7.23.1->ipykernel->jupyter<2.0.0,>=1.0.0->blackcellmagic) (0.2.2)\n",
      "Requirement already satisfied: idna>=2.8 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from anyio<4,>=3.1.0->jupyter-server>=1.8->nbclassic>=0.4.7->notebook->jupyter<2.0.0,>=1.0.0->blackcellmagic) (3.4)\n",
      "Requirement already satisfied: sniffio>=1.1 in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from anyio<4,>=3.1.0->jupyter-server>=1.8->nbclassic>=0.4.7->notebook->jupyter<2.0.0,>=1.0.0->blackcellmagic) (1.3.0)\n",
      "Requirement already satisfied: pycparser in c:\\users\\fulci\\anaconda3\\lib\\site-packages (from cffi>=1.0.1->argon2-cffi-bindings->argon2-cffi->notebook->jupyter<2.0.0,>=1.0.0->blackcellmagic) (2.21)\n",
      "Installing collected packages: tomli, black, blackcellmagic\n",
      "  Attempting uninstall: black\n",
      "    Found existing installation: black 0.0\n",
      "    Uninstalling black-0.0:\n",
      "      Successfully uninstalled black-0.0\n",
      "Successfully installed black-21.12b0 blackcellmagic-0.0.3 tomli-1.2.3\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "python-lsp-black 1.2.1 requires black>=22.3.0, but you have black 21.12b0 which is incompatible.\n",
      "spyder 5.4.3 requires jedi<0.19.0,>=0.17.2, but you have jedi 0.19.1 which is incompatible.\n"
     ]
    }
   ],
   "source": [
    "#!pip install blackcellmagic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "665ea443",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext blackcellmagic # to format code\n",
    "\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f89a0a48",
   "metadata": {},
   "source": [
    "**PEP 8**: Python Enhancement proposal, i.e. a Python Style Guide"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "id": "da30dadf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import this # see the Zen of Python with the most important style rules"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1b04c730",
   "metadata": {},
   "source": [
    "blackened cell / **black formatting**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "1d8445b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%black ## once we run this cell, the code is formatted (remove comment)\n",
    "seq='AUCUGUACUAAUGCUCAGCACGACGUACG'\n",
    "c='AUG'  # This is the start codon\n",
    "i =0  # Initialize sequence index\n",
    "while seq[ i : i + 3 ]!=c:\n",
    "    i+=1\n",
    "\n",
    "print('The start codon starts at index', i)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e0c8dbd",
   "metadata": {},
   "source": [
    "## Lesson 15: Comprehensions"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28da6203",
   "metadata": {},
   "source": [
    "List comprehendsions can be used to avoid nested loops and allocating memory more efficiently"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "c2aa5a04",
   "metadata": {},
   "outputs": [],
   "source": [
    "#!pip install sympy\n",
    "import sympy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "57ce0d10",
   "metadata": {},
   "source": [
    "from"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "8850a7a8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997]\n"
     ]
    }
   ],
   "source": [
    "# Largest number to consider\n",
    "n_max = 1000\n",
    "\n",
    "# Initialize list of primes\n",
    "primes = []\n",
    "\n",
    "# Loop through odd integers and add primes\n",
    "for x in range(n_max+1):\n",
    "    if sympy.isprime(x):\n",
    "        primes.append(x)\n",
    "\n",
    "# Take a look\n",
    "print(primes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e30923e9",
   "metadata": {},
   "source": [
    "to"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "3cdccd9e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997]\n"
     ]
    }
   ],
   "source": [
    "primes = [x for x in range(n_max) if sympy.isprime(x)]\n",
    "\n",
    "# Take a look\n",
    "print(primes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72b659b8",
   "metadata": {},
   "source": [
    "**=>** The first part, x, is an expression that will be inserted into the list. Next comes a for statement to produce the iterator. Finally, there is a conditional; if the conditional evaluates True, then the expression expression is included in the list"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e88c7b8e",
   "metadata": {},
   "source": [
    "Create a **list with list comprehensions**:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "ce902046",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99]\n",
      "\n",
      "[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(None, None, None)"
      ]
     },
     "execution_count": 73,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Give same result as list(range(100))\n",
    "my_list_of_ints = [i for i in range(100)] # without a conditional\n",
    "\n",
    "print(my_list_of_ints), print(), print(list(range(100)))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7aeb9e05",
   "metadata": {},
   "source": [
    "do a **list out of tuples**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "cf452759",
   "metadata": {},
   "outputs": [],
   "source": [
    "names = (\n",
    "    \"Frances Arnold\",\n",
    "    \"George Smith\",\n",
    "    \"Gregory Winter\",\n",
    "    \"postponed\",\n",
    "    \"Denis Mukwege\",\n",
    "    \"Nadia Murad\",\n",
    "    \"Arthur Ashkin\",\n",
    "    \"Gérard Mourou\",\n",
    "    \"Donna Strickland\",\n",
    "    \"James Allison\",\n",
    "    \"Tasuku Honjo\",\n",
    "    \"William Nordhaus\",\n",
    "    \"Paul Romer\",\n",
    ")\n",
    "\n",
    "nationalities = (\n",
    "    \"USA\",\n",
    "    \"USA\",\n",
    "    \"UK\",\n",
    "    \"---\",\n",
    "    \"DRC\",\n",
    "    \"Iraq\",\n",
    "    \"USA\",\n",
    "    \"France\",\n",
    "    \"Canada\",\n",
    "    \"USA\",\n",
    "    \"Japan\",\n",
    "    \"USA\",\n",
    "    \"USA\",\n",
    ")\n",
    "\n",
    "categories = (\n",
    "    \"Chemistry\",\n",
    "    \"Chemistry\",\n",
    "    \"Chemistry\",\n",
    "    \"Literature\",\n",
    "    \"Peace\",\n",
    "    \"Peace\",\n",
    "    \"Physics\",\n",
    "    \"Physics\",\n",
    "    \"Physics\",\n",
    "    \"Physiology or Medicine\",\n",
    "    \"Physiology or Medicine\",\n",
    "    \"Economics\",\n",
    "    \"Economics\",\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "3ff4d794",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('Chemistry', 'Frances Arnold', 'USA'),\n",
       " ('Chemistry', 'George Smith', 'USA'),\n",
       " ('Chemistry', 'Gregory Winter', 'UK'),\n",
       " ('Literature', 'postponed', '---'),\n",
       " ('Peace', 'Denis Mukwege', 'DRC'),\n",
       " ('Peace', 'Nadia Murad', 'Iraq'),\n",
       " ('Physics', 'Arthur Ashkin', 'USA'),\n",
       " ('Physics', 'Gérard Mourou', 'France'),\n",
       " ('Physics', 'Donna Strickland', 'Canada'),\n",
       " ('Physiology or Medicine', 'James Allison', 'USA'),\n",
       " ('Physiology or Medicine', 'Tasuku Honjo', 'Japan'),\n",
       " ('Economics', 'William Nordhaus', 'USA'),\n",
       " ('Economics', 'Paul Romer', 'USA')]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[(cat, name, nat) for name, nat, cat in zip(names, nationalities, categories)]\n",
    "# can use zip to iterate, as also range() and enumerate()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "ca2f227e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('Chemistry', 'Frances Arnold', 'USA'),\n",
       " ('Chemistry', 'George Smith', 'USA'),\n",
       " ('Chemistry', 'Gregory Winter', 'UK')]"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# or add a conditional e.g for only chemistry\n",
    "[\n",
    "    (cat, name, nat)\n",
    "    for name, nat, cat in zip(names, nationalities, categories)\n",
    "    if cat == \"Chemistry\"\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9546a10a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('Frances Arnold', 'George Smith'),\n",
       " ('Frances Arnold', 'Gregory Winter'),\n",
       " ('George Smith', 'Gregory Winter')]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# do a list with all possible combinations of two chemists nobel prize winners\n",
    "# First get list of chemistry laureates\n",
    "chem_names = [name for name, cat in zip(names, categories) if cat == \"Chemistry\"]\n",
    "\n",
    "# List of all possible pairs of chemistry laureates\n",
    "[\n",
    "    (n1, n2)\n",
    "    for i, n1 in enumerate(chem_names)\n",
    "    for j, n2 in enumerate(chem_names)\n",
    "    if i < j\n",
    "]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3936635a",
   "metadata": {},
   "source": [
    "General structure of List comprehensions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6773fe19",
   "metadata": {},
   "outputs": [],
   "source": [
    "[expression_to_put_in_list for i_1 in iterable_1 if condition_1\n",
    "                           for i_2 in iterable_2 if condition_2\n",
    "                                     ...\n",
    "                           for i_n in iterable_n if condition_n]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1de6c7fd",
   "metadata": {},
   "source": [
    "almost equivvalent to "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a711c2c5",
   "metadata": {},
   "outputs": [],
   "source": [
    "my_list = []\n",
    "for i_1 in iterable_1:\n",
    "    if condition_1:\n",
    "        for i_2 in iterable_2:\n",
    "            if condition_2:\n",
    "                ...\n",
    "                for i_n in iterable_n:\n",
    "                    if condition_n:\n",
    "                        my_list += [expression_to_put_in_list]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "36e3e37b",
   "metadata": {},
   "source": [
    "As else statements can not be used as seen from the general structure, we need to hide it into expression_to_put_in_list, which looks as follows:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "b5e4b2da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('Chemistry', 'Frances Arnold', 'USA'),\n",
       " ('Chemistry', 'George Smith', 'USA'),\n",
       " ('Chemistry', 'Gregory Winter', 'UK'),\n",
       " ('Literature', 'postponed', '---'),\n",
       " ('Peace', 'Denis Mukwege', 'DRC'),\n",
       " ('Peace', 'Nadia Murad', 'Iraq'),\n",
       " ('Physics', 'Arthur Ashkin', 'USA'),\n",
       " ('Physics', 'Gérard Mourou', 'France'),\n",
       " ('Physics', 'Donna Strickland', 'Canada'),\n",
       " ('Medicine', 'James Allison', 'USA'),\n",
       " ('Medicine', 'Tasuku Honjo', 'Japan'),\n",
       " ('Economics', 'William Nordhaus', 'USA'),\n",
       " ('Economics', 'Paul Romer', 'USA')]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[\n",
    "    (\"Medicine\", name, nat) if cat == \"Physiology or Medicine\" else (cat, name, nat)\n",
    "    for name, nat, cat in zip(names, nationalities, categories)\n",
    "]\n",
    "# where first we define that if the category is Ph. of medicine, then we want to change it to medicine,\n",
    "# else we want to keep the original category, and then we create a list out of them\n",
    "\n",
    "# this list comprehension is without a conditional!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6d485f5b",
   "metadata": {},
   "source": [
    "**Dictionary comprehensions** {key: (val1, val2) *for* var1, var2, var3 *in zip(*var1, var2, var3*)*}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e7585810",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Frances Arnold': ('Chemistry', 'USA'),\n",
       " 'George Smith': ('Chemistry', 'USA'),\n",
       " 'Gregory Winter': ('Chemistry', 'UK'),\n",
       " 'postponed': ('Literature', '---'),\n",
       " 'Denis Mukwege': ('Peace', 'DRC'),\n",
       " 'Nadia Murad': ('Peace', 'Iraq'),\n",
       " 'Arthur Ashkin': ('Physics', 'USA'),\n",
       " 'Gérard Mourou': ('Physics', 'France'),\n",
       " 'Donna Strickland': ('Physics', 'Canada'),\n",
       " 'James Allison': ('Physiology or Medicine', 'USA'),\n",
       " 'Tasuku Honjo': ('Physiology or Medicine', 'Japan'),\n",
       " 'William Nordhaus': ('Economics', 'USA'),\n",
       " 'Paul Romer': ('Economics', 'USA')}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "{name: (cat, nat) for name, nat, cat in zip(names, nationalities, categories)}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba0e7bfe",
   "metadata": {},
   "source": [
    "## Lesson 16: Introduction to Pandas"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "107f35aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "\n",
    "# Pandas, conventionally imported as pd\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "74dbfc67",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "participant number,gender,age,correct hit percentage,correct reject percentage,percent correct,confidence when correct hit,confidence when incorrect hit,confidence when correct reject,confidence when incorrect reject,confidence when correct,confidence when incorrect,sci,psqi,ess\n",
      "8,f,39,65,80,72.5,91,90,93,83.5,93,90,9,13,2\n",
      "16,m,42,90,90,90,75.5,55.5,70.5,50,75,50,4,11,7\n",
      "18,f,31,90,95,92.5,89.5,90,86,81,89,88,10,9,3\n",
      "22,f,35,100,75,87.5,89.5,*,71,80,88,80,13,8,20\n",
      "27,f,74,60,65,62.5,68.5,49,61,49,65,49,13,9,12\n",
      "28,f,61,80,20,50,71,63,31,72.5,64.5,70.5,15,14,2\n",
      "30,m,32,90,75,82.5,67,56.5,66,65,66,64,16,9,3\n",
      "33,m,62,45,90,67.5,54,37,65,81.5,62,61,14,9,9\n",
      "34,f,33,80,100,90,70.5,76.5,64.5,*,68,76.5,14,12,10\n",
      "35,f,53,100,50,75,74.5,*,60.5,65,71,65,14,8,7\n",
      "38,f,41,70,55,62.5,82,61.5,73,69,82,64,14,5,19\n",
      "41,f,36,90,100,95,76.5,75.5,75,*,76,75.5,15,7,0\n",
      "46,f,40,95,65,80,80,89,79,58.5,79.5,63,10,12,8\n",
      "49,f,24,85,75,80,58,50,49,68,55,59,14,13,4\n",
      "55,f,32,75,55,65,85,81,85,86,85,83.5,5,13,7\n",
      "71,f,40,40,100,70,69,56,70,*,70,56,0,11,14\n",
      "76,f,61,100,40,70,69.5,*,44.5,73,54.5,73,16,4,12\n",
      "77,f,42,70,90,80,87,72,90.5,43.5,88.5,64,11,10,10\n",
      "78,m,31,100,70,85,92,*,81,60,87.5,60,14,6,11\n",
      "80,m,28,100,50,75,100,*,100,100,100,100,12,7,12\n",
      "89,f,26,60,80,70,70,77,82,67.5,77,70.5,14,8,1\n",
      "90,m,45,100,95,97.5,100,*,100,100,100,100,14,9,6\n",
      "93,f,28,100,75,87.5,89.5,*,67,60,80,60,16,7,4\n",
      "100,f,44,65,25,45,62,72,87,77,69.5,73.5,1,15,6\n",
      "101,f,28,100,40,70,87,*,68,54,81,54,14,7,2\n",
      "1,f,42,80,65,72.5,51.5,44.5,43,49,51,49,29,1,5\n",
      "2,f,45,80,90,85,75,55.5,80,75,78.5,67,19,5,1\n",
      "3,f,16,70,80,75,70,57,54,53,57,54.5,23,1,3\n",
      "4,f,21,70,65,67.5,63.5,64,50,50,60,50,26,5,4\n",
      "5,f,18,90,100,95,76.5,83,80,*,80,83,21,7,5\n",
      "6,f,28,95,80,87.5,100,85,94,61,99,65,19,7,12\n",
      "7,f,38,90,95,92.5,77,43.5,79,21,78,36,28,3,4\n",
      "9,m,17,90,90,90,80.5,87.5,76.5,27,78.5,67.5,29,3,4\n",
      "10,f,25,100,100,100,90,*,85,*,90,*,17,10,11\n",
      "11,f,22,80,60,70,70,70,70,65,70,70,22,4,6\n",
      "12,m,41,90,80,85,76.5,55.5,67.5,52.5,74,55.5,28,5,3\n",
      "13,m,53,95,60,77.5,40,33,56,49,47,44,31,2,11\n",
      "14,m,43,95,90,92.5,52,29,49,36,52,29,31,2,10\n",
      "15,f,23,90,80,85,88,40,70.5,66.5,84,54.5,32,2,12\n",
      "17,m,19,55,60,57.5,62,50,66,50.5,63,50,25,5,6\n",
      "19,f,45,100,85,92.5,68,*,61,54,62,54,30,2,13\n",
      "20,f,43,65,65,65,59,55,64,59,59.5,57,28,3,12\n",
      "21,m,35,90,100,95,75.5,35.5,74.5,*,75.5,35.5,30,3,5\n",
      "23,m,24,55,100,77.5,68,67,80,*,78,67,22,4,10\n",
      "24,f,64,75,85,80,50,25,66,24,63,24.5,20,9,4\n",
      "25,f,36,100,80,90,88,*,66,63.5,81,63.5,26,5,4\n",
      "26,m,35,70,90,80,29.5,28.5,19,41.5,24,35,32,2,6\n",
      "29,f,43,90,85,87.5,65.5,27,41,45,50,45,32,1,3\n",
      "31,m,44,95,90,92.5,83,56,79,42.5,81,54,24,4,0\n",
      "32,f,29,95,55,75,67,53,65,53,67,53,26,8,12\n",
      "36,m,22,70,65,67.5,78,88,77,80,77,84,30,2,6\n",
      "37,f,46,80,75,77.5,72.5,72,100,91,100,91,28,2,5\n",
      "39,f,35,50,95,72.5,62.5,29,77,32,68,31,21,6,9\n",
      "40,m,53,65,80,72.5,54,32,47.5,78.5,48,51,29,3,7\n",
      "42,m,29,100,70,85,75,*,64.5,43,74,43,32,1,6\n",
      "43,f,31,85,90,87.5,82,49,81,36,82,49,26,5,10\n",
      "44,f,21,85,90,87.5,66,29,70,29,67,29,26,7,18\n",
      "45,f,42,90,90,90,83,83,80.5,36,82.5,76,23,3,11\n",
      "48,f,23,90,85,87.5,67,47,69,40,67,40,18,6,8\n",
      "50,m,54,90,70,80,90,83.5,77.5,69,88,79,22,6,16\n",
      "51,f,24,85,95,90,97,41,74,73,83,55.5,29,1,7\n",
      "52,f,21,85,75,80,65,73,56,68,63.5,68,20,6,9\n",
      "53,f,21,90,80,85,84,55.5,73.5,70,80.5,65,27,4,11\n",
      "54,f,43,95,75,85,74,89,68,65,71,68,19,4,4\n",
      "56,m,50,70,85,77.5,92.5,72.5,95,65,95,65,29,3,7\n",
      "57,f,53,95,75,85,84,55,68,61,78.5,58,24,5,4\n",
      "58,f,16,85,85,85,55,30,50,40,52.5,35,29,2,11\n",
      "59,f,67,95,75,85,70,7,69,60,69,59.5,17,7,12\n",
      "60,m,36,90,65,77.5,67.5,28.5,55,52,61,50,26,4,3\n",
      "61,f,34,90,90,90,58.5,43,73.5,53.5,66,47.5,30,0,3\n",
      "62,f,42,100,100,100,74.5,*,74,*,74,*,17,5,4\n",
      "63,f,46,80,90,85,92,75.5,92,63,92,73.5,25,1,11\n",
      "64,f,69,95,80,87.5,80,65,78.5,70.5,80,70,31,1,1\n",
      "65,f,31,100,95,97.5,98,*,90,40,92,40,27,4,4\n",
      "66,f,44,90,95,92.5,87,47.5,69,87,83,67,32,1,2\n",
      "67,f,25,100,100,100,61.5,*,58.5,*,60.5,*,28,8,9\n",
      "68,f,45,70,50,60,80.5,51.5,63,69,72.5,61.5,25,4,1\n",
      "69,f,47,90,100,95,100,*,71.5,83,97.5,83,30,2,2\n",
      "70,f,33,85,70,77.5,70,38,58.5,65,68,40,21,7,12\n",
      "72,f,18,80,75,77.5,67.5,51.5,66,57,67,53,29,4,6\n",
      "73,f,74,85,80,82.5,66,55,63,50.5,65,55,20,1,5\n",
      "74,m,21,40,40,40,90.5,80,74.5,83,82,81,22,7,5\n",
      "75,f,45,80,95,87.5,74,67,76,17,75,64,23,4,4\n",
      "79,f,37,90,80,85,95.5,68,83.5,83,94,71,20,5,9\n",
      "81,m,41,90,85,87.5,80,59.5,70,41,77,59,17,6,3\n",
      "82,f,41,80,75,77.5,94.5,61.5,86,74,92,67,27,4,8\n",
      "83,f,34,90,35,62.5,81,52,71,58,81,58,27,2,6\n",
      "84,f,39,75,70,72.5,57,57,59.5,50,58,50,22,3,10\n",
      "85,f,18,85,85,85,93,92,91,89,91.5,91,25,4,21\n",
      "86,f,31,100,85,92.5,100,*,100,50,100,50,30,3,5\n",
      "87,m,26,95,75,85,85,88,82,82,85,85,32,1,5\n",
      "88,m,66,60,85,72.5,67.5,66,74,57,74,64,30,5,9\n",
      "91,m,62,100,80,90,81,*,74.5,82,79.5,82,32,2,1\n",
      "92,m,22,85,95,90,66,56,72,63,70.5,59.5,28,1,8\n",
      "94,f,41,35,75,55,55,61,80,57,72,60,31,1,11\n",
      "95,m,46,95,80,87.5,90,75,80,80,85,75,29,3,5\n",
      "96,f,56,70,50,60,63,52.5,67.5,65.5,64,59.5,26,6,7\n",
      "97,f,23,70,85,77.5,77,66.5,77,77.5,77,74,20,8,10\n",
      "98,f,70,90,85,87.5,65.5,85.5,87,80,74,80,19,8,7\n",
      "99,f,24,70,80,75,61.5,81,70,61,65,81,31,2,15\n",
      "102,f,40,75,65,70,53,37,84,52,81,51,22,4,7\n",
      "103,f,33,85,40,62.5,80,27,31,82.5,81,73,24,5,7\n"
     ]
    }
   ],
   "source": [
    "!type data\\gfmt_sleep.csv "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "9aee4f67",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<class 'pandas.core.frame.DataFrame'>\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>participant number</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>correct hit percentage</th>\n",
       "      <th>correct reject percentage</th>\n",
       "      <th>percent correct</th>\n",
       "      <th>confidence when correct hit</th>\n",
       "      <th>confidence when incorrect hit</th>\n",
       "      <th>confidence when correct reject</th>\n",
       "      <th>confidence when incorrect reject</th>\n",
       "      <th>confidence when correct</th>\n",
       "      <th>confidence when incorrect</th>\n",
       "      <th>sci</th>\n",
       "      <th>psqi</th>\n",
       "      <th>ess</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>f</td>\n",
       "      <td>39</td>\n",
       "      <td>65</td>\n",
       "      <td>80</td>\n",
       "      <td>72.5</td>\n",
       "      <td>91.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>83.5</td>\n",
       "      <td>93.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>9</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16</td>\n",
       "      <td>m</td>\n",
       "      <td>42</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>90.0</td>\n",
       "      <td>75.5</td>\n",
       "      <td>55.5</td>\n",
       "      <td>70.5</td>\n",
       "      <td>50.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18</td>\n",
       "      <td>f</td>\n",
       "      <td>31</td>\n",
       "      <td>90</td>\n",
       "      <td>95</td>\n",
       "      <td>92.5</td>\n",
       "      <td>89.5</td>\n",
       "      <td>90.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>f</td>\n",
       "      <td>35</td>\n",
       "      <td>100</td>\n",
       "      <td>75</td>\n",
       "      <td>87.5</td>\n",
       "      <td>89.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>71.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>13</td>\n",
       "      <td>8</td>\n",
       "      <td>20</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>27</td>\n",
       "      <td>f</td>\n",
       "      <td>74</td>\n",
       "      <td>60</td>\n",
       "      <td>65</td>\n",
       "      <td>62.5</td>\n",
       "      <td>68.5</td>\n",
       "      <td>49.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   participant number gender  age  correct hit percentage  \\\n",
       "0                   8      f   39                      65   \n",
       "1                  16      m   42                      90   \n",
       "2                  18      f   31                      90   \n",
       "3                  22      f   35                     100   \n",
       "4                  27      f   74                      60   \n",
       "\n",
       "   correct reject percentage  percent correct  confidence when correct hit  \\\n",
       "0                         80             72.5                         91.0   \n",
       "1                         90             90.0                         75.5   \n",
       "2                         95             92.5                         89.5   \n",
       "3                         75             87.5                         89.5   \n",
       "4                         65             62.5                         68.5   \n",
       "\n",
       "   confidence when incorrect hit  confidence when correct reject  \\\n",
       "0                           90.0                            93.0   \n",
       "1                           55.5                            70.5   \n",
       "2                           90.0                            86.0   \n",
       "3                            NaN                            71.0   \n",
       "4                           49.0                            61.0   \n",
       "\n",
       "   confidence when incorrect reject  confidence when correct  \\\n",
       "0                              83.5                     93.0   \n",
       "1                              50.0                     75.0   \n",
       "2                              81.0                     89.0   \n",
       "3                              80.0                     88.0   \n",
       "4                              49.0                     65.0   \n",
       "\n",
       "   confidence when incorrect  sci  psqi  ess  \n",
       "0                       90.0    9    13    2  \n",
       "1                       50.0    4    11    7  \n",
       "2                       88.0   10     9    3  \n",
       "3                       80.0   13     8   20  \n",
       "4                       49.0   13     9   12  "
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df = pd.read_csv('data/gfmt_sleep.csv', na_values='*')\n",
    "\n",
    "# Check the type\n",
    "print(type(df))\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d85ae906",
   "metadata": {},
   "source": [
    "We normally index DataFrames by columnnames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "95b73be1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "62.5"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df['percent correct'][4]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e0abd514",
   "metadata": {},
   "source": [
    "But if we want a specific location in the dataframe, we preferrably use .loc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "9edde258",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "62.5"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.loc[4, 'percent correct'] # indexed by row, then column (row not necessarily an integer)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c359cb8c",
   "metadata": {},
   "source": [
    "**Booloan indexing**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "79dd207e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index and Percentage for Participant 42: 54    85.0\n",
      "Name: percent correct, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "print('Index and Percentage for Participant 42:', df.loc[df['participant number'] == 42, 'percent correct'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "a02df42c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>participant number</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>correct hit percentage</th>\n",
       "      <th>correct reject percentage</th>\n",
       "      <th>percent correct</th>\n",
       "      <th>confidence when correct hit</th>\n",
       "      <th>confidence when incorrect hit</th>\n",
       "      <th>confidence when correct reject</th>\n",
       "      <th>confidence when incorrect reject</th>\n",
       "      <th>confidence when correct</th>\n",
       "      <th>confidence when incorrect</th>\n",
       "      <th>sci</th>\n",
       "      <th>psqi</th>\n",
       "      <th>ess</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>54</th>\n",
       "      <td>42</td>\n",
       "      <td>m</td>\n",
       "      <td>29</td>\n",
       "      <td>100</td>\n",
       "      <td>70</td>\n",
       "      <td>85.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>64.5</td>\n",
       "      <td>43.0</td>\n",
       "      <td>74.0</td>\n",
       "      <td>43.0</td>\n",
       "      <td>32</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    participant number gender  age  correct hit percentage  \\\n",
       "54                  42      m   29                     100   \n",
       "\n",
       "    correct reject percentage  percent correct  confidence when correct hit  \\\n",
       "54                         70             85.0                         75.0   \n",
       "\n",
       "    confidence when incorrect hit  confidence when correct reject  \\\n",
       "54                            NaN                            64.5   \n",
       "\n",
       "    confidence when incorrect reject  confidence when correct  \\\n",
       "54                              43.0                     74.0   \n",
       "\n",
       "    confidence when incorrect  sci  psqi  ess  \n",
       "54                       43.0   32     1    6  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# full row\n",
    "df.loc[df['participant number'] == 42, :]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17fe7f01",
   "metadata": {},
   "source": [
    "Each boolean needs to be in parentheses when connected by a bitwise operator"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "2533436d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>participant number</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>correct hit percentage</th>\n",
       "      <th>correct reject percentage</th>\n",
       "      <th>percent correct</th>\n",
       "      <th>confidence when correct hit</th>\n",
       "      <th>confidence when incorrect hit</th>\n",
       "      <th>confidence when correct reject</th>\n",
       "      <th>confidence when incorrect reject</th>\n",
       "      <th>confidence when correct</th>\n",
       "      <th>confidence when incorrect</th>\n",
       "      <th>sci</th>\n",
       "      <th>psqi</th>\n",
       "      <th>ess</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>3</td>\n",
       "      <td>f</td>\n",
       "      <td>16</td>\n",
       "      <td>70</td>\n",
       "      <td>80</td>\n",
       "      <td>75.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>54.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>54.5</td>\n",
       "      <td>23</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>5</td>\n",
       "      <td>f</td>\n",
       "      <td>18</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>95.0</td>\n",
       "      <td>76.5</td>\n",
       "      <td>83.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>80.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>21</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>58</td>\n",
       "      <td>f</td>\n",
       "      <td>16</td>\n",
       "      <td>85</td>\n",
       "      <td>85</td>\n",
       "      <td>85.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>30.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>40.0</td>\n",
       "      <td>52.5</td>\n",
       "      <td>35.0</td>\n",
       "      <td>29</td>\n",
       "      <td>2</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>79</th>\n",
       "      <td>72</td>\n",
       "      <td>f</td>\n",
       "      <td>18</td>\n",
       "      <td>80</td>\n",
       "      <td>75</td>\n",
       "      <td>77.5</td>\n",
       "      <td>67.5</td>\n",
       "      <td>51.5</td>\n",
       "      <td>66.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>53.0</td>\n",
       "      <td>29</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>88</th>\n",
       "      <td>85</td>\n",
       "      <td>f</td>\n",
       "      <td>18</td>\n",
       "      <td>85</td>\n",
       "      <td>85</td>\n",
       "      <td>85.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>92.0</td>\n",
       "      <td>91.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>91.5</td>\n",
       "      <td>91.0</td>\n",
       "      <td>25</td>\n",
       "      <td>4</td>\n",
       "      <td>21</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    participant number gender  age  correct hit percentage  \\\n",
       "27                   3      f   16                      70   \n",
       "29                   5      f   18                      90   \n",
       "66                  58      f   16                      85   \n",
       "79                  72      f   18                      80   \n",
       "88                  85      f   18                      85   \n",
       "\n",
       "    correct reject percentage  percent correct  confidence when correct hit  \\\n",
       "27                         80             75.0                         70.0   \n",
       "29                        100             95.0                         76.5   \n",
       "66                         85             85.0                         55.0   \n",
       "79                         75             77.5                         67.5   \n",
       "88                         85             85.0                         93.0   \n",
       "\n",
       "    confidence when incorrect hit  confidence when correct reject  \\\n",
       "27                           57.0                            54.0   \n",
       "29                           83.0                            80.0   \n",
       "66                           30.0                            50.0   \n",
       "79                           51.5                            66.0   \n",
       "88                           92.0                            91.0   \n",
       "\n",
       "    confidence when incorrect reject  confidence when correct  \\\n",
       "27                              53.0                     57.0   \n",
       "29                               NaN                     80.0   \n",
       "66                              40.0                     52.5   \n",
       "79                              57.0                     67.0   \n",
       "88                              89.0                     91.5   \n",
       "\n",
       "    confidence when incorrect  sci  psqi  ess  \n",
       "27                       54.5   23     1    3  \n",
       "29                       83.0   21     7    5  \n",
       "66                       35.0   29     2   11  \n",
       "79                       53.0   29     4    6  \n",
       "88                       91.0   25     4   21  "
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for woman below 21\n",
    "df.loc[(df['age'] < 21) & (df['gender'] == 'f'), :]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "3e244cea",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>participant number</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>correct hit percentage</th>\n",
       "      <th>correct reject percentage</th>\n",
       "      <th>percent correct</th>\n",
       "      <th>confidence when correct hit</th>\n",
       "      <th>confidence when incorrect hit</th>\n",
       "      <th>confidence when correct reject</th>\n",
       "      <th>confidence when incorrect reject</th>\n",
       "      <th>confidence when correct</th>\n",
       "      <th>confidence when incorrect</th>\n",
       "      <th>sci</th>\n",
       "      <th>psqi</th>\n",
       "      <th>ess</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>93</td>\n",
       "      <td>f</td>\n",
       "      <td>28</td>\n",
       "      <td>100</td>\n",
       "      <td>75</td>\n",
       "      <td>87.5</td>\n",
       "      <td>89.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>67.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>16</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>5</td>\n",
       "      <td>f</td>\n",
       "      <td>18</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>95.0</td>\n",
       "      <td>76.5</td>\n",
       "      <td>83.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>80.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>21</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>6</td>\n",
       "      <td>f</td>\n",
       "      <td>28</td>\n",
       "      <td>95</td>\n",
       "      <td>80</td>\n",
       "      <td>87.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>19</td>\n",
       "      <td>7</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>10</td>\n",
       "      <td>f</td>\n",
       "      <td>25</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>85.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>90.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>44</td>\n",
       "      <td>f</td>\n",
       "      <td>21</td>\n",
       "      <td>85</td>\n",
       "      <td>90</td>\n",
       "      <td>87.5</td>\n",
       "      <td>66.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>26</td>\n",
       "      <td>7</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    participant number gender  age  correct hit percentage  \\\n",
       "22                  93      f   28                     100   \n",
       "29                   5      f   18                      90   \n",
       "30                   6      f   28                      95   \n",
       "33                  10      f   25                     100   \n",
       "56                  44      f   21                      85   \n",
       "\n",
       "    correct reject percentage  percent correct  confidence when correct hit  \\\n",
       "22                         75             87.5                         89.5   \n",
       "29                        100             95.0                         76.5   \n",
       "30                         80             87.5                        100.0   \n",
       "33                        100            100.0                         90.0   \n",
       "56                         90             87.5                         66.0   \n",
       "\n",
       "    confidence when incorrect hit  confidence when correct reject  \\\n",
       "22                            NaN                            67.0   \n",
       "29                           83.0                            80.0   \n",
       "30                           85.0                            94.0   \n",
       "33                            NaN                            85.0   \n",
       "56                           29.0                            70.0   \n",
       "\n",
       "    confidence when incorrect reject  confidence when correct  \\\n",
       "22                              60.0                     80.0   \n",
       "29                               NaN                     80.0   \n",
       "30                              61.0                     99.0   \n",
       "33                               NaN                     90.0   \n",
       "56                              29.0                     67.0   \n",
       "\n",
       "    confidence when incorrect  sci  psqi  ess  \n",
       "22                       60.0   16     7    4  \n",
       "29                       83.0   21     7    5  \n",
       "30                       65.0   19     7   12  \n",
       "33                        NaN   17    10   11  \n",
       "56                       29.0   26     7   18  "
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# for woman below 30 with a higher percentage than 85\n",
    "inds = (df[\"age\"] < 30) & (df[\"gender\"] == \"f\") & (df[\"percent correct\"] > 85)\n",
    "\n",
    "df.loc[inds, :].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db1a1a48",
   "metadata": {},
   "source": [
    "looping is done automatically. looks like this though (note iterrows to loop over a dataframe)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "337e5d0f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>participant number</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>correct hit percentage</th>\n",
       "      <th>correct reject percentage</th>\n",
       "      <th>percent correct</th>\n",
       "      <th>confidence when correct hit</th>\n",
       "      <th>confidence when incorrect hit</th>\n",
       "      <th>confidence when correct reject</th>\n",
       "      <th>confidence when incorrect reject</th>\n",
       "      <th>confidence when correct</th>\n",
       "      <th>confidence when incorrect</th>\n",
       "      <th>sci</th>\n",
       "      <th>psqi</th>\n",
       "      <th>ess</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>93</td>\n",
       "      <td>f</td>\n",
       "      <td>28</td>\n",
       "      <td>100</td>\n",
       "      <td>75</td>\n",
       "      <td>87.5</td>\n",
       "      <td>89.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>67.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>16</td>\n",
       "      <td>7</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>5</td>\n",
       "      <td>f</td>\n",
       "      <td>18</td>\n",
       "      <td>90</td>\n",
       "      <td>100</td>\n",
       "      <td>95.0</td>\n",
       "      <td>76.5</td>\n",
       "      <td>83.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>80.0</td>\n",
       "      <td>83.0</td>\n",
       "      <td>21</td>\n",
       "      <td>7</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>6</td>\n",
       "      <td>f</td>\n",
       "      <td>28</td>\n",
       "      <td>95</td>\n",
       "      <td>80</td>\n",
       "      <td>87.5</td>\n",
       "      <td>100.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>99.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>19</td>\n",
       "      <td>7</td>\n",
       "      <td>12</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td>10</td>\n",
       "      <td>f</td>\n",
       "      <td>25</td>\n",
       "      <td>100</td>\n",
       "      <td>100</td>\n",
       "      <td>100.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>85.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>90.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17</td>\n",
       "      <td>10</td>\n",
       "      <td>11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>44</td>\n",
       "      <td>f</td>\n",
       "      <td>21</td>\n",
       "      <td>85</td>\n",
       "      <td>90</td>\n",
       "      <td>87.5</td>\n",
       "      <td>66.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>70.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>67.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>26</td>\n",
       "      <td>7</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    participant number gender  age  correct hit percentage  \\\n",
       "22                  93      f   28                     100   \n",
       "29                   5      f   18                      90   \n",
       "30                   6      f   28                      95   \n",
       "33                  10      f   25                     100   \n",
       "56                  44      f   21                      85   \n",
       "\n",
       "    correct reject percentage  percent correct  confidence when correct hit  \\\n",
       "22                         75             87.5                         89.5   \n",
       "29                        100             95.0                         76.5   \n",
       "30                         80             87.5                        100.0   \n",
       "33                        100            100.0                         90.0   \n",
       "56                         90             87.5                         66.0   \n",
       "\n",
       "    confidence when incorrect hit  confidence when correct reject  \\\n",
       "22                            NaN                            67.0   \n",
       "29                           83.0                            80.0   \n",
       "30                           85.0                            94.0   \n",
       "33                            NaN                            85.0   \n",
       "56                           29.0                            70.0   \n",
       "\n",
       "    confidence when incorrect reject  confidence when correct  \\\n",
       "22                              60.0                     80.0   \n",
       "29                               NaN                     80.0   \n",
       "30                              61.0                     99.0   \n",
       "33                               NaN                     90.0   \n",
       "56                              29.0                     67.0   \n",
       "\n",
       "    confidence when incorrect  sci  psqi  ess  \n",
       "22                       60.0   16     7    4  \n",
       "29                       83.0   21     7    5  \n",
       "30                       65.0   19     7   12  \n",
       "33                        NaN   17    10   11  \n",
       "56                       29.0   26     7   18  "
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Initialize array of Boolean indices\n",
    "inds = [False] * len(df)\n",
    "\n",
    "# Iterate over the rows of the DataFrame to check if the row should be included\n",
    "for i, r in df.iterrows():\n",
    "    if r['age'] < 30 and r['gender'] == 'f' and r['percent correct'] > 85:\n",
    "        inds[i] = True\n",
    "\n",
    "# Make our selection with Boolean indexing\n",
    "df.loc[inds, :].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cac4b388",
   "metadata": {},
   "source": [
    "Add a new column insominia if the sci test was below 16"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "94730260",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0       True\n",
      "1       True\n",
      "2       True\n",
      "3       True\n",
      "4       True\n",
      "       ...  \n",
      "97     False\n",
      "98     False\n",
      "99     False\n",
      "100    False\n",
      "101    False\n",
      "Name: sci, Length: 102, dtype: bool\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>participant number</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>correct hit percentage</th>\n",
       "      <th>correct reject percentage</th>\n",
       "      <th>percent correct</th>\n",
       "      <th>confidence when correct hit</th>\n",
       "      <th>confidence when incorrect hit</th>\n",
       "      <th>confidence when correct reject</th>\n",
       "      <th>confidence when incorrect reject</th>\n",
       "      <th>confidence when correct</th>\n",
       "      <th>confidence when incorrect</th>\n",
       "      <th>sci</th>\n",
       "      <th>psqi</th>\n",
       "      <th>ess</th>\n",
       "      <th>insomnia</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>f</td>\n",
       "      <td>39</td>\n",
       "      <td>65</td>\n",
       "      <td>80</td>\n",
       "      <td>72.5</td>\n",
       "      <td>91.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>83.5</td>\n",
       "      <td>93.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>9</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16</td>\n",
       "      <td>m</td>\n",
       "      <td>42</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>90.0</td>\n",
       "      <td>75.5</td>\n",
       "      <td>55.5</td>\n",
       "      <td>70.5</td>\n",
       "      <td>50.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18</td>\n",
       "      <td>f</td>\n",
       "      <td>31</td>\n",
       "      <td>90</td>\n",
       "      <td>95</td>\n",
       "      <td>92.5</td>\n",
       "      <td>89.5</td>\n",
       "      <td>90.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>f</td>\n",
       "      <td>35</td>\n",
       "      <td>100</td>\n",
       "      <td>75</td>\n",
       "      <td>87.5</td>\n",
       "      <td>89.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>71.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>13</td>\n",
       "      <td>8</td>\n",
       "      <td>20</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>27</td>\n",
       "      <td>f</td>\n",
       "      <td>74</td>\n",
       "      <td>60</td>\n",
       "      <td>65</td>\n",
       "      <td>62.5</td>\n",
       "      <td>68.5</td>\n",
       "      <td>49.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   participant number gender  age  correct hit percentage  \\\n",
       "0                   8      f   39                      65   \n",
       "1                  16      m   42                      90   \n",
       "2                  18      f   31                      90   \n",
       "3                  22      f   35                     100   \n",
       "4                  27      f   74                      60   \n",
       "\n",
       "   correct reject percentage  percent correct  confidence when correct hit  \\\n",
       "0                         80             72.5                         91.0   \n",
       "1                         90             90.0                         75.5   \n",
       "2                         95             92.5                         89.5   \n",
       "3                         75             87.5                         89.5   \n",
       "4                         65             62.5                         68.5   \n",
       "\n",
       "   confidence when incorrect hit  confidence when correct reject  \\\n",
       "0                           90.0                            93.0   \n",
       "1                           55.5                            70.5   \n",
       "2                           90.0                            86.0   \n",
       "3                            NaN                            71.0   \n",
       "4                           49.0                            61.0   \n",
       "\n",
       "   confidence when incorrect reject  confidence when correct  \\\n",
       "0                              83.5                     93.0   \n",
       "1                              50.0                     75.0   \n",
       "2                              81.0                     89.0   \n",
       "3                              80.0                     88.0   \n",
       "4                              49.0                     65.0   \n",
       "\n",
       "   confidence when incorrect  sci  psqi  ess  insomnia  \n",
       "0                       90.0    9    13    2      True  \n",
       "1                       50.0    4    11    7      True  \n",
       "2                       88.0   10     9    3      True  \n",
       "3                       80.0   13     8   20      True  \n",
       "4                       49.0   13     9   12      True  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# this is an array, a 1d dataframe, also called a panda series\n",
    "print(df['sci'] <= 16) \n",
    "\n",
    "# Add the column to the DataFrame\n",
    "df['insomnia'] = df['sci'] <= 16\n",
    "\n",
    "# Take a look\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b97bb35",
   "metadata": {},
   "source": [
    "elementwise application of the <= operator means again that we can avoid the for loop"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "00168d00",
   "metadata": {},
   "outputs": [],
   "source": [
    "insomnia = []\n",
    "for sci in df['sci']:\n",
    "    insomnia.append(sci <= 16)\n",
    "\n",
    "# Add the column to the DataFrame\n",
    "df['insomnia'] = insomnia\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "545ca844",
   "metadata": {},
   "source": [
    "**Applying functions to df**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "id": "a7f2d4e9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Insomniacs: 76.1\n",
      "Control:    81.46103896103897\n"
     ]
    }
   ],
   "source": [
    "# e.g. mean of the correct percentage of matching of insomnics vs non insomniacs\n",
    "print('Insomniacs:', np.mean(df.loc[df['insomnia'], 'percent correct']))\n",
    "print('Control:   ', np.mean(df.loc[~df['insomnia'], 'percent correct']))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c2ed5271",
   "metadata": {},
   "source": [
    "~ is a bit switcher and represent a logical NOT."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c65a6e4f",
   "metadata": {},
   "source": [
    "**summary statistics**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "0300e51d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>participant number</th>\n",
       "      <th>age</th>\n",
       "      <th>correct hit percentage</th>\n",
       "      <th>correct reject percentage</th>\n",
       "      <th>percent correct</th>\n",
       "      <th>confidence when correct hit</th>\n",
       "      <th>confidence when incorrect hit</th>\n",
       "      <th>confidence when correct reject</th>\n",
       "      <th>confidence when incorrect reject</th>\n",
       "      <th>confidence when correct</th>\n",
       "      <th>confidence when incorrect</th>\n",
       "      <th>sci</th>\n",
       "      <th>psqi</th>\n",
       "      <th>ess</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>102.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>84.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>93.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>99.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>102.000000</td>\n",
       "      <td>102.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>52.049020</td>\n",
       "      <td>37.921569</td>\n",
       "      <td>83.088235</td>\n",
       "      <td>77.205882</td>\n",
       "      <td>80.147059</td>\n",
       "      <td>74.990196</td>\n",
       "      <td>58.565476</td>\n",
       "      <td>71.137255</td>\n",
       "      <td>61.220430</td>\n",
       "      <td>74.642157</td>\n",
       "      <td>61.979798</td>\n",
       "      <td>22.245098</td>\n",
       "      <td>5.274510</td>\n",
       "      <td>7.294118</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>30.020909</td>\n",
       "      <td>14.029450</td>\n",
       "      <td>15.091210</td>\n",
       "      <td>17.569854</td>\n",
       "      <td>12.047881</td>\n",
       "      <td>14.165916</td>\n",
       "      <td>19.560653</td>\n",
       "      <td>14.987479</td>\n",
       "      <td>17.671283</td>\n",
       "      <td>13.619725</td>\n",
       "      <td>15.921670</td>\n",
       "      <td>7.547128</td>\n",
       "      <td>3.404007</td>\n",
       "      <td>4.426715</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>16.000000</td>\n",
       "      <td>35.000000</td>\n",
       "      <td>20.000000</td>\n",
       "      <td>40.000000</td>\n",
       "      <td>29.500000</td>\n",
       "      <td>7.000000</td>\n",
       "      <td>19.000000</td>\n",
       "      <td>17.000000</td>\n",
       "      <td>24.000000</td>\n",
       "      <td>24.500000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>26.250000</td>\n",
       "      <td>26.500000</td>\n",
       "      <td>75.000000</td>\n",
       "      <td>70.000000</td>\n",
       "      <td>72.500000</td>\n",
       "      <td>66.000000</td>\n",
       "      <td>46.375000</td>\n",
       "      <td>64.625000</td>\n",
       "      <td>50.000000</td>\n",
       "      <td>66.000000</td>\n",
       "      <td>51.000000</td>\n",
       "      <td>17.000000</td>\n",
       "      <td>3.000000</td>\n",
       "      <td>4.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>52.500000</td>\n",
       "      <td>36.500000</td>\n",
       "      <td>90.000000</td>\n",
       "      <td>80.000000</td>\n",
       "      <td>83.750000</td>\n",
       "      <td>75.000000</td>\n",
       "      <td>56.250000</td>\n",
       "      <td>71.250000</td>\n",
       "      <td>61.000000</td>\n",
       "      <td>75.750000</td>\n",
       "      <td>61.500000</td>\n",
       "      <td>23.500000</td>\n",
       "      <td>5.000000</td>\n",
       "      <td>7.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>77.750000</td>\n",
       "      <td>45.000000</td>\n",
       "      <td>95.000000</td>\n",
       "      <td>90.000000</td>\n",
       "      <td>87.500000</td>\n",
       "      <td>86.500000</td>\n",
       "      <td>73.500000</td>\n",
       "      <td>80.000000</td>\n",
       "      <td>74.000000</td>\n",
       "      <td>82.375000</td>\n",
       "      <td>73.000000</td>\n",
       "      <td>29.000000</td>\n",
       "      <td>7.000000</td>\n",
       "      <td>10.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>103.000000</td>\n",
       "      <td>74.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>92.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>100.000000</td>\n",
       "      <td>32.000000</td>\n",
       "      <td>15.000000</td>\n",
       "      <td>21.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       participant number         age  correct hit percentage  \\\n",
       "count          102.000000  102.000000              102.000000   \n",
       "mean            52.049020   37.921569               83.088235   \n",
       "std             30.020909   14.029450               15.091210   \n",
       "min              1.000000   16.000000               35.000000   \n",
       "25%             26.250000   26.500000               75.000000   \n",
       "50%             52.500000   36.500000               90.000000   \n",
       "75%             77.750000   45.000000               95.000000   \n",
       "max            103.000000   74.000000              100.000000   \n",
       "\n",
       "       correct reject percentage  percent correct  \\\n",
       "count                 102.000000       102.000000   \n",
       "mean                   77.205882        80.147059   \n",
       "std                    17.569854        12.047881   \n",
       "min                    20.000000        40.000000   \n",
       "25%                    70.000000        72.500000   \n",
       "50%                    80.000000        83.750000   \n",
       "75%                    90.000000        87.500000   \n",
       "max                   100.000000       100.000000   \n",
       "\n",
       "       confidence when correct hit  confidence when incorrect hit  \\\n",
       "count                   102.000000                      84.000000   \n",
       "mean                     74.990196                      58.565476   \n",
       "std                      14.165916                      19.560653   \n",
       "min                      29.500000                       7.000000   \n",
       "25%                      66.000000                      46.375000   \n",
       "50%                      75.000000                      56.250000   \n",
       "75%                      86.500000                      73.500000   \n",
       "max                     100.000000                      92.000000   \n",
       "\n",
       "       confidence when correct reject  confidence when incorrect reject  \\\n",
       "count                      102.000000                         93.000000   \n",
       "mean                        71.137255                         61.220430   \n",
       "std                         14.987479                         17.671283   \n",
       "min                         19.000000                         17.000000   \n",
       "25%                         64.625000                         50.000000   \n",
       "50%                         71.250000                         61.000000   \n",
       "75%                         80.000000                         74.000000   \n",
       "max                        100.000000                        100.000000   \n",
       "\n",
       "       confidence when correct  confidence when incorrect         sci  \\\n",
       "count               102.000000                  99.000000  102.000000   \n",
       "mean                 74.642157                  61.979798   22.245098   \n",
       "std                  13.619725                  15.921670    7.547128   \n",
       "min                  24.000000                  24.500000    0.000000   \n",
       "25%                  66.000000                  51.000000   17.000000   \n",
       "50%                  75.750000                  61.500000   23.500000   \n",
       "75%                  82.375000                  73.000000   29.000000   \n",
       "max                 100.000000                 100.000000   32.000000   \n",
       "\n",
       "             psqi         ess  \n",
       "count  102.000000  102.000000  \n",
       "mean     5.274510    7.294118  \n",
       "std      3.404007    4.426715  \n",
       "min      0.000000    0.000000  \n",
       "25%      3.000000    4.000000  \n",
       "50%      5.000000    7.000000  \n",
       "75%      7.000000   10.000000  \n",
       "max     15.000000   21.000000  "
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46cad219",
   "metadata": {},
   "source": [
    "**extract only the median** of each entry with .loc"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "88422c52",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "participant number                  52.50\n",
       "age                                 36.50\n",
       "correct hit percentage              90.00\n",
       "correct reject percentage           80.00\n",
       "percent correct                     83.75\n",
       "confidence when correct hit         75.00\n",
       "confidence when incorrect hit       56.25\n",
       "confidence when correct reject      71.25\n",
       "confidence when incorrect reject    61.00\n",
       "confidence when correct             75.75\n",
       "confidence when incorrect           61.50\n",
       "sci                                 23.50\n",
       "psqi                                 5.00\n",
       "ess                                  7.00\n",
       "Name: 50%, dtype: float64"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.describe().loc['50%', :] # see the row name of the statistics df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "270edf25",
   "metadata": {},
   "source": [
    "**extract only the mean** of each entry"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "id": "628517dc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "participant number                  52.049020\n",
       "age                                 37.921569\n",
       "correct hit percentage              83.088235\n",
       "correct reject percentage           77.205882\n",
       "percent correct                     80.147059\n",
       "confidence when correct hit         74.990196\n",
       "confidence when incorrect hit       58.565476\n",
       "confidence when correct reject      71.137255\n",
       "confidence when incorrect reject    61.220430\n",
       "confidence when correct             74.642157\n",
       "confidence when incorrect           61.979798\n",
       "sci                                 22.245098\n",
       "psqi                                 5.274510\n",
       "ess                                  7.294118\n",
       "Name: mean, dtype: float64"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.describe().loc['mean', :]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d39a0b19",
   "metadata": {},
   "source": [
    "**Output a csv file**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "c10ede33",
   "metadata": {},
   "outputs": [],
   "source": [
    "df.to_csv('gfmt_sleep_with_insomnia.csv', index=False) # omit index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "8b254f05",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "participant number,gender,age,correct hit percentage,correct reject percentage,percent correct,confidence when correct hit,confidence when incorrect hit,confidence when correct reject,confidence when incorrect reject,confidence when correct,confidence when incorrect,sci,psqi,ess,insomnia\n",
      "8,f,39,65,80,72.5,91.0,90.0,93.0,83.5,93.0,90.0,9,13,2,True\n",
      "16,m,42,90,90,90.0,75.5,55.5,70.5,50.0,75.0,50.0,4,11,7,True\n",
      "18,f,31,90,95,92.5,89.5,90.0,86.0,81.0,89.0,88.0,10,9,3,True\n",
      "22,f,35,100,75,87.5,89.5,,71.0,80.0,88.0,80.0,13,8,20,True\n",
      "27,f,74,60,65,62.5,68.5,49.0,61.0,49.0,65.0,49.0,13,9,12,True\n",
      "28,f,61,80,20,50.0,71.0,63.0,31.0,72.5,64.5,70.5,15,14,2,True\n",
      "30,m,32,90,75,82.5,67.0,56.5,66.0,65.0,66.0,64.0,16,9,3,True\n",
      "33,m,62,45,90,67.5,54.0,37.0,65.0,81.5,62.0,61.0,14,9,9,True\n",
      "34,f,33,80,100,90.0,70.5,76.5,64.5,,68.0,76.5,14,12,10,True\n",
      "35,f,53,100,50,75.0,74.5,,60.5,65.0,71.0,65.0,14,8,7,True\n",
      "38,f,41,70,55,62.5,82.0,61.5,73.0,69.0,82.0,64.0,14,5,19,True\n",
      "41,f,36,90,100,95.0,76.5,75.5,75.0,,76.0,75.5,15,7,0,True\n",
      "46,f,40,95,65,80.0,80.0,89.0,79.0,58.5,79.5,63.0,10,12,8,True\n",
      "49,f,24,85,75,80.0,58.0,50.0,49.0,68.0,55.0,59.0,14,13,4,True\n",
      "55,f,32,75,55,65.0,85.0,81.0,85.0,86.0,85.0,83.5,5,13,7,True\n",
      "71,f,40,40,100,70.0,69.0,56.0,70.0,,70.0,56.0,0,11,14,True\n",
      "76,f,61,100,40,70.0,69.5,,44.5,73.0,54.5,73.0,16,4,12,True\n",
      "77,f,42,70,90,80.0,87.0,72.0,90.5,43.5,88.5,64.0,11,10,10,True\n",
      "78,m,31,100,70,85.0,92.0,,81.0,60.0,87.5,60.0,14,6,11,True\n",
      "80,m,28,100,50,75.0,100.0,,100.0,100.0,100.0,100.0,12,7,12,True\n",
      "89,f,26,60,80,70.0,70.0,77.0,82.0,67.5,77.0,70.5,14,8,1,True\n",
      "90,m,45,100,95,97.5,100.0,,100.0,100.0,100.0,100.0,14,9,6,True\n",
      "93,f,28,100,75,87.5,89.5,,67.0,60.0,80.0,60.0,16,7,4,True\n",
      "100,f,44,65,25,45.0,62.0,72.0,87.0,77.0,69.5,73.5,1,15,6,True\n",
      "101,f,28,100,40,70.0,87.0,,68.0,54.0,81.0,54.0,14,7,2,True\n",
      "1,f,42,80,65,72.5,51.5,44.5,43.0,49.0,51.0,49.0,29,1,5,False\n",
      "2,f,45,80,90,85.0,75.0,55.5,80.0,75.0,78.5,67.0,19,5,1,False\n",
      "3,f,16,70,80,75.0,70.0,57.0,54.0,53.0,57.0,54.5,23,1,3,False\n",
      "4,f,21,70,65,67.5,63.5,64.0,50.0,50.0,60.0,50.0,26,5,4,False\n",
      "5,f,18,90,100,95.0,76.5,83.0,80.0,,80.0,83.0,21,7,5,False\n",
      "6,f,28,95,80,87.5,100.0,85.0,94.0,61.0,99.0,65.0,19,7,12,False\n",
      "7,f,38,90,95,92.5,77.0,43.5,79.0,21.0,78.0,36.0,28,3,4,False\n",
      "9,m,17,90,90,90.0,80.5,87.5,76.5,27.0,78.5,67.5,29,3,4,False\n",
      "10,f,25,100,100,100.0,90.0,,85.0,,90.0,,17,10,11,False\n",
      "11,f,22,80,60,70.0,70.0,70.0,70.0,65.0,70.0,70.0,22,4,6,False\n",
      "12,m,41,90,80,85.0,76.5,55.5,67.5,52.5,74.0,55.5,28,5,3,False\n",
      "13,m,53,95,60,77.5,40.0,33.0,56.0,49.0,47.0,44.0,31,2,11,False\n",
      "14,m,43,95,90,92.5,52.0,29.0,49.0,36.0,52.0,29.0,31,2,10,False\n",
      "15,f,23,90,80,85.0,88.0,40.0,70.5,66.5,84.0,54.5,32,2,12,False\n",
      "17,m,19,55,60,57.5,62.0,50.0,66.0,50.5,63.0,50.0,25,5,6,False\n",
      "19,f,45,100,85,92.5,68.0,,61.0,54.0,62.0,54.0,30,2,13,False\n",
      "20,f,43,65,65,65.0,59.0,55.0,64.0,59.0,59.5,57.0,28,3,12,False\n",
      "21,m,35,90,100,95.0,75.5,35.5,74.5,,75.5,35.5,30,3,5,False\n",
      "23,m,24,55,100,77.5,68.0,67.0,80.0,,78.0,67.0,22,4,10,False\n",
      "24,f,64,75,85,80.0,50.0,25.0,66.0,24.0,63.0,24.5,20,9,4,False\n",
      "25,f,36,100,80,90.0,88.0,,66.0,63.5,81.0,63.5,26,5,4,False\n",
      "26,m,35,70,90,80.0,29.5,28.5,19.0,41.5,24.0,35.0,32,2,6,False\n",
      "29,f,43,90,85,87.5,65.5,27.0,41.0,45.0,50.0,45.0,32,1,3,False\n",
      "31,m,44,95,90,92.5,83.0,56.0,79.0,42.5,81.0,54.0,24,4,0,False\n",
      "32,f,29,95,55,75.0,67.0,53.0,65.0,53.0,67.0,53.0,26,8,12,False\n",
      "36,m,22,70,65,67.5,78.0,88.0,77.0,80.0,77.0,84.0,30,2,6,False\n",
      "37,f,46,80,75,77.5,72.5,72.0,100.0,91.0,100.0,91.0,28,2,5,False\n",
      "39,f,35,50,95,72.5,62.5,29.0,77.0,32.0,68.0,31.0,21,6,9,False\n",
      "40,m,53,65,80,72.5,54.0,32.0,47.5,78.5,48.0,51.0,29,3,7,False\n",
      "42,m,29,100,70,85.0,75.0,,64.5,43.0,74.0,43.0,32,1,6,False\n",
      "43,f,31,85,90,87.5,82.0,49.0,81.0,36.0,82.0,49.0,26,5,10,False\n",
      "44,f,21,85,90,87.5,66.0,29.0,70.0,29.0,67.0,29.0,26,7,18,False\n",
      "45,f,42,90,90,90.0,83.0,83.0,80.5,36.0,82.5,76.0,23,3,11,False\n",
      "48,f,23,90,85,87.5,67.0,47.0,69.0,40.0,67.0,40.0,18,6,8,False\n",
      "50,m,54,90,70,80.0,90.0,83.5,77.5,69.0,88.0,79.0,22,6,16,False\n",
      "51,f,24,85,95,90.0,97.0,41.0,74.0,73.0,83.0,55.5,29,1,7,False\n",
      "52,f,21,85,75,80.0,65.0,73.0,56.0,68.0,63.5,68.0,20,6,9,False\n",
      "53,f,21,90,80,85.0,84.0,55.5,73.5,70.0,80.5,65.0,27,4,11,False\n",
      "54,f,43,95,75,85.0,74.0,89.0,68.0,65.0,71.0,68.0,19,4,4,False\n",
      "56,m,50,70,85,77.5,92.5,72.5,95.0,65.0,95.0,65.0,29,3,7,False\n",
      "57,f,53,95,75,85.0,84.0,55.0,68.0,61.0,78.5,58.0,24,5,4,False\n",
      "58,f,16,85,85,85.0,55.0,30.0,50.0,40.0,52.5,35.0,29,2,11,False\n",
      "59,f,67,95,75,85.0,70.0,7.0,69.0,60.0,69.0,59.5,17,7,12,False\n",
      "60,m,36,90,65,77.5,67.5,28.5,55.0,52.0,61.0,50.0,26,4,3,False\n",
      "61,f,34,90,90,90.0,58.5,43.0,73.5,53.5,66.0,47.5,30,0,3,False\n",
      "62,f,42,100,100,100.0,74.5,,74.0,,74.0,,17,5,4,False\n",
      "63,f,46,80,90,85.0,92.0,75.5,92.0,63.0,92.0,73.5,25,1,11,False\n",
      "64,f,69,95,80,87.5,80.0,65.0,78.5,70.5,80.0,70.0,31,1,1,False\n",
      "65,f,31,100,95,97.5,98.0,,90.0,40.0,92.0,40.0,27,4,4,False\n",
      "66,f,44,90,95,92.5,87.0,47.5,69.0,87.0,83.0,67.0,32,1,2,False\n",
      "67,f,25,100,100,100.0,61.5,,58.5,,60.5,,28,8,9,False\n",
      "68,f,45,70,50,60.0,80.5,51.5,63.0,69.0,72.5,61.5,25,4,1,False\n",
      "69,f,47,90,100,95.0,100.0,,71.5,83.0,97.5,83.0,30,2,2,False\n",
      "70,f,33,85,70,77.5,70.0,38.0,58.5,65.0,68.0,40.0,21,7,12,False\n",
      "72,f,18,80,75,77.5,67.5,51.5,66.0,57.0,67.0,53.0,29,4,6,False\n",
      "73,f,74,85,80,82.5,66.0,55.0,63.0,50.5,65.0,55.0,20,1,5,False\n",
      "74,m,21,40,40,40.0,90.5,80.0,74.5,83.0,82.0,81.0,22,7,5,False\n",
      "75,f,45,80,95,87.5,74.0,67.0,76.0,17.0,75.0,64.0,23,4,4,False\n",
      "79,f,37,90,80,85.0,95.5,68.0,83.5,83.0,94.0,71.0,20,5,9,False\n",
      "81,m,41,90,85,87.5,80.0,59.5,70.0,41.0,77.0,59.0,17,6,3,False\n",
      "82,f,41,80,75,77.5,94.5,61.5,86.0,74.0,92.0,67.0,27,4,8,False\n",
      "83,f,34,90,35,62.5,81.0,52.0,71.0,58.0,81.0,58.0,27,2,6,False\n",
      "84,f,39,75,70,72.5,57.0,57.0,59.5,50.0,58.0,50.0,22,3,10,False\n",
      "85,f,18,85,85,85.0,93.0,92.0,91.0,89.0,91.5,91.0,25,4,21,False\n",
      "86,f,31,100,85,92.5,100.0,,100.0,50.0,100.0,50.0,30,3,5,False\n",
      "87,m,26,95,75,85.0,85.0,88.0,82.0,82.0,85.0,85.0,32,1,5,False\n",
      "88,m,66,60,85,72.5,67.5,66.0,74.0,57.0,74.0,64.0,30,5,9,False\n",
      "91,m,62,100,80,90.0,81.0,,74.5,82.0,79.5,82.0,32,2,1,False\n",
      "92,m,22,85,95,90.0,66.0,56.0,72.0,63.0,70.5,59.5,28,1,8,False\n",
      "94,f,41,35,75,55.0,55.0,61.0,80.0,57.0,72.0,60.0,31,1,11,False\n",
      "95,m,46,95,80,87.5,90.0,75.0,80.0,80.0,85.0,75.0,29,3,5,False\n",
      "96,f,56,70,50,60.0,63.0,52.5,67.5,65.5,64.0,59.5,26,6,7,False\n",
      "97,f,23,70,85,77.5,77.0,66.5,77.0,77.5,77.0,74.0,20,8,10,False\n",
      "98,f,70,90,85,87.5,65.5,85.5,87.0,80.0,74.0,80.0,19,8,7,False\n",
      "99,f,24,70,80,75.0,61.5,81.0,70.0,61.0,65.0,81.0,31,2,15,False\n",
      "102,f,40,75,65,70.0,53.0,37.0,84.0,52.0,81.0,51.0,22,4,7,False\n",
      "103,f,33,85,40,62.5,80.0,27.0,31.0,82.5,81.0,73.0,24,5,7,False\n"
     ]
    }
   ],
   "source": [
    "!type gfmt_sleep_with_insomnia.csv # NaN is a missing value, dont need to specify next time"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "18a6fe7c",
   "metadata": {},
   "source": [
    "**Styling Dataframes**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "5444ae82",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_800ef_row0_col0, #T_800ef_row0_col1, #T_800ef_row0_col2, #T_800ef_row0_col3, #T_800ef_row0_col4, #T_800ef_row0_col5, #T_800ef_row0_col6, #T_800ef_row0_col7, #T_800ef_row0_col8, #T_800ef_row0_col9, #T_800ef_row0_col10, #T_800ef_row0_col11, #T_800ef_row0_col12, #T_800ef_row0_col13, #T_800ef_row0_col14, #T_800ef_row0_col15, #T_800ef_row1_col0, #T_800ef_row1_col1, #T_800ef_row1_col2, #T_800ef_row1_col3, #T_800ef_row1_col4, #T_800ef_row1_col5, #T_800ef_row1_col6, #T_800ef_row1_col7, #T_800ef_row1_col8, #T_800ef_row1_col9, #T_800ef_row1_col10, #T_800ef_row1_col11, #T_800ef_row1_col12, #T_800ef_row1_col13, #T_800ef_row1_col14, #T_800ef_row1_col15, #T_800ef_row4_col0, #T_800ef_row4_col1, #T_800ef_row4_col2, #T_800ef_row4_col3, #T_800ef_row4_col4, #T_800ef_row4_col5, #T_800ef_row4_col6, #T_800ef_row4_col7, #T_800ef_row4_col8, #T_800ef_row4_col9, #T_800ef_row4_col10, #T_800ef_row4_col11, #T_800ef_row4_col12, #T_800ef_row4_col13, #T_800ef_row4_col14, #T_800ef_row4_col15 {\n",
       "  background-color: lightgray;\n",
       "}\n",
       "#T_800ef_row2_col0, #T_800ef_row2_col1, #T_800ef_row2_col2, #T_800ef_row2_col3, #T_800ef_row2_col4, #T_800ef_row2_col5, #T_800ef_row2_col6, #T_800ef_row2_col7, #T_800ef_row2_col8, #T_800ef_row2_col9, #T_800ef_row2_col10, #T_800ef_row2_col11, #T_800ef_row2_col12, #T_800ef_row2_col13, #T_800ef_row2_col14, #T_800ef_row2_col15, #T_800ef_row3_col0, #T_800ef_row3_col1, #T_800ef_row3_col2, #T_800ef_row3_col3, #T_800ef_row3_col4, #T_800ef_row3_col5, #T_800ef_row3_col6, #T_800ef_row3_col7, #T_800ef_row3_col8, #T_800ef_row3_col9, #T_800ef_row3_col10, #T_800ef_row3_col11, #T_800ef_row3_col12, #T_800ef_row3_col13, #T_800ef_row3_col14, #T_800ef_row3_col15 {\n",
       "  background-color: #7fc97f;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_800ef\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_800ef_level0_col0\" class=\"col_heading level0 col0\" >participant number</th>\n",
       "      <th id=\"T_800ef_level0_col1\" class=\"col_heading level0 col1\" >gender</th>\n",
       "      <th id=\"T_800ef_level0_col2\" class=\"col_heading level0 col2\" >age</th>\n",
       "      <th id=\"T_800ef_level0_col3\" class=\"col_heading level0 col3\" >correct hit percentage</th>\n",
       "      <th id=\"T_800ef_level0_col4\" class=\"col_heading level0 col4\" >correct reject percentage</th>\n",
       "      <th id=\"T_800ef_level0_col5\" class=\"col_heading level0 col5\" >percent correct</th>\n",
       "      <th id=\"T_800ef_level0_col6\" class=\"col_heading level0 col6\" >confidence when correct hit</th>\n",
       "      <th id=\"T_800ef_level0_col7\" class=\"col_heading level0 col7\" >confidence when incorrect hit</th>\n",
       "      <th id=\"T_800ef_level0_col8\" class=\"col_heading level0 col8\" >confidence when correct reject</th>\n",
       "      <th id=\"T_800ef_level0_col9\" class=\"col_heading level0 col9\" >confidence when incorrect reject</th>\n",
       "      <th id=\"T_800ef_level0_col10\" class=\"col_heading level0 col10\" >confidence when correct</th>\n",
       "      <th id=\"T_800ef_level0_col11\" class=\"col_heading level0 col11\" >confidence when incorrect</th>\n",
       "      <th id=\"T_800ef_level0_col12\" class=\"col_heading level0 col12\" >sci</th>\n",
       "      <th id=\"T_800ef_level0_col13\" class=\"col_heading level0 col13\" >psqi</th>\n",
       "      <th id=\"T_800ef_level0_col14\" class=\"col_heading level0 col14\" >ess</th>\n",
       "      <th id=\"T_800ef_level0_col15\" class=\"col_heading level0 col15\" >insomnia</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_800ef_level0_row0\" class=\"row_heading level0 row0\" >0</th>\n",
       "      <td id=\"T_800ef_row0_col0\" class=\"data row0 col0\" >8</td>\n",
       "      <td id=\"T_800ef_row0_col1\" class=\"data row0 col1\" >f</td>\n",
       "      <td id=\"T_800ef_row0_col2\" class=\"data row0 col2\" >39</td>\n",
       "      <td id=\"T_800ef_row0_col3\" class=\"data row0 col3\" >65</td>\n",
       "      <td id=\"T_800ef_row0_col4\" class=\"data row0 col4\" >80</td>\n",
       "      <td id=\"T_800ef_row0_col5\" class=\"data row0 col5\" >72.500000</td>\n",
       "      <td id=\"T_800ef_row0_col6\" class=\"data row0 col6\" >91.000000</td>\n",
       "      <td id=\"T_800ef_row0_col7\" class=\"data row0 col7\" >90.000000</td>\n",
       "      <td id=\"T_800ef_row0_col8\" class=\"data row0 col8\" >93.000000</td>\n",
       "      <td id=\"T_800ef_row0_col9\" class=\"data row0 col9\" >83.500000</td>\n",
       "      <td id=\"T_800ef_row0_col10\" class=\"data row0 col10\" >93.000000</td>\n",
       "      <td id=\"T_800ef_row0_col11\" class=\"data row0 col11\" >90.000000</td>\n",
       "      <td id=\"T_800ef_row0_col12\" class=\"data row0 col12\" >9</td>\n",
       "      <td id=\"T_800ef_row0_col13\" class=\"data row0 col13\" >13</td>\n",
       "      <td id=\"T_800ef_row0_col14\" class=\"data row0 col14\" >2</td>\n",
       "      <td id=\"T_800ef_row0_col15\" class=\"data row0 col15\" >True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_800ef_level0_row1\" class=\"row_heading level0 row1\" >1</th>\n",
       "      <td id=\"T_800ef_row1_col0\" class=\"data row1 col0\" >16</td>\n",
       "      <td id=\"T_800ef_row1_col1\" class=\"data row1 col1\" >m</td>\n",
       "      <td id=\"T_800ef_row1_col2\" class=\"data row1 col2\" >42</td>\n",
       "      <td id=\"T_800ef_row1_col3\" class=\"data row1 col3\" >90</td>\n",
       "      <td id=\"T_800ef_row1_col4\" class=\"data row1 col4\" >90</td>\n",
       "      <td id=\"T_800ef_row1_col5\" class=\"data row1 col5\" >90.000000</td>\n",
       "      <td id=\"T_800ef_row1_col6\" class=\"data row1 col6\" >75.500000</td>\n",
       "      <td id=\"T_800ef_row1_col7\" class=\"data row1 col7\" >55.500000</td>\n",
       "      <td id=\"T_800ef_row1_col8\" class=\"data row1 col8\" >70.500000</td>\n",
       "      <td id=\"T_800ef_row1_col9\" class=\"data row1 col9\" >50.000000</td>\n",
       "      <td id=\"T_800ef_row1_col10\" class=\"data row1 col10\" >75.000000</td>\n",
       "      <td id=\"T_800ef_row1_col11\" class=\"data row1 col11\" >50.000000</td>\n",
       "      <td id=\"T_800ef_row1_col12\" class=\"data row1 col12\" >4</td>\n",
       "      <td id=\"T_800ef_row1_col13\" class=\"data row1 col13\" >11</td>\n",
       "      <td id=\"T_800ef_row1_col14\" class=\"data row1 col14\" >7</td>\n",
       "      <td id=\"T_800ef_row1_col15\" class=\"data row1 col15\" >True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_800ef_level0_row2\" class=\"row_heading level0 row2\" >2</th>\n",
       "      <td id=\"T_800ef_row2_col0\" class=\"data row2 col0\" >18</td>\n",
       "      <td id=\"T_800ef_row2_col1\" class=\"data row2 col1\" >f</td>\n",
       "      <td id=\"T_800ef_row2_col2\" class=\"data row2 col2\" >31</td>\n",
       "      <td id=\"T_800ef_row2_col3\" class=\"data row2 col3\" >90</td>\n",
       "      <td id=\"T_800ef_row2_col4\" class=\"data row2 col4\" >95</td>\n",
       "      <td id=\"T_800ef_row2_col5\" class=\"data row2 col5\" >92.500000</td>\n",
       "      <td id=\"T_800ef_row2_col6\" class=\"data row2 col6\" >89.500000</td>\n",
       "      <td id=\"T_800ef_row2_col7\" class=\"data row2 col7\" >90.000000</td>\n",
       "      <td id=\"T_800ef_row2_col8\" class=\"data row2 col8\" >86.000000</td>\n",
       "      <td id=\"T_800ef_row2_col9\" class=\"data row2 col9\" >81.000000</td>\n",
       "      <td id=\"T_800ef_row2_col10\" class=\"data row2 col10\" >89.000000</td>\n",
       "      <td id=\"T_800ef_row2_col11\" class=\"data row2 col11\" >88.000000</td>\n",
       "      <td id=\"T_800ef_row2_col12\" class=\"data row2 col12\" >10</td>\n",
       "      <td id=\"T_800ef_row2_col13\" class=\"data row2 col13\" >9</td>\n",
       "      <td id=\"T_800ef_row2_col14\" class=\"data row2 col14\" >3</td>\n",
       "      <td id=\"T_800ef_row2_col15\" class=\"data row2 col15\" >True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_800ef_level0_row3\" class=\"row_heading level0 row3\" >3</th>\n",
       "      <td id=\"T_800ef_row3_col0\" class=\"data row3 col0\" >22</td>\n",
       "      <td id=\"T_800ef_row3_col1\" class=\"data row3 col1\" >f</td>\n",
       "      <td id=\"T_800ef_row3_col2\" class=\"data row3 col2\" >35</td>\n",
       "      <td id=\"T_800ef_row3_col3\" class=\"data row3 col3\" >100</td>\n",
       "      <td id=\"T_800ef_row3_col4\" class=\"data row3 col4\" >75</td>\n",
       "      <td id=\"T_800ef_row3_col5\" class=\"data row3 col5\" >87.500000</td>\n",
       "      <td id=\"T_800ef_row3_col6\" class=\"data row3 col6\" >89.500000</td>\n",
       "      <td id=\"T_800ef_row3_col7\" class=\"data row3 col7\" >nan</td>\n",
       "      <td id=\"T_800ef_row3_col8\" class=\"data row3 col8\" >71.000000</td>\n",
       "      <td id=\"T_800ef_row3_col9\" class=\"data row3 col9\" >80.000000</td>\n",
       "      <td id=\"T_800ef_row3_col10\" class=\"data row3 col10\" >88.000000</td>\n",
       "      <td id=\"T_800ef_row3_col11\" class=\"data row3 col11\" >80.000000</td>\n",
       "      <td id=\"T_800ef_row3_col12\" class=\"data row3 col12\" >13</td>\n",
       "      <td id=\"T_800ef_row3_col13\" class=\"data row3 col13\" >8</td>\n",
       "      <td id=\"T_800ef_row3_col14\" class=\"data row3 col14\" >20</td>\n",
       "      <td id=\"T_800ef_row3_col15\" class=\"data row3 col15\" >True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_800ef_level0_row4\" class=\"row_heading level0 row4\" >4</th>\n",
       "      <td id=\"T_800ef_row4_col0\" class=\"data row4 col0\" >27</td>\n",
       "      <td id=\"T_800ef_row4_col1\" class=\"data row4 col1\" >f</td>\n",
       "      <td id=\"T_800ef_row4_col2\" class=\"data row4 col2\" >74</td>\n",
       "      <td id=\"T_800ef_row4_col3\" class=\"data row4 col3\" >60</td>\n",
       "      <td id=\"T_800ef_row4_col4\" class=\"data row4 col4\" >65</td>\n",
       "      <td id=\"T_800ef_row4_col5\" class=\"data row4 col5\" >62.500000</td>\n",
       "      <td id=\"T_800ef_row4_col6\" class=\"data row4 col6\" >68.500000</td>\n",
       "      <td id=\"T_800ef_row4_col7\" class=\"data row4 col7\" >49.000000</td>\n",
       "      <td id=\"T_800ef_row4_col8\" class=\"data row4 col8\" >61.000000</td>\n",
       "      <td id=\"T_800ef_row4_col9\" class=\"data row4 col9\" >49.000000</td>\n",
       "      <td id=\"T_800ef_row4_col10\" class=\"data row4 col10\" >65.000000</td>\n",
       "      <td id=\"T_800ef_row4_col11\" class=\"data row4 col11\" >49.000000</td>\n",
       "      <td id=\"T_800ef_row4_col12\" class=\"data row4 col12\" >13</td>\n",
       "      <td id=\"T_800ef_row4_col13\" class=\"data row4 col13\" >9</td>\n",
       "      <td id=\"T_800ef_row4_col14\" class=\"data row4 col14\" >12</td>\n",
       "      <td id=\"T_800ef_row4_col15\" class=\"data row4 col15\" >True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x15f141d08d0>"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# female and percentage > 75 get a green bg color, else grey\n",
    "def highlight_high_scoring_females(s):\n",
    "    if s[\"gender\"] == \"f\" and s[\"percent correct\"] >= 75:\n",
    "        return [\"background-color: #7fc97f\"] * len(s)\n",
    "    else:\n",
    "        return [\"background-color: lightgray\"] * len(s) # len(s) is the number of columns\n",
    "\n",
    "df.head().style.apply(highlight_high_scoring_females, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "d18c20cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_af936_row0_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 72.5%, transparent 72.5%);\n",
       "}\n",
       "#T_af936_row1_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 90.0%, transparent 90.0%);\n",
       "}\n",
       "#T_af936_row2_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 92.5%, transparent 92.5%);\n",
       "}\n",
       "#T_af936_row3_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 87.5%, transparent 87.5%);\n",
       "}\n",
       "#T_af936_row4_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 62.5%, transparent 62.5%);\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_af936\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_af936_level0_col0\" class=\"col_heading level0 col0\" >participant number</th>\n",
       "      <th id=\"T_af936_level0_col1\" class=\"col_heading level0 col1\" >gender</th>\n",
       "      <th id=\"T_af936_level0_col2\" class=\"col_heading level0 col2\" >age</th>\n",
       "      <th id=\"T_af936_level0_col3\" class=\"col_heading level0 col3\" >correct hit percentage</th>\n",
       "      <th id=\"T_af936_level0_col4\" class=\"col_heading level0 col4\" >correct reject percentage</th>\n",
       "      <th id=\"T_af936_level0_col5\" class=\"col_heading level0 col5\" >percent correct</th>\n",
       "      <th id=\"T_af936_level0_col6\" class=\"col_heading level0 col6\" >confidence when correct hit</th>\n",
       "      <th id=\"T_af936_level0_col7\" class=\"col_heading level0 col7\" >confidence when incorrect hit</th>\n",
       "      <th id=\"T_af936_level0_col8\" class=\"col_heading level0 col8\" >confidence when correct reject</th>\n",
       "      <th id=\"T_af936_level0_col9\" class=\"col_heading level0 col9\" >confidence when incorrect reject</th>\n",
       "      <th id=\"T_af936_level0_col10\" class=\"col_heading level0 col10\" >confidence when correct</th>\n",
       "      <th id=\"T_af936_level0_col11\" class=\"col_heading level0 col11\" >confidence when incorrect</th>\n",
       "      <th id=\"T_af936_level0_col12\" class=\"col_heading level0 col12\" >sci</th>\n",
       "      <th id=\"T_af936_level0_col13\" class=\"col_heading level0 col13\" >psqi</th>\n",
       "      <th id=\"T_af936_level0_col14\" class=\"col_heading level0 col14\" >ess</th>\n",
       "      <th id=\"T_af936_level0_col15\" class=\"col_heading level0 col15\" >insomnia</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_af936_level0_row0\" class=\"row_heading level0 row0\" >0</th>\n",
       "      <td id=\"T_af936_row0_col0\" class=\"data row0 col0\" >8</td>\n",
       "      <td id=\"T_af936_row0_col1\" class=\"data row0 col1\" >f</td>\n",
       "      <td id=\"T_af936_row0_col2\" class=\"data row0 col2\" >39</td>\n",
       "      <td id=\"T_af936_row0_col3\" class=\"data row0 col3\" >65</td>\n",
       "      <td id=\"T_af936_row0_col4\" class=\"data row0 col4\" >80</td>\n",
       "      <td id=\"T_af936_row0_col5\" class=\"data row0 col5\" >72.500000</td>\n",
       "      <td id=\"T_af936_row0_col6\" class=\"data row0 col6\" >91.000000</td>\n",
       "      <td id=\"T_af936_row0_col7\" class=\"data row0 col7\" >90.000000</td>\n",
       "      <td id=\"T_af936_row0_col8\" class=\"data row0 col8\" >93.000000</td>\n",
       "      <td id=\"T_af936_row0_col9\" class=\"data row0 col9\" >83.500000</td>\n",
       "      <td id=\"T_af936_row0_col10\" class=\"data row0 col10\" >93.000000</td>\n",
       "      <td id=\"T_af936_row0_col11\" class=\"data row0 col11\" >90.000000</td>\n",
       "      <td id=\"T_af936_row0_col12\" class=\"data row0 col12\" >9</td>\n",
       "      <td id=\"T_af936_row0_col13\" class=\"data row0 col13\" >13</td>\n",
       "      <td id=\"T_af936_row0_col14\" class=\"data row0 col14\" >2</td>\n",
       "      <td id=\"T_af936_row0_col15\" class=\"data row0 col15\" >True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_af936_level0_row1\" class=\"row_heading level0 row1\" >1</th>\n",
       "      <td id=\"T_af936_row1_col0\" class=\"data row1 col0\" >16</td>\n",
       "      <td id=\"T_af936_row1_col1\" class=\"data row1 col1\" >m</td>\n",
       "      <td id=\"T_af936_row1_col2\" class=\"data row1 col2\" >42</td>\n",
       "      <td id=\"T_af936_row1_col3\" class=\"data row1 col3\" >90</td>\n",
       "      <td id=\"T_af936_row1_col4\" class=\"data row1 col4\" >90</td>\n",
       "      <td id=\"T_af936_row1_col5\" class=\"data row1 col5\" >90.000000</td>\n",
       "      <td id=\"T_af936_row1_col6\" class=\"data row1 col6\" >75.500000</td>\n",
       "      <td id=\"T_af936_row1_col7\" class=\"data row1 col7\" >55.500000</td>\n",
       "      <td id=\"T_af936_row1_col8\" class=\"data row1 col8\" >70.500000</td>\n",
       "      <td id=\"T_af936_row1_col9\" class=\"data row1 col9\" >50.000000</td>\n",
       "      <td id=\"T_af936_row1_col10\" class=\"data row1 col10\" >75.000000</td>\n",
       "      <td id=\"T_af936_row1_col11\" class=\"data row1 col11\" >50.000000</td>\n",
       "      <td id=\"T_af936_row1_col12\" class=\"data row1 col12\" >4</td>\n",
       "      <td id=\"T_af936_row1_col13\" class=\"data row1 col13\" >11</td>\n",
       "      <td id=\"T_af936_row1_col14\" class=\"data row1 col14\" >7</td>\n",
       "      <td id=\"T_af936_row1_col15\" class=\"data row1 col15\" >True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_af936_level0_row2\" class=\"row_heading level0 row2\" >2</th>\n",
       "      <td id=\"T_af936_row2_col0\" class=\"data row2 col0\" >18</td>\n",
       "      <td id=\"T_af936_row2_col1\" class=\"data row2 col1\" >f</td>\n",
       "      <td id=\"T_af936_row2_col2\" class=\"data row2 col2\" >31</td>\n",
       "      <td id=\"T_af936_row2_col3\" class=\"data row2 col3\" >90</td>\n",
       "      <td id=\"T_af936_row2_col4\" class=\"data row2 col4\" >95</td>\n",
       "      <td id=\"T_af936_row2_col5\" class=\"data row2 col5\" >92.500000</td>\n",
       "      <td id=\"T_af936_row2_col6\" class=\"data row2 col6\" >89.500000</td>\n",
       "      <td id=\"T_af936_row2_col7\" class=\"data row2 col7\" >90.000000</td>\n",
       "      <td id=\"T_af936_row2_col8\" class=\"data row2 col8\" >86.000000</td>\n",
       "      <td id=\"T_af936_row2_col9\" class=\"data row2 col9\" >81.000000</td>\n",
       "      <td id=\"T_af936_row2_col10\" class=\"data row2 col10\" >89.000000</td>\n",
       "      <td id=\"T_af936_row2_col11\" class=\"data row2 col11\" >88.000000</td>\n",
       "      <td id=\"T_af936_row2_col12\" class=\"data row2 col12\" >10</td>\n",
       "      <td id=\"T_af936_row2_col13\" class=\"data row2 col13\" >9</td>\n",
       "      <td id=\"T_af936_row2_col14\" class=\"data row2 col14\" >3</td>\n",
       "      <td id=\"T_af936_row2_col15\" class=\"data row2 col15\" >True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_af936_level0_row3\" class=\"row_heading level0 row3\" >3</th>\n",
       "      <td id=\"T_af936_row3_col0\" class=\"data row3 col0\" >22</td>\n",
       "      <td id=\"T_af936_row3_col1\" class=\"data row3 col1\" >f</td>\n",
       "      <td id=\"T_af936_row3_col2\" class=\"data row3 col2\" >35</td>\n",
       "      <td id=\"T_af936_row3_col3\" class=\"data row3 col3\" >100</td>\n",
       "      <td id=\"T_af936_row3_col4\" class=\"data row3 col4\" >75</td>\n",
       "      <td id=\"T_af936_row3_col5\" class=\"data row3 col5\" >87.500000</td>\n",
       "      <td id=\"T_af936_row3_col6\" class=\"data row3 col6\" >89.500000</td>\n",
       "      <td id=\"T_af936_row3_col7\" class=\"data row3 col7\" >nan</td>\n",
       "      <td id=\"T_af936_row3_col8\" class=\"data row3 col8\" >71.000000</td>\n",
       "      <td id=\"T_af936_row3_col9\" class=\"data row3 col9\" >80.000000</td>\n",
       "      <td id=\"T_af936_row3_col10\" class=\"data row3 col10\" >88.000000</td>\n",
       "      <td id=\"T_af936_row3_col11\" class=\"data row3 col11\" >80.000000</td>\n",
       "      <td id=\"T_af936_row3_col12\" class=\"data row3 col12\" >13</td>\n",
       "      <td id=\"T_af936_row3_col13\" class=\"data row3 col13\" >8</td>\n",
       "      <td id=\"T_af936_row3_col14\" class=\"data row3 col14\" >20</td>\n",
       "      <td id=\"T_af936_row3_col15\" class=\"data row3 col15\" >True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_af936_level0_row4\" class=\"row_heading level0 row4\" >4</th>\n",
       "      <td id=\"T_af936_row4_col0\" class=\"data row4 col0\" >27</td>\n",
       "      <td id=\"T_af936_row4_col1\" class=\"data row4 col1\" >f</td>\n",
       "      <td id=\"T_af936_row4_col2\" class=\"data row4 col2\" >74</td>\n",
       "      <td id=\"T_af936_row4_col3\" class=\"data row4 col3\" >60</td>\n",
       "      <td id=\"T_af936_row4_col4\" class=\"data row4 col4\" >65</td>\n",
       "      <td id=\"T_af936_row4_col5\" class=\"data row4 col5\" >62.500000</td>\n",
       "      <td id=\"T_af936_row4_col6\" class=\"data row4 col6\" >68.500000</td>\n",
       "      <td id=\"T_af936_row4_col7\" class=\"data row4 col7\" >49.000000</td>\n",
       "      <td id=\"T_af936_row4_col8\" class=\"data row4 col8\" >61.000000</td>\n",
       "      <td id=\"T_af936_row4_col9\" class=\"data row4 col9\" >49.000000</td>\n",
       "      <td id=\"T_af936_row4_col10\" class=\"data row4 col10\" >65.000000</td>\n",
       "      <td id=\"T_af936_row4_col11\" class=\"data row4 col11\" >49.000000</td>\n",
       "      <td id=\"T_af936_row4_col12\" class=\"data row4 col12\" >13</td>\n",
       "      <td id=\"T_af936_row4_col13\" class=\"data row4 col13\" >9</td>\n",
       "      <td id=\"T_af936_row4_col14\" class=\"data row4 col14\" >12</td>\n",
       "      <td id=\"T_af936_row4_col15\" class=\"data row4 col15\" >True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x15f141cc950>"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# percent correct represented as a bar\n",
    "df.head().style.bar(subset=[\"percent correct\"], vmin=0, vmax=100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "568c10e3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_367fa_row0_col5 {\n",
       "  background-color: #fca082;\n",
       "  color: #000000;\n",
       "}\n",
       "#T_367fa_row1_col5 {\n",
       "  background-color: #900a12;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_367fa_row2_col5 {\n",
       "  background-color: #67000d;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_367fa_row3_col5 {\n",
       "  background-color: #b11218;\n",
       "  color: #f1f1f1;\n",
       "}\n",
       "#T_367fa_row4_col5 {\n",
       "  background-color: #fff5f0;\n",
       "  color: #000000;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_367fa\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_367fa_level0_col0\" class=\"col_heading level0 col0\" >participant number</th>\n",
       "      <th id=\"T_367fa_level0_col1\" class=\"col_heading level0 col1\" >gender</th>\n",
       "      <th id=\"T_367fa_level0_col2\" class=\"col_heading level0 col2\" >age</th>\n",
       "      <th id=\"T_367fa_level0_col3\" class=\"col_heading level0 col3\" >correct hit percentage</th>\n",
       "      <th id=\"T_367fa_level0_col4\" class=\"col_heading level0 col4\" >correct reject percentage</th>\n",
       "      <th id=\"T_367fa_level0_col5\" class=\"col_heading level0 col5\" >percent correct</th>\n",
       "      <th id=\"T_367fa_level0_col6\" class=\"col_heading level0 col6\" >confidence when correct hit</th>\n",
       "      <th id=\"T_367fa_level0_col7\" class=\"col_heading level0 col7\" >confidence when incorrect hit</th>\n",
       "      <th id=\"T_367fa_level0_col8\" class=\"col_heading level0 col8\" >confidence when correct reject</th>\n",
       "      <th id=\"T_367fa_level0_col9\" class=\"col_heading level0 col9\" >confidence when incorrect reject</th>\n",
       "      <th id=\"T_367fa_level0_col10\" class=\"col_heading level0 col10\" >confidence when correct</th>\n",
       "      <th id=\"T_367fa_level0_col11\" class=\"col_heading level0 col11\" >confidence when incorrect</th>\n",
       "      <th id=\"T_367fa_level0_col12\" class=\"col_heading level0 col12\" >sci</th>\n",
       "      <th id=\"T_367fa_level0_col13\" class=\"col_heading level0 col13\" >psqi</th>\n",
       "      <th id=\"T_367fa_level0_col14\" class=\"col_heading level0 col14\" >ess</th>\n",
       "      <th id=\"T_367fa_level0_col15\" class=\"col_heading level0 col15\" >insomnia</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_367fa_level0_row0\" class=\"row_heading level0 row0\" >0</th>\n",
       "      <td id=\"T_367fa_row0_col0\" class=\"data row0 col0\" >8</td>\n",
       "      <td id=\"T_367fa_row0_col1\" class=\"data row0 col1\" >f</td>\n",
       "      <td id=\"T_367fa_row0_col2\" class=\"data row0 col2\" >39</td>\n",
       "      <td id=\"T_367fa_row0_col3\" class=\"data row0 col3\" >65</td>\n",
       "      <td id=\"T_367fa_row0_col4\" class=\"data row0 col4\" >80</td>\n",
       "      <td id=\"T_367fa_row0_col5\" class=\"data row0 col5\" >72.500000</td>\n",
       "      <td id=\"T_367fa_row0_col6\" class=\"data row0 col6\" >91.000000</td>\n",
       "      <td id=\"T_367fa_row0_col7\" class=\"data row0 col7\" >90.000000</td>\n",
       "      <td id=\"T_367fa_row0_col8\" class=\"data row0 col8\" >93.000000</td>\n",
       "      <td id=\"T_367fa_row0_col9\" class=\"data row0 col9\" >83.500000</td>\n",
       "      <td id=\"T_367fa_row0_col10\" class=\"data row0 col10\" >93.000000</td>\n",
       "      <td id=\"T_367fa_row0_col11\" class=\"data row0 col11\" >90.000000</td>\n",
       "      <td id=\"T_367fa_row0_col12\" class=\"data row0 col12\" >9</td>\n",
       "      <td id=\"T_367fa_row0_col13\" class=\"data row0 col13\" >13</td>\n",
       "      <td id=\"T_367fa_row0_col14\" class=\"data row0 col14\" >2</td>\n",
       "      <td id=\"T_367fa_row0_col15\" class=\"data row0 col15\" >True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_367fa_level0_row1\" class=\"row_heading level0 row1\" >1</th>\n",
       "      <td id=\"T_367fa_row1_col0\" class=\"data row1 col0\" >16</td>\n",
       "      <td id=\"T_367fa_row1_col1\" class=\"data row1 col1\" >m</td>\n",
       "      <td id=\"T_367fa_row1_col2\" class=\"data row1 col2\" >42</td>\n",
       "      <td id=\"T_367fa_row1_col3\" class=\"data row1 col3\" >90</td>\n",
       "      <td id=\"T_367fa_row1_col4\" class=\"data row1 col4\" >90</td>\n",
       "      <td id=\"T_367fa_row1_col5\" class=\"data row1 col5\" >90.000000</td>\n",
       "      <td id=\"T_367fa_row1_col6\" class=\"data row1 col6\" >75.500000</td>\n",
       "      <td id=\"T_367fa_row1_col7\" class=\"data row1 col7\" >55.500000</td>\n",
       "      <td id=\"T_367fa_row1_col8\" class=\"data row1 col8\" >70.500000</td>\n",
       "      <td id=\"T_367fa_row1_col9\" class=\"data row1 col9\" >50.000000</td>\n",
       "      <td id=\"T_367fa_row1_col10\" class=\"data row1 col10\" >75.000000</td>\n",
       "      <td id=\"T_367fa_row1_col11\" class=\"data row1 col11\" >50.000000</td>\n",
       "      <td id=\"T_367fa_row1_col12\" class=\"data row1 col12\" >4</td>\n",
       "      <td id=\"T_367fa_row1_col13\" class=\"data row1 col13\" >11</td>\n",
       "      <td id=\"T_367fa_row1_col14\" class=\"data row1 col14\" >7</td>\n",
       "      <td id=\"T_367fa_row1_col15\" class=\"data row1 col15\" >True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_367fa_level0_row2\" class=\"row_heading level0 row2\" >2</th>\n",
       "      <td id=\"T_367fa_row2_col0\" class=\"data row2 col0\" >18</td>\n",
       "      <td id=\"T_367fa_row2_col1\" class=\"data row2 col1\" >f</td>\n",
       "      <td id=\"T_367fa_row2_col2\" class=\"data row2 col2\" >31</td>\n",
       "      <td id=\"T_367fa_row2_col3\" class=\"data row2 col3\" >90</td>\n",
       "      <td id=\"T_367fa_row2_col4\" class=\"data row2 col4\" >95</td>\n",
       "      <td id=\"T_367fa_row2_col5\" class=\"data row2 col5\" >92.500000</td>\n",
       "      <td id=\"T_367fa_row2_col6\" class=\"data row2 col6\" >89.500000</td>\n",
       "      <td id=\"T_367fa_row2_col7\" class=\"data row2 col7\" >90.000000</td>\n",
       "      <td id=\"T_367fa_row2_col8\" class=\"data row2 col8\" >86.000000</td>\n",
       "      <td id=\"T_367fa_row2_col9\" class=\"data row2 col9\" >81.000000</td>\n",
       "      <td id=\"T_367fa_row2_col10\" class=\"data row2 col10\" >89.000000</td>\n",
       "      <td id=\"T_367fa_row2_col11\" class=\"data row2 col11\" >88.000000</td>\n",
       "      <td id=\"T_367fa_row2_col12\" class=\"data row2 col12\" >10</td>\n",
       "      <td id=\"T_367fa_row2_col13\" class=\"data row2 col13\" >9</td>\n",
       "      <td id=\"T_367fa_row2_col14\" class=\"data row2 col14\" >3</td>\n",
       "      <td id=\"T_367fa_row2_col15\" class=\"data row2 col15\" >True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_367fa_level0_row3\" class=\"row_heading level0 row3\" >3</th>\n",
       "      <td id=\"T_367fa_row3_col0\" class=\"data row3 col0\" >22</td>\n",
       "      <td id=\"T_367fa_row3_col1\" class=\"data row3 col1\" >f</td>\n",
       "      <td id=\"T_367fa_row3_col2\" class=\"data row3 col2\" >35</td>\n",
       "      <td id=\"T_367fa_row3_col3\" class=\"data row3 col3\" >100</td>\n",
       "      <td id=\"T_367fa_row3_col4\" class=\"data row3 col4\" >75</td>\n",
       "      <td id=\"T_367fa_row3_col5\" class=\"data row3 col5\" >87.500000</td>\n",
       "      <td id=\"T_367fa_row3_col6\" class=\"data row3 col6\" >89.500000</td>\n",
       "      <td id=\"T_367fa_row3_col7\" class=\"data row3 col7\" >nan</td>\n",
       "      <td id=\"T_367fa_row3_col8\" class=\"data row3 col8\" >71.000000</td>\n",
       "      <td id=\"T_367fa_row3_col9\" class=\"data row3 col9\" >80.000000</td>\n",
       "      <td id=\"T_367fa_row3_col10\" class=\"data row3 col10\" >88.000000</td>\n",
       "      <td id=\"T_367fa_row3_col11\" class=\"data row3 col11\" >80.000000</td>\n",
       "      <td id=\"T_367fa_row3_col12\" class=\"data row3 col12\" >13</td>\n",
       "      <td id=\"T_367fa_row3_col13\" class=\"data row3 col13\" >8</td>\n",
       "      <td id=\"T_367fa_row3_col14\" class=\"data row3 col14\" >20</td>\n",
       "      <td id=\"T_367fa_row3_col15\" class=\"data row3 col15\" >True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_367fa_level0_row4\" class=\"row_heading level0 row4\" >4</th>\n",
       "      <td id=\"T_367fa_row4_col0\" class=\"data row4 col0\" >27</td>\n",
       "      <td id=\"T_367fa_row4_col1\" class=\"data row4 col1\" >f</td>\n",
       "      <td id=\"T_367fa_row4_col2\" class=\"data row4 col2\" >74</td>\n",
       "      <td id=\"T_367fa_row4_col3\" class=\"data row4 col3\" >60</td>\n",
       "      <td id=\"T_367fa_row4_col4\" class=\"data row4 col4\" >65</td>\n",
       "      <td id=\"T_367fa_row4_col5\" class=\"data row4 col5\" >62.500000</td>\n",
       "      <td id=\"T_367fa_row4_col6\" class=\"data row4 col6\" >68.500000</td>\n",
       "      <td id=\"T_367fa_row4_col7\" class=\"data row4 col7\" >49.000000</td>\n",
       "      <td id=\"T_367fa_row4_col8\" class=\"data row4 col8\" >61.000000</td>\n",
       "      <td id=\"T_367fa_row4_col9\" class=\"data row4 col9\" >49.000000</td>\n",
       "      <td id=\"T_367fa_row4_col10\" class=\"data row4 col10\" >65.000000</td>\n",
       "      <td id=\"T_367fa_row4_col11\" class=\"data row4 col11\" >49.000000</td>\n",
       "      <td id=\"T_367fa_row4_col12\" class=\"data row4 col12\" >13</td>\n",
       "      <td id=\"T_367fa_row4_col13\" class=\"data row4 col13\" >9</td>\n",
       "      <td id=\"T_367fa_row4_col14\" class=\"data row4 col14\" >12</td>\n",
       "      <td id=\"T_367fa_row4_col15\" class=\"data row4 col15\" >True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x15f15f16c90>"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# color map for percent correct\n",
    "df.head().style.background_gradient(subset=[\"percent correct\"], cmap=\"Reds\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "245e9fe4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style type=\"text/css\">\n",
       "#T_8a77d_row0_col0, #T_8a77d_row0_col1, #T_8a77d_row0_col2, #T_8a77d_row0_col3, #T_8a77d_row0_col4, #T_8a77d_row0_col6, #T_8a77d_row0_col7, #T_8a77d_row0_col8, #T_8a77d_row0_col9, #T_8a77d_row0_col10, #T_8a77d_row0_col11, #T_8a77d_row0_col12, #T_8a77d_row0_col13, #T_8a77d_row0_col14, #T_8a77d_row0_col15, #T_8a77d_row1_col0, #T_8a77d_row1_col1, #T_8a77d_row1_col2, #T_8a77d_row1_col3, #T_8a77d_row1_col4, #T_8a77d_row1_col6, #T_8a77d_row1_col7, #T_8a77d_row1_col8, #T_8a77d_row1_col9, #T_8a77d_row1_col10, #T_8a77d_row1_col11, #T_8a77d_row1_col12, #T_8a77d_row1_col13, #T_8a77d_row1_col14, #T_8a77d_row1_col15, #T_8a77d_row4_col0, #T_8a77d_row4_col1, #T_8a77d_row4_col2, #T_8a77d_row4_col3, #T_8a77d_row4_col4, #T_8a77d_row4_col6, #T_8a77d_row4_col7, #T_8a77d_row4_col8, #T_8a77d_row4_col9, #T_8a77d_row4_col10, #T_8a77d_row4_col11, #T_8a77d_row4_col12, #T_8a77d_row4_col13, #T_8a77d_row4_col14, #T_8a77d_row4_col15 {\n",
       "  background-color: lightgray;\n",
       "}\n",
       "#T_8a77d_row0_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 72.5%, transparent 72.5%);\n",
       "  background-color: lightgray;\n",
       "}\n",
       "#T_8a77d_row1_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 90.0%, transparent 90.0%);\n",
       "  background-color: lightgray;\n",
       "}\n",
       "#T_8a77d_row2_col0, #T_8a77d_row2_col1, #T_8a77d_row2_col2, #T_8a77d_row2_col3, #T_8a77d_row2_col4, #T_8a77d_row2_col6, #T_8a77d_row2_col7, #T_8a77d_row2_col8, #T_8a77d_row2_col9, #T_8a77d_row2_col10, #T_8a77d_row2_col11, #T_8a77d_row2_col12, #T_8a77d_row2_col13, #T_8a77d_row2_col14, #T_8a77d_row2_col15, #T_8a77d_row3_col0, #T_8a77d_row3_col1, #T_8a77d_row3_col2, #T_8a77d_row3_col3, #T_8a77d_row3_col4, #T_8a77d_row3_col6, #T_8a77d_row3_col7, #T_8a77d_row3_col8, #T_8a77d_row3_col9, #T_8a77d_row3_col10, #T_8a77d_row3_col11, #T_8a77d_row3_col12, #T_8a77d_row3_col13, #T_8a77d_row3_col14, #T_8a77d_row3_col15 {\n",
       "  background-color: #7fc97f;\n",
       "}\n",
       "#T_8a77d_row2_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 92.5%, transparent 92.5%);\n",
       "  background-color: #7fc97f;\n",
       "}\n",
       "#T_8a77d_row3_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 87.5%, transparent 87.5%);\n",
       "  background-color: #7fc97f;\n",
       "}\n",
       "#T_8a77d_row4_col5 {\n",
       "  width: 10em;\n",
       "  background: linear-gradient(90deg, #d65f5f 62.5%, transparent 62.5%);\n",
       "  background-color: lightgray;\n",
       "}\n",
       "</style>\n",
       "<table id=\"T_8a77d\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th class=\"blank level0\" >&nbsp;</th>\n",
       "      <th id=\"T_8a77d_level0_col0\" class=\"col_heading level0 col0\" >participant number</th>\n",
       "      <th id=\"T_8a77d_level0_col1\" class=\"col_heading level0 col1\" >gender</th>\n",
       "      <th id=\"T_8a77d_level0_col2\" class=\"col_heading level0 col2\" >age</th>\n",
       "      <th id=\"T_8a77d_level0_col3\" class=\"col_heading level0 col3\" >correct hit percentage</th>\n",
       "      <th id=\"T_8a77d_level0_col4\" class=\"col_heading level0 col4\" >correct reject percentage</th>\n",
       "      <th id=\"T_8a77d_level0_col5\" class=\"col_heading level0 col5\" >percent correct</th>\n",
       "      <th id=\"T_8a77d_level0_col6\" class=\"col_heading level0 col6\" >confidence when correct hit</th>\n",
       "      <th id=\"T_8a77d_level0_col7\" class=\"col_heading level0 col7\" >confidence when incorrect hit</th>\n",
       "      <th id=\"T_8a77d_level0_col8\" class=\"col_heading level0 col8\" >confidence when correct reject</th>\n",
       "      <th id=\"T_8a77d_level0_col9\" class=\"col_heading level0 col9\" >confidence when incorrect reject</th>\n",
       "      <th id=\"T_8a77d_level0_col10\" class=\"col_heading level0 col10\" >confidence when correct</th>\n",
       "      <th id=\"T_8a77d_level0_col11\" class=\"col_heading level0 col11\" >confidence when incorrect</th>\n",
       "      <th id=\"T_8a77d_level0_col12\" class=\"col_heading level0 col12\" >sci</th>\n",
       "      <th id=\"T_8a77d_level0_col13\" class=\"col_heading level0 col13\" >psqi</th>\n",
       "      <th id=\"T_8a77d_level0_col14\" class=\"col_heading level0 col14\" >ess</th>\n",
       "      <th id=\"T_8a77d_level0_col15\" class=\"col_heading level0 col15\" >insomnia</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th id=\"T_8a77d_level0_row0\" class=\"row_heading level0 row0\" >0</th>\n",
       "      <td id=\"T_8a77d_row0_col0\" class=\"data row0 col0\" >8</td>\n",
       "      <td id=\"T_8a77d_row0_col1\" class=\"data row0 col1\" >f</td>\n",
       "      <td id=\"T_8a77d_row0_col2\" class=\"data row0 col2\" >39</td>\n",
       "      <td id=\"T_8a77d_row0_col3\" class=\"data row0 col3\" >65</td>\n",
       "      <td id=\"T_8a77d_row0_col4\" class=\"data row0 col4\" >80</td>\n",
       "      <td id=\"T_8a77d_row0_col5\" class=\"data row0 col5\" >72.500000</td>\n",
       "      <td id=\"T_8a77d_row0_col6\" class=\"data row0 col6\" >91.000000</td>\n",
       "      <td id=\"T_8a77d_row0_col7\" class=\"data row0 col7\" >90.000000</td>\n",
       "      <td id=\"T_8a77d_row0_col8\" class=\"data row0 col8\" >93.000000</td>\n",
       "      <td id=\"T_8a77d_row0_col9\" class=\"data row0 col9\" >83.500000</td>\n",
       "      <td id=\"T_8a77d_row0_col10\" class=\"data row0 col10\" >93.000000</td>\n",
       "      <td id=\"T_8a77d_row0_col11\" class=\"data row0 col11\" >90.000000</td>\n",
       "      <td id=\"T_8a77d_row0_col12\" class=\"data row0 col12\" >9</td>\n",
       "      <td id=\"T_8a77d_row0_col13\" class=\"data row0 col13\" >13</td>\n",
       "      <td id=\"T_8a77d_row0_col14\" class=\"data row0 col14\" >2</td>\n",
       "      <td id=\"T_8a77d_row0_col15\" class=\"data row0 col15\" >True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_8a77d_level0_row1\" class=\"row_heading level0 row1\" >1</th>\n",
       "      <td id=\"T_8a77d_row1_col0\" class=\"data row1 col0\" >16</td>\n",
       "      <td id=\"T_8a77d_row1_col1\" class=\"data row1 col1\" >m</td>\n",
       "      <td id=\"T_8a77d_row1_col2\" class=\"data row1 col2\" >42</td>\n",
       "      <td id=\"T_8a77d_row1_col3\" class=\"data row1 col3\" >90</td>\n",
       "      <td id=\"T_8a77d_row1_col4\" class=\"data row1 col4\" >90</td>\n",
       "      <td id=\"T_8a77d_row1_col5\" class=\"data row1 col5\" >90.000000</td>\n",
       "      <td id=\"T_8a77d_row1_col6\" class=\"data row1 col6\" >75.500000</td>\n",
       "      <td id=\"T_8a77d_row1_col7\" class=\"data row1 col7\" >55.500000</td>\n",
       "      <td id=\"T_8a77d_row1_col8\" class=\"data row1 col8\" >70.500000</td>\n",
       "      <td id=\"T_8a77d_row1_col9\" class=\"data row1 col9\" >50.000000</td>\n",
       "      <td id=\"T_8a77d_row1_col10\" class=\"data row1 col10\" >75.000000</td>\n",
       "      <td id=\"T_8a77d_row1_col11\" class=\"data row1 col11\" >50.000000</td>\n",
       "      <td id=\"T_8a77d_row1_col12\" class=\"data row1 col12\" >4</td>\n",
       "      <td id=\"T_8a77d_row1_col13\" class=\"data row1 col13\" >11</td>\n",
       "      <td id=\"T_8a77d_row1_col14\" class=\"data row1 col14\" >7</td>\n",
       "      <td id=\"T_8a77d_row1_col15\" class=\"data row1 col15\" >True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_8a77d_level0_row2\" class=\"row_heading level0 row2\" >2</th>\n",
       "      <td id=\"T_8a77d_row2_col0\" class=\"data row2 col0\" >18</td>\n",
       "      <td id=\"T_8a77d_row2_col1\" class=\"data row2 col1\" >f</td>\n",
       "      <td id=\"T_8a77d_row2_col2\" class=\"data row2 col2\" >31</td>\n",
       "      <td id=\"T_8a77d_row2_col3\" class=\"data row2 col3\" >90</td>\n",
       "      <td id=\"T_8a77d_row2_col4\" class=\"data row2 col4\" >95</td>\n",
       "      <td id=\"T_8a77d_row2_col5\" class=\"data row2 col5\" >92.500000</td>\n",
       "      <td id=\"T_8a77d_row2_col6\" class=\"data row2 col6\" >89.500000</td>\n",
       "      <td id=\"T_8a77d_row2_col7\" class=\"data row2 col7\" >90.000000</td>\n",
       "      <td id=\"T_8a77d_row2_col8\" class=\"data row2 col8\" >86.000000</td>\n",
       "      <td id=\"T_8a77d_row2_col9\" class=\"data row2 col9\" >81.000000</td>\n",
       "      <td id=\"T_8a77d_row2_col10\" class=\"data row2 col10\" >89.000000</td>\n",
       "      <td id=\"T_8a77d_row2_col11\" class=\"data row2 col11\" >88.000000</td>\n",
       "      <td id=\"T_8a77d_row2_col12\" class=\"data row2 col12\" >10</td>\n",
       "      <td id=\"T_8a77d_row2_col13\" class=\"data row2 col13\" >9</td>\n",
       "      <td id=\"T_8a77d_row2_col14\" class=\"data row2 col14\" >3</td>\n",
       "      <td id=\"T_8a77d_row2_col15\" class=\"data row2 col15\" >True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_8a77d_level0_row3\" class=\"row_heading level0 row3\" >3</th>\n",
       "      <td id=\"T_8a77d_row3_col0\" class=\"data row3 col0\" >22</td>\n",
       "      <td id=\"T_8a77d_row3_col1\" class=\"data row3 col1\" >f</td>\n",
       "      <td id=\"T_8a77d_row3_col2\" class=\"data row3 col2\" >35</td>\n",
       "      <td id=\"T_8a77d_row3_col3\" class=\"data row3 col3\" >100</td>\n",
       "      <td id=\"T_8a77d_row3_col4\" class=\"data row3 col4\" >75</td>\n",
       "      <td id=\"T_8a77d_row3_col5\" class=\"data row3 col5\" >87.500000</td>\n",
       "      <td id=\"T_8a77d_row3_col6\" class=\"data row3 col6\" >89.500000</td>\n",
       "      <td id=\"T_8a77d_row3_col7\" class=\"data row3 col7\" >nan</td>\n",
       "      <td id=\"T_8a77d_row3_col8\" class=\"data row3 col8\" >71.000000</td>\n",
       "      <td id=\"T_8a77d_row3_col9\" class=\"data row3 col9\" >80.000000</td>\n",
       "      <td id=\"T_8a77d_row3_col10\" class=\"data row3 col10\" >88.000000</td>\n",
       "      <td id=\"T_8a77d_row3_col11\" class=\"data row3 col11\" >80.000000</td>\n",
       "      <td id=\"T_8a77d_row3_col12\" class=\"data row3 col12\" >13</td>\n",
       "      <td id=\"T_8a77d_row3_col13\" class=\"data row3 col13\" >8</td>\n",
       "      <td id=\"T_8a77d_row3_col14\" class=\"data row3 col14\" >20</td>\n",
       "      <td id=\"T_8a77d_row3_col15\" class=\"data row3 col15\" >True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th id=\"T_8a77d_level0_row4\" class=\"row_heading level0 row4\" >4</th>\n",
       "      <td id=\"T_8a77d_row4_col0\" class=\"data row4 col0\" >27</td>\n",
       "      <td id=\"T_8a77d_row4_col1\" class=\"data row4 col1\" >f</td>\n",
       "      <td id=\"T_8a77d_row4_col2\" class=\"data row4 col2\" >74</td>\n",
       "      <td id=\"T_8a77d_row4_col3\" class=\"data row4 col3\" >60</td>\n",
       "      <td id=\"T_8a77d_row4_col4\" class=\"data row4 col4\" >65</td>\n",
       "      <td id=\"T_8a77d_row4_col5\" class=\"data row4 col5\" >62.500000</td>\n",
       "      <td id=\"T_8a77d_row4_col6\" class=\"data row4 col6\" >68.500000</td>\n",
       "      <td id=\"T_8a77d_row4_col7\" class=\"data row4 col7\" >49.000000</td>\n",
       "      <td id=\"T_8a77d_row4_col8\" class=\"data row4 col8\" >61.000000</td>\n",
       "      <td id=\"T_8a77d_row4_col9\" class=\"data row4 col9\" >49.000000</td>\n",
       "      <td id=\"T_8a77d_row4_col10\" class=\"data row4 col10\" >65.000000</td>\n",
       "      <td id=\"T_8a77d_row4_col11\" class=\"data row4 col11\" >49.000000</td>\n",
       "      <td id=\"T_8a77d_row4_col12\" class=\"data row4 col12\" >13</td>\n",
       "      <td id=\"T_8a77d_row4_col13\" class=\"data row4 col13\" >9</td>\n",
       "      <td id=\"T_8a77d_row4_col14\" class=\"data row4 col14\" >12</td>\n",
       "      <td id=\"T_8a77d_row4_col15\" class=\"data row4 col15\" >True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n"
      ],
      "text/plain": [
       "<pandas.io.formats.style.Styler at 0x15f1418aa50>"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# multiple effects\n",
    "df.head().style.bar(\n",
    "    subset=[\"percent correct\"], vmin=0, vmax=100\n",
    ").apply(\n",
    "    highlight_high_scoring_females, axis=1\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17ac978a",
   "metadata": {},
   "source": [
    "## Lesson 17: Tidy data and split-apply-combine"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "b7a1978a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3455d47",
   "metadata": {},
   "source": [
    "tidy format data: each row represents a signle test on a single participant\n",
    "\n",
    "-> also: Each type of observation will have its own separate data frame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "f0989cb0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>participant number</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>correct hit percentage</th>\n",
       "      <th>correct reject percentage</th>\n",
       "      <th>percent correct</th>\n",
       "      <th>confidence when correct hit</th>\n",
       "      <th>confidence when incorrect hit</th>\n",
       "      <th>confidence when correct reject</th>\n",
       "      <th>confidence when incorrect reject</th>\n",
       "      <th>confidence when correct</th>\n",
       "      <th>confidence when incorrect</th>\n",
       "      <th>sci</th>\n",
       "      <th>psqi</th>\n",
       "      <th>ess</th>\n",
       "      <th>insomnia</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>f</td>\n",
       "      <td>39</td>\n",
       "      <td>65</td>\n",
       "      <td>80</td>\n",
       "      <td>72.5</td>\n",
       "      <td>91.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>83.5</td>\n",
       "      <td>93.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>9</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16</td>\n",
       "      <td>m</td>\n",
       "      <td>42</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>90.0</td>\n",
       "      <td>75.5</td>\n",
       "      <td>55.5</td>\n",
       "      <td>70.5</td>\n",
       "      <td>50.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18</td>\n",
       "      <td>f</td>\n",
       "      <td>31</td>\n",
       "      <td>90</td>\n",
       "      <td>95</td>\n",
       "      <td>92.5</td>\n",
       "      <td>89.5</td>\n",
       "      <td>90.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>f</td>\n",
       "      <td>35</td>\n",
       "      <td>100</td>\n",
       "      <td>75</td>\n",
       "      <td>87.5</td>\n",
       "      <td>89.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>71.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>13</td>\n",
       "      <td>8</td>\n",
       "      <td>20</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>27</td>\n",
       "      <td>f</td>\n",
       "      <td>74</td>\n",
       "      <td>60</td>\n",
       "      <td>65</td>\n",
       "      <td>62.5</td>\n",
       "      <td>68.5</td>\n",
       "      <td>49.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   participant number gender  age  correct hit percentage  \\\n",
       "0                   8      f   39                      65   \n",
       "1                  16      m   42                      90   \n",
       "2                  18      f   31                      90   \n",
       "3                  22      f   35                     100   \n",
       "4                  27      f   74                      60   \n",
       "\n",
       "   correct reject percentage  percent correct  confidence when correct hit  \\\n",
       "0                         80             72.5                         91.0   \n",
       "1                         90             90.0                         75.5   \n",
       "2                         95             92.5                         89.5   \n",
       "3                         75             87.5                         89.5   \n",
       "4                         65             62.5                         68.5   \n",
       "\n",
       "   confidence when incorrect hit  confidence when correct reject  \\\n",
       "0                           90.0                            93.0   \n",
       "1                           55.5                            70.5   \n",
       "2                           90.0                            86.0   \n",
       "3                            NaN                            71.0   \n",
       "4                           49.0                            61.0   \n",
       "\n",
       "   confidence when incorrect reject  confidence when correct  \\\n",
       "0                              83.5                     93.0   \n",
       "1                              50.0                     75.0   \n",
       "2                              81.0                     89.0   \n",
       "3                              80.0                     88.0   \n",
       "4                              49.0                     65.0   \n",
       "\n",
       "   confidence when incorrect  sci  psqi  ess  insomnia  \n",
       "0                       90.0    9    13    2      True  \n",
       "1                       50.0    4    11    7      True  \n",
       "2                       88.0   10     9    3      True  \n",
       "3                       80.0   13     8   20      True  \n",
       "4                       49.0   13     9   12      True  "
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# reload the df with the insomnia column\n",
    "df = pd.read_csv('data/gfmt_sleep.csv', na_values='*')\n",
    "df['insomnia'] = df['sci'] <= 16\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d9f0d85",
   "metadata": {},
   "source": [
    "**Split-apply-combine**\n",
    "\n",
    "-- Split the data by some criterion. Apply some function to the split-up data. Combine the results into a new data frame --\n",
    "\n",
    "For a df where we havve median percent correct face matching for insomnia and control, we 1.st split by insomnia, then apply median function, then combine the results into a summary dataset with insomnia and median.\n",
    "\n",
    "given by [panda's groupby](http://pandas.pydata.org/pandas-docs/stable/user_guide/groupby.html)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "9eb51636",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<pandas.core.groupby.generic.DataFrameGroupBy object at 0x0000015F16020D50>"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "grouped = df.groupby('insomnia')\n",
    "\n",
    "grouped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "2b61bf97",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>participant number</th>\n",
       "      <th>age</th>\n",
       "      <th>correct hit percentage</th>\n",
       "      <th>correct reject percentage</th>\n",
       "      <th>percent correct</th>\n",
       "      <th>confidence when correct hit</th>\n",
       "      <th>confidence when incorrect hit</th>\n",
       "      <th>confidence when correct reject</th>\n",
       "      <th>confidence when incorrect reject</th>\n",
       "      <th>confidence when correct</th>\n",
       "      <th>confidence when incorrect</th>\n",
       "      <th>sci</th>\n",
       "      <th>psqi</th>\n",
       "      <th>ess</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>insomnia</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>False</th>\n",
       "      <td>54.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>74.5</td>\n",
       "      <td>55.5</td>\n",
       "      <td>71.5</td>\n",
       "      <td>59.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>59.25</td>\n",
       "      <td>26.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <td>46.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>76.5</td>\n",
       "      <td>72.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>68.5</td>\n",
       "      <td>77.0</td>\n",
       "      <td>65.00</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          participant number   age  correct hit percentage  \\\n",
       "insomnia                                                     \n",
       "False                   54.0  36.0                    90.0   \n",
       "True                    46.0  39.0                    90.0   \n",
       "\n",
       "          correct reject percentage  percent correct  \\\n",
       "insomnia                                               \n",
       "False                          80.0             85.0   \n",
       "True                           75.0             75.0   \n",
       "\n",
       "          confidence when correct hit  confidence when incorrect hit  \\\n",
       "insomnia                                                               \n",
       "False                            74.5                           55.5   \n",
       "True                             76.5                           72.0   \n",
       "\n",
       "          confidence when correct reject  confidence when incorrect reject  \\\n",
       "insomnia                                                                     \n",
       "False                               71.5                              59.0   \n",
       "True                                71.0                              68.5   \n",
       "\n",
       "          confidence when correct  confidence when incorrect   sci  psqi  ess  \n",
       "insomnia                                                                       \n",
       "False                        75.0                      59.25  26.0   4.0  6.0  \n",
       "True                         77.0                      65.00  14.0   9.0  7.0  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_median = grouped.median(numeric_only=True)\n",
    "\n",
    "# df has insomnia as name of the index \n",
    "df_median"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "6184cc89",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>level_0</th>\n",
       "      <th>index</th>\n",
       "      <th>insomnia</th>\n",
       "      <th>participant number</th>\n",
       "      <th>age</th>\n",
       "      <th>correct hit percentage</th>\n",
       "      <th>correct reject percentage</th>\n",
       "      <th>percent correct</th>\n",
       "      <th>confidence when correct hit</th>\n",
       "      <th>confidence when incorrect hit</th>\n",
       "      <th>confidence when correct reject</th>\n",
       "      <th>confidence when incorrect reject</th>\n",
       "      <th>confidence when correct</th>\n",
       "      <th>confidence when incorrect</th>\n",
       "      <th>sci</th>\n",
       "      <th>psqi</th>\n",
       "      <th>ess</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>False</td>\n",
       "      <td>54.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>74.5</td>\n",
       "      <td>55.5</td>\n",
       "      <td>71.5</td>\n",
       "      <td>59.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>59.25</td>\n",
       "      <td>26.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "      <td>46.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>76.5</td>\n",
       "      <td>72.0</td>\n",
       "      <td>71.0</td>\n",
       "      <td>68.5</td>\n",
       "      <td>77.0</td>\n",
       "      <td>65.00</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   level_0  index  insomnia  participant number   age  correct hit percentage  \\\n",
       "0        0      0     False                54.0  36.0                    90.0   \n",
       "1        1      1      True                46.0  39.0                    90.0   \n",
       "\n",
       "   correct reject percentage  percent correct  confidence when correct hit  \\\n",
       "0                       80.0             85.0                         74.5   \n",
       "1                       75.0             75.0                         76.5   \n",
       "\n",
       "   confidence when incorrect hit  confidence when correct reject  \\\n",
       "0                           55.5                            71.5   \n",
       "1                           72.0                            71.0   \n",
       "\n",
       "   confidence when incorrect reject  confidence when correct  \\\n",
       "0                              59.0                     75.0   \n",
       "1                              68.5                     77.0   \n",
       "\n",
       "   confidence when incorrect   sci  psqi  ess  \n",
       "0                      59.25  26.0   4.0  6.0  \n",
       "1                      65.00  14.0   9.0  7.0  "
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# give again nrs to the indexes\n",
    "df_median.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "bd7b7470",
   "metadata": {},
   "outputs": [],
   "source": [
    "# save the df without the insomnia index\n",
    "df_median = df_median.reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "681d5ec9",
   "metadata": {},
   "source": [
    "Look at a function for more than 1 group, e.g. median for insomnia + gender"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "971a1b7f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gender</th>\n",
       "      <th>insomnia</th>\n",
       "      <th>participant number</th>\n",
       "      <th>age</th>\n",
       "      <th>correct hit percentage</th>\n",
       "      <th>correct reject percentage</th>\n",
       "      <th>percent correct</th>\n",
       "      <th>confidence when correct hit</th>\n",
       "      <th>confidence when incorrect hit</th>\n",
       "      <th>confidence when correct reject</th>\n",
       "      <th>confidence when incorrect reject</th>\n",
       "      <th>confidence when correct</th>\n",
       "      <th>confidence when incorrect</th>\n",
       "      <th>sci</th>\n",
       "      <th>psqi</th>\n",
       "      <th>ess</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>f</td>\n",
       "      <td>False</td>\n",
       "      <td>58.0</td>\n",
       "      <td>36.0</td>\n",
       "      <td>85.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>85.00</td>\n",
       "      <td>74.00</td>\n",
       "      <td>55.00</td>\n",
       "      <td>70.50</td>\n",
       "      <td>60.00</td>\n",
       "      <td>74.00</td>\n",
       "      <td>58.75</td>\n",
       "      <td>26.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>f</td>\n",
       "      <td>True</td>\n",
       "      <td>46.0</td>\n",
       "      <td>39.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>72.50</td>\n",
       "      <td>76.50</td>\n",
       "      <td>73.75</td>\n",
       "      <td>71.00</td>\n",
       "      <td>68.50</td>\n",
       "      <td>77.00</td>\n",
       "      <td>70.50</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>m</td>\n",
       "      <td>False</td>\n",
       "      <td>41.0</td>\n",
       "      <td>38.5</td>\n",
       "      <td>90.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>82.50</td>\n",
       "      <td>76.00</td>\n",
       "      <td>57.75</td>\n",
       "      <td>74.25</td>\n",
       "      <td>54.75</td>\n",
       "      <td>76.25</td>\n",
       "      <td>59.25</td>\n",
       "      <td>29.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>m</td>\n",
       "      <td>True</td>\n",
       "      <td>55.5</td>\n",
       "      <td>37.0</td>\n",
       "      <td>95.0</td>\n",
       "      <td>82.5</td>\n",
       "      <td>83.75</td>\n",
       "      <td>83.75</td>\n",
       "      <td>55.50</td>\n",
       "      <td>75.75</td>\n",
       "      <td>73.25</td>\n",
       "      <td>81.25</td>\n",
       "      <td>62.50</td>\n",
       "      <td>14.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gender  insomnia  participant number   age  correct hit percentage  \\\n",
       "0      f     False                58.0  36.0                    85.0   \n",
       "1      f      True                46.0  39.0                    80.0   \n",
       "2      m     False                41.0  38.5                    90.0   \n",
       "3      m      True                55.5  37.0                    95.0   \n",
       "\n",
       "   correct reject percentage  percent correct  confidence when correct hit  \\\n",
       "0                       80.0            85.00                        74.00   \n",
       "1                       75.0            72.50                        76.50   \n",
       "2                       80.0            82.50                        76.00   \n",
       "3                       82.5            83.75                        83.75   \n",
       "\n",
       "   confidence when incorrect hit  confidence when correct reject  \\\n",
       "0                          55.00                           70.50   \n",
       "1                          73.75                           71.00   \n",
       "2                          57.75                           74.25   \n",
       "3                          55.50                           75.75   \n",
       "\n",
       "   confidence when incorrect reject  confidence when correct  \\\n",
       "0                             60.00                    74.00   \n",
       "1                             68.50                    77.00   \n",
       "2                             54.75                    76.25   \n",
       "3                             73.25                    81.25   \n",
       "\n",
       "   confidence when incorrect   sci  psqi  ess  \n",
       "0                      58.75  26.0   4.0  7.0  \n",
       "1                      70.50  14.0   9.0  7.0  \n",
       "2                      59.25  29.0   3.0  6.0  \n",
       "3                      62.50  14.0   9.0  8.0  "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# aggregation\n",
    "df.groupby(['gender', 'insomnia']).median(numeric_only=True).reset_index()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e8da6391",
   "metadata": {},
   "source": [
    "**Transformation**"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29acbd82",
   "metadata": {},
   "source": [
    "Achieve a column of ranks with the indexing of the original dataframe"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "d33ef5f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0      11.0\n",
       "1      21.5\n",
       "2      23.0\n",
       "3      19.5\n",
       "4       3.5\n",
       "       ... \n",
       "97     25.0\n",
       "98     52.5\n",
       "99     19.0\n",
       "100    11.5\n",
       "101     6.5\n",
       "Name: percent correct, Length: 102, dtype: float64"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# rank for both groups the amount of correct matching percentage\n",
    "grouped['percent correct'].rank()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 95,
   "id": "ab0ea276",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>participant number</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>correct hit percentage</th>\n",
       "      <th>correct reject percentage</th>\n",
       "      <th>percent correct</th>\n",
       "      <th>confidence when correct hit</th>\n",
       "      <th>confidence when incorrect hit</th>\n",
       "      <th>confidence when correct reject</th>\n",
       "      <th>confidence when incorrect reject</th>\n",
       "      <th>confidence when correct</th>\n",
       "      <th>confidence when incorrect</th>\n",
       "      <th>sci</th>\n",
       "      <th>psqi</th>\n",
       "      <th>ess</th>\n",
       "      <th>insomnia</th>\n",
       "      <th>rank grouped by insomnia</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>f</td>\n",
       "      <td>39</td>\n",
       "      <td>65</td>\n",
       "      <td>80</td>\n",
       "      <td>72.5</td>\n",
       "      <td>91.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>83.5</td>\n",
       "      <td>93.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>9</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16</td>\n",
       "      <td>m</td>\n",
       "      <td>42</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>90.0</td>\n",
       "      <td>75.5</td>\n",
       "      <td>55.5</td>\n",
       "      <td>70.5</td>\n",
       "      <td>50.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>True</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18</td>\n",
       "      <td>f</td>\n",
       "      <td>31</td>\n",
       "      <td>90</td>\n",
       "      <td>95</td>\n",
       "      <td>92.5</td>\n",
       "      <td>89.5</td>\n",
       "      <td>90.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>f</td>\n",
       "      <td>35</td>\n",
       "      <td>100</td>\n",
       "      <td>75</td>\n",
       "      <td>87.5</td>\n",
       "      <td>89.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>71.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>13</td>\n",
       "      <td>8</td>\n",
       "      <td>20</td>\n",
       "      <td>True</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>27</td>\n",
       "      <td>f</td>\n",
       "      <td>74</td>\n",
       "      <td>60</td>\n",
       "      <td>65</td>\n",
       "      <td>62.5</td>\n",
       "      <td>68.5</td>\n",
       "      <td>49.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   participant number gender  age  correct hit percentage  \\\n",
       "0                   8      f   39                      65   \n",
       "1                  16      m   42                      90   \n",
       "2                  18      f   31                      90   \n",
       "3                  22      f   35                     100   \n",
       "4                  27      f   74                      60   \n",
       "\n",
       "   correct reject percentage  percent correct  confidence when correct hit  \\\n",
       "0                         80             72.5                         91.0   \n",
       "1                         90             90.0                         75.5   \n",
       "2                         95             92.5                         89.5   \n",
       "3                         75             87.5                         89.5   \n",
       "4                         65             62.5                         68.5   \n",
       "\n",
       "   confidence when incorrect hit  confidence when correct reject  \\\n",
       "0                           90.0                            93.0   \n",
       "1                           55.5                            70.5   \n",
       "2                           90.0                            86.0   \n",
       "3                            NaN                            71.0   \n",
       "4                           49.0                            61.0   \n",
       "\n",
       "   confidence when incorrect reject  confidence when correct  \\\n",
       "0                              83.5                     93.0   \n",
       "1                              50.0                     75.0   \n",
       "2                              81.0                     89.0   \n",
       "3                              80.0                     88.0   \n",
       "4                              49.0                     65.0   \n",
       "\n",
       "   confidence when incorrect  sci  psqi  ess  insomnia  \\\n",
       "0                       90.0    9    13    2      True   \n",
       "1                       50.0    4    11    7      True   \n",
       "2                       88.0   10     9    3      True   \n",
       "3                       80.0   13     8   20      True   \n",
       "4                       49.0   13     9   12      True   \n",
       "\n",
       "   rank grouped by insomnia  \n",
       "0                      11.0  \n",
       "1                      21.0  \n",
       "2                      23.0  \n",
       "3                      19.0  \n",
       "4                       3.0  "
      ]
     },
     "execution_count": 95,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add to the original df\n",
    "# method='first': ranks assigned in order they appear in the array\n",
    "df['rank grouped by insomnia'] = grouped['percent correct'].rank(method='first')\n",
    "\n",
    "# Take a look\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e17df825",
   "metadata": {},
   "source": [
    "**Sort by ranks** (rank 1 == most insomnia)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "24ac8b1a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>insomnia</th>\n",
       "      <th>percent correct</th>\n",
       "      <th>rank grouped by insomnia</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>81</th>\n",
       "      <td>False</td>\n",
       "      <td>40.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>94</th>\n",
       "      <td>False</td>\n",
       "      <td>55.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39</th>\n",
       "      <td>False</td>\n",
       "      <td>57.5</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>76</th>\n",
       "      <td>False</td>\n",
       "      <td>60.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>96</th>\n",
       "      <td>False</td>\n",
       "      <td>60.0</td>\n",
       "      <td>5.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>True</td>\n",
       "      <td>90.0</td>\n",
       "      <td>21.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>True</td>\n",
       "      <td>90.0</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>True</td>\n",
       "      <td>92.5</td>\n",
       "      <td>23.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>True</td>\n",
       "      <td>95.0</td>\n",
       "      <td>24.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>True</td>\n",
       "      <td>97.5</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>102 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    insomnia  percent correct  rank grouped by insomnia\n",
       "81     False             40.0                       1.0\n",
       "94     False             55.0                       2.0\n",
       "39     False             57.5                       3.0\n",
       "76     False             60.0                       4.0\n",
       "96     False             60.0                       5.0\n",
       "..       ...              ...                       ...\n",
       "1       True             90.0                      21.0\n",
       "8       True             90.0                      22.0\n",
       "2       True             92.5                      23.0\n",
       "11      True             95.0                      24.0\n",
       "21      True             97.5                      25.0\n",
       "\n",
       "[102 rows x 3 columns]"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_sorted = df.sort_values(by=['insomnia', 'percent correct'])\n",
    "\n",
    "# Look at relevant columns\n",
    "df_sorted[['insomnia', 'percent correct', 'rank grouped by insomnia']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ef74ba91",
   "metadata": {},
   "source": [
    "<u>Aggregating and transforming with custom functions</u>"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d5747edf",
   "metadata": {},
   "source": [
    "compute the coefficient of variation (CoV, the standard deviation divided by the mean)\n",
    "*  values of each column are denoted by data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "a2db13ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "def coeff_of_var(data):\n",
    "    \"\"\"Compute coefficient of variation from an array of data.\"\"\"\n",
    "    return np.std(data) / np.mean(data)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7aca75f8",
   "metadata": {},
   "source": [
    "**apply aggregating function** for all but the categorical gender and insomnia columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "id": "242a25b5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>insomnia</th>\n",
       "      <th>participant number</th>\n",
       "      <th>age</th>\n",
       "      <th>correct hit percentage</th>\n",
       "      <th>correct reject percentage</th>\n",
       "      <th>percent correct</th>\n",
       "      <th>confidence when correct hit</th>\n",
       "      <th>confidence when incorrect hit</th>\n",
       "      <th>confidence when correct reject</th>\n",
       "      <th>confidence when incorrect reject</th>\n",
       "      <th>confidence when correct</th>\n",
       "      <th>confidence when incorrect</th>\n",
       "      <th>sci</th>\n",
       "      <th>psqi</th>\n",
       "      <th>ess</th>\n",
       "      <th>rank grouped by insomnia</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>False</td>\n",
       "      <td>0.586191</td>\n",
       "      <td>0.384262</td>\n",
       "      <td>0.166784</td>\n",
       "      <td>0.184061</td>\n",
       "      <td>0.138785</td>\n",
       "      <td>0.195978</td>\n",
       "      <td>0.350286</td>\n",
       "      <td>0.204312</td>\n",
       "      <td>0.298216</td>\n",
       "      <td>0.187304</td>\n",
       "      <td>0.262509</td>\n",
       "      <td>0.175245</td>\n",
       "      <td>0.577869</td>\n",
       "      <td>0.571566</td>\n",
       "      <td>0.5699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>True</td>\n",
       "      <td>0.536117</td>\n",
       "      <td>0.313853</td>\n",
       "      <td>0.218834</td>\n",
       "      <td>0.325760</td>\n",
       "      <td>0.171856</td>\n",
       "      <td>0.156219</td>\n",
       "      <td>0.225440</td>\n",
       "      <td>0.222827</td>\n",
       "      <td>0.211512</td>\n",
       "      <td>0.160061</td>\n",
       "      <td>0.197484</td>\n",
       "      <td>0.381907</td>\n",
       "      <td>0.299741</td>\n",
       "      <td>0.681514</td>\n",
       "      <td>0.5547</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   insomnia  participant number       age  correct hit percentage  \\\n",
       "0     False            0.586191  0.384262                0.166784   \n",
       "1      True            0.536117  0.313853                0.218834   \n",
       "\n",
       "   correct reject percentage  percent correct  confidence when correct hit  \\\n",
       "0                   0.184061         0.138785                     0.195978   \n",
       "1                   0.325760         0.171856                     0.156219   \n",
       "\n",
       "   confidence when incorrect hit  confidence when correct reject  \\\n",
       "0                       0.350286                        0.204312   \n",
       "1                       0.225440                        0.222827   \n",
       "\n",
       "   confidence when incorrect reject  confidence when correct  \\\n",
       "0                          0.298216                 0.187304   \n",
       "1                          0.211512                 0.160061   \n",
       "\n",
       "   confidence when incorrect       sci      psqi       ess  \\\n",
       "0                   0.262509  0.175245  0.577869  0.571566   \n",
       "1                   0.197484  0.381907  0.299741  0.681514   \n",
       "\n",
       "   rank grouped by insomnia  \n",
       "0                    0.5699  \n",
       "1                    0.5547  "
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cols = df.columns[(df.columns != 'gender') & (df.columns != 'insomnia')] # the cols are selected from original df\n",
    "grouped[cols].agg(coeff_of_var).reset_index() # apply selection to the grouped data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73ef0b48",
   "metadata": {},
   "source": [
    "**transformation** for empirical cumulative distribution function (ECDF) => ECDF(x) = fraction of measurements ≤ x.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "64827518",
   "metadata": {},
   "source": [
    "-> The y-values of an ECDF are the rankings of the entries, divided by the total number of entries. We will want to plot ECDFs for each group, so we apply the function for each group."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "8b2bc600",
   "metadata": {},
   "outputs": [],
   "source": [
    "def ecdf_y(data):\n",
    "    \"\"\"Give y-values of an ECDF for an unsorted column in a data frame.\"\"\"\n",
    "    return data.rank(method='first') / len(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "id": "d490df58",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>participant number</th>\n",
       "      <th>gender</th>\n",
       "      <th>age</th>\n",
       "      <th>correct hit percentage</th>\n",
       "      <th>correct reject percentage</th>\n",
       "      <th>percent correct</th>\n",
       "      <th>confidence when correct hit</th>\n",
       "      <th>confidence when incorrect hit</th>\n",
       "      <th>confidence when correct reject</th>\n",
       "      <th>confidence when incorrect reject</th>\n",
       "      <th>confidence when correct</th>\n",
       "      <th>confidence when incorrect</th>\n",
       "      <th>sci</th>\n",
       "      <th>psqi</th>\n",
       "      <th>ess</th>\n",
       "      <th>insomnia</th>\n",
       "      <th>rank grouped by insomnia</th>\n",
       "      <th>ecdf_y grouped by insomnia</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>8</td>\n",
       "      <td>f</td>\n",
       "      <td>39</td>\n",
       "      <td>65</td>\n",
       "      <td>80</td>\n",
       "      <td>72.5</td>\n",
       "      <td>91.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>93.0</td>\n",
       "      <td>83.5</td>\n",
       "      <td>93.0</td>\n",
       "      <td>90.0</td>\n",
       "      <td>9</td>\n",
       "      <td>13</td>\n",
       "      <td>2</td>\n",
       "      <td>True</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.44</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>16</td>\n",
       "      <td>m</td>\n",
       "      <td>42</td>\n",
       "      <td>90</td>\n",
       "      <td>90</td>\n",
       "      <td>90.0</td>\n",
       "      <td>75.5</td>\n",
       "      <td>55.5</td>\n",
       "      <td>70.5</td>\n",
       "      <td>50.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>4</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>True</td>\n",
       "      <td>21.0</td>\n",
       "      <td>0.84</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>18</td>\n",
       "      <td>f</td>\n",
       "      <td>31</td>\n",
       "      <td>90</td>\n",
       "      <td>95</td>\n",
       "      <td>92.5</td>\n",
       "      <td>89.5</td>\n",
       "      <td>90.0</td>\n",
       "      <td>86.0</td>\n",
       "      <td>81.0</td>\n",
       "      <td>89.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>10</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>True</td>\n",
       "      <td>23.0</td>\n",
       "      <td>0.92</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>22</td>\n",
       "      <td>f</td>\n",
       "      <td>35</td>\n",
       "      <td>100</td>\n",
       "      <td>75</td>\n",
       "      <td>87.5</td>\n",
       "      <td>89.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>71.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>80.0</td>\n",
       "      <td>13</td>\n",
       "      <td>8</td>\n",
       "      <td>20</td>\n",
       "      <td>True</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>27</td>\n",
       "      <td>f</td>\n",
       "      <td>74</td>\n",
       "      <td>60</td>\n",
       "      <td>65</td>\n",
       "      <td>62.5</td>\n",
       "      <td>68.5</td>\n",
       "      <td>49.0</td>\n",
       "      <td>61.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>65.0</td>\n",
       "      <td>49.0</td>\n",
       "      <td>13</td>\n",
       "      <td>9</td>\n",
       "      <td>12</td>\n",
       "      <td>True</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.12</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   participant number gender  age  correct hit percentage  \\\n",
       "0                   8      f   39                      65   \n",
       "1                  16      m   42                      90   \n",
       "2                  18      f   31                      90   \n",
       "3                  22      f   35                     100   \n",
       "4                  27      f   74                      60   \n",
       "\n",
       "   correct reject percentage  percent correct  confidence when correct hit  \\\n",
       "0                         80             72.5                         91.0   \n",
       "1                         90             90.0                         75.5   \n",
       "2                         95             92.5                         89.5   \n",
       "3                         75             87.5                         89.5   \n",
       "4                         65             62.5                         68.5   \n",
       "\n",
       "   confidence when incorrect hit  confidence when correct reject  \\\n",
       "0                           90.0                            93.0   \n",
       "1                           55.5                            70.5   \n",
       "2                           90.0                            86.0   \n",
       "3                            NaN                            71.0   \n",
       "4                           49.0                            61.0   \n",
       "\n",
       "   confidence when incorrect reject  confidence when correct  \\\n",
       "0                              83.5                     93.0   \n",
       "1                              50.0                     75.0   \n",
       "2                              81.0                     89.0   \n",
       "3                              80.0                     88.0   \n",
       "4                              49.0                     65.0   \n",
       "\n",
       "   confidence when incorrect  sci  psqi  ess  insomnia  \\\n",
       "0                       90.0    9    13    2      True   \n",
       "1                       50.0    4    11    7      True   \n",
       "2                       88.0   10     9    3      True   \n",
       "3                       80.0   13     8   20      True   \n",
       "4                       49.0   13     9   12      True   \n",
       "\n",
       "   rank grouped by insomnia  ecdf_y grouped by insomnia  \n",
       "0                      11.0                        0.44  \n",
       "1                      21.0                        0.84  \n",
       "2                      23.0                        0.92  \n",
       "3                      19.0                        0.76  \n",
       "4                       3.0                        0.12  "
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# apply transformation\n",
    "df['ecdf_y grouped by insomnia'] = grouped['percent correct'].transform(ecdf_y)\n",
    "\n",
    "# Take a look\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29368301",
   "metadata": {},
   "source": [
    "Now splitting by insomnia and gender, getting the quartiles and extrema, as well as mean and standard deviation, in percent of correct facial matchings."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "08a5c6ed",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gender</th>\n",
       "      <th>insomnia</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">f</th>\n",
       "      <th>False</th>\n",
       "      <td>55.0</td>\n",
       "      <td>82.045455</td>\n",
       "      <td>10.844982</td>\n",
       "      <td>55.0</td>\n",
       "      <td>75.000</td>\n",
       "      <td>85.00</td>\n",
       "      <td>88.750</td>\n",
       "      <td>100.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <td>19.0</td>\n",
       "      <td>73.947368</td>\n",
       "      <td>13.624829</td>\n",
       "      <td>45.0</td>\n",
       "      <td>67.500</td>\n",
       "      <td>72.50</td>\n",
       "      <td>83.750</td>\n",
       "      <td>95.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">m</th>\n",
       "      <th>False</th>\n",
       "      <td>22.0</td>\n",
       "      <td>80.000000</td>\n",
       "      <td>12.770875</td>\n",
       "      <td>40.0</td>\n",
       "      <td>77.500</td>\n",
       "      <td>82.50</td>\n",
       "      <td>89.375</td>\n",
       "      <td>95.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>True</th>\n",
       "      <td>6.0</td>\n",
       "      <td>82.916667</td>\n",
       "      <td>10.655593</td>\n",
       "      <td>67.5</td>\n",
       "      <td>76.875</td>\n",
       "      <td>83.75</td>\n",
       "      <td>88.750</td>\n",
       "      <td>97.5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 count       mean        std   min     25%    50%     75%  \\\n",
       "gender insomnia                                                             \n",
       "f      False      55.0  82.045455  10.844982  55.0  75.000  85.00  88.750   \n",
       "       True       19.0  73.947368  13.624829  45.0  67.500  72.50  83.750   \n",
       "m      False      22.0  80.000000  12.770875  40.0  77.500  82.50  89.375   \n",
       "       True        6.0  82.916667  10.655593  67.5  76.875  83.75  88.750   \n",
       "\n",
       "                   max  \n",
       "gender insomnia         \n",
       "f      False     100.0  \n",
       "       True       95.0  \n",
       "m      False      95.0  \n",
       "       True       97.5  "
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby(['gender', 'insomnia'])['percent correct'].describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "95d66813",
   "metadata": {},
   "source": [
    "**Loop over a groupby object**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "af6fc4af",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('f', False) :  85.0\n",
      "('f', True) :  72.5\n",
      "('m', False) :  82.5\n",
      "('m', True) :  83.75\n"
     ]
    }
   ],
   "source": [
    "# name of index, group like the correct percent median\n",
    "for name, group in df.groupby(['gender', 'insomnia']):\n",
    "    print(name, ': ', group['percent correct'].median())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3bc88070",
   "metadata": {},
   "source": [
    "**Tidying a dataset**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "806b8afd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "# Raw data from Reeves, et al., Dev. Cell, 22, 544, 2012.\n",
      "#\n",
      "# These data were used to create figure 1F.\n",
      "#\n",
      "# The first row of column headings contains the genotypes.\n",
      "# The second row contains the staining/visualization method.\n",
      "# The data give the computed gradient width as parametrized by\n",
      "# sigma (see the original paper).\n",
      "#\n",
      "# Kindly provided by Greg Reeves (NC State) and Nathanie Trisnadi and Angela Stathopoulos (Caltech)\n",
      "wt,wt,dl1/+; dl-venus/+,dl1/+; dl-venus/+,dl1/+; dl-venus/+,dl1/+; dl-gfp/+,dl1/+; dl-gfp/+,dl1/+; dl-gfp/+\n",
      "wholemounts,cross-sections,anti-Dorsal,anti-Venus,Venus (live),anti-Dorsal,anti-GFP,GFP (live)\n",
      "0.1288,0.1327,0.1482,0.1632,0.1666,0.2248,0.2389,0.2412\n",
      "0.1554,0.1457,0.1503,0.1671,0.1753,0.1891,0.2035,0.1942\n",
      "0.1306,0.1447,0.1577,0.1704,0.1705,0.1705,0.1943,0.2186\n",
      "0.1413,0.1282,0.1711,0.1779,,0.1735,0.2,0.2104\n",
      "0.1557,0.1487,0.1342,0.1483,,0.2135,0.256,0.2463\n",
      "0.1689,0.1203,0.1773,0.1831,,0.2129,0.2309,0.202\n",
      "0.1417,0.1315,0.1705,0.1783,,0.1782,0.1949,\n",
      "0.1315,0.1463,0.1596,0.1667,,0.2011,0.2216,\n",
      "0.1664,0.1458,0.1601,0.178,,0.1931,0.23,\n",
      "0.156,0.1402,0.1553,0.1653,,0.2146,0.2682,\n",
      "0.1479,0.133,0.1675,0.1708,,0.2398,0.2633,\n",
      "0.1475,0.1375,0.1692,0.178,,0.2193,0.2495,\n",
      "0.1347,0.1576,0.1584,0.174,,0.2112,0.2353,\n",
      "0.1129,0.1498,0.1546,0.1739,,0.2516,0.2642,\n",
      "0.125,0.1532,0.1605,0.1812,,0.1759,0.2238,\n",
      "0.1648,0.1352,0.1754,0.1902,,0.1787,0.2115,\n",
      "0.1555,0.1346,0.161,0.178,,0.164,0.1934,\n",
      "0.1403,0.1364,0.1491,0.1624,,0.1756,0.2041,\n",
      "0.1096,0.143,0.1442,0.1553,,0.1977,0.2286,\n",
      "0.1779,0.144,0.1728,0.1768,,0.1809,0.2104,\n",
      "0.1244,0.1805,0.1534,0.1701,,,,\n",
      "0.1708,0.1688,0.1576,0.1724,,,,\n",
      "0.1691,0.1532,0.1701,0.1738,,,,\n",
      "0.13,0.1681,0.1436,0.1703,,,,\n",
      ",0.1442,0.1746,0.1842,,,,\n",
      ",0.1362,0.1698,0.172,,,,\n",
      ",0.1391,0.171,0.1807,,,,\n",
      ",0.1606,0.1555,0.1796,,,,\n",
      ",0.1742,0.1616,0.1759,,,,\n",
      ",0.1335,,,,,,\n",
      ",0.1347,,,,,,\n",
      ",0.1404,,,,,,\n",
      ",0.1524,,,,,,\n",
      ",0.1469,,,,,,\n",
      ",0.1365,,,,,,\n",
      ",0.1433,,,,,,\n",
      ",0.1358,,,,,,\n",
      ",0.1643,,,,,,\n",
      ",0.1371,,,,,,\n",
      ",0.1141,,,,,,\n",
      ",0.1436,,,,,,\n",
      ",0.1223,,,,,,\n",
      ",0.1459,,,,,,\n",
      ",0.1368,,,,,,\n",
      ",0.1376,,,,,,\n",
      ",0.1601,,,,,,\n",
      ",0.1309,,,,,,\n",
      ",0.1305,,,,,,\n",
      ",0.1562,,,,,,\n",
      ",0.1577,,,,,,\n",
      ",0.154,,,,,,\n",
      ",0.1443,,,,,,\n",
      ",0.1364,,,,,,\n",
      ",0.1344,,,,,,\n",
      ",0.1315,,,,,,\n",
      ",0.1572,,,,,,\n",
      ",0.1441,,,,,,\n",
      ",0.1399,,,,,,\n",
      ",0.1497,,,,,,\n",
      ",0.1647,,,,,,\n",
      ",0.1629,,,,,,\n",
      ",0.1194,,,,,,\n",
      ",0.1296,,,,,,\n",
      ",0.1574,,,,,,\n",
      ",0.1365,,,,,,\n",
      ",0.1759,,,,,,\n",
      ",0.1522,,,,,,\n",
      ",0.1535,,,,,,\n",
      ",0.1398,,,,,,\n",
      ",0.1453,,,,,,\n",
      ",0.1569,,,,,,\n",
      ",0.1582,,,,,,\n",
      ",0.174,,,,,,\n",
      ",0.1781,,,,,,\n",
      ",0.1488,,,,,,\n",
      ",0.1663,,,,,,\n",
      ",0.1699,,,,,,\n",
      ",0.173,,,,,,\n",
      ",0.1655,,,,,,\n",
      ",0.1341,,,,,,\n",
      ",0.1014,,,,,,\n",
      ",0.162,,,,,,\n",
      ",0.1704,,,,,,\n",
      ",0.1499,,,,,,\n",
      ",0.1648,,,,,,\n",
      ",0.1397,,,,,,\n",
      ",0.1745,,,,,,\n",
      ",0.1684,,,,,,\n",
      ",0.1399,,,,,,\n",
      ",0.1512,,,,,,\n",
      ",0.1793,,,,,,\n",
      ",0.1581,,,,,,\n",
      ",0.1515,,,,,,\n",
      ",0.1465,,,,,,\n",
      ",0.162,,,,,,\n",
      ",0.1581,,,,,,\n",
      ",0.1559,,,,,,\n",
      ",0.147,,,,,,\n",
      ",0.1445,,,,,,\n",
      ",0.1673,,,,,,\n",
      ",0.1703,,,,,,\n",
      ",0.154,,,,,,\n",
      ",0.16,,,,,,\n",
      ",0.1147,,,,,,\n",
      ",0.1317,,,,,,\n",
      ",0.1685,,,,,,\n",
      ",0.1415,,,,,,\n",
      ",0.1575,,,,,,\n",
      ",0.1524,,,,,,\n",
      ",0.1374,,,,,,\n",
      ",0.1596,,,,,,\n",
      ",0.1159,,,,,,\n",
      ",0.1104,,,,,,\n",
      ",0.1368,,,,,,\n",
      ",0.1621,,,,,,\n",
      ",0.1117,,,,,,\n",
      ",0.1563,,,,,,\n",
      ",0.158,,,,,,\n",
      ",0.146,,,,,,\n",
      ",0.1174,,,,,,\n",
      ",0.155,,,,,,\n",
      ",0.1327,,,,,,\n",
      ",0.1443,,,,,,\n",
      ",0.1329,,,,,,\n",
      ",0.1567,,,,,,\n",
      ",0.1388,,,,,,\n",
      ",0.1449,,,,,,\n",
      ",0.1548,,,,,,\n",
      ",0.1727,,,,,,\n",
      ",0.142,,,,,,\n",
      ",0.1735,,,,,,\n",
      ",0.1641,,,,,,\n",
      ",0.1756,,,,,,\n",
      ",0.1691,,,,,,\n",
      ",0.1599,,,,,,\n",
      ",0.1271,,,,,,\n",
      ",0.1468,,,,,,\n",
      ",0.149,,,,,,\n",
      ",0.1706,,,,,,\n",
      ",0.1283,,,,,,\n",
      ",0.1325,,,,,,\n",
      ",0.1663,,,,,,\n",
      ",0.1325,,,,,,\n",
      ",0.1529,,,,,,\n",
      ",0.1682,,,,,,\n",
      ",0.1557,,,,,,\n",
      ",0.1577,,,,,,\n",
      ",0.1466,,,,,,\n",
      ",0.1671,,,,,,\n",
      ",0.1265,,,,,,\n",
      ",0.1448,,,,,,\n",
      ",0.174,,,,,,\n"
     ]
    }
   ],
   "source": [
    "!type data\\reeves_gradient_width_various_methods.csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "108fc0c7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th colspan=\"2\" halign=\"left\">wt</th>\n",
       "      <th colspan=\"3\" halign=\"left\">dl1/+; dl-venus/+</th>\n",
       "      <th colspan=\"3\" halign=\"left\">dl1/+; dl-gfp/+</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>wholemounts</th>\n",
       "      <th>cross-sections</th>\n",
       "      <th>anti-Dorsal</th>\n",
       "      <th>anti-Venus</th>\n",
       "      <th>Venus (live)</th>\n",
       "      <th>anti-Dorsal</th>\n",
       "      <th>anti-GFP</th>\n",
       "      <th>GFP (live)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.1288</td>\n",
       "      <td>0.1327</td>\n",
       "      <td>0.1482</td>\n",
       "      <td>0.1632</td>\n",
       "      <td>0.1666</td>\n",
       "      <td>0.2248</td>\n",
       "      <td>0.2389</td>\n",
       "      <td>0.2412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.1554</td>\n",
       "      <td>0.1457</td>\n",
       "      <td>0.1503</td>\n",
       "      <td>0.1671</td>\n",
       "      <td>0.1753</td>\n",
       "      <td>0.1891</td>\n",
       "      <td>0.2035</td>\n",
       "      <td>0.1942</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.1306</td>\n",
       "      <td>0.1447</td>\n",
       "      <td>0.1577</td>\n",
       "      <td>0.1704</td>\n",
       "      <td>0.1705</td>\n",
       "      <td>0.1705</td>\n",
       "      <td>0.1943</td>\n",
       "      <td>0.2186</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.1413</td>\n",
       "      <td>0.1282</td>\n",
       "      <td>0.1711</td>\n",
       "      <td>0.1779</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.1735</td>\n",
       "      <td>0.2000</td>\n",
       "      <td>0.2104</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.1557</td>\n",
       "      <td>0.1487</td>\n",
       "      <td>0.1342</td>\n",
       "      <td>0.1483</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.2135</td>\n",
       "      <td>0.2560</td>\n",
       "      <td>0.2463</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           wt                dl1/+; dl-venus/+                          \\\n",
       "  wholemounts cross-sections       anti-Dorsal anti-Venus Venus (live)   \n",
       "0      0.1288         0.1327            0.1482     0.1632       0.1666   \n",
       "1      0.1554         0.1457            0.1503     0.1671       0.1753   \n",
       "2      0.1306         0.1447            0.1577     0.1704       0.1705   \n",
       "3      0.1413         0.1282            0.1711     0.1779          NaN   \n",
       "4      0.1557         0.1487            0.1342     0.1483          NaN   \n",
       "\n",
       "  dl1/+; dl-gfp/+                      \n",
       "      anti-Dorsal anti-GFP GFP (live)  \n",
       "0          0.2248   0.2389     0.2412  \n",
       "1          0.1891   0.2035     0.1942  \n",
       "2          0.1705   0.1943     0.2186  \n",
       "3          0.1735   0.2000     0.2104  \n",
       "4          0.2135   0.2560     0.2463  "
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# read in and remove comment, set multiindex\n",
    "df = pd.read_csv('data/reeves_gradient_width_various_methods.csv', comment='#', header=[0,1])\n",
    "\n",
    "df.head() # rows have no meaning"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d1faedc9",
   "metadata": {},
   "source": [
    "As the columns, but the rows too need to have a meaning, this is not yet a tidy dataset and we need to melt the df into one that fullfilles tidyness"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "id": "6a6ef966",
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.melt(df, value_name='gradient width', var_name=['genotype', 'method'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "26217fad",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(  genotype       method  gradient width\n",
       " 0       wt  wholemounts          0.1288\n",
       " 1       wt  wholemounts          0.1554\n",
       " 2       wt  wholemounts          0.1306\n",
       " 3       wt  wholemounts          0.1413\n",
       " 4       wt  wholemounts          0.1557,\n",
       "              genotype      method  gradient width\n",
       " 1211  dl1/+; dl-gfp/+  GFP (live)             NaN\n",
       " 1212  dl1/+; dl-gfp/+  GFP (live)             NaN\n",
       " 1213  dl1/+; dl-gfp/+  GFP (live)             NaN\n",
       " 1214  dl1/+; dl-gfp/+  GFP (live)             NaN\n",
       " 1215  dl1/+; dl-gfp/+  GFP (live)             NaN)"
      ]
     },
     "execution_count": 131,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head(), df.tail()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "f83ab258",
   "metadata": {},
   "outputs": [],
   "source": [
    "# drop NaN\n",
    "df = df.dropna()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7b3528f9",
   "metadata": {},
   "source": [
    "can group by genotype and method again and show a summary statistic with groupby"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "7ee41d1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th colspan=\"8\" halign=\"left\">gradient width</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>count</th>\n",
       "      <th>mean</th>\n",
       "      <th>std</th>\n",
       "      <th>min</th>\n",
       "      <th>25%</th>\n",
       "      <th>50%</th>\n",
       "      <th>75%</th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>genotype</th>\n",
       "      <th>method</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">dl1/+; dl-gfp/+</th>\n",
       "      <th>GFP (live)</th>\n",
       "      <td>6.0</td>\n",
       "      <td>0.218783</td>\n",
       "      <td>0.021052</td>\n",
       "      <td>0.1942</td>\n",
       "      <td>0.204100</td>\n",
       "      <td>0.21450</td>\n",
       "      <td>0.235550</td>\n",
       "      <td>0.2463</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>anti-Dorsal</th>\n",
       "      <td>20.0</td>\n",
       "      <td>0.198300</td>\n",
       "      <td>0.024468</td>\n",
       "      <td>0.1640</td>\n",
       "      <td>0.177625</td>\n",
       "      <td>0.19540</td>\n",
       "      <td>0.213775</td>\n",
       "      <td>0.2516</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>anti-GFP</th>\n",
       "      <td>20.0</td>\n",
       "      <td>0.226120</td>\n",
       "      <td>0.024664</td>\n",
       "      <td>0.1934</td>\n",
       "      <td>0.203950</td>\n",
       "      <td>0.22620</td>\n",
       "      <td>0.241550</td>\n",
       "      <td>0.2682</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">dl1/+; dl-venus/+</th>\n",
       "      <th>Venus (live)</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.170800</td>\n",
       "      <td>0.004358</td>\n",
       "      <td>0.1666</td>\n",
       "      <td>0.168550</td>\n",
       "      <td>0.17050</td>\n",
       "      <td>0.172900</td>\n",
       "      <td>0.1753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>anti-Dorsal</th>\n",
       "      <td>29.0</td>\n",
       "      <td>0.160490</td>\n",
       "      <td>0.010805</td>\n",
       "      <td>0.1342</td>\n",
       "      <td>0.154600</td>\n",
       "      <td>0.16010</td>\n",
       "      <td>0.170100</td>\n",
       "      <td>0.1773</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>anti-Venus</th>\n",
       "      <td>29.0</td>\n",
       "      <td>0.173031</td>\n",
       "      <td>0.008742</td>\n",
       "      <td>0.1483</td>\n",
       "      <td>0.170100</td>\n",
       "      <td>0.17390</td>\n",
       "      <td>0.178000</td>\n",
       "      <td>0.1902</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">wt</th>\n",
       "      <th>cross-sections</th>\n",
       "      <td>152.0</td>\n",
       "      <td>0.148458</td>\n",
       "      <td>0.016141</td>\n",
       "      <td>0.1014</td>\n",
       "      <td>0.136725</td>\n",
       "      <td>0.14695</td>\n",
       "      <td>0.159925</td>\n",
       "      <td>0.1805</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>wholemounts</th>\n",
       "      <td>24.0</td>\n",
       "      <td>0.145279</td>\n",
       "      <td>0.019059</td>\n",
       "      <td>0.1096</td>\n",
       "      <td>0.130450</td>\n",
       "      <td>0.14460</td>\n",
       "      <td>0.158200</td>\n",
       "      <td>0.1779</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                 gradient width                              \\\n",
       "                                          count      mean       std     min   \n",
       "genotype          method                                                      \n",
       "dl1/+; dl-gfp/+   GFP (live)                6.0  0.218783  0.021052  0.1942   \n",
       "                  anti-Dorsal              20.0  0.198300  0.024468  0.1640   \n",
       "                  anti-GFP                 20.0  0.226120  0.024664  0.1934   \n",
       "dl1/+; dl-venus/+ Venus (live)              3.0  0.170800  0.004358  0.1666   \n",
       "                  anti-Dorsal              29.0  0.160490  0.010805  0.1342   \n",
       "                  anti-Venus               29.0  0.173031  0.008742  0.1483   \n",
       "wt                cross-sections          152.0  0.148458  0.016141  0.1014   \n",
       "                  wholemounts              24.0  0.145279  0.019059  0.1096   \n",
       "\n",
       "                                                                       \n",
       "                                       25%      50%       75%     max  \n",
       "genotype          method                                               \n",
       "dl1/+; dl-gfp/+   GFP (live)      0.204100  0.21450  0.235550  0.2463  \n",
       "                  anti-Dorsal     0.177625  0.19540  0.213775  0.2516  \n",
       "                  anti-GFP        0.203950  0.22620  0.241550  0.2682  \n",
       "dl1/+; dl-venus/+ Venus (live)    0.168550  0.17050  0.172900  0.1753  \n",
       "                  anti-Dorsal     0.154600  0.16010  0.170100  0.1773  \n",
       "                  anti-Venus      0.170100  0.17390  0.178000  0.1902  \n",
       "wt                cross-sections  0.136725  0.14695  0.159925  0.1805  \n",
       "                  wholemounts     0.130450  0.14460  0.158200  0.1779  "
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.groupby(['genotype', 'method']).describe()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "73ad8593",
   "metadata": {},
   "source": [
    "## Lesson 18: Making plots"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "5e093345",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       "        .bk-notebook-logo {\n",
       "            display: block;\n",
       "            width: 20px;\n",
       "            height: 20px;\n",
       "            background-image: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABQAAAAUCAYAAACNiR0NAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAABx0RVh0U29mdHdhcmUAQWRvYmUgRmlyZXdvcmtzIENTNui8sowAAAOkSURBVDiNjZRtaJVlGMd/1/08zzln5zjP1LWcU9N0NkN8m2CYjpgQYQXqSs0I84OLIC0hkEKoPtiH3gmKoiJDU7QpLgoLjLIQCpEsNJ1vqUOdO7ppbuec5+V+rj4ctwzd8IIbbi6u+8f1539dt3A78eXC7QizUF7gyV1fD1Yqg4JWz84yffhm0qkFqBogB9rM8tZdtwVsPUhWhGcFJngGeWrPzHm5oaMmkfEg1usvLFyc8jLRqDOMru7AyC8saQr7GG7f5fvDeH7Ej8CM66nIF+8yngt6HWaKh7k49Soy9nXurCi1o3qUbS3zWfrYeQDTB/Qj6kX6Ybhw4B+bOYoLKCC9H3Nu/leUTZ1JdRWkkn2ldcCamzrcf47KKXdAJllSlxAOkRgyHsGC/zRday5Qld9DyoM4/q/rUoy/CXh3jzOu3bHUVZeU+DEn8FInkPBFlu3+nW3Nw0mk6vCDiWg8CeJaxEwuHS3+z5RgY+YBR6V1Z1nxSOfoaPa4LASWxxdNp+VWTk7+4vzaou8v8PN+xo+KY2xsw6une2frhw05CTYOmQvsEhjhWjn0bmXPjpE1+kplmmkP3suftwTubK9Vq22qKmrBhpY4jvd5afdRA3wGjFAgcnTK2s4hY0/GPNIb0nErGMCRxWOOX64Z8RAC4oCXdklmEvcL8o0BfkNK4lUg9HTl+oPlQxdNo3Mg4Nv175e/1LDGzZen30MEjRUtmXSfiTVu1kK8W4txyV6BMKlbgk3lMwYCiusNy9fVfvvwMxv8Ynl6vxoByANLTWplvuj/nF9m2+PDtt1eiHPBr1oIfhCChQMBw6Aw0UulqTKZdfVvfG7VcfIqLG9bcldL/+pdWTLxLUy8Qq38heUIjh4XlzZxzQm19lLFlr8vdQ97rjZVOLf8nclzckbcD4wxXMidpX30sFd37Fv/GtwwhzhxGVAprjbg0gCAEeIgwCZyTV2Z1REEW8O4py0wsjeloKoMr6iCY6dP92H6Vw/oTyICIthibxjm/DfN9lVz8IqtqKYLUXfoKVMVQVVJOElGjrnnUt9T9wbgp8AyYKaGlqingHZU/uG2NTZSVqwHQTWkx9hxjkpWDaCg6Ckj5qebgBVbT3V3NNXMSiWSDdGV3hrtzla7J+duwPOToIg42ChPQOQjspnSlp1V+Gjdged7+8UN5CRAV7a5EdFNwCjEaBR27b3W890TE7g24NAP/mMDXRWrGoFPQI9ls/MWO2dWFAar/xcOIImbbpA3zgAAAABJRU5ErkJggg==);\n",
       "        }\n",
       "    </style>\n",
       "    <div>\n",
       "        <a href=\"https://bokeh.org\" target=\"_blank\" class=\"bk-notebook-logo\"></a>\n",
       "        <span id=\"a9c1c414-c152-4e33-bf22-43b200e759f0\">Loading BokehJS ...</span>\n",
       "    </div>\n"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "application/javascript": "(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\nconst JS_MIME_TYPE = 'application/javascript';\n  const HTML_MIME_TYPE = 'text/html';\n  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n  const CLASS_NAME = 'output_bokeh rendered_html';\n\n  /**\n   * Render data to the DOM node\n   */\n  function render(props, node) {\n    const script = document.createElement(\"script\");\n    node.appendChild(script);\n  }\n\n  /**\n   * Handle when an output is cleared or removed\n   */\n  function handleClearOutput(event, handle) {\n    const cell = handle.cell;\n\n    const id = cell.output_area._bokeh_element_id;\n    const server_id = cell.output_area._bokeh_server_id;\n    // Clean up Bokeh references\n    if (id != null && id in Bokeh.index) {\n      Bokeh.index[id].model.document.clear();\n      delete Bokeh.index[id];\n    }\n\n    if (server_id !== undefined) {\n      // Clean up Bokeh references\n      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n      cell.notebook.kernel.execute(cmd_clean, {\n        iopub: {\n          output: function(msg) {\n            const id = msg.content.text.trim();\n            if (id in Bokeh.index) {\n              Bokeh.index[id].model.document.clear();\n              delete Bokeh.index[id];\n            }\n          }\n        }\n      });\n      // Destroy server and session\n      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n      cell.notebook.kernel.execute(cmd_destroy);\n    }\n  }\n\n  /**\n   * Handle when a new output is added\n   */\n  function handleAddOutput(event, handle) {\n    const output_area = handle.output_area;\n    const output = handle.output;\n\n    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n      return\n    }\n\n    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n\n    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n      // store reference to embed id on output_area\n      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n    }\n    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n      const bk_div = document.createElement(\"div\");\n      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n      const script_attrs = bk_div.children[0].attributes;\n      for (let i = 0; i < script_attrs.length; i++) {\n        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n      }\n      // store reference to server id on output_area\n      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n    }\n  }\n\n  function register_renderer(events, OutputArea) {\n\n    function append_mime(data, metadata, element) {\n      // create a DOM node to render to\n      const toinsert = this.create_output_subarea(\n        metadata,\n        CLASS_NAME,\n        EXEC_MIME_TYPE\n      );\n      this.keyboard_manager.register_events(toinsert);\n      // Render to node\n      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n      render(props, toinsert[toinsert.length - 1]);\n      element.append(toinsert);\n      return toinsert\n    }\n\n    /* Handle when an output is cleared or removed */\n    events.on('clear_output.CodeCell', handleClearOutput);\n    events.on('delete.Cell', handleClearOutput);\n\n    /* Handle when a new output is added */\n    events.on('output_added.OutputArea', handleAddOutput);\n\n    /**\n     * Register the mime type and append_mime function with output_area\n     */\n    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n      /* Is output safe? */\n      safe: true,\n      /* Index of renderer in `output_area.display_order` */\n      index: 0\n    });\n  }\n\n  // register the mime type if in Jupyter Notebook environment and previously unregistered\n  if (root.Jupyter !== undefined) {\n    const events = require('base/js/events');\n    const OutputArea = require('notebook/js/outputarea').OutputArea;\n\n    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n      register_renderer(events, OutputArea);\n    }\n  }\n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(\"a9c1c414-c152-4e33-bf22-43b200e759f0\");\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-3.2.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-3.2.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-3.2.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-3.2.1.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-3.2.1.min.js\"];\n  const css_urls = [];\n\n  const inline_js = [    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\nfunction(Bokeh) {\n    }\n  ];\n\n  function run_inline_js() {\n    if (root.Bokeh !== undefined || force === true) {\n          for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\nif (force === true) {\n        display_loaded();\n      }} else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(\"a9c1c414-c152-4e33-bf22-43b200e759f0\")).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));",
      "application/vnd.bokehjs_load.v0+json": ""
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "# Our main plotting package (must have explicit import of submodules)\n",
    "import bokeh.io\n",
    "import bokeh.models\n",
    "import bokeh.plotting\n",
    "\n",
    "# Enable viewing Bokeh plots in the notebook\n",
    "bokeh.io.output_notebook()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9dbba4de",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
